// @sanity/workbench-cli/dist/actions/build/vite/workbench-vite-plugins.js
// The build-facing entry of the workbench federation stack: turn a workbench
// app's build inputs into the Vite plugins that produce its module-federation
// remote. `@sanity/cli-build`'s `getViteConfig` calls this instead of
// assembling `federation`'s options itself, so the discriminated app-vs-studio
// option shape, the no-app-view rule, and the studio-config requirement all
// live here next to `federation` — the build package just hands over its inputs.
import path from 'node:path';
import { readPackageJson } from '@sanity/cli-core';
import { federation } from './plugin.js';
import { sanityAppId } from './plugins/plugin-sanity-app-id.js';
/**
 * A workbench studio renders from its `sanity.config.*`, so the build needs one.
 * An explicit `applicationType: 'studio'` wins over detection, so a studio can
 * reach here with no config file — fail with the fix rather than a cryptic build
 * error downstream.
 */ function requireStudioConfigPath(relativeConfigLocation) {
    if (relativeConfigLocation === null) {
        throw new Error('Workbench studios need a sanity.config.js or sanity.config.ts file. ' + "Add one, or remove `applicationType: 'studio'` from `unstable_defineApp` " + 'to let the CLI infer the application type.');
    }
    return relativeConfigLocation;
}
/** Build the Vite plugins for a workbench app's module-federation remote. */ export async function workbenchVitePlugins(options) {
    const { appId, cwd, entries, exposes, isApp } = options;
    const pkgJson = await readPackageJson(path.join(cwd, 'package.json'));
    const federationPlugin = federation({
        ...isApp ? {
            // `null` relativeEntry (a branded app with no `entry`) → omit `appEntry`,
            // so the remote exposes no `./App`, only its views.
            ...entries.relativeEntry ? {
                appEntry: entries.relativeEntry
            } : {},
            isApp: true
        } : {
            isApp: false,
            studioConfigPath: requireStudioConfigPath(entries.relativeConfigLocation)
        },
        exposes,
        pkgJson,
        workDir: cwd
    });
    return appId === undefined ? federationPlugin : [
        federationPlugin,
        sanityAppId(appId)
    ];
}
