// @sanity/workbench-cli/dist/resolveWorkbenchApp.js
// Package-internal shared resolver: turn a CLI config's branded
// `unstable_defineApp` app into its declared interfaces, or `null` for a plain
// project. The build and deploy accessors (actions/build, actions/deploy) each
// build their command-specific view on top of this one brand-check +
// extraction, so the discrimination lives in exactly one place.
import { isWorkbenchApp, readConfig } from './defineApp.js';
/**
 * Resolve the workbench app for a CLI config, or `null` for a plain project.
 * @public
 */ export function resolveWorkbenchApp(cliConfig) {
    const app = cliConfig?.app;
    if (!isWorkbenchApp(app)) return null;
    return {
        applicationType: app.applicationType,
        config: readConfig(app),
        entry: app.entry,
        icon: app.icon,
        isSingleton: app.isSingleton,
        name: app.name,
        organizationId: app.organizationId,
        services: app.services ?? [],
        slug: app.slug,
        views: app.views ?? [],
        visibility: app.visibility
    };
}
