// @sanity/workbench/dist/_internal.js
import { BehaviorSubject } from "rxjs";
import { getSanityEnv } from "./_chunks-es/studio.js";
import { logger } from "./_chunks-es/index.js";
import { createInstance, log } from "./_chunks-es/module-federation.js";
import { connect, defineStateTopics, disconnectApp } from "./_chunks-es/bus.js";
const REMOTE_NAME = "workbench-remote", REMOTE_MODULE = "App", LOCAL_APPS_HMR_EVENT = "sanity:workbench:local-applications", LOCAL_APPS_HMR_REQUEST = "sanity:workbench:get-local-applications";
async function renderWorkbench(rootElement, config, options) {
  if (!rootElement)
    throw new Error("Missing root element to mount application into");
  const deployedWorkbenchUrl = getSanityEnv() === "staging" ? "workbench-apps-of5p8qpku.run.sanity.work" : "workbench-apps-oSyH1iET5.sanity.run", remoteUrl = import.meta.env.SANITY_INTERNAL_WORKBENCH_REMOTE_URL ?? `https://${deployedWorkbenchUrl}/mf-manifest.json`;
  let remoteModule = await createInstance({
    name: "sanity-workbench",
    plugins: [log(logger.debug)],
    remotes: [
      {
        name: REMOTE_NAME,
        entry: remoteUrl
      }
    ]
  }).loadRemote(
    `${REMOTE_NAME}/${REMOTE_MODULE}`
  );
  if (!remoteModule || typeof remoteModule.render != "function")
    throw new Error(
      `Remote module "${REMOTE_NAME}/${REMOTE_MODULE}" did not expose a render function`
    );
  let localApplications, appConfigs, cleanupHmr = () => {
  };
  if (import.meta.hot) {
    const localApps$ = new BehaviorSubject([]), configs$ = new BehaviorSubject([]), handler = (payload) => {
      localApps$.next(payload.applications), configs$.next(payload.configs ?? []);
    };
    import.meta.hot.on(LOCAL_APPS_HMR_EVENT, handler), import.meta.hot.send(LOCAL_APPS_HMR_REQUEST), localApplications = localApps$, appConfigs = configs$, cleanupHmr = () => import.meta.hot?.off(LOCAL_APPS_HMR_EVENT, handler);
  }
  const unmount = remoteModule.render(
    rootElement,
    { config, appConfigs, localApplications },
    options
  );
  return () => {
    cleanupHmr(), unmount();
  };
}
export {
  connect,
  defineStateTopics,
  disconnectApp,
  renderWorkbench
};
