// @sanity/workbench/dist/core.js
import { brandApplicationId, InstalledApplication, getApplicationDomain } from "./_chunks-es/studio.js";
import { AbstractApplication, ActiveDeployment, AppConfigSchema, ApplicationBase, ApplicationId, BrettApplication, BrettWorkspace, CanvasApplication, CoreApplication, ExternalApplication, InstallationId, InterfaceSchema, InternalApplication, LocalApplication, LocalCoreApplicationManifest, LocalInterfaceSchema, LocalStudioApplicationManifest, Organization, OrganizationId, Project, ProjectId, Studio, StudioApplication, StudioWorkspace, brandCanvasId, brandOrganizationId, brandProjectId, getApiHost, getSanityDomain, getSanityEnv, joinUrlPaths, parseCanvas, parseCoreApplication, parseInstalledApplication, parseOrganization, parseProject, parseStudioApplication, resolveAppConfig } from "./_chunks-es/studio.js";
import { createLogger, logger } from "./_chunks-es/index.js";
class ApplicationList {
  applications;
  /**
   * @param applications - The applications to initialize the list with.
   */
  constructor(applications) {
    this.applications = [...applications];
  }
  findApplicationsByType(_type) {
    const types = Array.isArray(_type) ? _type : [_type];
    return this.applications.filter((r) => types.includes(r.type));
  }
  findApplication(type, id) {
    switch (type) {
      case "media-library":
      case "canvas":
        return this.applications.find((r) => r.type === type);
      case "coreApp":
      case "studio":
      case "workspace":
        return this.applications.find((r) => r.type === type && r.id === id);
    }
  }
  /**
   * The one application a location puts in the foreground — media library and
   * canvas by route (one per org), applications by the id in the URL.
   * `undefined` when nothing matches or the match has no app view (not
   * navigable, e.g. an SDK app that only docks panels).
   */
  findByPath(pathname) {
    const [root, id] = pathname.split("/").filter(Boolean), application = (() => {
      switch (root) {
        case "media":
          return this.findApplication("media-library");
        case "canvas":
          return this.findApplication("canvas");
        case "studio":
          return id ? this.findApplication("studio", brandApplicationId(id)) : void 0;
        case "application":
          return id ? this.findApplication("coreApp", brandApplicationId(id)) : void 0;
        case "local":
          return this.applications.find(
            (candidate) => candidate.isLocal && (candidate.type === "studio" || candidate.type === "coreApp") && candidate.id === id
          );
        default:
          return;
      }
    })();
    return application?.hasAppView ? application : void 0;
  }
  static areApplicationsEqual(application1, application2) {
    return !application1 || !application2 || application1.type !== application2.type ? !1 : application1.id === application2.id;
  }
}
class MediaLibraryApplication extends InstalledApplication {
  constructor(id, organizationId) {
    super({ type: "media-library", slug: "media-library", id, organizationId });
  }
  get href() {
    return "media";
  }
  get title() {
    return "Media Library";
  }
  get url() {
    return new URL(
      `https://${this.slug}-apps-${this.organizationId}.${getApplicationDomain()}`
    );
  }
}
export {
  AbstractApplication,
  ActiveDeployment,
  AppConfigSchema,
  ApplicationBase,
  ApplicationId,
  ApplicationList,
  BrettApplication,
  BrettWorkspace,
  CanvasApplication,
  CoreApplication,
  ExternalApplication,
  InstallationId,
  InstalledApplication,
  InterfaceSchema,
  InternalApplication,
  LocalApplication,
  LocalCoreApplicationManifest,
  LocalInterfaceSchema,
  LocalStudioApplicationManifest,
  MediaLibraryApplication,
  Organization,
  OrganizationId,
  Project,
  ProjectId,
  Studio,
  StudioApplication,
  StudioWorkspace,
  brandApplicationId,
  brandCanvasId,
  brandOrganizationId,
  brandProjectId,
  createLogger,
  getApiHost,
  getApplicationDomain,
  getSanityDomain,
  getSanityEnv,
  joinUrlPaths,
  logger,
  parseCanvas,
  parseCoreApplication,
  parseInstalledApplication,
  parseOrganization,
  parseProject,
  parseStudioApplication,
  resolveAppConfig
};
