// @sanity/workbench/src/system/auth.machine.ts
import {
  type AuthState,
  AuthStateType,
  type CurrentUser,
  getAuthState,
  type LoggedInAuthState,
  logout,
  type OrganizationBase,
  resolveOrganization,
  type SanityInstance,
} from "@sanity/sdk";
import {
  and,
  assign,
  fromCallback,
  fromObservable,
  fromPromise,
  setup,
  stateIn,
} from "xstate";

import { defineStateTopics, os } from "../runtime/bus";
import type { OSBaseInput } from "./root.machine";

/**
 * The session context the auth machine establishes: login also resolves the
 * organization everything else is fetched against.
 * @internal
 */
export interface AuthInput extends OSBaseInput {
  organizationId: string;
}

const authStateLogic = fromObservable<AuthState, OSBaseInput>(
  ({ input }) => getAuthState(input.instance).observable,
);

type CurrentOrganization = Pick<OrganizationBase, "id" | "name" | "slug">;

const organizationResolverLogic = fromPromise<CurrentOrganization, AuthInput>(
  async ({ input }) => {
    const organization = await resolveOrganization(input.instance, {
      organizationId: input.organizationId,
    });
    return {
      id: organization.id,
      name: organization.name,
      slug: organization.slug,
    };
  },
);

// Forwards `auth.token.refresh` requests into the machine, so `publishToken`
// stays the sole `auth.token` emitter: the machine re-publishes the token and
// replies to the caller, modelling a genuine token request. Per-app re-issuance
// slots into that handler later — today it resends the session's own token.
const tokenRefreshResponder = fromCallback(({ sendBack }) => {
  const stop = new AbortController();

  os.subscribe(
    "auth.token.refresh",
    (message) => {
      sendBack({ type: "token.refresh", reply: message.reply });
    },
    { signal: stop.signal },
  );

  return () => stop.abort();
});

/**
 * @internal
 */
export interface LogoutInput extends OSBaseInput {}

const logoutActorLogic = fromPromise<void, LogoutInput>(async ({ input }) => {
  await logout(input.instance);
});

type AuthContext = {
  instance: SanityInstance;
  organizationId: string;
  token: string | null;
  currentUser: CurrentUser | null;
  error: unknown;
};

type AuthEvent =
  | { type: "auth.logout" }
  | { type: "token.refresh"; reply: (token: string) => void };

const isLoggedInComplete = (state: AuthState | undefined): boolean =>
  state?.type === AuthStateType.LOGGED_IN &&
  Boolean((state as LoggedInAuthState).token) &&
  (state as LoggedInAuthState).currentUser !== null;

const sessionChanged = (
  context: AuthContext,
  state: LoggedInAuthState,
): boolean =>
  state.token !== context.token ||
  state.currentUser?.id !== context.currentUser?.id ||
  state.currentUser?.name !== context.currentUser?.name ||
  state.currentUser?.email !== context.currentUser?.email ||
  state.currentUser?.profileImage !== context.currentUser?.profileImage;

export const authLogic = setup({
  types: {
    input: {} as AuthInput,
    context: {} as AuthContext,
    events: {} as AuthEvent,
    tags: {} as "authenticating" | "authenticated" | "error",
  },
  actors: {
    authState: authStateLogic,
    logoutActor: logoutActorLogic,
    organizationResolver: organizationResolverLogic,
    tokenRefreshResponder,
  },
  delays: {
    authTimeout: 30_000,
  },
  guards: {
    isLoggedInComplete: (_, params: { state: AuthState | undefined }) =>
      isLoggedInComplete(params.state),
    isAuthState: (
      _,
      params: { state: AuthState | undefined; type: AuthStateType },
    ) => params.state?.type === params.type,
  },
  actions: {
    // Seed the (suspending) topic before any publish, or a state emit would be
    // misrouted as an event.
    seedTopics: () => {
      defineStateTopics(os, {
        "auth.token": undefined,
        "organizations.current": undefined,
        "users.current": undefined,
      });
    },
    publishToken: ({ context }) => {
      os.emit("auth.token", context.token);
    },
    replyToken: ({ context }, params: { reply: (token: string) => void }) => {
      if (context.token) params.reply(context.token);
    },
    publishOrganization: (
      _,
      params: { organization: CurrentOrganization | null },
    ) => {
      os.emit("organizations.current", params.organization);
    },
    publishCurrentUser: ({ context }) => {
      const user = context.currentUser;
      os.emit(
        "users.current",
        user && {
          id: user.id,
          name: user.name,
          email: user.email,
          profileImage: user.profileImage,
        },
      );
    },
    setLoggedIn: assign({
      token: (_, params: { token: string; currentUser: CurrentUser }) =>
        params.token,
      currentUser: (_, params: { token: string; currentUser: CurrentUser }) =>
        params.currentUser,
      error: () => null,
    }),
    clearAuth: assign({
      token: () => null,
      currentUser: () => null,
      error: () => null,
    }),
    setError: assign({
      token: () => null,
      currentUser: () => null,
      error: (_, params: { error: unknown }) => params.error,
    }),
  },
}).createMachine({
  id: "auth",
  entry: [{ type: "seedTopics" }],
  initial: "init",
  context: ({ input }) => ({
    instance: input.instance,
    organizationId: input.organizationId,
    token: null,
    currentUser: null,
    error: null,
  }),
  invoke: {
    src: "authState",
    input: ({ context }) => ({ instance: context.instance }),
    onSnapshot: [
      // Already logged in and the session actually changed (token refresh, user
      // update): update in place — no re-entry, no republish of an unchanged
      // user.
      {
        guard: and([
          ({ event }) => isLoggedInComplete(event.snapshot.context),
          stateIn(`#auth.${AuthStateType.LOGGED_IN}`),
          ({ context, event }) =>
            sessionChanged(
              context,
              event.snapshot.context as LoggedInAuthState,
            ),
        ]),
        actions: [
          {
            type: "setLoggedIn",
            params: ({ event }) => {
              const state = event.snapshot.context as LoggedInAuthState;
              return {
                token: state.token,
                currentUser: state.currentUser!,
              };
            },
          },
          { type: "publishCurrentUser" },
          { type: "publishToken" },
        ],
      },
      // Already logged in, nothing changed: swallow the snapshot.
      {
        guard: and([
          ({ event }) => isLoggedInComplete(event.snapshot.context),
          stateIn(`#auth.${AuthStateType.LOGGED_IN}`),
        ]),
      },
      {
        guard: {
          type: "isLoggedInComplete",
          params: ({ event }) => ({
            state: event.snapshot.context,
          }),
        },
        actions: [
          {
            type: "setLoggedIn",
            params: ({ event }) => {
              const state = event.snapshot.context as LoggedInAuthState;
              return {
                token: state.token,
                currentUser: state.currentUser!,
              };
            },
          },
        ],
        target: `.${AuthStateType.LOGGED_IN}`,
      },
      {
        guard: {
          type: "isAuthState",
          params: ({ event }) => ({
            state: event.snapshot.context,
            type: AuthStateType.LOGGING_IN,
          }),
        },
        actions: [{ type: "clearAuth" }],
        target: `.${AuthStateType.LOGGING_IN}`,
      },
      {
        guard: {
          type: "isAuthState",
          params: ({ event }) => ({
            state: event.snapshot.context,
            type: AuthStateType.ERROR,
          }),
        },
        actions: [
          {
            type: "setError",
            params: ({ event }) => ({
              error:
                event.snapshot.context?.type === AuthStateType.ERROR
                  ? event.snapshot.context.error
                  : null,
            }),
          },
        ],
        target: `.${AuthStateType.ERROR}`,
      },
      {
        guard: {
          type: "isAuthState",
          params: ({ event }) => ({
            state: event.snapshot.context,
            type: AuthStateType.LOGGED_OUT,
          }),
        },
        actions: [{ type: "clearAuth" }],
        target: `.${AuthStateType.LOGGED_OUT}`,
      },
    ],
  },
  states: {
    init: {
      tags: ["authenticating"],
      after: {
        authTimeout: {
          actions: [
            {
              type: "setError",
              params: () => ({
                error: new Error("Authentication timed out"),
              }),
            },
          ],
          target: AuthStateType.ERROR,
        },
      },
    },
    [AuthStateType.LOGGING_IN]: {
      tags: ["authenticating"],
      after: {
        authTimeout: {
          actions: [
            {
              type: "setError",
              params: () => ({
                error: new Error("Authentication timed out"),
              }),
            },
          ],
          target: AuthStateType.ERROR,
        },
      },
    },
    [AuthStateType.LOGGED_IN]: {
      entry: [{ type: "publishCurrentUser" }, { type: "publishToken" }],
      on: {
        "auth.logout": { target: "logging-out" },
      },
      // The session is only established once its organization context resolved:
      // `authenticated` (and with it boot readiness) waits for the resolve, so
      // a failed resolve fails boot exactly like a failed login.
      initial: "resolving-organization",
      states: {
        "resolving-organization": {
          tags: ["authenticating"],
          invoke: {
            src: "organizationResolver",
            input: ({ context }) => ({
              instance: context.instance,
              organizationId: context.organizationId,
            }),
            onDone: {
              target: "ready",
              actions: [
                {
                  type: "publishOrganization",
                  params: ({ event }) => ({ organization: event.output }),
                },
              ],
            },
            onError: {
              target: `#auth.${AuthStateType.ERROR}`,
              actions: [
                {
                  type: "setError",
                  params: ({ event }) => ({
                    error: new Error(
                      "Failed to resolve the session's organization",
                      { cause: event.error },
                    ),
                  }),
                },
              ],
            },
          },
        },
        ready: {
          tags: ["authenticated"],
          // Leaving `ready` unsubscribes the responder, so signed-out callers
          // get NO_RESPONDER instead of a token.
          invoke: { src: "tokenRefreshResponder" },
          on: {
            // A refresh runs the publish path (deduped today, since the token is
            // unchanged) and replies to the caller with the live token.
            "token.refresh": {
              actions: [
                { type: "publishToken" },
                {
                  type: "replyToken",
                  params: ({ event }) => ({ reply: event.reply }),
                },
              ],
            },
          },
        },
      },
    },
    ["logging-out"]: {
      invoke: {
        src: "logoutActor",
        input: ({ context }) => ({ instance: context.instance }),
        onDone: {
          target: AuthStateType.LOGGED_OUT,
        },
        onError: {
          actions: [
            {
              type: "setError",
              params: ({ event }) => ({ error: event.error }),
            },
          ],
          target: AuthStateType.ERROR,
        },
      },
    },
    [AuthStateType.LOGGED_OUT]: {
      entry: [
        { type: "publishCurrentUser" },
        { type: "publishToken" },
        { type: "publishOrganization", params: { organization: null } },
      ],
    },
    [AuthStateType.ERROR]: {
      tags: ["error"],
      entry: [
        { type: "publishCurrentUser" },
        { type: "publishToken" },
        { type: "publishOrganization", params: { organization: null } },
      ],
    },
  },
});
