// @sanity/workbench/src/system/load-federated-module.ts
import { logger } from "../core/log";
import type { FederationInstance } from "./module-federation";

/**
 * A module exposed by an application's federation remote.
 *
 * @public
 */
export interface FederatedModuleRef {
  /**
   * Fully-qualified module id — `${appId}/${expose}`, e.g.
   * `studio-1/views/feed/panel` (a view component) or `studio-1/App` (the
   * full-page view). The remote's name is its first segment.
   */
  moduleId: string;
  /** The remote's `mf-manifest.json` URL (served from the app's origin). */
  entry: string;
}

/**
 * Register an app's federation remote (idempotent) and load one of its exposes
 * by its fully-qualified `moduleId`. Failures propagate, so the caller drives
 * its own error handling (e.g. a machine's error state with the original
 * cause). For the render-empty-on-failure policy, use {@link loadFederatedModule}.
 */
export async function loadRemoteModule<T>(
  instance: FederationInstance,
  { moduleId, entry }: FederatedModuleRef,
): Promise<T | null> {
  // The remote name is the first segment; the expose is the remainder.
  const name = moduleId.slice(0, moduleId.indexOf("/"));
  instance.registerRemotes([{ name, entry }]);
  return (await instance.loadRemote<T | null>(moduleId)) ?? null;
}

/**
 * {@link loadRemoteModule} with the render-empty-on-failure policy: any failure
 * resolves to `null` (and logs), so every caller handles a missing interface
 * the same way — a panel renders empty, a worker doesn't spawn.
 *
 * @public
 */
export async function loadFederatedModule<T>(
  instance: FederationInstance,
  ref: FederatedModuleRef,
): Promise<T | null> {
  try {
    return await loadRemoteModule<T>(instance, ref);
  } catch (error) {
    logger.error(`Failed to load federated module "${ref.moduleId}"`, error);
    return null;
  }
}
