// @sanity/workbench/src/system/module-federation.ts
import { createInstance as createMFInstance } from "@module-federation/runtime";
import type {
  ModuleFederationRuntimePlugin,
  UserOptions,
} from "@module-federation/runtime/types";

/** Make some properties of a type optional. */
type MakeOptional<T, K extends keyof T> = Omit<T, K> & Partial<Pick<T, K>>;

/**
 * @public
 */
export type FederationRemote = {
  name: string;
  entry: string;
};

/**
 * @public
 */
export type CreateInstanceOptions = MakeOptional<
  Pick<UserOptions, "name" | "remotes" | "shared" | "plugins">,
  "remotes" | "plugins"
>;

/**
 * @public
 */
export interface FederationInstance {
  registerRemotes(remotes: FederationRemote[]): void;
  loadRemote<T>(id: string): Promise<T | null>;
}

/**
 * @public
 */
export function createInstance(
  options: CreateInstanceOptions,
): FederationInstance {
  const instance = createMFInstance({
    remotes: [],
    ...options,
  });

  return {
    registerRemotes: (remotes) =>
      instance.registerRemotes(remotes, { force: false }),
    loadRemote: async <T>(id: string) => {
      let remoteModule: T | null;

      try {
        remoteModule = await instance.loadRemote<T>(id);
      } catch (error) {
        throw new Error(`Failed to load remote module "${id}"`, {
          cause: error,
        });
      }

      return remoteModule;
    },
  };
}

/**
 * Forward Module Federation lifecycle events to a provided logging function
 * from workbench.
 *
 * @public
 */
export function log(
  onDebug: (message: string, context?: Record<string, unknown>) => void,
): ModuleFederationRuntimePlugin {
  const logEvent = (eventName: string, args: unknown) => {
    // `origin` logs the whole instance, which is noisy
    const { origin: _origin, ...data } =
      typeof args === "object" && args !== null
        ? (args as Record<string, unknown>)
        : { args };

    onDebug(`[Lifecycle] ${eventName}`, data);
  };

  return {
    name: "sanity-logger",
    beforeInit(args) {
      logEvent("beforeInit", args);
      return args;
    },
    beforeRegisterRemote(args) {
      logEvent("beforeRegisterRemote", args);
      return args;
    },
    beforePreloadRemote(args) {
      logEvent("beforePreloadRemote", args);
    },
    beforeRequest(args) {
      logEvent("beforeRequest", args);
      return args;
    },
    afterResolve(args) {
      logEvent("afterResolve", args);
      return args;
    },
    onLoad(args) {
      logEvent("onLoad", args);
      return args;
    },
    loadShare(args) {
      logEvent("loadShare", args);
    },
    beforeLoadShare(args) {
      logEvent("beforeLoadShare", args);
      return args;
    },
  };
}
