// @sanity/workbench/src/system/remotes.machine.ts
import { type ActorRefFrom, assign, setup } from "xstate";

import { type FederationInstance } from "./module-federation";
import { remoteLogic } from "./remote.machine";

/** The shared federation instance, supplied by the root machine. */
type RemotesInput = {
  instance: FederationInstance;
};

type RemotesContext = {
  instance: FederationInstance;
  children: Map<string, ActorRefFrom<typeof remoteLogic>>;
};

type RemotesEvent =
  | { type: "remote.load.request"; moduleId: string; entry: string }
  | { type: "remote.settled"; moduleId: string; status: "loaded" | "error" };

/**
 * Supervisor machine for federated modules.
 *
 * Uses the shared {@link FederationInstance} (the `workbench-applications`
 * instance, supplied by the root machine and shared with `services`) and spawns
 * one {@link remoteLogic} child per remote id (`${appId}/${moduleId}`) — so an
 * application's `App` entry and each of its dock-panel view components load (and
 * are tracked) independently. The
 * supervisor is invoked at the root of the OS machine for inspector visibility
 * (`[sanity-workbench:os:remotes]`) — boot does not gate on it.
 *
 * @internal
 */
export const remotesLogic = setup({
  types: {
    input: {} as RemotesInput,
    context: {} as RemotesContext,
    events: {} as RemotesEvent,
  },
  actors: {
    remote: remoteLogic,
  },
  guards: {
    remoteNotKnown: ({ context }, params: { moduleId: string }) =>
      !context.children.has(params.moduleId),
  },
  actions: {
    spawnRemote: assign({
      children: (
        { context, spawn },
        params: { moduleId: string; entry: string },
      ) => {
        const ref = spawn("remote", {
          id: params.moduleId,
          systemId: `remotes.${params.moduleId}`,
          input: {
            moduleId: params.moduleId,
            entry: params.entry,
            instance: context.instance,
          },
        });
        return new Map(context.children).set(params.moduleId, ref);
      },
    }),
  },
}).createMachine({
  id: "remotes",
  context: ({ input }) => ({
    instance: input.instance,
    children: new Map(),
  }),
  on: {
    "remote.load.request": {
      guard: {
        type: "remoteNotKnown",
        params: ({ event }) => ({ moduleId: event.moduleId }),
      },
      actions: [
        {
          type: "spawnRemote",
          params: ({ event }) => ({
            moduleId: event.moduleId,
            entry: event.entry,
          }),
        },
      ],
    },
    // Reserved for future supervision/retry. Forwarded by per-remote
    // children on entry to `loaded` or `error`; currently a no-op so the
    // event is part of the typed surface rather than a silent unknown.
    "remote.settled": {},
  },
});
