// @sanity/workbench/src/system/root.machine.ts
import { createSanityInstance, type SanityInstance } from "@sanity/sdk";
import { raise, sendTo, setup, type ActorOptions } from "xstate";

import { logger } from "../core/log";
import { authLogic } from "./auth.machine";
import { inspect } from "./inspect";
import {
  createInstance,
  type FederationInstance,
  log,
} from "./module-federation";
import { remotesLogic } from "./remotes.machine";
import { servicesLogic } from "./services.machine";
import {
  type SystemPreferencesAdapter,
  systemPreferencesLogic,
} from "./system-preferences.machine";
import {
  telemetryLogic,
  type WorkbenchUserProperties,
} from "./telemetry.machine";

type OSInput = WorkbenchUserProperties & {
  systemPreferencesAdapter?: SystemPreferencesAdapter;
};

type OSContext = {
  instance: SanityInstance;
  /**
   * The single module-federation instance every interface loads through —
   * shared by the `remotes` (views) and `services` (workers) supervisors so an
   * app's remote is registered once for all its interfaces.
   */
  federationInstance: FederationInstance;
  userProperties: WorkbenchUserProperties;
  systemPreferencesAdapter: SystemPreferencesAdapter | undefined;
};

/**
 * The base inputs for the OS machine.
 * @internal
 */
export interface OSBaseInput extends Pick<OSContext, "instance"> {}

/**
 * The sanity OS machine, responsible for managing the global state of the OS.
 * @public
 * @example
 * ```ts
 * import { os, createOSOptions } from "@sanity/workbench/system";
 * import { useActor } from "@xstate/react";
 *
 * const [state, send] = useActor(os, createOSOptions({
 *   version: "1.0.0",
 *   organizationId: "...",
 *   environment: "production",
 *   userAgent: navigator.userAgent,
 * }));
 * ```
 */
export const os = setup({
  types: {
    input: {} as OSInput,
    context: {} as OSContext,
    events: {} as
      | { type: "boot.auth.ready" }
      | { type: "boot.auth.failed" }
      | { type: "boot.telemetry.ready" },
    // https://github.com/statelyai/xstate/issues/5515
    children: {} as {
      auth: "auth";
      telemetry: "telemetry";
      "system-preferences": "systemPreferences";
      remotes: "remotes";
      services: "services";
    },
  },
  actors: {
    auth: authLogic,
    telemetry: telemetryLogic,
    systemPreferences: systemPreferencesLogic,
    remotes: remotesLogic,
    services: servicesLogic,
  },
  guards: {
    hasTag: (_, params: { hasTag: boolean }) => params.hasTag,
  },
  actions: {
    raiseAuthReady: raise({ type: "boot.auth.ready" }),
    raiseAuthFailed: raise({ type: "boot.auth.failed" }),
    raiseTelemetryReady: raise({
      type: "boot.telemetry.ready",
    }),
    startTelemetry: sendTo("telemetry", {
      type: "telemetry.start",
    }),
  },
}).createMachine({
  id: "os",
  context: ({ input }) => ({
    instance: createSanityInstance(),
    federationInstance: createInstance({
      name: "workbench-applications",
      plugins: [log(logger.debug)],
    }),
    userProperties: {
      version: input.version,
      organizationId: input.organizationId,
      environment: input.environment,
      userAgent: input.userAgent,
    },
    systemPreferencesAdapter: input.systemPreferencesAdapter,
  }),
  initial: "booting",
  invoke: [
    {
      id: "auth",
      systemId: "auth",
      src: "auth",
      // The auth machine establishes the session context, so it also gets the
      // organization the session is scoped to.
      input: ({ context }) => ({
        instance: context.instance,
        organizationId: context.userProperties.organizationId,
      }),
      onSnapshot: [
        {
          guard: {
            type: "hasTag",
            params: ({ event }) => ({
              hasTag: event.snapshot.hasTag("authenticated"),
            }),
          },
          actions: [{ type: "raiseAuthReady" }],
        },
        {
          guard: {
            type: "hasTag",
            params: ({ event }) => ({
              hasTag: event.snapshot.hasTag("error"),
            }),
          },
          actions: [{ type: "raiseAuthFailed" }],
        },
      ],
    },
    {
      id: "telemetry",
      systemId: "telemetry",
      src: "telemetry",
      input: ({ context }) => ({
        instance: context.instance,
        ...context.userProperties,
      }),
      onSnapshot: [
        {
          guard: {
            type: "hasTag",
            params: ({ event }) => ({
              hasTag: event.snapshot.hasTag("telemetry-resolved"),
            }),
          },
          actions: [{ type: "raiseTelemetryReady" }],
        },
      ],
    },
    {
      id: "system-preferences",
      systemId: "system-preferences",
      src: "systemPreferences",
      input: ({ context }) => ({ adapter: context.systemPreferencesAdapter }),
    },
    {
      id: "remotes",
      systemId: "remotes",
      src: "remotes",
      input: ({ context }) => ({ instance: context.federationInstance }),
    },
    {
      id: "services",
      systemId: "services",
      src: "services",
      input: ({ context }) => ({ instance: context.federationInstance }),
    },
  ],
  states: {
    booting: {
      initial: "auth",
      states: {
        auth: {
          on: {
            "boot.auth.ready": {
              target: "telemetry",
              actions: [{ type: "startTelemetry" }],
            },
            "boot.auth.failed": { target: "error" },
          },
        },
        telemetry: {
          on: {
            "boot.telemetry.ready": { target: "done" },
          },
        },
        error: {},
        done: { type: "final" },
      },
      onDone: { target: "running" },
    },
    running: {
      type: "parallel",
    },
  },
});

/**
 * Creates a set of default options for the OS machine. Forwards the workbench
 * user properties to the machine's input and wires the structured-logging
 * `inspect` callback. Browser-specific concerns (color-scheme seeding,
 * persistence) live in the {@link SystemPreferencesAdapter} the host passes
 * via `systemPreferencesAdapter`.
 * @public
 */
export function createOSOptions(input: OSInput) {
  return {
    id: "os",
    input,
    inspect,
  } satisfies ActorOptions<typeof os>;
}
