// @sanity/workbench-cli/dist/actions/dev/exposesSetId.js
import { deriveConfigEntries } from './deriveInterfaces.js';
/**
 * Order-independent id of an app's exposed-module set. Keys each interface by
 * its type, name, and source file, and each config by its module
 * identity plus one key per entry (a named source file), so the id changes on
 * any rebuild-worthy edit — add, remove, rename, repoint, or gaining/losing the
 * config module — while reordering and HMR content edits keep it stable.
 */ export function exposesSetId({ configs, interfaces }) {
    const keys = [
        ...(interfaces ?? []).map((iface)=>[
                iface.type,
                iface.name,
                iface.src
            ].join('::')),
        ...(configs ?? []).flatMap((config)=>[
                [
                    'config',
                    config.appType
                ].join('::'),
                ...deriveConfigEntries(config).map((entry)=>[
                        'config',
                        config.appType,
                        entry.name,
                        entry.src
                    ].join('::'))
            ])
    ];
    if (keys.length === 0) return '';
    return keys.toSorted().join('|');
}
/**
 * `changed`/`commit` are split so the caller commits only after the rebuild that
 * depends on the new set succeeds — a thrown rebuild retries on the next save.
 */ export function trackExposesSet(initial) {
    let lastId = exposesSetId(initial);
    return {
        changed: (next)=>exposesSetId(next) !== lastId,
        commit: (next)=>{
            lastId = exposesSetId(next);
        }
    };
}
const serverKey = (server)=>`${server.id ?? ''}@${server.host ?? ''}:${server.port}`;
const serverExposesId = (server)=>exposesSetId({
        configs: server.configs,
        interfaces: server.interfaces
    });
/**
 * Multi-app counterpart to {@link trackExposesSet}: true when a *known* app's
 * set changed — its remote was rebuilt, so the workbench must full-reload to
 * drop the stale remote-entry. A new/removed app isn't a rebuild.
 */ export function createExposesTracker() {
    let known = new Map();
    return {
        hasChanged (servers) {
            const rebuilt = servers.some((server)=>{
                const key = serverKey(server);
                return known.has(key) && known.get(key) !== serverExposesId(server);
            });
            known = new Map(servers.map((server)=>[
                    serverKey(server),
                    serverExposesId(server)
                ]));
            return rebuilt;
        }
    };
}
