// @sanity/workbench-cli/dist/actions/dev/startDevServerRegistration.js
import { getCliConfigUncached } from '@sanity/cli-core';
import { resolveAppId } from '../../appId.js';
import { deriveConfigs, deriveInterfaces } from './deriveInterfaces.js';
import { trackExposesSet } from './exposesSetId.js';
import { registerDevServer } from './registry.js';
import { startDevManifestWatcher } from './startDevManifestWatcher.js';
/** The address the server actually bound — the live socket, which can differ from the configured port under non-strict ports. */ function serverAddress(server) {
    const resolvedHost = server.config.server.host;
    const addr = server.httpServer?.address();
    return {
        host: typeof resolvedHost === 'string' ? resolvedHost : 'localhost',
        port: typeof addr === 'object' && addr ? addr.port : server.config.server.port
    };
}
/**
 * Register the dev server in the registry and watch its config for manifest +
 * interface changes. The workbench reads the entry to locate and render the
 * server; the watcher keeps it current as `sanity.cli.ts` is edited.
 */ export async function startDevServerRegistration(options) {
    const { cliConfig, extractManifest, isApp, onInterfaceSetChange, output, server, workDir } = options;
    const { host: appHost, port: appPort } = serverAddress(server);
    // Forwarded alongside (not inside) the manifest so the workbench renders local
    // panels/workers and reads the configs without a deploy.
    const interfaces = deriveInterfaces(cliConfig.app, {
        isApp
    });
    const configs = await deriveConfigs(cliConfig.app);
    const registration = registerDevServer({
        configs,
        host: appHost,
        // Keyed by where it's served (not the deployment id), so a running app can't
        // collide with its deployed twin — on the configured port, not the bound one,
        // to match `__SANITY_APP_ID__`, compiled before any non-strict shift.
        id: resolveAppId({
            host: appHost,
            port: server.config.server.port ?? appPort
        }),
        interfaces,
        port: appPort,
        projectId: cliConfig?.api?.projectId,
        type: isApp ? 'coreApp' : 'studio',
        workDir
    });
    const exposesSet = trackExposesSet({
        configs,
        interfaces
    });
    const watcher = await startDevManifestWatcher({
        // Re-derive every pass (don't omit): the registry patch is a shallow merge,
        // so omitting would wipe the registered set.
        extract: async (params)=>{
            const app = (await getCliConfigUncached(params.workDir)).app;
            return {
                configs: await deriveConfigs(app),
                interfaces: deriveInterfaces(app, {
                    isApp
                }),
                manifest: await extractManifest(params)
            };
        },
        // A studio's root resolves to `sanity.config.*` but its interfaces live in
        // `sanity.cli.*` — watch that too. Apps already root at `sanity.cli.*`.
        extraWatchFilenames: isApp ? undefined : [
            'sanity.cli.js',
            'sanity.cli.ts'
        ],
        output,
        update: async (patch)=>{
            if (!exposesSet.changed({
                configs: patch.configs,
                interfaces: patch.interfaces
            })) {
                registration.update(patch);
                return;
            }
            // Rebuild the remote *before* patching the registry — the patch reloads the
            // page, which must re-fetch a remote that already exposes the new interface.
            const rebuiltServer = await onInterfaceSetChange?.();
            // Commit only after a successful rebuild, so a thrown one retries next pass.
            exposesSet.commit({
                configs: patch.configs,
                interfaces: patch.interfaces
            });
            // The recreated server can bind a different port (non-strict ports).
            registration.update(rebuiltServer ? {
                ...patch,
                ...serverAddress(rebuiltServer)
            } : patch);
        },
        workDir
    });
    return {
        close: async ()=>{
            registration.release();
            await watcher.close();
        }
    };
}
