// @portabletext/plugin-table/dist/ui/index.js
import { jsxs, jsx, Fragment } from "react/jsx-runtime";
import { useEditor, useEditorSelector, defineContainer } from "@portabletext/editor";
import { c } from "react/compiler-runtime";
import { getChildren, getText, getFirstChild, getEnclosingBlock, getParent, getBlock } from "@portabletext/editor/traversal";
import { isEqualPaths, getBlockStartPoint } from "@portabletext/editor/utils";
import { useState, useLayoutEffect, useRef, useEffect, createContext, useContext } from "react";
import { resolveTableConfig, tableRows, rowCells, createTableGuards, getTableSelection } from "../_chunks-es/get-table-selection.js";
import { autoUpdate, computePosition, offset, hide } from "@floating-ui/dom";
import { createPortal } from "react-dom";
const BLUE = "var(--pt-plugin-table-accent)", BORDER = "var(--pt-plugin-table-border)", FG = "var(--pt-plugin-table-fg)", HEADER_BG = "var(--pt-plugin-table-header-bg)", HEADER_WEIGHT = "var(--pt-plugin-table-header-weight, 600)", SELECTED_BG = "var(--pt-plugin-table-selected-bg)", CELL_PADDING = "var(--pt-plugin-table-cell-padding)", TABLE_RADIUS = "var(--pt-plugin-table-radius)";
function cellRangeEdges(row, col, range) {
  return !range || row < range.r0 || row > range.r1 || col < range.c0 || col > range.c1 ? {
    top: !1,
    bottom: !1,
    left: !1,
    right: !1,
    inside: !1
  } : {
    top: row === range.r0,
    bottom: row === range.r1,
    left: col === range.c0,
    right: col === range.c1,
    inside: !0
  };
}
function cornerRadius(rowIdx, colIdx, rowCount, colCount) {
  const style = {};
  return rowIdx === 0 && colIdx === 0 && (style.borderTopLeftRadius = TABLE_RADIUS), rowIdx === 0 && colIdx === colCount - 1 && (style.borderTopRightRadius = TABLE_RADIUS), rowIdx === rowCount - 1 && colIdx === 0 && (style.borderBottomLeftRadius = TABLE_RADIUS), rowIdx === rowCount - 1 && colIdx === colCount - 1 && (style.borderBottomRightRadius = TABLE_RADIUS), style;
}
function suppressRightGrid(rowIdx, colIdx, range) {
  return !!range && colIdx + 1 === range.c0 && rowIdx >= range.r0 && rowIdx <= range.r1;
}
function suppressBottomGrid(rowIdx, colIdx, range) {
  return !!range && rowIdx + 1 === range.r0 && colIdx >= range.c0 && colIdx <= range.c1;
}
function cellStyle({
  rowIdx,
  colIdx,
  rowCount,
  colCount,
  isHeader,
  range
}) {
  const outline = cellRangeEdges(rowIdx, colIdx, range), showOverlay = outline.inside, gridSide = `1px solid ${BORDER}`, selSide = `${1.5}px solid ${BLUE}`, shadows = [], borderTop = outline.top ? selSide : rowIdx === 0 ? gridSide : "none", borderLeft = outline.left ? selSide : colIdx === 0 ? gridSide : "none";
  let borderBottom = outline.bottom ? selSide : gridSide, borderRight = outline.right ? selSide : gridSide;
  !outline.right && suppressRightGrid(rowIdx, colIdx, range) && (borderRight = "none"), !outline.bottom && suppressBottomGrid(rowIdx, colIdx, range) && (borderBottom = "none"), showOverlay && (outline.bottom || (borderBottom = "none"), outline.right || (borderRight = "none"), !outline.bottom && rowIdx < rowCount - 1 && shadows.push(`inset 0 -1px 0 0 ${BORDER}`), !outline.right && colIdx < colCount - 1 && shadows.push(`inset -1px 0 0 0 ${BORDER}`));
  let background = isHeader ? HEADER_BG : "var(--pt-plugin-table-bg)";
  return showOverlay && (background = isHeader ? `linear-gradient(${SELECTED_BG}, ${SELECTED_BG}), ${HEADER_BG}` : SELECTED_BG), {
    ...cornerRadius(rowIdx, colIdx, rowCount, colCount),
    borderTop,
    borderLeft,
    borderBottom,
    borderRight,
    boxShadow: shadows.length > 0 ? shadows.join(", ") : void 0,
    padding: CELL_PADDING,
    verticalAlign: "top",
    background,
    fontWeight: isHeader ? HEADER_WEIGHT : 400,
    color: FG,
    wordBreak: "break-word",
    ...showOverlay ? {
      position: "relative",
      zIndex: 1
    } : {}
  };
}
function EllipsisIcon(props) {
  const $ = c(5);
  let t0, t1, t2;
  $[0] === /* @__PURE__ */ Symbol.for("react.memo_cache_sentinel") ? (t0 = /* @__PURE__ */ jsx("circle", { cx: "12", cy: "12", r: "1" }), t1 = /* @__PURE__ */ jsx("circle", { cx: "19", cy: "12", r: "1" }), t2 = /* @__PURE__ */ jsx("circle", { cx: "5", cy: "12", r: "1" }), $[0] = t0, $[1] = t1, $[2] = t2) : (t0 = $[0], t1 = $[1], t2 = $[2]);
  let t3;
  return $[3] !== props ? (t3 = /* @__PURE__ */ jsxs(Icon, { ...props, children: [
    t0,
    t1,
    t2
  ] }), $[3] = props, $[4] = t3) : t3 = $[4], t3;
}
function PanelTopIcon(props) {
  const $ = c(4);
  let t0, t1;
  $[0] === /* @__PURE__ */ Symbol.for("react.memo_cache_sentinel") ? (t0 = /* @__PURE__ */ jsx("rect", { width: "18", height: "18", x: "3", y: "3", rx: "2" }), t1 = /* @__PURE__ */ jsx("path", { d: "M3 9h18" }), $[0] = t0, $[1] = t1) : (t0 = $[0], t1 = $[1]);
  let t2;
  return $[2] !== props ? (t2 = /* @__PURE__ */ jsxs(Icon, { ...props, children: [
    t0,
    t1
  ] }), $[2] = props, $[3] = t2) : t2 = $[3], t2;
}
function TableIcon(props) {
  const $ = c(6);
  let t0, t1, t2, t3;
  $[0] === /* @__PURE__ */ Symbol.for("react.memo_cache_sentinel") ? (t0 = /* @__PURE__ */ jsx("path", { d: "M12 3v18" }), t1 = /* @__PURE__ */ jsx("rect", { width: "18", height: "18", x: "3", y: "3", rx: "2" }), t2 = /* @__PURE__ */ jsx("path", { d: "M3 9h18" }), t3 = /* @__PURE__ */ jsx("path", { d: "M3 15h18" }), $[0] = t0, $[1] = t1, $[2] = t2, $[3] = t3) : (t0 = $[0], t1 = $[1], t2 = $[2], t3 = $[3]);
  let t4;
  return $[4] !== props ? (t4 = /* @__PURE__ */ jsxs(Icon, { ...props, children: [
    t0,
    t1,
    t2,
    t3
  ] }), $[4] = props, $[5] = t4) : t4 = $[5], t4;
}
function Trash2Icon(props) {
  const $ = c(7);
  let t0, t1, t2, t3, t4;
  $[0] === /* @__PURE__ */ Symbol.for("react.memo_cache_sentinel") ? (t0 = /* @__PURE__ */ jsx("path", { d: "M10 11v6" }), t1 = /* @__PURE__ */ jsx("path", { d: "M14 11v6" }), t2 = /* @__PURE__ */ jsx("path", { d: "M19 6v14a2 2 0 0 1-2 2H7a2 2 0 0 1-2-2V6" }), t3 = /* @__PURE__ */ jsx("path", { d: "M3 6h18" }), t4 = /* @__PURE__ */ jsx("path", { d: "M8 6V4a2 2 0 0 1 2-2h4a2 2 0 0 1 2 2v2" }), $[0] = t0, $[1] = t1, $[2] = t2, $[3] = t3, $[4] = t4) : (t0 = $[0], t1 = $[1], t2 = $[2], t3 = $[3], t4 = $[4]);
  let t5;
  return $[5] !== props ? (t5 = /* @__PURE__ */ jsxs(Icon, { ...props, children: [
    t0,
    t1,
    t2,
    t3,
    t4
  ] }), $[5] = props, $[6] = t5) : t5 = $[6], t5;
}
function Icon(t0) {
  const $ = c(3), {
    size,
    children
  } = t0;
  let t1;
  return $[0] !== children || $[1] !== size ? (t1 = /* @__PURE__ */ jsx("svg", { xmlns: "http://www.w3.org/2000/svg", width: size, height: size, viewBox: "0 0 24 24", fill: "none", stroke: "currentColor", strokeWidth: 2, strokeLinecap: "round", strokeLinejoin: "round", "aria-hidden": "true", children }), $[0] = children, $[1] = size, $[2] = t1) : t1 = $[2], t1;
}
function snapPx(px) {
  if (typeof window > "u")
    return px;
  const dpr = window.devicePixelRatio || 1;
  return Math.round(px * dpr) / dpr;
}
function snapPxCenter(px) {
  if (typeof window > "u")
    return px;
  const dpr = window.devicePixelRatio || 1;
  return (Math.round(px * dpr - 0.5) + 0.5) / dpr;
}
function useTableMetrics(tableRef, revision) {
  const $ = c(5), [metrics, setMetrics] = useState(null);
  let t0;
  $[0] !== tableRef ? (t0 = () => {
    const table = tableRef.current;
    if (!table)
      return;
    const measure = () => {
      const rect = table.getBoundingClientRect(), trs = [...table.querySelectorAll("tbody tr")], firstRowCells = trs[0]?.querySelectorAll("td") ?? [], rows = trs.map((tr) => {
        const r = tr.getBoundingClientRect(), top = snapPx(r.top - rect.top), height = snapPx(r.height);
        return {
          top,
          height,
          centerY: snapPx(top + height / 2)
        };
      }), cols = [...firstRowCells].map((td) => {
        const c2 = td.getBoundingClientRect(), left = snapPx(c2.left - rect.left), width = snapPx(c2.width);
        return {
          left,
          width,
          centerX: snapPx(left + width / 2)
        };
      });
      setMetrics({
        width: snapPx(rect.width),
        height: snapPx(rect.height),
        rows,
        cols
      });
    };
    measure();
    const observer = new ResizeObserver(measure);
    return observer.observe(table), window.addEventListener("resize", measure), () => {
      observer.disconnect(), window.removeEventListener("resize", measure);
    };
  }, $[0] = tableRef, $[1] = t0) : t0 = $[1];
  let t1;
  return $[2] !== revision || $[3] !== tableRef ? (t1 = [tableRef, revision], $[2] = revision, $[3] = tableRef, $[4] = t1) : t1 = $[4], useLayoutEffect(t0, t1), metrics;
}
const GUTTER_TOP = 20, GUTTER_LEFT = 12, WRAPPER_PAD_TOP = 20, HANDLE_REST_ROW = {
  w: 3,
  h: 16
}, HANDLE_REST_COL = {
  w: 16,
  h: 3
}, HANDLE_BTN_ROW = {
  w: 12,
  h: 16
}, HANDLE_BTN_COL = {
  w: 16,
  h: 12
}, HANDLE_BTN_PAD = 3, HANDLE_REST_BG = "var(--pt-plugin-table-handle-rest)", HANDLE_HIT = 24, HANDLE_EXPANDED_SHADOW = `inset 0 0 0 0.5px var(--pt-plugin-table-handle-ring), 0 0 0 1px ${BORDER}`, HANDLE_EXPANDED_SHADOW_SELECTED = "inset 0 0 0 0.5px var(--pt-plugin-table-handle-ring), 0 0 0 1px var(--pt-plugin-table-handle-bg)", HANDLE_DOT = 2, HANDLE_DOT_GAP = 2, HANDLE_GREY = "var(--pt-plugin-table-boundary-dot)", BOUNDARY_DOT = 4, BOUNDARY_PLUS = 18, GRID_LINE_HALF = 0.5, INSERT_GUIDE = 1.5, EXTEND_SIZE = 17, EXTEND_GAP = 3, EXTEND_LANE = EXTEND_SIZE + EXTEND_GAP, EXTEND_BAR_BG = "var(--pt-plugin-table-lane-bg)", EXTEND_BAR_BG_HOVER = "var(--pt-plugin-table-lane-bg-hover)", EXTEND_ICON = "var(--pt-plugin-table-lane-icon)", EXTEND_ICON_HOVER = "var(--pt-plugin-table-lane-icon-hover)", TRASH_SIZE = 26, TRASH_GAP = 8, TRASH_Z_INDEX = 10050;
function colBorderX(metrics, index) {
  const {
    cols,
    width
  } = metrics;
  return index <= 0 ? snapPxCenter((cols[0]?.left ?? 0) + GRID_LINE_HALF) : index >= cols.length ? snapPxCenter(width - GRID_LINE_HALF) : snapPxCenter(cols[index].left - GRID_LINE_HALF);
}
function rowBorderY(metrics, index) {
  const {
    rows,
    height
  } = metrics;
  return index <= 0 ? snapPxCenter(GRID_LINE_HALF) : index >= rows.length ? snapPxCenter(height - GRID_LINE_HALF) : snapPxCenter(rows[index].top - GRID_LINE_HALF);
}
const defaultLabels = {
  "add-column": "Add column at end",
  "add-row": "Add row at end",
  "column-handle": "Column handle",
  "delete-column": "Delete column",
  "delete-row": "Delete row",
  "insert-here": "Insert here",
  "menu-delete-table": "Delete table",
  "menu-header-row": "Header row",
  "menu-select-table": "Select table",
  "row-handle": "Row handle",
  "table-options": "Table options"
};
function TableChrome(t0) {
  const $ = c(94), {
    metrics,
    active,
    hoverCell,
    selectedRow,
    selectedCol,
    onHandlePointerDown,
    onHandleKeyboardSelect,
    drag,
    onInsertRow,
    onInsertCol,
    labels
  } = t0, [boundary, setBoundary] = useState(null);
  if (!metrics)
    return null;
  const dragging = !!drag?.active, lastRow = metrics.rows.length - 1, lastCol = metrics.cols.length - 1;
  let t1;
  $[0] === /* @__PURE__ */ Symbol.for("react.memo_cache_sentinel") ? (t1 = {
    position: "absolute",
    inset: 0,
    pointerEvents: "none"
  }, $[0] = t1) : t1 = $[0];
  let t2;
  $[1] !== metrics.width ? (t2 = /* @__PURE__ */ jsx("div", { style: {
    position: "absolute",
    top: 0,
    left: GUTTER_LEFT,
    width: metrics.width,
    height: GUTTER_TOP,
    pointerEvents: "auto"
  }, onPointerDown: _temp$1 }), $[1] = metrics.width, $[2] = t2) : t2 = $[2];
  const t3 = labels["add-row"], t4 = active && !dragging && hoverCell?.row === lastRow, t5 = GUTTER_TOP + metrics.height + EXTEND_GAP;
  let t6;
  $[3] !== metrics.width || $[4] !== t5 ? (t6 = {
    left: GUTTER_LEFT,
    top: t5,
    width: metrics.width,
    height: EXTEND_SIZE
  }, $[3] = metrics.width, $[4] = t5, $[5] = t6) : t6 = $[5];
  let t7;
  $[6] !== metrics.rows.length || $[7] !== onInsertRow ? (t7 = () => onInsertRow(metrics.rows.length), $[6] = metrics.rows.length, $[7] = onInsertRow, $[8] = t7) : t7 = $[8];
  let t8;
  $[9] !== t3 || $[10] !== t4 || $[11] !== t6 || $[12] !== t7 ? (t8 = /* @__PURE__ */ jsx(ExtendBar, { label: t3, visible: t4, style: t6, onClick: t7 }), $[9] = t3, $[10] = t4, $[11] = t6, $[12] = t7, $[13] = t8) : t8 = $[13];
  const t9 = labels["add-column"], t10 = active && !dragging && hoverCell?.col === lastCol, t11 = GUTTER_LEFT + metrics.width + EXTEND_GAP;
  let t12;
  $[14] !== metrics.height || $[15] !== t11 ? (t12 = {
    left: t11,
    top: GUTTER_TOP,
    width: EXTEND_SIZE,
    height: metrics.height
  }, $[14] = metrics.height, $[15] = t11, $[16] = t12) : t12 = $[16];
  let t13;
  $[17] !== metrics.cols.length || $[18] !== onInsertCol ? (t13 = () => onInsertCol(metrics.cols.length), $[17] = metrics.cols.length, $[18] = onInsertCol, $[19] = t13) : t13 = $[19];
  let t14;
  $[20] !== t10 || $[21] !== t12 || $[22] !== t13 || $[23] !== t9 ? (t14 = /* @__PURE__ */ jsx(ExtendBar, { label: t9, visible: t10, style: t12, onClick: t13 }), $[20] = t10, $[21] = t12, $[22] = t13, $[23] = t9, $[24] = t14) : t14 = $[24];
  let t15;
  if ($[25] !== active || $[26] !== drag?.kind || $[27] !== dragging || $[28] !== hoverCell?.col || $[29] !== labels || $[30] !== metrics.cols || $[31] !== onHandleKeyboardSelect || $[32] !== onHandlePointerDown || $[33] !== selectedCol) {
    let t162;
    $[35] !== active || $[36] !== drag?.kind || $[37] !== dragging || $[38] !== hoverCell?.col || $[39] !== labels || $[40] !== onHandleKeyboardSelect || $[41] !== onHandlePointerDown || $[42] !== selectedCol ? (t162 = (col, index) => /* @__PURE__ */ jsx(Handle, { kind: "column", index, x: GUTTER_LEFT + col.centerX, y: GUTTER_TOP, active, cellHot: hoverCell?.col === index, selected: selectedCol === index, dragging: dragging && drag?.kind === "column", onPointerDown: (event_0) => onHandlePointerDown("column", index, event_0), label: labels["column-handle"], onKeyboardSelect: () => onHandleKeyboardSelect("column", index) }, `col-${index}`), $[35] = active, $[36] = drag?.kind, $[37] = dragging, $[38] = hoverCell?.col, $[39] = labels, $[40] = onHandleKeyboardSelect, $[41] = onHandlePointerDown, $[42] = selectedCol, $[43] = t162) : t162 = $[43], t15 = metrics.cols.map(t162), $[25] = active, $[26] = drag?.kind, $[27] = dragging, $[28] = hoverCell?.col, $[29] = labels, $[30] = metrics.cols, $[31] = onHandleKeyboardSelect, $[32] = onHandlePointerDown, $[33] = selectedCol, $[34] = t15;
  } else
    t15 = $[34];
  let t16;
  if ($[44] !== active || $[45] !== drag?.kind || $[46] !== dragging || $[47] !== hoverCell?.row || $[48] !== labels || $[49] !== metrics.rows || $[50] !== onHandleKeyboardSelect || $[51] !== onHandlePointerDown || $[52] !== selectedRow) {
    let t172;
    $[54] !== active || $[55] !== drag?.kind || $[56] !== dragging || $[57] !== hoverCell?.row || $[58] !== labels || $[59] !== onHandleKeyboardSelect || $[60] !== onHandlePointerDown || $[61] !== selectedRow ? (t172 = (row, index_0) => /* @__PURE__ */ jsx(Handle, { kind: "row", index: index_0, x: GUTTER_LEFT, y: GUTTER_TOP + row.centerY, active, cellHot: hoverCell?.row === index_0, selected: selectedRow === index_0, dragging: dragging && drag?.kind === "row", onPointerDown: (event_1) => onHandlePointerDown("row", index_0, event_1), label: labels["row-handle"], onKeyboardSelect: () => onHandleKeyboardSelect("row", index_0) }, `row-${index_0}`), $[54] = active, $[55] = drag?.kind, $[56] = dragging, $[57] = hoverCell?.row, $[58] = labels, $[59] = onHandleKeyboardSelect, $[60] = onHandlePointerDown, $[61] = selectedRow, $[62] = t172) : t172 = $[62], t16 = metrics.rows.map(t172), $[44] = active, $[45] = drag?.kind, $[46] = dragging, $[47] = hoverCell?.row, $[48] = labels, $[49] = metrics.rows, $[50] = onHandleKeyboardSelect, $[51] = onHandlePointerDown, $[52] = selectedRow, $[53] = t16;
  } else
    t16 = $[53];
  let t17;
  $[63] !== drag || $[64] !== metrics ? (t17 = drag?.active ? /* @__PURE__ */ jsx(ReorderInsertLine, { metrics, kind: drag.kind, insertIndex: drag.insertIndex }) : null, $[63] = drag, $[64] = metrics, $[65] = t17) : t17 = $[65];
  let t18;
  $[66] !== active || $[67] !== boundary || $[68] !== dragging || $[69] !== labels || $[70] !== metrics || $[71] !== onInsertCol ? (t18 = Array.from({
    length: Math.max(metrics.cols.length - 1, 0)
  }, (_, k) => {
    const index_1 = k + 1;
    return /* @__PURE__ */ jsx(BoundaryControl, { x: GUTTER_LEFT + colBorderX(metrics, index_1), y: GUTTER_TOP, visible: active && !dragging, hot: boundary?.kind === "column" && boundary.index === index_1, onEnter: () => setBoundary({
      kind: "column",
      index: index_1
    }), onLeave: () => setBoundary(null), onInsert: () => onInsertCol(index_1), label: labels["insert-here"] }, `col-boundary-${index_1}`);
  }), $[66] = active, $[67] = boundary, $[68] = dragging, $[69] = labels, $[70] = metrics, $[71] = onInsertCol, $[72] = t18) : t18 = $[72];
  let t19;
  $[73] !== active || $[74] !== boundary || $[75] !== dragging || $[76] !== labels || $[77] !== metrics || $[78] !== onInsertRow ? (t19 = Array.from({
    length: Math.max(metrics.rows.length - 1, 0)
  }, (__0, k_0) => {
    const index_2 = k_0 + 1;
    return /* @__PURE__ */ jsx(BoundaryControl, { x: GUTTER_LEFT, y: GUTTER_TOP + rowBorderY(metrics, index_2), visible: active && !dragging, hot: boundary?.kind === "row" && boundary.index === index_2, onEnter: () => setBoundary({
      kind: "row",
      index: index_2
    }), onLeave: () => setBoundary(null), onInsert: () => onInsertRow(index_2), label: labels["insert-here"] }, `row-boundary-${index_2}`);
  }), $[73] = active, $[74] = boundary, $[75] = dragging, $[76] = labels, $[77] = metrics, $[78] = onInsertRow, $[79] = t19) : t19 = $[79];
  let t20;
  $[80] !== active || $[81] !== boundary || $[82] !== metrics ? (t20 = active && boundary ? /* @__PURE__ */ jsx(InsertGuideline, { metrics, hovered: boundary }) : null, $[80] = active, $[81] = boundary, $[82] = metrics, $[83] = t20) : t20 = $[83];
  let t21;
  return $[84] !== t14 || $[85] !== t15 || $[86] !== t16 || $[87] !== t17 || $[88] !== t18 || $[89] !== t19 || $[90] !== t2 || $[91] !== t20 || $[92] !== t8 ? (t21 = /* @__PURE__ */ jsxs("div", { contentEditable: !1, style: t1, children: [
    t2,
    t8,
    t14,
    t15,
    t16,
    t17,
    t18,
    t19,
    t20
  ] }), $[84] = t14, $[85] = t15, $[86] = t16, $[87] = t17, $[88] = t18, $[89] = t19, $[90] = t2, $[91] = t20, $[92] = t8, $[93] = t21) : t21 = $[93], t21;
}
function _temp$1(event) {
  event.preventDefault();
}
function ExtendBar(t0) {
  const $ = c(14), {
    label,
    visible,
    style,
    onClick
  } = t0, [hovered, setHovered] = useState(!1), t1 = visible ? 0 : -1;
  let t2, t3;
  $[0] === /* @__PURE__ */ Symbol.for("react.memo_cache_sentinel") ? (t2 = () => setHovered(!0), t3 = () => setHovered(!1), $[0] = t2, $[1] = t3) : (t2 = $[0], t3 = $[1]);
  const t4 = visible ? "auto" : "none", t5 = visible ? 1 : 0, t6 = hovered ? EXTEND_BAR_BG_HOVER : EXTEND_BAR_BG, t7 = hovered ? EXTEND_ICON_HOVER : EXTEND_ICON;
  let t8;
  $[2] !== style || $[3] !== t4 || $[4] !== t5 || $[5] !== t6 || $[6] !== t7 ? (t8 = {
    position: "absolute",
    ...style,
    display: "flex",
    alignItems: "center",
    justifyContent: "center",
    border: "none",
    borderRadius: 3,
    padding: 0,
    pointerEvents: t4,
    opacity: t5,
    transition: "opacity 100ms ease, background 100ms ease",
    background: t6,
    color: t7,
    cursor: "pointer",
    zIndex: 1
  }, $[2] = style, $[3] = t4, $[4] = t5, $[5] = t6, $[6] = t7, $[7] = t8) : t8 = $[7];
  let t9;
  $[8] === /* @__PURE__ */ Symbol.for("react.memo_cache_sentinel") ? (t9 = /* @__PURE__ */ jsx("svg", { "aria-hidden": "true", width: "14", height: "14", viewBox: "0 0 14 14", fill: "none", children: /* @__PURE__ */ jsx("path", { d: "M7 2.5v9M2.5 7h9", stroke: "currentColor", strokeWidth: "1.15", strokeLinecap: "round" }) }), $[8] = t9) : t9 = $[8];
  let t10;
  return $[9] !== label || $[10] !== onClick || $[11] !== t1 || $[12] !== t8 ? (t10 = /* @__PURE__ */ jsx("button", { type: "button", "aria-label": label, tabIndex: t1, onPointerDown: _temp2$1, onClick, onMouseEnter: t2, onMouseLeave: t3, style: t8, children: t9 }), $[9] = label, $[10] = onClick, $[11] = t1, $[12] = t8, $[13] = t10) : t10 = $[13], t10;
}
function _temp2$1(event) {
  event.preventDefault();
}
function BoundaryControl(t0) {
  const $ = c(15), {
    x,
    y,
    visible,
    hot,
    onEnter,
    onLeave,
    onInsert,
    label
  } = t0, t1 = visible ? 0 : -1, t2 = x - 10, t3 = y - 10, t4 = visible ? "auto" : "none";
  let t5;
  $[0] !== t2 || $[1] !== t3 || $[2] !== t4 ? (t5 = {
    position: "absolute",
    left: t2,
    top: t3,
    width: 20,
    height: 20,
    display: "flex",
    alignItems: "center",
    justifyContent: "center",
    pointerEvents: t4,
    cursor: "pointer",
    zIndex: 6,
    border: "none",
    background: "transparent",
    padding: 0
  }, $[0] = t2, $[1] = t3, $[2] = t4, $[3] = t5) : t5 = $[3];
  let t6;
  $[4] !== hot || $[5] !== visible ? (t6 = hot ? /* @__PURE__ */ jsx("div", { style: {
    width: BOUNDARY_PLUS,
    height: BOUNDARY_PLUS,
    borderRadius: "50%",
    background: BLUE,
    display: "flex",
    alignItems: "center",
    justifyContent: "center",
    boxShadow: "0 1px 4px color-mix(in srgb, var(--pt-plugin-table-accent) 40%, transparent)"
  }, children: /* @__PURE__ */ jsx("svg", { "aria-hidden": "true", width: "10", height: "10", viewBox: "0 0 10 10", fill: "none", children: /* @__PURE__ */ jsx("path", { d: "M5 1v8M1 5h8", stroke: "var(--pt-plugin-table-accent-fg)", strokeWidth: "1.5", strokeLinecap: "round" }) }) }) : /* @__PURE__ */ jsx("div", { style: {
    width: BOUNDARY_DOT,
    height: BOUNDARY_DOT,
    borderRadius: "50%",
    background: HANDLE_GREY,
    opacity: visible ? 1 : 0,
    transition: "opacity 100ms ease"
  } }), $[4] = hot, $[5] = visible, $[6] = t6) : t6 = $[6];
  let t7;
  return $[7] !== label || $[8] !== onEnter || $[9] !== onInsert || $[10] !== onLeave || $[11] !== t1 || $[12] !== t5 || $[13] !== t6 ? (t7 = /* @__PURE__ */ jsx("button", { type: "button", "aria-label": label, tabIndex: t1, onMouseEnter: onEnter, onMouseLeave: onLeave, onPointerDown: _temp3$1, onClick: onInsert, style: t5, children: t6 }), $[7] = label, $[8] = onEnter, $[9] = onInsert, $[10] = onLeave, $[11] = t1, $[12] = t5, $[13] = t6, $[14] = t7) : t7 = $[14], t7;
}
function _temp3$1(event) {
  event.preventDefault(), event.stopPropagation();
}
function InsertGuideline(t0) {
  const $ = c(6), {
    metrics,
    hovered
  } = t0;
  if (hovered.kind === "column") {
    const t12 = GUTTER_LEFT + colBorderX(metrics, hovered.index);
    let t22;
    return $[0] !== metrics.height || $[1] !== t12 ? (t22 = /* @__PURE__ */ jsx("div", { style: {
      position: "absolute",
      left: t12,
      top: GUTTER_TOP,
      width: INSERT_GUIDE,
      height: metrics.height,
      transform: "translateX(-50%)",
      background: BLUE,
      pointerEvents: "none",
      zIndex: 3
    } }), $[0] = metrics.height, $[1] = t12, $[2] = t22) : t22 = $[2], t22;
  }
  const t1 = GUTTER_TOP + rowBorderY(metrics, hovered.index);
  let t2;
  return $[3] !== metrics.width || $[4] !== t1 ? (t2 = /* @__PURE__ */ jsx("div", { style: {
    position: "absolute",
    left: GUTTER_LEFT,
    top: t1,
    width: metrics.width,
    height: INSERT_GUIDE,
    transform: "translateY(-50%)",
    background: BLUE,
    pointerEvents: "none",
    zIndex: 3
  } }), $[3] = metrics.width, $[4] = t1, $[5] = t2) : t2 = $[5], t2;
}
function Handle(t0) {
  const $ = c(36), {
    kind,
    index,
    x,
    y,
    active,
    cellHot,
    selected,
    dragging,
    onPointerDown,
    onKeyboardSelect,
    label
  } = t0, [hot, setHot] = useState(!1), isColumn = kind === "column", rest = isColumn ? HANDLE_REST_COL : HANDLE_REST_ROW, btn = isColumn ? HANDLE_BTN_COL : HANDLE_BTN_ROW, showRest = cellHot && !hot && !selected, showExpanded = hot || selected, lit = active && (cellHot || hot || selected), t1 = lit ? 0 : -1;
  let t2, t3;
  $[0] === /* @__PURE__ */ Symbol.for("react.memo_cache_sentinel") ? (t2 = () => setHot(!0), t3 = () => setHot(!1), $[0] = t2, $[1] = t3) : (t2 = $[0], t3 = $[1]);
  let t4;
  $[2] !== onKeyboardSelect ? (t4 = keyboardActivation(onKeyboardSelect), $[2] = onKeyboardSelect, $[3] = t4) : t4 = $[3];
  const t5 = lit ? "auto" : "none", t6 = lit ? 1 : 0, t7 = dragging ? "grabbing" : showExpanded ? "grab" : "pointer";
  let t8;
  $[4] !== t5 || $[5] !== t6 || $[6] !== t7 || $[7] !== x || $[8] !== y ? (t8 = {
    position: "absolute",
    left: x,
    top: y,
    transform: "translate(-50%, -50%)",
    width: HANDLE_HIT,
    height: HANDLE_HIT,
    display: "flex",
    alignItems: "center",
    justifyContent: "center",
    pointerEvents: t5,
    opacity: t6,
    transition: "opacity 100ms ease",
    cursor: t7,
    zIndex: 5,
    border: "none",
    background: "transparent",
    padding: 0
  }, $[4] = t5, $[5] = t6, $[6] = t7, $[7] = x, $[8] = y, $[9] = t8) : t8 = $[9];
  const t9 = showExpanded ? btn.w : rest.w, t10 = showExpanded ? btn.h : rest.h, t11 = showExpanded ? 4 : 3, t12 = showExpanded ? HANDLE_BTN_PAD : 0;
  let t13;
  $[10] !== selected || $[11] !== showExpanded || $[12] !== showRest ? (t13 = showExpanded ? {
    background: selected ? BLUE : "var(--pt-plugin-table-handle-bg)",
    boxShadow: selected ? HANDLE_EXPANDED_SHADOW_SELECTED : HANDLE_EXPANDED_SHADOW
  } : showRest ? {
    background: HANDLE_REST_BG,
    boxShadow: "inset 0 0 0 0.5px var(--pt-plugin-table-handle-ring)"
  } : {
    background: "transparent"
  }, $[10] = selected, $[11] = showExpanded, $[12] = showRest, $[13] = t13) : t13 = $[13];
  let t14;
  $[14] !== t10 || $[15] !== t11 || $[16] !== t12 || $[17] !== t13 || $[18] !== t9 ? (t14 = {
    width: t9,
    height: t10,
    boxSizing: "border-box",
    display: "flex",
    alignItems: "center",
    justifyContent: "center",
    borderRadius: t11,
    padding: t12,
    ...t13
  }, $[14] = t10, $[15] = t11, $[16] = t12, $[17] = t13, $[18] = t9, $[19] = t14) : t14 = $[19];
  let t15;
  $[20] !== isColumn || $[21] !== selected || $[22] !== showExpanded ? (t15 = showExpanded ? /* @__PURE__ */ jsx(DragDots, { isColumn, selected }) : null, $[20] = isColumn, $[21] = selected, $[22] = showExpanded, $[23] = t15) : t15 = $[23];
  let t16;
  $[24] !== t14 || $[25] !== t15 ? (t16 = /* @__PURE__ */ jsx("div", { style: t14, children: t15 }), $[24] = t14, $[25] = t15, $[26] = t16) : t16 = $[26];
  let t17;
  return $[27] !== index || $[28] !== kind || $[29] !== label || $[30] !== onPointerDown || $[31] !== t1 || $[32] !== t16 || $[33] !== t4 || $[34] !== t8 ? (t17 = /* @__PURE__ */ jsx("button", { type: "button", "aria-label": label, "data-pt-plugin-table-handle": kind, "data-pt-plugin-table-handle-index": index, tabIndex: t1, onMouseEnter: t2, onMouseLeave: t3, onPointerDown, onClick: t4, style: t8, children: t16 }), $[27] = index, $[28] = kind, $[29] = label, $[30] = onPointerDown, $[31] = t1, $[32] = t16, $[33] = t4, $[34] = t8, $[35] = t17) : t17 = $[35], t17;
}
function DragDots(t0) {
  const $ = c(7), {
    isColumn,
    selected
  } = t0, t1 = `repeat(${isColumn ? 3 : 2}, ${HANDLE_DOT}px)`;
  let t2;
  $[0] !== t1 ? (t2 = {
    display: "grid",
    gridTemplateColumns: t1,
    gap: HANDLE_DOT_GAP
  }, $[0] = t1, $[1] = t2) : t2 = $[1];
  let t3;
  $[2] !== selected ? (t3 = Array.from({
    length: 6
  }, (_, index) => /* @__PURE__ */ jsx("div", { style: {
    width: HANDLE_DOT,
    height: HANDLE_DOT,
    borderRadius: "50%",
    background: selected ? "var(--pt-plugin-table-accent-fg)" : "var(--pt-plugin-table-handle-dots)"
  } }, index)), $[2] = selected, $[3] = t3) : t3 = $[3];
  let t4;
  return $[4] !== t2 || $[5] !== t3 ? (t4 = /* @__PURE__ */ jsx("div", { "aria-hidden": !0, style: t2, children: t3 }), $[4] = t2, $[5] = t3, $[6] = t4) : t4 = $[6], t4;
}
function TableTrashLayer(t0) {
  const $ = c(19), {
    tableRef,
    metrics,
    selectedRow,
    selectedCol,
    canDeleteRow,
    canDeleteCol,
    onDeleteRow,
    onDeleteCol,
    portalElement,
    trashIcon,
    labels,
    tokenStyle
  } = t0, rowIndex = selectedRow !== null && canDeleteRow ? selectedRow : null, colIndex = selectedCol !== null && canDeleteCol ? selectedCol : null;
  if (!metrics || rowIndex === null && colIndex === null)
    return null;
  let t1;
  $[0] !== labels || $[1] !== onDeleteRow || $[2] !== rowIndex || $[3] !== tableRef || $[4] !== trashIcon ? (t1 = rowIndex !== null ? /* @__PURE__ */ jsx(TrashButton, { icon: trashIcon, label: labels["delete-row"], placement: "left", handleKind: "row", handleIndex: rowIndex, tableRef, onClick: () => onDeleteRow(rowIndex) }) : null, $[0] = labels, $[1] = onDeleteRow, $[2] = rowIndex, $[3] = tableRef, $[4] = trashIcon, $[5] = t1) : t1 = $[5];
  let t2;
  $[6] !== colIndex || $[7] !== labels || $[8] !== onDeleteCol || $[9] !== tableRef || $[10] !== trashIcon ? (t2 = colIndex !== null ? /* @__PURE__ */ jsx(TrashButton, { icon: trashIcon, label: labels["delete-column"], placement: "top", handleKind: "column", handleIndex: colIndex, tableRef, onClick: () => onDeleteCol(colIndex) }) : null, $[6] = colIndex, $[7] = labels, $[8] = onDeleteCol, $[9] = tableRef, $[10] = trashIcon, $[11] = t2) : t2 = $[11];
  let t3;
  $[12] !== t1 || $[13] !== t2 || $[14] !== tokenStyle ? (t3 = /* @__PURE__ */ jsxs("div", { className: "pt-plugin-table-portal", style: tokenStyle, children: [
    t1,
    t2
  ] }), $[12] = t1, $[13] = t2, $[14] = tokenStyle, $[15] = t3) : t3 = $[15];
  const t4 = portalElement ?? document.body;
  let t5;
  return $[16] !== t3 || $[17] !== t4 ? (t5 = createPortal(t3, t4), $[16] = t3, $[17] = t4, $[18] = t5) : t5 = $[18], t5;
}
function TrashButton(t0) {
  const $ = c(21), {
    icon,
    label,
    placement,
    handleKind,
    handleIndex,
    tableRef,
    onClick
  } = t0, [hovered, setHovered] = useState(!1), floatingRef = useRef(null), [position, setPosition] = useState(null);
  let t1, t2;
  $[0] !== handleIndex || $[1] !== handleKind || $[2] !== placement || $[3] !== tableRef ? (t1 = () => {
    const floating = floatingRef.current, reference = tableRef.current?.parentElement?.querySelector(`[data-pt-plugin-table-handle="${handleKind}"][data-pt-plugin-table-handle-index="${handleIndex}"]`);
    return !floating || !reference ? void 0 : autoUpdate(reference, floating, () => {
      computePosition(reference, floating, {
        strategy: "fixed",
        placement,
        middleware: [offset(TRASH_GAP), hide()]
      }).then((t32) => {
        const {
          x,
          y,
          middlewareData
        } = t32;
        setPosition({
          left: x,
          top: y,
          hidden: middlewareData.hide?.referenceHidden ?? !1
        });
      });
    });
  }, t2 = [placement, handleKind, handleIndex, tableRef], $[0] = handleIndex, $[1] = handleKind, $[2] = placement, $[3] = tableRef, $[4] = t1, $[5] = t2) : (t1 = $[4], t2 = $[5]), useLayoutEffect(t1, t2);
  const hidden = !position || position.hidden;
  let t3, t4;
  $[6] === /* @__PURE__ */ Symbol.for("react.memo_cache_sentinel") ? (t3 = () => setHovered(!0), t4 = () => setHovered(!1), $[6] = t3, $[7] = t4) : (t3 = $[6], t4 = $[7]);
  const t5 = position?.left ?? 0, t6 = position?.top ?? 0, t7 = hidden ? "hidden" : "visible", t8 = hidden ? "none" : "auto", t9 = hovered ? "var(--pt-plugin-table-danger)" : "var(--pt-plugin-table-trash-bg)";
  let t10;
  $[8] !== t5 || $[9] !== t6 || $[10] !== t7 || $[11] !== t8 || $[12] !== t9 ? (t10 = {
    fontSize: 15,
    position: "fixed",
    left: t5,
    top: t6,
    visibility: t7,
    pointerEvents: t8,
    width: TRASH_SIZE,
    height: TRASH_SIZE,
    display: "flex",
    alignItems: "center",
    justifyContent: "center",
    background: t9,
    color: "var(--pt-plugin-table-trash-fg)",
    border: "none",
    borderRadius: 4,
    padding: 0,
    cursor: "pointer",
    zIndex: TRASH_Z_INDEX,
    transition: "background 100ms ease"
  }, $[8] = t5, $[9] = t6, $[10] = t7, $[11] = t8, $[12] = t9, $[13] = t10) : t10 = $[13];
  let t11;
  $[14] !== icon ? (t11 = icon ?? /* @__PURE__ */ jsx(Trash2Icon, { size: 14 }), $[14] = icon, $[15] = t11) : t11 = $[15];
  let t12;
  return $[16] !== label || $[17] !== onClick || $[18] !== t10 || $[19] !== t11 ? (t12 = /* @__PURE__ */ jsx("button", { ref: floatingRef, type: "button", "aria-label": label, onMouseEnter: t3, onMouseLeave: t4, onPointerDown: _temp4, onClick, style: t10, children: t11 }), $[16] = label, $[17] = onClick, $[18] = t10, $[19] = t11, $[20] = t12) : t12 = $[20], t12;
}
function _temp4(event) {
  event.preventDefault();
}
const MENU_BTN = 25, MENU_MIN_WIDTH = 200, MENU_Z_INDEX = 10100, MENU_ABOVE_GAP = 2;
function TableMenuAnchor(t0) {
  const $ = c(9), {
    right,
    visible,
    children
  } = t0, anchorRef = useRef(null), [clipped, setClipped] = useState(!1);
  let t1, t2;
  $[0] === /* @__PURE__ */ Symbol.for("react.memo_cache_sentinel") ? (t1 = () => {
    const anchor = anchorRef.current;
    if (!anchor)
      return;
    const observer = new IntersectionObserver((t32) => {
      const [entry] = t32;
      setClipped(!(entry?.isIntersecting ?? !0));
    });
    return observer.observe(anchor), () => observer.disconnect();
  }, t2 = [], $[0] = t1, $[1] = t2) : (t1 = $[0], t2 = $[1]), useLayoutEffect(t1, t2);
  const t3 = visible && !clipped ? "auto" : "none", t4 = visible && !clipped ? 1 : 0;
  let t5;
  $[2] !== right || $[3] !== t3 || $[4] !== t4 ? (t5 = {
    position: "absolute",
    right,
    top: WRAPPER_PAD_TOP + GUTTER_TOP - MENU_ABOVE_GAP,
    transform: "translateY(-100%)",
    pointerEvents: t3,
    opacity: t4,
    transition: "opacity 100ms ease",
    zIndex: 6
  }, $[2] = right, $[3] = t3, $[4] = t4, $[5] = t5) : t5 = $[5];
  const t6 = clipped ? null : children;
  let t7;
  return $[6] !== t5 || $[7] !== t6 ? (t7 = /* @__PURE__ */ jsx("div", { ref: anchorRef, contentEditable: !1, onPointerDown: _temp5, style: t5, children: t6 }), $[6] = t5, $[7] = t6, $[8] = t7) : t7 = $[8], t7;
}
function _temp5(event) {
  event.preventDefault();
}
function TableMenu(t0) {
  const $ = c(33), {
    right,
    active,
    handlers,
    portalElement,
    labels,
    tokenStyle
  } = t0, triggerRef = useRef(null), menuRef = useRef(null), [open, setOpen] = useState(!1), [hovered, setHovered] = useState(!1), [menuPos, setMenuPos] = useState(null);
  let t1;
  $[0] === /* @__PURE__ */ Symbol.for("react.memo_cache_sentinel") ? (t1 = () => {
    setOpen(!1), setMenuPos(null);
  }, $[0] = t1) : t1 = $[0];
  const closeMenu = t1;
  let t2, t3;
  $[1] !== open ? (t2 = () => {
    if (!open)
      return;
    const trigger = triggerRef.current, menu = menuRef.current;
    return !trigger || !menu ? void 0 : autoUpdate(trigger, menu, () => {
      computePosition(trigger, menu, {
        strategy: "fixed",
        placement: "bottom-end",
        middleware: [offset(6), hide()]
      }).then((t42) => {
        const {
          x,
          y,
          middlewareData
        } = t42;
        if (middlewareData.hide?.referenceHidden) {
          closeMenu();
          return;
        }
        setMenuPos({
          left: x,
          top: y
        });
      });
    });
  }, t3 = [open, closeMenu], $[1] = open, $[2] = t2, $[3] = t3) : (t2 = $[2], t3 = $[3]), useLayoutEffect(t2, t3);
  let t4, t5;
  $[4] !== open ? (t4 = () => {
    if (!open)
      return;
    const onPointerDown = (event) => {
      const target = event.target;
      target instanceof Node && (menuRef.current?.contains(target) || triggerRef.current?.contains(target)) || closeMenu();
    }, onKeyDown = (event_0) => {
      event_0.key === "Escape" && closeMenu();
    };
    return document.addEventListener("pointerdown", onPointerDown), document.addEventListener("keydown", onKeyDown), () => {
      document.removeEventListener("pointerdown", onPointerDown), document.removeEventListener("keydown", onKeyDown);
    };
  }, t5 = [open, closeMenu], $[4] = open, $[5] = t4, $[6] = t5) : (t4 = $[5], t5 = $[6]), useLayoutEffect(t4, t5);
  const visible = active || open, t6 = labels["table-options"], t7 = visible ? 0 : -1;
  let t8, t9;
  $[7] === /* @__PURE__ */ Symbol.for("react.memo_cache_sentinel") ? (t8 = () => setHovered(!0), t9 = () => setHovered(!1), $[7] = t8, $[8] = t9) : (t8 = $[7], t9 = $[8]);
  let t10;
  $[9] !== open ? (t10 = () => {
    open ? closeMenu() : setOpen(!0);
  }, $[9] = open, $[10] = t10) : t10 = $[10];
  const t11 = open ? "var(--pt-plugin-table-lane-bg-hover)" : hovered ? "var(--pt-plugin-table-header-bg)" : "transparent", t12 = visible ? "auto" : "none", t13 = visible ? 1 : 0;
  let t14;
  $[11] !== right || $[12] !== t11 || $[13] !== t12 || $[14] !== t13 ? (t14 = {
    position: "absolute",
    right,
    top: WRAPPER_PAD_TOP + GUTTER_TOP - MENU_BTN - MENU_ABOVE_GAP,
    width: MENU_BTN,
    height: MENU_BTN,
    display: "flex",
    alignItems: "center",
    justifyContent: "center",
    border: "none",
    borderRadius: 3,
    padding: 0,
    background: t11,
    color: "var(--pt-plugin-table-fg)",
    cursor: "pointer",
    pointerEvents: t12,
    opacity: t13,
    transition: "opacity 100ms ease, background 100ms ease",
    zIndex: 6
  }, $[11] = right, $[12] = t11, $[13] = t12, $[14] = t13, $[15] = t14) : t14 = $[15];
  let t15;
  $[16] === /* @__PURE__ */ Symbol.for("react.memo_cache_sentinel") ? (t15 = /* @__PURE__ */ jsx(EllipsisIcon, { size: 20 }), $[16] = t15) : t15 = $[16];
  let t16;
  $[17] !== open || $[18] !== t10 || $[19] !== t14 || $[20] !== t6 || $[21] !== t7 ? (t16 = /* @__PURE__ */ jsx("button", { ref: triggerRef, type: "button", contentEditable: !1, "aria-label": t6, "aria-haspopup": "menu", "aria-expanded": open, tabIndex: t7, onMouseEnter: t8, onMouseLeave: t9, onPointerDown: _temp6, onClick: t10, style: t14, children: t15 }), $[17] = open, $[18] = t10, $[19] = t14, $[20] = t6, $[21] = t7, $[22] = t16) : t16 = $[22];
  let t17;
  $[23] !== handlers || $[24] !== labels || $[25] !== menuPos || $[26] !== open || $[27] !== portalElement || $[28] !== tokenStyle ? (t17 = open ? createPortal(/* @__PURE__ */ jsxs("div", { ref: menuRef, role: "menu", className: "pt-plugin-table-portal", style: {
    ...tokenStyle,
    position: "fixed",
    left: menuPos?.left ?? 0,
    top: menuPos?.top ?? 0,
    visibility: menuPos ? "visible" : "hidden",
    width: MENU_MIN_WIDTH,
    background: "var(--pt-plugin-table-menu-bg)",
    border: "1px solid var(--pt-plugin-table-menu-border)",
    borderRadius: "var(--pt-plugin-table-radius)",
    padding: 5,
    zIndex: MENU_Z_INDEX
  }, children: [
    /* @__PURE__ */ jsx(MenuRow, { label: labels["menu-header-row"], icon: /* @__PURE__ */ jsx(PanelTopIcon, { size: 16 }), trailing: /* @__PURE__ */ jsx(ToggleSwitch, { checked: handlers.hasHeader }), onClick: handlers.onToggleHeader }),
    /* @__PURE__ */ jsx(MenuRow, { label: labels["menu-select-table"], icon: /* @__PURE__ */ jsx(TableIcon, { size: 16 }), onClick: () => {
      closeMenu(), handlers.onSelectTable();
    } }),
    /* @__PURE__ */ jsx(MenuRow, { label: labels["menu-delete-table"], icon: /* @__PURE__ */ jsx(Trash2Icon, { size: 16 }), destructive: !0, onClick: () => {
      closeMenu(), handlers.onDeleteTable();
    } })
  ] }), portalElement ?? document.body) : null, $[23] = handlers, $[24] = labels, $[25] = menuPos, $[26] = open, $[27] = portalElement, $[28] = tokenStyle, $[29] = t17) : t17 = $[29];
  let t18;
  return $[30] !== t16 || $[31] !== t17 ? (t18 = /* @__PURE__ */ jsxs(Fragment, { children: [
    t16,
    t17
  ] }), $[30] = t16, $[31] = t17, $[32] = t18) : t18 = $[32], t18;
}
function _temp6(event_1) {
  event_1.preventDefault(), event_1.stopPropagation();
}
function MenuRow(t0) {
  const $ = c(17), {
    label,
    icon,
    trailing,
    destructive,
    onClick
  } = t0, [hovered, setHovered] = useState(!1);
  let t1, t2;
  $[0] === /* @__PURE__ */ Symbol.for("react.memo_cache_sentinel") ? (t1 = () => setHovered(!0), t2 = () => setHovered(!1), $[0] = t1, $[1] = t2) : (t1 = $[0], t2 = $[1]);
  const t3 = hovered ? "var(--pt-plugin-table-menu-hover)" : "transparent", t4 = destructive ? "var(--pt-plugin-table-danger)" : "var(--pt-plugin-table-fg)";
  let t5;
  $[2] !== t3 || $[3] !== t4 ? (t5 = {
    display: "flex",
    alignItems: "center",
    gap: 10,
    width: "100%",
    padding: "8px 10px",
    border: "none",
    background: t3,
    cursor: "pointer",
    borderRadius: 4,
    textAlign: "left",
    fontSize: 13,
    color: t4
  }, $[2] = t3, $[3] = t4, $[4] = t5) : t5 = $[4];
  let t6;
  $[5] === /* @__PURE__ */ Symbol.for("react.memo_cache_sentinel") ? (t6 = {
    width: 21,
    height: 21,
    display: "inline-flex",
    alignItems: "center",
    justifyContent: "center",
    flexShrink: 0
  }, $[5] = t6) : t6 = $[5];
  let t7;
  $[6] !== icon ? (t7 = /* @__PURE__ */ jsx("span", { style: t6, children: icon }), $[6] = icon, $[7] = t7) : t7 = $[7];
  let t8;
  $[8] === /* @__PURE__ */ Symbol.for("react.memo_cache_sentinel") ? (t8 = {
    flex: 1,
    whiteSpace: "nowrap"
  }, $[8] = t8) : t8 = $[8];
  let t9;
  $[9] !== label ? (t9 = /* @__PURE__ */ jsx("span", { style: t8, children: label }), $[9] = label, $[10] = t9) : t9 = $[10];
  let t10;
  return $[11] !== onClick || $[12] !== t5 || $[13] !== t7 || $[14] !== t9 || $[15] !== trailing ? (t10 = /* @__PURE__ */ jsxs("button", { type: "button", role: "menuitem", onPointerDown: _temp7, onClick, onMouseEnter: t1, onMouseLeave: t2, style: t5, children: [
    t7,
    t9,
    trailing
  ] }), $[11] = onClick, $[12] = t5, $[13] = t7, $[14] = t9, $[15] = trailing, $[16] = t10) : t10 = $[16], t10;
}
function _temp7(event) {
  return event.preventDefault();
}
function ToggleSwitch(t0) {
  const $ = c(6), {
    checked
  } = t0, t1 = checked ? "var(--pt-plugin-table-toggle-track-on)" : "var(--pt-plugin-table-toggle-track)", t2 = checked ? "flex-end" : "flex-start";
  let t3;
  $[0] !== t1 || $[1] !== t2 ? (t3 = {
    width: 28,
    height: 16,
    borderRadius: 8,
    padding: 2,
    background: t1,
    display: "flex",
    alignItems: "center",
    justifyContent: t2,
    flexShrink: 0,
    transition: "background 100ms ease"
  }, $[0] = t1, $[1] = t2, $[2] = t3) : t3 = $[2];
  let t4;
  $[3] === /* @__PURE__ */ Symbol.for("react.memo_cache_sentinel") ? (t4 = /* @__PURE__ */ jsx("span", { style: {
    width: 12,
    height: 12,
    borderRadius: "50%",
    background: "var(--pt-plugin-table-toggle-knob)",
    boxShadow: "0 1px 2px rgba(0,0,0,0.15)"
  } }), $[3] = t4) : t4 = $[3];
  let t5;
  return $[4] !== t3 ? (t5 = /* @__PURE__ */ jsx("span", { "aria-hidden": "true", style: t3, children: t4 }), $[4] = t3, $[5] = t5) : t5 = $[5], t5;
}
function ReorderInsertLine(t0) {
  const $ = c(6), {
    metrics,
    kind,
    insertIndex
  } = t0;
  if (kind === "column") {
    const t12 = GUTTER_LEFT + colBorderX(metrics, insertIndex);
    let t22;
    return $[0] !== metrics.height || $[1] !== t12 ? (t22 = /* @__PURE__ */ jsx("div", { style: {
      position: "absolute",
      left: t12,
      top: GUTTER_TOP,
      width: INSERT_GUIDE,
      height: metrics.height,
      transform: "translateX(-50%)",
      background: BLUE,
      pointerEvents: "none",
      zIndex: 12
    } }), $[0] = metrics.height, $[1] = t12, $[2] = t22) : t22 = $[2], t22;
  }
  const t1 = GUTTER_TOP + rowBorderY(metrics, insertIndex);
  let t2;
  return $[3] !== metrics.width || $[4] !== t1 ? (t2 = /* @__PURE__ */ jsx("div", { style: {
    position: "absolute",
    left: GUTTER_LEFT,
    top: t1,
    width: metrics.width,
    height: INSERT_GUIDE,
    transform: "translateY(-50%)",
    background: BLUE,
    pointerEvents: "none",
    zIndex: 12
  } }), $[3] = metrics.width, $[4] = t1, $[5] = t2) : t2 = $[5], t2;
}
function keyboardActivation(action) {
  return (event) => {
    event.detail === 0 && action();
  };
}
const GHOST_SHADOW = "0 10px 32px rgba(0, 0, 0, 0.14), 0 2px 6px rgba(0, 0, 0, 0.06)";
function ReorderGhost(t0) {
  const $ = c(32), {
    drag,
    metrics,
    hasHeader,
    cellTexts
  } = t0;
  if (!drag?.active || !metrics || !cellTexts)
    return null;
  const left = drag.clientX - drag.grabOffsetX, top = drag.clientY - drag.grabOffsetY;
  if (drag.kind === "row") {
    const rowMetrics = metrics.rows[drag.index];
    if (!rowMetrics)
      return null;
    const isHeader = hasHeader && drag.index === 0, t12 = isHeader ? "var(--pt-plugin-table-header-bg)" : "var(--pt-plugin-table-bg)", t22 = isHeader ? HEADER_WEIGHT : 400;
    let t32;
    $[0] !== left || $[1] !== metrics.width || $[2] !== rowMetrics.height || $[3] !== t12 || $[4] !== t22 || $[5] !== top ? (t32 = {
      position: "fixed",
      left,
      top,
      width: metrics.width,
      height: rowMetrics.height,
      display: "flex",
      pointerEvents: "none",
      zIndex: 1e4,
      borderRadius: "var(--pt-plugin-table-radius)",
      border: `1px solid ${BORDER}`,
      background: t12,
      boxShadow: GHOST_SHADOW,
      overflow: "hidden",
      fontWeight: t22
    }, $[0] = left, $[1] = metrics.width, $[2] = rowMetrics.height, $[3] = t12, $[4] = t22, $[5] = top, $[6] = t32) : t32 = $[6];
    let t4;
    if ($[7] !== cellTexts || $[8] !== metrics.cols) {
      let t52;
      $[10] !== cellTexts || $[11] !== metrics.cols.length ? (t52 = (col, index) => /* @__PURE__ */ jsx("div", { style: {
        width: col.width,
        padding: "8px 12px",
        borderRight: index < metrics.cols.length - 1 ? `1px solid ${BORDER}` : "none",
        overflow: "hidden",
        whiteSpace: "nowrap",
        textOverflow: "ellipsis"
      }, children: cellTexts[index] }, index), $[10] = cellTexts, $[11] = metrics.cols.length, $[12] = t52) : t52 = $[12], t4 = metrics.cols.map(t52), $[7] = cellTexts, $[8] = metrics.cols, $[9] = t4;
    } else
      t4 = $[9];
    let t5;
    return $[13] !== t32 || $[14] !== t4 ? (t5 = /* @__PURE__ */ jsx("div", { contentEditable: !1, style: t32, children: t4 }), $[13] = t32, $[14] = t4, $[15] = t5) : t5 = $[15], t5;
  }
  const colMetrics = metrics.cols[drag.index];
  if (!colMetrics)
    return null;
  let t1;
  $[16] !== colMetrics.width || $[17] !== left || $[18] !== metrics.height || $[19] !== top ? (t1 = {
    position: "fixed",
    left,
    top,
    width: colMetrics.width,
    height: metrics.height,
    display: "flex",
    flexDirection: "column",
    pointerEvents: "none",
    zIndex: 1e4,
    borderRadius: "var(--pt-plugin-table-radius)",
    border: `1px solid ${BORDER}`,
    background: "var(--pt-plugin-table-bg)",
    boxShadow: GHOST_SHADOW,
    overflow: "hidden"
  }, $[16] = colMetrics.width, $[17] = left, $[18] = metrics.height, $[19] = top, $[20] = t1) : t1 = $[20];
  let t2;
  if ($[21] !== cellTexts || $[22] !== hasHeader || $[23] !== metrics.rows) {
    let t32;
    $[25] !== cellTexts || $[26] !== hasHeader || $[27] !== metrics.rows.length ? (t32 = (row, index_0) => /* @__PURE__ */ jsx("div", { style: {
      height: row.height,
      padding: "8px 12px",
      borderBottom: index_0 < metrics.rows.length - 1 ? `1px solid ${BORDER}` : "none",
      background: hasHeader && index_0 === 0 ? "var(--pt-plugin-table-header-bg)" : void 0,
      fontWeight: hasHeader && index_0 === 0 ? HEADER_WEIGHT : 400,
      overflow: "hidden",
      whiteSpace: "nowrap",
      textOverflow: "ellipsis"
    }, children: cellTexts[index_0] }, index_0), $[25] = cellTexts, $[26] = hasHeader, $[27] = metrics.rows.length, $[28] = t32) : t32 = $[28], t2 = metrics.rows.map(t32), $[21] = cellTexts, $[22] = hasHeader, $[23] = metrics.rows, $[24] = t2;
  } else
    t2 = $[24];
  let t3;
  return $[29] !== t1 || $[30] !== t2 ? (t3 = /* @__PURE__ */ jsx("div", { contentEditable: !1, style: t1, children: t2 }), $[29] = t1, $[30] = t2, $[31] = t3) : t3 = $[31], t3;
}
function TableScrollFade(t0) {
  const $ = c(8), {
    left,
    right
  } = t0;
  if (!left && !right)
    return null;
  let t1;
  $[0] === /* @__PURE__ */ Symbol.for("react.memo_cache_sentinel") ? (t1 = {
    position: "absolute",
    top: WRAPPER_PAD_TOP + GUTTER_TOP,
    bottom: EXTEND_SIZE + EXTEND_GAP + 12,
    width: 36,
    pointerEvents: "none",
    zIndex: 7
  }, $[0] = t1) : t1 = $[0];
  const fadeBase = t1;
  let t2;
  $[1] !== left ? (t2 = left ? /* @__PURE__ */ jsx("div", { contentEditable: !1, "aria-hidden": "true", style: {
    ...fadeBase,
    left: 0,
    background: "linear-gradient(to right, var(--pt-plugin-table-bg) 0%, transparent 100%)"
  } }) : null, $[1] = left, $[2] = t2) : t2 = $[2];
  let t3;
  $[3] !== right ? (t3 = right ? /* @__PURE__ */ jsx("div", { contentEditable: !1, "aria-hidden": "true", style: {
    ...fadeBase,
    right: 0,
    background: "linear-gradient(to left, var(--pt-plugin-table-bg) 0%, transparent 100%)"
  } }) : null, $[3] = right, $[4] = t3) : t3 = $[4];
  let t4;
  return $[5] !== t2 || $[6] !== t3 ? (t4 = /* @__PURE__ */ jsxs(Fragment, { children: [
    t2,
    t3
  ] }), $[5] = t2, $[6] = t3, $[7] = t4) : t4 = $[7], t4;
}
const DRAG_THRESHOLD_PX = 4;
function computeGhostGrabOffset(kind, index, clientX, clientY, tableRect, metrics) {
  if (!tableRect || !metrics)
    return {
      grabOffsetX: 0,
      grabOffsetY: 0
    };
  if (kind === "row") {
    const rowMetrics = metrics.rows[index];
    return rowMetrics ? {
      grabOffsetX: clientX - tableRect.left,
      grabOffsetY: clientY - (tableRect.top + rowMetrics.top)
    } : {
      grabOffsetX: 0,
      grabOffsetY: 0
    };
  }
  const colMetrics = metrics.cols[index];
  return colMetrics ? {
    grabOffsetX: clientX - (tableRect.left + colMetrics.left),
    grabOffsetY: clientY - tableRect.top
  } : {
    grabOffsetX: 0,
    grabOffsetY: 0
  };
}
function computeRowInsertIndex(clientY, tableRect, rows) {
  if (rows.length === 0)
    return 0;
  const y = clientY - tableRect.top, lastRow = rows[rows.length - 1], boundaries = [...rows.map((row) => row.top), lastRow.top + lastRow.height];
  return nearestBoundary(y, boundaries);
}
function computeColInsertIndex(clientX, tableRect, cols) {
  if (cols.length === 0)
    return 0;
  const x = clientX - tableRect.left, lastCol = cols[cols.length - 1], boundaries = [...cols.map((col) => col.left), lastCol.left + lastCol.width];
  return nearestBoundary(x, boundaries);
}
function nearestBoundary(position, boundaries) {
  let nearest = 0, best = Number.POSITIVE_INFINITY;
  return boundaries.forEach((boundary, index) => {
    const distance = Math.abs(position - boundary);
    distance < best && (best = distance, nearest = index);
  }), nearest;
}
function reorderIndex(from, insertBefore) {
  return insertBefore === from || insertBefore === from + 1 ? from : insertBefore > from ? insertBefore - 1 : insertBefore;
}
function useTableDragReorder(t0) {
  const $ = c(14), {
    tableRef,
    metrics,
    onCommitRow,
    onCommitCol,
    onSelectRow,
    onSelectCol
  } = t0, sessionRef = useRef(null), [drag, setDrag] = useState(null);
  let t1;
  $[0] === /* @__PURE__ */ Symbol.for("react.memo_cache_sentinel") ? (t1 = () => {
    sessionRef.current = null, setDrag(null), document.body.style.userSelect = "", document.body.style.cursor = "";
  }, $[0] = t1) : t1 = $[0];
  const endSession = t1;
  let t2;
  $[1] !== metrics || $[2] !== onCommitCol || $[3] !== onCommitRow || $[4] !== onSelectCol || $[5] !== onSelectRow || $[6] !== tableRef ? (t2 = (kind, index, event) => {
    if (event.button !== 0)
      return;
    event.preventDefault(), event.stopPropagation();
    const target = event.currentTarget;
    target.setPointerCapture(event.pointerId);
    const tableRect = tableRef.current?.getBoundingClientRect() ?? null, {
      grabOffsetX,
      grabOffsetY
    } = computeGhostGrabOffset(kind, index, event.clientX, event.clientY, tableRect, metrics);
    sessionRef.current = {
      kind,
      index,
      startX: event.clientX,
      startY: event.clientY,
      clientX: event.clientX,
      clientY: event.clientY,
      grabOffsetX,
      grabOffsetY,
      active: !1,
      insertIndex: index,
      pointerId: event.pointerId
    }, setDrag({
      ...sessionRef.current
    }), document.body.style.userSelect = "none";
    const onMove = (moveEvent) => {
      const session = sessionRef.current;
      if (!session || moveEvent.pointerId !== session.pointerId)
        return;
      const dx = moveEvent.clientX - session.startX, dy = moveEvent.clientY - session.startY, active = session.active || Math.hypot(dx, dy) >= DRAG_THRESHOLD_PX, table = tableRef.current;
      let insertIndex = session.insertIndex;
      if (active && table && metrics) {
        const rect = table.getBoundingClientRect();
        insertIndex = session.kind === "row" ? computeRowInsertIndex(moveEvent.clientY, rect, metrics.rows) : computeColInsertIndex(moveEvent.clientX, rect, metrics.cols);
      }
      sessionRef.current = {
        ...session,
        active,
        clientX: moveEvent.clientX,
        clientY: moveEvent.clientY,
        insertIndex
      }, setDrag({
        ...sessionRef.current
      }), active && (document.body.style.cursor = "grabbing");
    }, onUp = (upEvent) => {
      const session_0 = sessionRef.current;
      if (!(!session_0 || upEvent.pointerId !== session_0.pointerId)) {
        try {
          target.releasePointerCapture(session_0.pointerId);
        } catch {
        }
        session_0.active ? session_0.kind === "row" ? onCommitRow(session_0.index, session_0.insertIndex) : onCommitCol(session_0.index, session_0.insertIndex) : session_0.kind === "row" ? onSelectRow(session_0.index) : onSelectCol(session_0.index), window.removeEventListener("pointermove", onMove), window.removeEventListener("pointerup", onUp), window.removeEventListener("pointercancel", onUp), endSession();
      }
    };
    window.addEventListener("pointermove", onMove), window.addEventListener("pointerup", onUp), window.addEventListener("pointercancel", onUp);
  }, $[1] = metrics, $[2] = onCommitCol, $[3] = onCommitRow, $[4] = onSelectCol, $[5] = onSelectRow, $[6] = tableRef, $[7] = t2) : t2 = $[7];
  const onHandlePointerDown = t2;
  let t3, t4;
  $[8] === /* @__PURE__ */ Symbol.for("react.memo_cache_sentinel") ? (t3 = () => () => endSession(), t4 = [endSession], $[8] = t3, $[9] = t4) : (t3 = $[8], t4 = $[9]), useEffect(t3, t4);
  const t5 = !!drag?.active;
  let t6;
  return $[10] !== drag || $[11] !== onHandlePointerDown || $[12] !== t5 ? (t6 = {
    drag,
    onHandlePointerDown,
    isDragging: t5
  }, $[10] = drag, $[11] = onHandlePointerDown, $[12] = t5, $[13] = t6) : t6 = $[13], t6;
}
const MIN_COL_PX = 140, SCROLL_PADDING_BOTTOM = 12, EDGE_EPSILON = 2;
function useTableHorizontalLayout(scrollRef, colCount, revision) {
  const $ = c(10), [enforceMinColumnWidth, setEnforceMinColumnWidth] = useState(!1);
  let t0;
  $[0] !== colCount || $[1] !== scrollRef ? (t0 = () => {
    const element = scrollRef.current;
    if (!element || colCount < 1) {
      setEnforceMinColumnWidth(!1);
      return;
    }
    const next = colCount * MIN_COL_PX > element.clientWidth + EDGE_EPSILON;
    setEnforceMinColumnWidth((prev) => prev === next ? prev : next);
  }, $[0] = colCount, $[1] = scrollRef, $[2] = t0) : t0 = $[2];
  const measure = t0;
  let t1;
  $[3] !== measure || $[4] !== scrollRef ? (t1 = () => {
    const element_0 = scrollRef.current;
    if (!element_0)
      return;
    measure(), window.addEventListener("resize", measure);
    const observer = new ResizeObserver(measure);
    return observer.observe(element_0), element_0.firstElementChild && observer.observe(element_0.firstElementChild), () => {
      window.removeEventListener("resize", measure), observer.disconnect();
    };
  }, $[3] = measure, $[4] = scrollRef, $[5] = t1) : t1 = $[5];
  let t2;
  return $[6] !== measure || $[7] !== revision || $[8] !== scrollRef ? (t2 = [scrollRef, measure, revision], $[6] = measure, $[7] = revision, $[8] = scrollRef, $[9] = t2) : t2 = $[9], useLayoutEffect(t1, t2), enforceMinColumnWidth;
}
function useScrollFade(scrollRef, revision) {
  const $ = c(10);
  let t0;
  $[0] === /* @__PURE__ */ Symbol.for("react.memo_cache_sentinel") ? (t0 = {
    left: !1,
    right: !1
  }, $[0] = t0) : t0 = $[0];
  const [fade, setFade] = useState(t0);
  let t1;
  $[1] !== scrollRef ? (t1 = () => {
    const element = scrollRef.current;
    if (!element) {
      setFade({
        left: !1,
        right: !1
      });
      return;
    }
    const {
      scrollLeft,
      scrollWidth,
      clientWidth
    } = element, overflows = scrollWidth > clientWidth + EDGE_EPSILON;
    setFade((prev) => {
      const next = {
        left: overflows && scrollLeft > EDGE_EPSILON,
        right: overflows && scrollLeft + clientWidth < scrollWidth - EDGE_EPSILON
      };
      return prev.left === next.left && prev.right === next.right ? prev : next;
    });
  }, $[1] = scrollRef, $[2] = t1) : t1 = $[2];
  const measure = t1;
  let t2;
  $[3] !== measure || $[4] !== scrollRef ? (t2 = () => {
    const element_0 = scrollRef.current;
    if (!element_0)
      return;
    measure(), element_0.addEventListener("scroll", measure, {
      passive: !0
    }), window.addEventListener("resize", measure);
    const observer = new ResizeObserver(measure);
    return observer.observe(element_0), element_0.firstElementChild && observer.observe(element_0.firstElementChild), () => {
      element_0.removeEventListener("scroll", measure), window.removeEventListener("resize", measure), observer.disconnect();
    };
  }, $[3] = measure, $[4] = scrollRef, $[5] = t2) : t2 = $[5];
  let t3;
  return $[6] !== measure || $[7] !== revision || $[8] !== scrollRef ? (t3 = [scrollRef, measure, revision], $[6] = measure, $[7] = revision, $[8] = scrollRef, $[9] = t3) : t3 = $[9], useEffect(t2, t3), fade;
}
const TableDragContext = createContext(null);
function Table(t0) {
  const $ = c(143), {
    attributes,
    children,
    node,
    path,
    portalElement,
    renderMenu,
    icons,
    labels,
    tokens
  } = t0;
  let t1;
  $[0] !== labels ? (t1 = {
    ...defaultLabels,
    ...labels
  }, $[0] = labels, $[1] = t1) : t1 = $[1];
  const resolvedLabels = t1, tokenStyle = tokens, editor = useEditor(), config = resolveTableConfig(node._type), {
    isTable
  } = createTableGuards(config), table = isTable(node) ? node : void 0, rows = table ? tableRows(config, table) : [], rowCount = rows.length, columnCount = rows[0] ? rowCells(config, rows[0]).length : 0, tableRef = useRef(null), scrollRef = useRef(null), structureRevision = rows.map((row) => `${row._key}:${rowCells(config, row).map(_temp).join(",")}`).join("|"), metrics = useTableMetrics(tableRef, structureRevision), scrollWide = useTableHorizontalLayout(scrollRef, columnCount, structureRevision), fade = useScrollFade(scrollRef, structureRevision), [active, setActive] = useState(!1), [hoverCell, setHoverCell] = useState(null), [menuOpen, setMenuOpen] = useState(!1), readOnly = useEditorSelector(editor, _temp2);
  let t2;
  $[2] !== config || $[3] !== path ? (t2 = (snapshot_0) => {
    const selection = getTableSelection(snapshot_0, config);
    return selection && isEqualPaths(selection.tablePath, path) ? selection : null;
  }, $[2] = config, $[3] = path, $[4] = t2) : t2 = $[4];
  const tableSelection = useEditorSelector(editor, t2, isEqualTableSelection), hasCellRange = tableSelection !== null, selectedRow = tableSelection && tableSelection.rowRange[0] === tableSelection.rowRange[1] && tableSelection.colRange[0] === 0 && tableSelection.colRange[1] === columnCount - 1 ? tableSelection.rowRange[0] : null, selectedCol = tableSelection && tableSelection.colRange[0] === tableSelection.colRange[1] && tableSelection.rowRange[0] === 0 && tableSelection.rowRange[1] === rowCount - 1 ? tableSelection.colRange[0] : null;
  let t3;
  $[5] !== editor ? (t3 = (anchorCellPath, focusCellPath) => {
    const snapshot_1 = editor.getSnapshot(), anchor = cellStartPoint(snapshot_1, anchorCellPath), focus = cellStartPoint(snapshot_1, focusCellPath);
    anchor && focus && editor.send({
      type: "select",
      at: {
        anchor,
        focus
      }
    });
  }, $[5] = editor, $[6] = t3) : t3 = $[6];
  const selectCells = t3;
  let t4;
  $[7] !== editor || $[8] !== path || $[9] !== selectCells ? (t4 = (index) => {
    const snapshot_2 = editor.getSnapshot(), row_0 = getChildren(snapshot_2, path).at(index), cells = row_0 ? getChildren(snapshot_2, row_0.path) : [], first = cells.at(0), last = cells.at(-1);
    first && last && selectCells(first.path, last.path);
  }, $[7] = editor, $[8] = path, $[9] = selectCells, $[10] = t4) : t4 = $[10];
  const selectRow = t4;
  let t5;
  $[11] !== editor || $[12] !== path || $[13] !== selectCells ? (t5 = (index_0) => {
    const snapshot_3 = editor.getSnapshot(), rows_0 = getChildren(snapshot_3, path), firstRow = rows_0.at(0), lastRow = rows_0.at(-1), first_0 = firstRow && getChildren(snapshot_3, firstRow.path).at(index_0), last_0 = lastRow && getChildren(snapshot_3, lastRow.path).at(index_0);
    first_0 && last_0 && selectCells(first_0.path, last_0.path);
  }, $[11] = editor, $[12] = path, $[13] = selectCells, $[14] = t5) : t5 = $[14];
  const selectCol = t5;
  let t6;
  $[15] !== editor || $[16] !== path ? (t6 = (boundary) => {
    const snapshot_4 = editor.getSnapshot(), rows_1 = getChildren(snapshot_4, path), referenceRow = rows_1.at(Math.min(boundary, rows_1.length - 1)), referenceCell = referenceRow && getChildren(snapshot_4, referenceRow.path).at(0), insideCell = referenceCell && getFirstChild(snapshot_4, referenceCell.path);
    insideCell && editor.send({
      type: "custom.insert.row",
      at: insideCell.path,
      position: boundary < rows_1.length ? "before" : "after"
    });
  }, $[15] = editor, $[16] = path, $[17] = t6) : t6 = $[17];
  const insertRow = t6;
  let t7;
  $[18] !== editor || $[19] !== path ? (t7 = (boundary_0) => {
    const snapshot_5 = editor.getSnapshot(), firstRow_0 = getChildren(snapshot_5, path).at(0), cells_0 = firstRow_0 ? getChildren(snapshot_5, firstRow_0.path) : [], referenceCell_0 = cells_0.at(Math.min(boundary_0, cells_0.length - 1)), insideCell_0 = referenceCell_0 && getFirstChild(snapshot_5, referenceCell_0.path);
    insideCell_0 && editor.send({
      type: "custom.insert.column",
      at: insideCell_0.path,
      position: boundary_0 < cells_0.length ? "before" : "after"
    });
  }, $[18] = editor, $[19] = path, $[20] = t7) : t7 = $[20];
  const insertCol = t7, hasHeader = (Number(table?.headerRows) || 0) >= 1;
  let t8;
  $[21] !== editor || $[22] !== hasHeader || $[23] !== path ? (t8 = () => {
    editor.send({
      type: "block.set",
      at: path,
      props: {
        headerRows: hasHeader ? 0 : 1
      }
    });
  }, $[21] = editor, $[22] = hasHeader, $[23] = path, $[24] = t8) : t8 = $[24];
  const toggleHeader = t8;
  let t9;
  $[25] !== editor || $[26] !== path || $[27] !== selectCells ? (t9 = () => {
    const snapshot_6 = editor.getSnapshot(), rows_2 = getChildren(snapshot_6, path), firstRow_1 = rows_2.at(0), lastRow_0 = rows_2.at(-1), first_1 = firstRow_1 && getChildren(snapshot_6, firstRow_1.path).at(0), last_1 = lastRow_0 && getChildren(snapshot_6, lastRow_0.path).at(-1);
    first_1 && last_1 && selectCells(first_1.path, last_1.path);
  }, $[25] = editor, $[26] = path, $[27] = selectCells, $[28] = t9) : t9 = $[28];
  const selectTable = t9;
  let t10;
  $[29] !== editor || $[30] !== path ? (t10 = () => {
    const snapshot_7 = editor.getSnapshot(), firstRow_2 = getChildren(snapshot_7, path).at(0), cell = firstRow_2 && getChildren(snapshot_7, firstRow_2.path).at(0), insideCell_1 = cell && getFirstChild(snapshot_7, cell.path);
    insideCell_1 && editor.send({
      type: "custom.unset.table",
      at: insideCell_1.path
    });
  }, $[29] = editor, $[30] = path, $[31] = t10) : t10 = $[31];
  const deleteTable = t10;
  let t11;
  $[32] !== editor || $[33] !== path ? (t11 = (index_1) => {
    const snapshot_8 = editor.getSnapshot(), row_1 = getChildren(snapshot_8, path).at(index_1), cell_0 = row_1 && getChildren(snapshot_8, row_1.path).at(0), insideCell_2 = cell_0 && getFirstChild(snapshot_8, cell_0.path);
    insideCell_2 && editor.send({
      type: "custom.unset.row",
      at: insideCell_2.path
    });
  }, $[32] = editor, $[33] = path, $[34] = t11) : t11 = $[34];
  const deleteRow = t11;
  let t12;
  $[35] !== editor || $[36] !== path ? (t12 = (index_2) => {
    const snapshot_9 = editor.getSnapshot(), firstRow_3 = getChildren(snapshot_9, path).at(0), cell_1 = firstRow_3 && getChildren(snapshot_9, firstRow_3.path).at(index_2), insideCell_3 = cell_1 && getFirstChild(snapshot_9, cell_1.path);
    insideCell_3 && editor.send({
      type: "custom.unset.column",
      at: insideCell_3.path
    });
  }, $[35] = editor, $[36] = path, $[37] = t12) : t12 = $[37];
  const deleteCol = t12;
  let t13;
  $[38] !== editor || $[39] !== path ? (t13 = (from, insertBefore) => {
    const finalIndex = reorderIndex(from, insertBefore);
    if (finalIndex === from)
      return;
    const snapshot_10 = editor.getSnapshot(), rows_3 = getChildren(snapshot_10, path), origin = rows_3.at(from), destination = rows_3.at(finalIndex);
    origin && destination && editor.send({
      type: "custom.move.row",
      at: origin.path,
      to: destination.path
    });
  }, $[38] = editor, $[39] = path, $[40] = t13) : t13 = $[40];
  const commitRowDrag = t13;
  let t14;
  $[41] !== editor || $[42] !== path ? (t14 = (from_0, insertBefore_0) => {
    const finalIndex_0 = reorderIndex(from_0, insertBefore_0);
    if (finalIndex_0 === from_0)
      return;
    const snapshot_11 = editor.getSnapshot(), firstRow_4 = getChildren(snapshot_11, path).at(0), cells_1 = firstRow_4 ? getChildren(snapshot_11, firstRow_4.path) : [], origin_0 = cells_1.at(from_0), destination_0 = cells_1.at(finalIndex_0);
    origin_0 && destination_0 && editor.send({
      type: "custom.move.column",
      at: origin_0.path,
      to: destination_0.path
    });
  }, $[41] = editor, $[42] = path, $[43] = t14) : t14 = $[43];
  const commitColDrag = t14;
  let t15;
  $[44] !== commitColDrag || $[45] !== commitRowDrag || $[46] !== metrics || $[47] !== selectCol || $[48] !== selectRow ? (t15 = {
    tableRef,
    metrics,
    onCommitRow: commitRowDrag,
    onCommitCol: commitColDrag,
    onSelectRow: selectRow,
    onSelectCol: selectCol
  }, $[44] = commitColDrag, $[45] = commitRowDrag, $[46] = metrics, $[47] = selectCol, $[48] = selectRow, $[49] = t15) : t15 = $[49];
  const {
    drag,
    onHandlePointerDown
  } = useTableDragReorder(t15), activeDragKind = drag?.active ? drag.kind : null, activeDragIndex = drag?.active ? drag.index : null, dragKind = drag?.kind ?? null, dragIndex = drag?.index ?? null;
  let t16;
  $[50] !== activeDragIndex || $[51] !== activeDragKind ? (t16 = activeDragKind !== null && activeDragIndex !== null ? {
    kind: activeDragKind,
    index: activeDragIndex
  } : null, $[50] = activeDragIndex, $[51] = activeDragKind, $[52] = t16) : t16 = $[52];
  const dragContextValue = t16;
  let t17;
  if ($[53] !== dragIndex || $[54] !== dragKind || $[55] !== editor || $[56] !== path) {
    bb0: {
      if (dragKind === null || dragIndex === null) {
        t17 = null;
        break bb0;
      }
      const snapshot_12 = editor.getSnapshot(), rows_4 = getChildren(snapshot_12, path);
      if (dragKind === "row") {
        const row_2 = rows_4.at(dragIndex);
        if (!row_2) {
          t17 = null;
          break bb0;
        }
        t17 = getChildren(snapshot_12, row_2.path).map((cell_2) => getText(snapshot_12, cell_2.path) ?? "");
        break bb0;
      }
      t17 = rows_4.map((row_3) => {
        const cell_3 = getChildren(snapshot_12, row_3.path).at(dragIndex);
        return cell_3 ? getText(snapshot_12, cell_3.path) ?? "" : "";
      });
    }
    $[53] = dragIndex, $[54] = dragKind, $[55] = editor, $[56] = path, $[57] = t17;
  } else
    t17 = $[57];
  const ghostCellTexts = t17;
  let t18;
  $[58] !== metrics ? (t18 = (event) => {
    const rect = tableRef.current?.getBoundingClientRect();
    if (!rect || !metrics)
      return;
    const relY = event.clientY - rect.top, relX = event.clientX - rect.left;
    let row_4 = metrics.rows.findIndex((candidate) => relY >= candidate.top && relY < candidate.top + candidate.height), col = metrics.cols.findIndex((candidate_0) => relX >= candidate_0.left && relX < candidate_0.left + candidate_0.width);
    const lastRow_1 = metrics.rows.at(-1);
    row_4 === -1 && lastRow_1 && relY >= lastRow_1.top && relY <= metrics.height + EXTEND_LANE && (row_4 = metrics.rows.length - 1);
    const lastCol = metrics.cols.at(-1);
    col === -1 && lastCol && relX >= lastCol.left && relX <= metrics.width + EXTEND_LANE && (col = metrics.cols.length - 1), setHoverCell(row_4 >= 0 && col >= 0 ? {
      row: row_4,
      col
    } : null);
  }, $[58] = metrics, $[59] = t18) : t18 = $[59];
  const onMouseMove = t18;
  let t19, t20;
  $[60] === /* @__PURE__ */ Symbol.for("react.memo_cache_sentinel") ? (t19 = () => setActive(!0), t20 = () => {
    setActive(!1), setHoverCell(null);
  }, $[60] = t19, $[61] = t20) : (t19 = $[60], t20 = $[61]);
  const t21 = scrollWide ? "auto" : void 0, t22 = scrollWide ? SCROLL_PADDING_BOTTOM : void 0;
  let t23;
  $[62] !== t21 || $[63] !== t22 ? (t23 = {
    overflowX: t21,
    paddingBottom: t22
  }, $[62] = t21, $[63] = t22, $[64] = t23) : t23 = $[64];
  const t24 = scrollWide ? columnCount * MIN_COL_PX + 32 : void 0;
  let t25;
  $[65] !== t24 ? (t25 = {
    position: "relative",
    padding: "20px 20px 20px 12px",
    width: t24
  }, $[65] = t24, $[66] = t25) : t25 = $[66];
  const t26 = hasCellRange ? "" : void 0;
  let t27;
  $[67] !== columnCount ? (t27 = /* @__PURE__ */ jsx("colgroup", { contentEditable: !1, children: Array.from({
    length: columnCount
  }, _temp3) }), $[67] = columnCount, $[68] = t27) : t27 = $[68];
  let t28;
  $[69] !== children || $[70] !== dragContextValue ? (t28 = /* @__PURE__ */ jsx("tbody", { children: /* @__PURE__ */ jsx(TableDragContext.Provider, { value: dragContextValue, children }) }), $[69] = children, $[70] = dragContextValue, $[71] = t28) : t28 = $[71];
  let t29;
  $[72] !== t26 || $[73] !== t27 || $[74] !== t28 ? (t29 = /* @__PURE__ */ jsxs("table", { ref: tableRef, className: "pt-plugin-table", "data-pt-plugin-table-cell-range": t26, children: [
    t27,
    t28
  ] }), $[72] = t26, $[73] = t27, $[74] = t28, $[75] = t29) : t29 = $[75];
  let t30;
  $[76] !== active || $[77] !== drag || $[78] !== hoverCell || $[79] !== insertCol || $[80] !== insertRow || $[81] !== metrics || $[82] !== onHandlePointerDown || $[83] !== readOnly || $[84] !== resolvedLabels || $[85] !== selectCol || $[86] !== selectRow || $[87] !== selectedCol || $[88] !== selectedRow ? (t30 = readOnly ? null : /* @__PURE__ */ jsx(TableChrome, { labels: resolvedLabels, metrics, active, hoverCell, selectedRow, selectedCol, onHandlePointerDown, onHandleKeyboardSelect: (kind, index_4) => {
    kind === "row" ? selectRow(index_4) : selectCol(index_4);
  }, drag, onInsertRow: insertRow, onInsertCol: insertCol }), $[76] = active, $[77] = drag, $[78] = hoverCell, $[79] = insertCol, $[80] = insertRow, $[81] = metrics, $[82] = onHandlePointerDown, $[83] = readOnly, $[84] = resolvedLabels, $[85] = selectCol, $[86] = selectRow, $[87] = selectedCol, $[88] = selectedRow, $[89] = t30) : t30 = $[89];
  let t31;
  $[90] !== t25 || $[91] !== t29 || $[92] !== t30 ? (t31 = /* @__PURE__ */ jsxs("div", { className: "pt-plugin-table-inner", style: t25, children: [
    t29,
    t30
  ] }), $[90] = t25, $[91] = t29, $[92] = t30, $[93] = t31) : t31 = $[93];
  let t32;
  $[94] !== t23 || $[95] !== t31 ? (t32 = /* @__PURE__ */ jsx("div", { ref: scrollRef, className: "pt-plugin-table-scroll", style: t23, children: t31 }), $[94] = t23, $[95] = t31, $[96] = t32) : t32 = $[96];
  let t33;
  $[97] !== drag || $[98] !== ghostCellTexts || $[99] !== hasHeader || $[100] !== metrics || $[101] !== readOnly ? (t33 = readOnly ? null : /* @__PURE__ */ jsx(ReorderGhost, { drag, metrics, hasHeader, cellTexts: ghostCellTexts }), $[97] = drag, $[98] = ghostCellTexts, $[99] = hasHeader, $[100] = metrics, $[101] = readOnly, $[102] = t33) : t33 = $[102];
  let t34;
  $[103] !== active || $[104] !== deleteTable || $[105] !== hasHeader || $[106] !== menuOpen || $[107] !== portalElement || $[108] !== readOnly || $[109] !== renderMenu || $[110] !== resolvedLabels || $[111] !== scrollWide || $[112] !== selectTable || $[113] !== toggleHeader || $[114] !== tokenStyle ? (t34 = readOnly ? null : renderMenu ? /* @__PURE__ */ jsx(TableMenuAnchor, { right: scrollWide ? 0 : 20, visible: active || menuOpen, children: renderMenu({
    hasHeader,
    onToggleHeader: toggleHeader,
    onSelectTable: selectTable,
    onDeleteTable: deleteTable,
    onOpenChange: setMenuOpen
  }) }) : /* @__PURE__ */ jsx(TableMenu, { labels: resolvedLabels, tokenStyle, right: scrollWide ? 0 : 20, active, portalElement, handlers: {
    hasHeader,
    onToggleHeader: toggleHeader,
    onSelectTable: selectTable,
    onDeleteTable: deleteTable
  } }), $[103] = active, $[104] = deleteTable, $[105] = hasHeader, $[106] = menuOpen, $[107] = portalElement, $[108] = readOnly, $[109] = renderMenu, $[110] = resolvedLabels, $[111] = scrollWide, $[112] = selectTable, $[113] = toggleHeader, $[114] = tokenStyle, $[115] = t34) : t34 = $[115];
  let t35;
  $[116] !== fade.left || $[117] !== fade.right ? (t35 = /* @__PURE__ */ jsx(TableScrollFade, { left: fade.left, right: fade.right }), $[116] = fade.left, $[117] = fade.right, $[118] = t35) : t35 = $[118];
  let t36;
  $[119] !== columnCount || $[120] !== deleteCol || $[121] !== deleteRow || $[122] !== icons?.trash || $[123] !== metrics || $[124] !== portalElement || $[125] !== readOnly || $[126] !== resolvedLabels || $[127] !== rowCount || $[128] !== selectedCol || $[129] !== selectedRow || $[130] !== tokenStyle ? (t36 = readOnly ? null : /* @__PURE__ */ jsx(TableTrashLayer, { labels: resolvedLabels, tokenStyle, tableRef, metrics, portalElement, trashIcon: icons?.trash, selectedRow, selectedCol, canDeleteRow: rowCount > 1, canDeleteCol: columnCount > 1, onDeleteRow: deleteRow, onDeleteCol: deleteCol }), $[119] = columnCount, $[120] = deleteCol, $[121] = deleteRow, $[122] = icons?.trash, $[123] = metrics, $[124] = portalElement, $[125] = readOnly, $[126] = resolvedLabels, $[127] = rowCount, $[128] = selectedCol, $[129] = selectedRow, $[130] = tokenStyle, $[131] = t36) : t36 = $[131];
  let t37;
  $[132] !== onMouseMove || $[133] !== t32 || $[134] !== t33 || $[135] !== t34 || $[136] !== t35 || $[137] !== t36 || $[138] !== tokenStyle ? (t37 = /* @__PURE__ */ jsxs("div", { className: "pt-plugin-table-chrome", style: tokenStyle, onMouseEnter: t19, onMouseLeave: t20, onMouseMove, children: [
    t32,
    t33,
    t34,
    t35,
    t36
  ] }), $[132] = onMouseMove, $[133] = t32, $[134] = t33, $[135] = t34, $[136] = t35, $[137] = t36, $[138] = tokenStyle, $[139] = t37) : t37 = $[139];
  let t38;
  return $[140] !== attributes || $[141] !== t37 ? (t38 = /* @__PURE__ */ jsx("div", { ...attributes, children: t37 }), $[140] = attributes, $[141] = t37, $[142] = t38) : t38 = $[142], t38;
}
function _temp3(_, index_3) {
  return /* @__PURE__ */ jsx("col", {}, index_3);
}
function _temp2(snapshot) {
  return snapshot.context.readOnly;
}
function _temp(rowCell) {
  return rowCell._key;
}
function TableRow(t0) {
  const $ = c(4), {
    attributes,
    children,
    selected
  } = t0, t1 = selected ? "" : void 0;
  let t2;
  return $[0] !== attributes || $[1] !== children || $[2] !== t1 ? (t2 = /* @__PURE__ */ jsx("tr", { ...attributes, "data-pt-plugin-table-selected": t1, children }), $[0] = attributes, $[1] = children, $[2] = t1, $[3] = t2) : t2 = $[3], t2;
}
function TableCell(t0) {
  const $ = c(15), {
    attributes,
    children,
    node,
    path
  } = t0, editor = useEditor();
  let t1;
  $[0] !== node._type ? (t1 = resolveTableConfig(node._type), $[0] = node._type, $[1] = t1) : t1 = $[1];
  const config = t1;
  let t2;
  $[2] !== config || $[3] !== path ? (t2 = (snapshot) => {
    const {
      isCell,
      isRow,
      isTable
    } = createTableGuards(config), cell = getEnclosingBlock(snapshot, path, {
      match: isCell
    }), row = cell && getParent(snapshot, cell.path, {
      match: isRow
    }), table = row && getParent(snapshot, row.path, {
      match: isTable
    });
    if (!cell || !row || !table)
      return null;
    const tableRowsList = tableRows(config, table.node), rowIdx = tableRowsList.findIndex((candidate) => candidate._key === row.node._key), colIdx = rowCells(config, row.node).findIndex((candidate_0) => candidate_0._key === cell.node._key), selection = getTableSelection(snapshot, config), range = selection && isEqualPaths(selection.tablePath, table.path) ? {
      r0: selection.rowRange[0],
      r1: selection.rowRange[1],
      c0: selection.colRange[0],
      c1: selection.colRange[1]
    } : null, firstRow = tableRowsList[0];
    return {
      rowIdx,
      colIdx,
      rowCount: tableRowsList.length,
      colCount: firstRow ? rowCells(config, firstRow).length : 0,
      isHeader: (Number(table.node.headerRows) || 0) >= 1 && rowIdx === 0,
      range
    };
  }, $[2] = config, $[3] = path, $[4] = t2) : t2 = $[4];
  const descriptor = useEditorSelector(editor, t2, isEqualCellDescriptor), dragContext = useContext(TableDragContext), dimmed = descriptor !== null && dragContext !== null && (dragContext.kind === "row" ? descriptor.rowIdx === dragContext.index : descriptor.colIdx === dragContext.index);
  let t3;
  $[5] !== descriptor ? (t3 = descriptor ? cellStyle(descriptor) : void 0, $[5] = descriptor, $[6] = t3) : t3 = $[6];
  const style = t3, t4 = descriptor?.isHeader ? "" : void 0;
  let t5;
  $[7] !== dimmed || $[8] !== style ? (t5 = dimmed ? {
    ...style,
    opacity: 0.35
  } : style, $[7] = dimmed, $[8] = style, $[9] = t5) : t5 = $[9];
  let t6;
  return $[10] !== attributes || $[11] !== children || $[12] !== t4 || $[13] !== t5 ? (t6 = /* @__PURE__ */ jsx("td", { ...attributes, "data-pt-plugin-table-header": t4, style: t5, children }), $[10] = attributes, $[11] = children, $[12] = t4, $[13] = t5, $[14] = t6) : t6 = $[14], t6;
}
function cellStartPoint(snapshot, cellPath) {
  const firstChild = getFirstChild(snapshot, cellPath), block = firstChild && getBlock(snapshot, firstChild.path);
  return block ? getBlockStartPoint({
    context: snapshot.context,
    block
  }) : null;
}
function isEqualTableSelection(a, b) {
  return a === b ? !0 : !a || !b ? !1 : a.rowRange[0] === b.rowRange[0] && a.rowRange[1] === b.rowRange[1] && a.colRange[0] === b.colRange[0] && a.colRange[1] === b.colRange[1] && isEqualPaths(a.tablePath, b.tablePath);
}
function isEqualCellDescriptor(a, b) {
  return a === b ? !0 : !a || !b ? !1 : a.rowIdx === b.rowIdx && a.colIdx === b.colIdx && a.rowCount === b.rowCount && a.colCount === b.colCount && a.isHeader === b.isHeader && isEqualCellRange(a.range, b.range);
}
function isEqualCellRange(a, b) {
  return a === b ? !0 : !a || !b ? !1 : a.r0 === b.r0 && a.r1 === b.r1 && a.c0 === b.c0 && a.c1 === b.c1;
}
const referenceContainers = {
  table: defineContainer({
    type: "table",
    arrayField: "rows",
    render: (props) => /* @__PURE__ */ jsx(Table, { ...props })
  }),
  row: defineContainer({
    type: "row",
    arrayField: "cells",
    render: (props) => /* @__PURE__ */ jsx(TableRow, { ...props })
  }),
  cell: defineContainer({
    type: "cell",
    arrayField: "value",
    render: (props) => /* @__PURE__ */ jsx(TableCell, { ...props })
  })
};
export {
  Table,
  TableCell,
  TableRow,
  referenceContainers
};
