// @sanity/workbench-cli/dist/actions/deploy/deployWorkbenchApp.js
import { basename, dirname } from 'node:path';
import { createGzip } from 'node:zlib';
import { spinner } from '@sanity/cli-core/ux';
import { pack } from 'tar-fs';
import { createApplication, createDeployment, deleteApplication, updateApplication } from '../../services/applications.js';
/**
 * Create a coreApp record (no deployment), so the CLI can build with its id
 * before shipping the first deployment. First deploy only.
 * @internal
 */ export async function createCoreApp(options) {
    const spin = spinner('Creating application...').start();
    try {
        const { id } = await createApplication({
            ...options,
            type: 'coreApp'
        });
        spin.succeed();
        return {
            applicationId: id,
            rollback: ()=>deleteApplication(id)
        };
    } catch (error) {
        spin.fail();
        throw error;
    }
}
/**
 * Create a studio record (no deployment).
 * @internal
 */ export async function createStudio(options) {
    const spin = spinner('Creating studio...').start();
    try {
        const { id } = await createApplication({
            ...options,
            type: 'studio'
        });
        spin.succeed();
        return {
            applicationId: id,
            rollback: ()=>deleteApplication(id)
        };
    } catch (error) {
        spin.fail();
        throw error;
    }
}
/**
 * Ship a deployment to an already-created (or `deployment.appId`) application,
 * then sync its mutable metadata (`title`, and `icon`/`visibility` when set)
 * from config. The deploy endpoint ignores these, so a redeploy patches them
 * here alongside the new deployment.
 *
 * `onDeployed` fires the instant the deployment is live, before the metadata
 * sync — so a caller can disarm a create-time rollback that must not delete an
 * application once it has an active deployment.
 * @internal
 */ export async function deployWorkbenchApp(options) {
    const { applicationId, icon, interfaces, isAutoUpdating, label = 'Deploying...', onDeployed, sourceDir, title, version, visibility, workspaces } = options;
    const tarball = pack(dirname(sourceDir), {
        entries: [
            basename(sourceDir)
        ]
    }).pipe(createGzip());
    const spin = spinner(label).start();
    try {
        await createDeployment({
            applicationId,
            interfaces,
            isAutoUpdating,
            tarball,
            version,
            workspaces
        });
        onDeployed?.();
        await updateApplication(applicationId, {
            title,
            ...icon ? {
                icon
            } : {},
            ...visibility ? {
                visibility
            } : {}
        });
        spin.succeed();
    } catch (error) {
        spin.clear();
        throw error;
    }
}
