// @sanity/workbench/src/core/applications/application-list.ts
import type { CanvasApplication } from "../canvases";
import type { MediaLibraryApplication } from "../media-libraries";
import { brandApplicationId, type ApplicationId } from "./brett-application";
import type { CoreApplication } from "./core-app";
import type { StudioApplication, StudioWorkspace } from "./studios";

/**
 * @public
 */
export type Application =
  | CoreApplication
  | StudioApplication
  | StudioWorkspace
  | CanvasApplication
  | MediaLibraryApplication;

/**
 * @public
 */
export class ApplicationList<TApplication extends Application = Application> {
  readonly applications: TApplication[];

  /**
   * @param applications - The applications to initialize the list with.
   */
  constructor(applications: TApplication[]) {
    this.applications = [...applications];
  }

  findApplicationsByType<T extends TApplication["type"]>(
    type: T,
  ): Extract<TApplication, { type: T }>[];
  findApplicationsByType<T extends TApplication["type"][]>(
    types: T,
  ): Extract<TApplication, { type: T[number] }>[];
  findApplicationsByType(
    _type: TApplication["type"] | TApplication["type"][],
  ): TApplication[] {
    const types = Array.isArray(_type) ? _type : [_type];

    return this.applications.filter((r) => types.includes(r.type));
  }

  findApplication(
    type: "media-library",
  ): Extract<TApplication, { type: "media-library" }> | undefined;
  findApplication(
    type: "canvas",
  ): Extract<TApplication, { type: "canvas" }> | undefined;
  findApplication(
    type: "coreApp",
    id: ApplicationId,
  ): Extract<TApplication, { type: "coreApp" }> | undefined;
  findApplication(
    type: "studio",
    id: ApplicationId,
  ): Extract<TApplication, { type: "studio" }> | undefined;
  findApplication(
    type: "workspace",
    id: string,
  ): Extract<TApplication, { type: "workspace" }> | undefined;
  findApplication(
    type: TApplication["type"],
    id?: ApplicationId | string,
  ): TApplication | undefined {
    switch (type) {
      case "media-library":
      case "canvas":
        return this.applications.find((r) => r.type === type);
      case "coreApp":
      case "studio":
      case "workspace": {
        return this.applications.find((r) => r.type === type && r.id === id);
      }
    }
  }

  /**
   * The one application a location puts in the foreground — media library and
   * canvas by route (one per org), applications by the id in the URL.
   * `undefined` when nothing matches or the match has no app view (not
   * navigable, e.g. an SDK app that only docks panels).
   */
  findByPath(pathname: string): TApplication | undefined {
    const [root, id] = pathname.split("/").filter(Boolean);
    const application = (() => {
      switch (root) {
        case "media":
          return this.findApplication("media-library");
        case "canvas":
          return this.findApplication("canvas");
        case "studio":
          return id
            ? this.findApplication("studio", brandApplicationId(id))
            : undefined;
        case "application":
          return id
            ? this.findApplication("coreApp", brandApplicationId(id))
            : undefined;
        case "local":
          // A local app can share its id with its deployed twin; the /local
          // route always renders the local instance.
          return this.applications.find(
            (candidate) =>
              candidate.isLocal &&
              (candidate.type === "studio" || candidate.type === "coreApp") &&
              candidate.id === id,
          );
        default:
          return undefined;
      }
    })();

    return application?.hasAppView ? application : undefined;
  }

  static areApplicationsEqual(
    application1?: Application,
    application2?: Application,
  ): boolean {
    if (
      !application1 ||
      !application2 ||
      application1.type !== application2.type
    ) {
      return false;
    }

    return application1.id === application2.id;
  }
}
