// @sanity/workbench/src/core/applications/application.ts
import { z } from "zod";

import {
  type AppConfig,
  AppConfigSchema,
  type LocalAppConfig,
  type ResolvedAppConfig,
  resolveAppConfig,
} from "./config";
import type { LocalCoreApplicationManifest } from "./core-app";
import {
  type DockGroup,
  type Interface,
  type InterfaceOfType,
  InterfaceSchema,
  type LocalInterface,
  LocalInterfaceSchema,
  type Panel,
  type PanelComponent,
} from "./interface";
import type { LocalStudioApplicationManifest } from "./studios/schemas";

// Re-export the wire model (interfaces + app config) so it stays part
// of the workbench app-model surface; the authoring half lives with the CLI.
export {
  type AppConfig,
  AppConfigSchema,
  type DockGroup,
  type Interface,
  type InterfaceOfType,
  InterfaceSchema,
  type LocalAppConfig,
  type LocalInterface,
  LocalInterfaceSchema,
  type Panel,
  type PanelComponent,
  resolveAppConfig,
  type ResolvedAppConfig,
};

const LocalApplicationBase = z.object({
  host: z.string(),
  port: z.number(),
  /**
   * Interfaces the app exposes — dock panels (`type: "panel"`) and
   * background workers (`type: "worker"`), forwarded from the dev
   * server in one list.
   */
  interfaces: z.array(LocalInterfaceSchema).optional(),
  /** The `deployment.appId` from the application's `sanity.cli.ts`, when set. */
  id: z.string().optional(),
  /**
   * The `api.projectId` from the application's `sanity.cli.ts`. Available
   * synchronously at dev-server startup (no manifest extraction required), so
   * the studio's primary project is resolvable from the very first local
   * application event.
   */
  projectId: z.string().optional(),
});

/**
 * Raw data for a local application discovered by the CLI dev server.
 *
 * The CLI forwards the full studio or app manifest so the workbench can
 * derive icons, titles, workspaces and schema references the same way it
 * does for deployed applications. The manifest shape is discriminated on
 * `type`: studios receive a `LocalStudioApplicationManifest`, core apps a `LocalCoreApplicationManifest`.
 *
 * When set on an application, `isLocal` is `true` and `url`/`isFederated`
 * honour the local dev-server instead of the deployed application's host.
 *
 * @public
 */
export const LocalApplication = z.discriminatedUnion("type", [
  LocalApplicationBase.extend({
    type: z.literal("studio"),
    manifest: z.custom<LocalStudioApplicationManifest>().optional(),
  }),
  LocalApplicationBase.extend({
    type: z.literal("coreApp"),
    manifest: z.custom<LocalCoreApplicationManifest>().optional(),
  }),
]);

/**
 * @public
 */
export type LocalApplication = z.output<typeof LocalApplication>;

/**
 * @public
 */
export type AbstractApplicationType =
  | "studio"
  | "coreApp"
  | "canvas"
  | "media-library"
  | "workspace";

/**
 * @public
 */
export abstract class AbstractApplication<
  TType extends AbstractApplicationType,
> {
  readonly type: TType;

  constructor(type: TType) {
    this.type = type;
  }

  /**
   * The in-app route this application navigates to, or `null` when it isn't
   * navigable as a full-page app (US5, spec 002-workbench-extension-api — e.g. an SDK app with no `app` view).
   * The dock renders a navigable item only for a non-null `href`; the app
   * routes 404 a `null`-href app on direct visit.
   */
  abstract get href(): string | null;

  abstract get title(): string;

  abstract get id(): string;

  /**
   * Whether the application is federated or not. This is used to determine
   * if the application should be rendered in an iframe or not.
   */
  abstract get isFederated(): boolean;

  /**
   * Whether the application is served by a local CLI dev server rather than a
   * deployed remote. Defaults to `false`; applications flip this when
   * constructed from a `LocalApplication` payload.
   */
  get isLocal(): boolean {
    return false;
  }

  abstract get url(): URL;

  /**
   * Interfaces the application exposes (e.g. dock panels) — all of them, or
   * only those of the given `type`. Defaults to none; applications
   * surface what they declared via `unstable_defineApp`.
   */
  interfaces(): readonly Interface[];
  interfaces<T extends Interface["type"]>(
    type: T,
  ): readonly InterfaceOfType<T>[];
  interfaces(_type?: Interface["type"]): readonly Interface[] {
    return [];
  }

  get config(): ResolvedAppConfig | undefined {
    throw new Error("App configs are not supported yet");
  }

  /**
   * Whether the application has a navigable full-page `app` view (US5, spec 002-workbench-extension-api). Defaults
   * to `true` — studios, Canvas, and Media Library are always navigable. An SDK
   * app overrides this to derive it from whether it declares an `app` interface.
   */
  get hasAppView(): boolean {
    return true;
  }

  /**
   * Federation module id of the app's full-page view, or `null` when it can't
   * be federation-loaded: the app isn't federated (Canvas, Media Library,
   * deployed apps shown in an iframe) or has no app view. Only a non-null
   * `moduleId` is fetched and rendered by the remotes machine.
   */
  get moduleId(): string | null {
    return this.isFederated && this.hasAppView ? `${this.id}/App` : null;
  }

  /**
   * Federation module id of one of this app's panel view components, or `null`
   * when the app isn't federated (a non-federated app's panels can't be loaded
   * as remotes).
   */
  resolveViewModuleId(view: Panel, component: PanelComponent): string | null {
    return this.isFederated
      ? `${this.id}/views/${view.name}/${component}`
      : null;
  }

  get initials(): string {
    const SYMBOLS = /[^\p{Alpha}\p{N}\p{White_Space}]/gu;
    const WHITESPACE = /\p{White_Space}+/u;
    const ALPHANUMERIC_SEGMENTS = /(\p{N}+|\p{Alpha}+)/gu;
    const IS_NUMERIC = /^\p{N}+$/u;

    if (!this.title) return "";

    const namesArray = this.title
      .replace(SYMBOLS, "")
      .split(WHITESPACE)
      .filter(Boolean);

    if (namesArray.length === 0) return "";

    if (namesArray.length === 1) {
      const word = namesArray[0];
      const segments = word.match(ALPHANUMERIC_SEGMENTS) || [];

      if (segments.length === 0) return "";
      if (segments.length === 1) {
        if (word.length === 1) {
          return word.toUpperCase();
        }

        if (IS_NUMERIC.test(word)) {
          return `${word.charAt(0)}${word.charAt(1)}`.toUpperCase();
        }

        return word.charAt(0).toUpperCase();
      }

      return `${segments[0]!.charAt(0)}${segments[1].charAt(0)}`.toUpperCase();
    }

    return `${namesArray[0].charAt(0)}${namesArray[namesArray.length - 1].charAt(0)}`.toUpperCase();
  }
}
