// @sanity/workbench/src/core/applications/config.ts
import { z } from "zod";

// A field's `defineField` value can't serialize, so a media-library config
// splits in two: metadata over the wire, live values from the federation
// module, merged back by `name`.
const MediaLibraryConfigFieldSchema = z.object({
  name: z.string(),
  title: z.string(),
  public: z.boolean().optional(),
});

const MediaLibraryConfigModuleSchema = z.object({
  config: z.object({
    fields: z.array(z.object({ name: z.string(), config: z.unknown() })),
  }),
});

/**
 * An app config tagged by the `appType` that owns it; `config` holds the
 * type-specific payload. A discriminated union — different app types carry
 * differently-shaped configs.
 * @public
 */
export const AppConfigSchema = z.discriminatedUnion("appType", [
  z.object({
    appType: z.literal("media-library"),
    config: z.object({
      fields: z.array(MediaLibraryConfigFieldSchema),
    }),
  }),
]);

/** @public */
export type AppConfig = z.output<typeof AppConfigSchema>;

/**
 * An app config a local dev server forwards, plus how to reach the
 * app's federation remote: `moduleName` is its module-federation alias and
 * `remoteURL` its origin (e.g. `http://localhost:3337`).
 * @public
 */
export type LocalAppConfig = AppConfig & {
  /** Content hash the CLI derives — the change-detection key. */
  id: string;
  moduleName: string;
  remoteURL: string;
};

/**
 * An app config with each field's live value attached. `appType` stays the
 * union of known types, like its wire counterpart.
 * @public
 */
export type ResolvedAppConfig = {
  appType: AppConfig["appType"];
  fields: (z.output<typeof MediaLibraryConfigFieldSchema> & {
    config?: unknown;
  })[];
};

/**
 * Merge a remote's federation-module values into its wire config by field
 * name. Every app type's payload is field-shaped, so one merge covers the
 * whole union — a future type whose payload diverges breaks the
 * `config.fields` access below and forces its own branch. Returns `null`
 * when the module doesn't match the expected shape.
 * @public
 */
export function resolveAppConfig(
  appConfig: AppConfig,
  module: unknown,
): ResolvedAppConfig | null {
  const parsed = MediaLibraryConfigModuleSchema.safeParse(module);
  if (!parsed.success) return null;

  const valuesByName = new Map(
    parsed.data.config.fields.map((field) => [field.name, field.config]),
  );
  return {
    appType: appConfig.appType,
    fields: appConfig.config.fields.map((field) => ({
      ...field,
      config: valuesByName.get(field.name),
    })),
  };
}
