// @sanity/workbench/src/core/applications/core-app.ts
import { z } from "zod";

import {
  BrettApplication,
  ExternalApplication,
  InternalApplication,
} from "./brett-application";

/**
 * @public
 */
export const LocalCoreApplicationManifest = z.object({
  version: z.string(),
  icon: z.string().optional(),
  title: z.string().optional(),
});

/**
 * @public
 */
export type LocalCoreApplicationManifest = z.output<
  typeof LocalCoreApplicationManifest
>;

const internalCoreApplication = InternalApplication.extend({
  type: z.literal("coreApp"),
  title: z.string(),
  config: z.object({ mfManifest: z.unknown().optional() }).optional(),
  manifest: LocalCoreApplicationManifest.nullable().optional(),
});

const externalCoreApplication = ExternalApplication.extend({
  type: z.literal("coreApp"),
  title: z.string(),
  config: z.object({ mfManifest: z.unknown().optional() }).optional(),
  manifest: LocalCoreApplicationManifest.nullable().optional(),
});

/**
 * A Sanity-hosted SDK app (`slug` set, `externalUrl: null`) — the only kind
 * brett hosts.
 * @public
 */
export type InternalCoreApplication = z.output<typeof internalCoreApplication>;

/**
 * An externally served SDK app (`externalUrl` set, `slug: null`). Never on
 * the wire — brett doesn't host external SDK apps — but a local CLI dev
 * server serves its app from `http://localhost`.
 * @public
 */
export type ExternalCoreApplication = z.output<typeof externalCoreApplication>;

/**
 * An SDK app from `GET /applications?type=coreApp`, brett's shape verbatim.
 * `manifest` is the one addition: the channel a local CLI dev server forwards
 * its extracted manifest on — never present on the wire.
 * @public
 */
export type CoreApp = InternalCoreApplication | ExternalCoreApplication;

/**
 * Validates and parses a raw API response into a branded
 * CoreApp. Parses the internal variant only — brett doesn't host
 * external SDK apps.
 * @public
 */
export function parseCoreApplication(data: unknown): CoreApp {
  return internalCoreApplication.parse(data);
}

/**
 * @public
 */
export class CoreApplication extends BrettApplication<CoreApp, "coreApp"> {
  readonly activeDeployment: CoreApp["activeDeployment"];

  constructor(
    application: CoreApp,
    options: {
      isLocal?: boolean;
      remoteApplication?: CoreApplication | null;
    } = {},
  ) {
    super(application, "coreApp", options);

    this.activeDeployment = application.activeDeployment;
  }

  // Typed `string | null` so the UI-aware remote subclass can override it to
  // `null` for a non-navigable app (US5, spec 002-workbench-extension-api); the core route itself is always set.
  get href(): string | null {
    return this.isLocal
      ? `/local/${this.id}`
      : `/application/${this.application.id}`;
  }

  get title() {
    return this.manifest?.title ?? this.application.title;
  }

  /**
   * The manifest a local CLI dev server forwards. Deployed apps carry none —
   * their dock metadata rides the `app` interface (see {@link group} and
   * {@link priority}), and icons have no slot on brett's payload yet, so
   * deployed apps fall back to the avatar.
   */
  get manifest(): LocalCoreApplicationManifest | null {
    return this.application.manifest ?? null;
  }

  get subtitle() {
    return undefined;
  }

  get updatedAt() {
    return this.application.updatedAt;
  }

  get<TKey extends keyof CoreApp>(attr: TKey): CoreApp[TKey] {
    if (!(attr in this.application)) {
      throw new Error(
        `Attribute ${attr.toString()} does not exist on application ${this.application.id}`,
      );
    }

    return this.application[attr];
  }
}
