// @sanity/workbench/src/core/applications/interface.ts
import { z } from "zod";

/**
 * The interface records the workbench renders from — brett's deployment shape: a
 * discriminating `type`, a federation `moduleId`, and (on `app`) dock
 * `metadata`. A local dev server forwards the same records plus the `src` file
 * it serves (see {@link LocalInterface}); the create-from-local helpers strip
 * `src` to read them as one shape. The authoring half (`unstable_defineView` /
 * `unstable_defineService`) lives with the CLI build helpers.
 */

const InterfaceBase = z.object({
  id: z.string(),
  name: z.string(),
  title: z.string(),
  version: z.string().optional(),
  moduleId: z.string(),
});

const AppInterfaceMetadata = z
  .object({ group: z.string().optional(), priority: z.number().optional() })
  .nullable();

const PanelInterfaceSchema = InterfaceBase.extend({ type: z.literal("panel") });
const WorkerInterfaceSchema = InterfaceBase.extend({
  type: z.literal("worker"),
});

// Only the navigable `app` interface docks today, so it alone carries
// `metadata`; adding it to another type is a per-type opt-in here.
const AppInterfaceSchema = InterfaceBase.extend({
  type: z.literal("app"),
  metadata: AppInterfaceMetadata,
});

/**
 * The shape the workbench renders from — a `panel` renders in the dock, a
 * `worker` runs a background service, an `app` is the navigable full-page view.
 * @public
 */
export const InterfaceSchema = z.discriminatedUnion("type", [
  PanelInterfaceSchema,
  WorkerInterfaceSchema,
  AppInterfaceSchema,
]);

/** @public */
export type Interface = z.output<typeof InterfaceSchema>;

/** The interface record for a given `type`. @public */
export type InterfaceOfType<T extends Interface["type"]> = Extract<
  Interface,
  { type: T }
>;

/** @public */
export type Panel = InterfaceOfType<"panel">;

/**
 * The interface a local CLI dev server forwards — a deployed interface plus the
 * `src` file it serves. The create-from-local helpers strip `src` to read it as
 * an {@link Interface}.
 * @public
 */
export const LocalInterfaceSchema = z.discriminatedUnion("type", [
  PanelInterfaceSchema.extend({ src: z.string() }),
  WorkerInterfaceSchema.extend({ src: z.string() }),
  AppInterfaceSchema.extend({ src: z.string() }),
]);

/** @public */
export type LocalInterface = z.output<typeof LocalInterfaceSchema>;

/**
 * A panel's view-component slots, in render order — each its own federation
 * island (`${id}/views/${name}/${component}`).
 * @public
 */
export type PanelComponent = "title" | "panel";

/**
 * Where an application docks — the `group` value the workbench reads to place an
 * app in the dock.
 * @public
 */
export type DockGroup = "dock.system" | "dock.applications" | "dock.user";
