// @sanity/workbench/src/runtime/topics.ts
import type { CurrentUser, OrganizationBase } from "@sanity/sdk";

import type { Application } from "../core/applications/application-list";
import type { ResolvedAppConfig } from "../core/applications/config";
import type { SystemPreferencesContext } from "../system/system-preferences.machine";

/**
 * Declares a state topic: holds a current value, replayed to new subscribers.
 * @public
 */
export type StateTopicDef<T> = { kind: "state"; value: T };

/**
 * Declares an event topic: a stream of occurrences with no memory. Declaring a
 * `reply` makes it a request topic — an awaited `emit` resolves with that reply.
 * @public
 */
export type EventTopicDef<P, R = never> = {
  kind: "event";
  payload: P;
  reply?: R;
};

/**
 * The value of a state topic that can fail independently of the session. One
 * `ok` discriminant across all such topics, so consumers can share failure
 * handling (e.g. a retry utility) without per-topic shapes.
 * @public
 */
export type TopicResult<T> = { ok: true; value: T } | { ok: false };

/**
 * The topics the workbench itself owns. Every topic lands here with its owner's
 * publish wiring and, if bridge-bound, a clonability assertion in
 * `topics.test.ts`. A session-derived value must not outlive the session: such
 * topics clear (publish `null`) when the user signs out.
 * @public
 */
export interface WorkbenchTopics {
  /**
   * The application the URL currently puts in the foreground, as a bridgeable
   * summary; `null` when none is — including before the location and the
   * application list are known, and once the session ends. Seeded `null`, then
   * publishes only on actual changes: in-app navigation stays silent.
   * Published by the applications machine.
   */
  "applications.foreground": StateTopicDef<Pick<
    Application,
    "id" | "title" | "type"
  > | null>;
  /**
   * The applications available in this workbench, as bridgeable summaries —
   * `ok: false` when the list cannot be loaded. Suspends until the first
   * application sync; `null` once the session ends. Published by the
   * applications machine.
   */
  "applications.list": StateTopicDef<TopicResult<
    readonly Pick<Application, "id" | "title" | "type">[]
  > | null>;
  /**
   * The session token the requesting app reads; `null` while signed out. Each
   * connection carries its app id, so per-app tokens can land later without
   * changing the contract — today every app reads the same session token.
   * Read-only for apps: published by the auth machine, cleared to `null` on
   * sign-out. Suspends until auth settles.
   */
  "auth.token": StateTopicDef<string | null>;
  /**
   * Requests the session token on demand, for an app that wants to re-fetch
   * rather than wait on its `auth.token` subscription (e.g. after a rejected
   * request). Carries the caller's app id. Today it resends the session's
   * current token — the seam where per-app re-issuance lands later. Fails
   * `NO_RESPONDER` while signed out.
   */
  "auth.token.refresh": EventTopicDef<void, string>;
  /**
   * The media library's resolved deployment configuration, published whole: the
   * field metadata together with each field's live `defineField` value —
   * `value: null` when no config is deployed, `ok: false` when resolution
   * failed. Suspends until the host publishes it; `null` once the session ends.
   * In-process only — it carries non-serializable code, so it has no
   * clonability assertion in `topics.test.ts`. Published by the applications
   * machine.
   */
  "media-libraries.config": StateTopicDef<TopicResult<ResolvedAppConfig | null> | null>;
  /**
   * The session's organization context; `null` when signed out or the session
   * failed. Resolved on login — everything else is fetched against it. The
   * bridgeable projection of the SDK's `OrganizationBase`. Suspends until auth
   * settles; published by the auth machine.
   */
  "organizations.current": StateTopicDef<Pick<
    OrganizationBase,
    "id" | "name" | "slug"
  > | null>;
  /**
   * The open panel — which app owns it and its mode; `value: null` when
   * closed, including at boot; `null` once the authed shell unmounts. The
   * value carries the owning app id: several apps can own panels, and a
   * sender's app id is matched against it — an app only ever changes its own
   * panel. `ok: false` signals a mode the panel cannot honor — e.g. a view
   * that forbids expanding to `full`. Seeded, then publishes only on actual
   * changes. Published by the applications machine.
   */
  "panels.mode": StateTopicDef<TopicResult<{
    appId: string;
    mode: "aside" | "full";
  } | null> | null>;
  /**
   * The resolved color scheme the workbench renders with — the user's
   * preference, falling back to the OS scheme. Environment-derived, not
   * session data: it stays published across sign-out (the login screen renders
   * with it too). Seeded synchronously at boot; published by the
   * system-preferences machine.
   */
  "preferences.color-scheme": StateTopicDef<
    SystemPreferencesContext["colorScheme"]
  >;
  /**
   * The signed-in user, or `null` when signed out. The bridgeable projection of
   * the SDK's `CurrentUser`. Suspends until auth settles; published by the auth
   * machine.
   */
  "users.current": StateTopicDef<Pick<
    CurrentUser,
    "id" | "name" | "email" | "profileImage"
  > | null>;
}

/**
 * The central topic registry — every call site is type-checked against it.
 * @public
 */
export interface Topics extends WorkbenchTopics {}

/**
 * Every declared topic name.
 * @public
 */
export type TopicName = keyof Topics;

/**
 * Names of all declared state topics.
 * @public
 */
export type StateTopic = {
  [K in TopicName]: Topics[K] extends { kind: "state" } ? K : never;
}[TopicName];

/**
 * Names of all declared event topics.
 * @public
 */
export type EventTopic = {
  [K in TopicName]: Topics[K] extends { kind: "event" } ? K : never;
}[TopicName];

/**
 * The value type of a state topic.
 * @public
 */
export type ValueOf<K extends StateTopic> =
  Topics[K] extends StateTopicDef<infer T> ? T : never;

/**
 * The payload type of an event topic.
 * @public
 */
export type PayloadOf<K extends EventTopic> =
  Topics[K] extends EventTopicDef<infer P, infer _R> ? P : never;

/**
 * The reply type of an event topic (`never` if it declares none).
 * @public
 */
export type ReplyOf<K extends EventTopic> =
  Topics[K] extends EventTopicDef<infer _P, infer R> ? R : never;

/**
 * A value that survives `structuredClone` — a function can ride an in-process
 * topic, but it can never cross the iframe bridge. Bridge-bound topics are
 * asserted against this in `topics.test.ts`.
 * @internal
 */
export type Bridgeable =
  | string
  | number
  | boolean
  | null
  | undefined
  // A `void` payload crosses as `undefined`.
  | void
  | readonly Bridgeable[]
  | { readonly [key: string]: Bridgeable };

/**
 * One adjacent version step: `up` lifts the older shape, `down` projects back.
 * Request replies are not migrated — evolve them additively.
 * @internal
 */
export interface TopicMigration {
  /** The older version this step lifts from / projects back to. */
  readonly from: number;
  /** The newer version (`from + 1`). */
  readonly to: number;
  up(older: unknown): unknown;
  down(newer: unknown): unknown;
}

/**
 * Per-topic migration chains; a topic's version is its chain depth. On a shape
 * change, keep `Topics` at the newest shape and append the adjacent step here.
 * @internal
 */
export const topicMigrations: Partial<
  Record<TopicName, readonly TopicMigration[]>
> = {};
