// @sanity/workbench/src/core/installations.ts
import { z } from "zod";

import {
  AbstractApplication,
  type AbstractApplicationType,
} from "./applications/application";
import { ApplicationId } from "./applications/brett-application";
import { OrganizationId } from "./organizations";

/**
 * Installation ID schema, branded to keep it distinct from the application id
 * an installation points at.
 * @public
 */
export const InstallationId = z.string().nonempty().brand("InstallationId");

/**
 * Installation ID type, branded for type safety.
 * @public
 */
export type InstallationId = z.output<typeof InstallationId>;

const InstallationSchema = z.object({
  id: InstallationId,
  applicationId: ApplicationId,
  organizationId: OrganizationId,
  application: z.object({
    title: z.string(),
    slug: z.string().nullable(),
  }),
  access: z.array(
    z.object({ resourceType: z.string(), resourceId: z.string() }),
  ),
});

/**
 * An organization's active installation of a singleton application, as returned
 * from `GET /installations?organizationId=&include=access`.
 * @public
 */
export type Installation = z.output<typeof InstallationSchema>;

/**
 * Validates and parses a raw `/installations` record into a branded
 * Installation.
 * @public
 */
export function parseInstalledApplication(data: unknown): Installation {
  return InstallationSchema.parse(data);
}

/**
 * @public
 */
export abstract class InstalledApplication<
  TType extends AbstractApplicationType,
> extends AbstractApplication<TType> {
  readonly id: ApplicationId;

  readonly organizationId: OrganizationId;

  readonly slug: string;

  constructor({
    type,
    slug,
    id,
    organizationId,
  }: {
    type: TType;
    slug: string;
    id: ApplicationId;
    organizationId: OrganizationId;
  }) {
    super(type);
    this.slug = slug;
    this.id = id;
    this.organizationId = organizationId;
  }

  get isFederated(): boolean {
    return false;
  }
}
