// @sanity/cli/dist/actions/build/eventListenerFactory.js
import { styleText } from 'node:util';
import { select } from '@sanity/cli-core/ux';
/**
 * Creates check dependency event handlers that are shared between buildStudio and startStudioDevServer.
 */ export function checkDependenciesEventListenerFactory(output) {
    return {
        onIncompatibleDeclaredStyledComponentsVersionRange ({ message }) {
            output.warn(message);
        },
        onIncompatibleInstalledStyledComponentsVersionRange ({ message }) {
            output.warn(message);
        },
        onInvalidStyledComponentsVersionRange ({ message }) {
            output.error(message, {
                exit: 1
            });
        },
        onNoDeclaredStyledComponentsVersion ({ message }) {
            output.error(message, {
                exit: 1
            });
        },
        onNoInstalledSanityVersion ({ message }) {
            output.error(message, {
                exit: 1
            });
        },
        onNoInstalledStyledComponentsVersion ({ message }) {
            output.error(message, {
                exit: 1
            });
        }
    };
}
/**
 * Creates pre-release event handlers that are shared between buildApp and buildStudio.
 */ export function preReleaseEventListenerFactory(output) {
    return {
        async onPreReleaseInInteractiveAutoUpdate ({ prereleaseMessage }) {
            const choice = await select({
                choices: [
                    {
                        name: 'Disable auto-updates for this build and continue',
                        value: 'disable-auto-updates'
                    },
                    {
                        name: 'Cancel build',
                        value: 'cancel'
                    }
                ],
                default: 'disable-auto-updates',
                message: styleText('yellow', prereleaseMessage)
            });
            if (choice === 'cancel') {
                output.error('Declined to continue with build', {
                    exit: 1
                });
                return;
            }
            output.warn('Auto-updates disabled for this build');
        },
        onPreReleaseInNonInteractiveAutoUpdate ({ message }) {
            output.error(message, {
                exit: 1
            });
        }
    };
}
