// @portabletext/sanity-bridge/dist/index.js
import { isKeySegment, isArraySchemaType } from "@sanity/types";
import { Schema } from "@sanity/schema";
import { builtinTypes } from "@sanity/schema/_internal";
function createPortableTextMemberSchemaTypes(portableTextType) {
  if (!portableTextType)
    throw new Error("Parameter 'portabletextType' missing (required)");
  return createPortableTextMemberSchemaTypesFromOf(
    portableTextType,
    portableTextType.of ?? []
  );
}
function createPortableTextMemberSchemaTypesFromOf(rootPortableTextType, of) {
  const blockType = of.find(findBlockType$1);
  if (!blockType)
    throw new Error("Block type is not defined in this schema (required)");
  const childrenField = blockType.fields?.find(
    (field) => field.name === "children"
  );
  if (!childrenField)
    throw new Error("Children field for block type found in schema (required)");
  const ofType = childrenField.type.of;
  if (!ofType)
    throw new Error(
      "Valid types for block children not found in schema (required)"
    );
  const spanType = ofType.find((memberType) => memberType.name === "span");
  if (!spanType)
    throw new Error("Span type not found in schema (required)");
  const inlineObjectTypes = ofType.filter(
    (memberType) => memberType.name !== "span"
  ) || [], blockObjectTypes = of.filter(
    (field) => field.name !== blockType.name
  ) || [];
  return {
    styles: resolveEnabledStyles$1(blockType),
    decorators: resolveEnabledDecorators$1(spanType),
    lists: resolveEnabledListItems$1(blockType),
    block: blockType,
    span: spanType,
    portableText: rootPortableTextType,
    inlineObjects: inlineObjectTypes,
    blockObjects: blockObjectTypes,
    annotations: spanType.annotations
  };
}
function resolveEnabledStyles$1(blockType) {
  const styleField = blockType.fields?.find(
    (btField) => btField.name === "style"
  );
  if (!styleField)
    throw new Error(
      "A field with name 'style' is not defined in the block type (required)."
    );
  const textStyles = styleField.type.options?.list && styleField.type.options.list?.filter(
    (style) => style.value
  );
  if (!textStyles || textStyles.length === 0)
    throw new Error(
      "The style fields need at least one style defined. I.e: {title: 'Normal', value: 'normal'}."
    );
  return textStyles;
}
function resolveEnabledDecorators$1(spanType) {
  return spanType.decorators;
}
function resolveEnabledListItems$1(blockType) {
  const listField = blockType.fields?.find(
    (btField) => btField.name === "listItem"
  );
  if (!listField)
    throw new Error(
      "A field with name 'listItem' is not defined in the block type (required)."
    );
  const listItems = listField.type.options?.list && listField.type.options.list.filter((list) => list.value);
  if (!listItems)
    throw new Error("The list field need at least to be an empty array");
  return listItems;
}
function findBlockType$1(type) {
  return type.type ? findBlockType$1(type.type) : type.name === "block" ? type : null;
}
function getSanitySubSchema(rootPortableTextType, value, path) {
  const enclosingOf = resolveEnclosingOf(rootPortableTextType, value, path);
  return createPortableTextMemberSchemaTypesFromOf(
    rootPortableTextType,
    enclosingOf ?? rootPortableTextType.of ?? []
  );
}
function resolveEnclosingOf(rootPortableTextType, value, path) {
  const parentKeyedIndex = findParentKeyedIndex(path);
  if (parentKeyedIndex < 0)
    return;
  let currentOf = rootPortableTextType.of ?? [], currentChildren = value, containerOf;
  for (let segmentIndex = 0; segmentIndex <= parentKeyedIndex; segmentIndex++) {
    const segment = path[segmentIndex];
    if (typeof segment == "string")
      continue;
    let childNode;
    if (isKeySegment(segment))
      childNode = currentChildren.find((child) => child._key === segment._key);
    else if (typeof segment == "number")
      childNode = currentChildren.at(segment);
    else
      return containerOf;
    if (!childNode)
      return containerOf;
    const member = currentOf.find((entry) => entry.name === childNode._type);
    if (!member)
      return containerOf;
    const childField = findChildArrayField(member);
    if (!childField)
      return containerOf;
    const nextOf = safeGetOf$1(childField.type) ?? [];
    isPortableTextShaped(nextOf) && (containerOf = nextOf), currentOf = nextOf;
    const rawChildren = childNode[childField.name];
    currentChildren = Array.isArray(rawChildren) ? rawChildren : [];
  }
  return containerOf;
}
function isPortableTextShaped(of) {
  return of.some((entry) => walkToRoot(entry).name === "block");
}
function walkToRoot(type) {
  return type.type ? walkToRoot(type.type) : type;
}
function findParentKeyedIndex(path) {
  let lastKeyedIndex = -1, parentKeyedIndex = -1;
  for (let index = 0; index < path.length; index++) {
    const segment = path[index];
    segment !== void 0 && isKeySegment(segment) && (parentKeyedIndex = lastKeyedIndex, lastKeyedIndex = index);
  }
  return parentKeyedIndex;
}
function findChildArrayField(member) {
  const fields = member.fields;
  if (fields) {
    for (const field of fields)
      if (isArraySchemaType(field.type))
        return { name: field.name, type: field.type };
  }
}
function safeGetOf$1(schemaType) {
  try {
    if (schemaType.jsonType === "array") {
      const arrayOf = schemaType.of;
      return Array.isArray(arrayOf) ? arrayOf : void 0;
    }
  } catch {
  }
}
function sanitySchemaToPortableTextSchema(sanitySchema) {
  const compiled = sanitySchema.hasOwnProperty("jsonType") ? sanitySchema : compileType(sanitySchema);
  return sanitySchemaTypeToSchema(compiled);
}
function sanitySchemaTypeToSchema(portableTextType) {
  if (!portableTextType)
    throw new Error("Parameter 'portableTextType' missing (required)");
  const blockType = portableTextType.of?.find(findBlockType);
  if (!blockType)
    throw new Error("Block type is not defined in this schema (required)");
  const childrenField = blockType.fields?.find(
    (field) => field.name === "children"
  );
  if (!childrenField)
    throw new Error("Children field for block type found in schema (required)");
  const ofType = childrenField.type.of;
  if (!ofType)
    throw new Error(
      "Valid types for block children not found in schema (required)"
    );
  const spanType = ofType.find((memberType) => memberType.name === "span");
  if (!spanType)
    throw new Error("Span type not found in schema (required)");
  const inlineObjectTypes = ofType.filter(
    (memberType) => memberType.name !== "span"
  ) || [], blockObjectTypes = portableTextType.of?.filter(
    (field) => field.name !== blockType.name
  ) || [], styles = resolveEnabledStyles(blockType), decorators = resolveEnabledDecorators(spanType), lists = resolveEnabledListItems(blockType), annotations = spanType.annotations, conversion = {
    work: [],
    ancestors: /* @__PURE__ */ new Map(),
    distinctAncestorCount: 0,
    memo: /* @__PURE__ */ new Map(),
    inFlight: /* @__PURE__ */ new Map(),
    rootBlockObjects: new Set(blockObjectTypes)
  }, pendingWork = [], result = {
    block: {
      name: blockType.name
    },
    span: {
      name: spanType.name
    },
    styles: styles.map((style) => ({
      name: style.value,
      title: style.title,
      value: style.value
    })),
    lists: lists.map((list) => ({
      name: list.value,
      title: list.title,
      value: list.value
    })),
    decorators: decorators.map((decorator) => ({
      name: decorator.value,
      title: decorator.title,
      value: decorator.value
    })),
    annotations: annotations.map((annotation) => {
      const built = buildFields(
        conversion,
        annotation.fields,
        void 0,
        annotation
      );
      return pendingWork.push(built.work), {
        name: annotation.name,
        title: annotation.title,
        fields: built.holes
      };
    }),
    blockObjects: blockObjectTypes.map((blockObject) => {
      const built = buildFields(
        conversion,
        blockObject.fields,
        blockObject.name
      );
      return pendingWork.push(built.work), {
        name: blockObject.name,
        title: blockObject.title,
        fields: built.holes
      };
    }),
    inlineObjects: inlineObjectTypes.map((inlineObject) => {
      const built = buildFields(
        conversion,
        inlineObject.fields,
        inlineObject.name,
        inlineObject
      );
      return pendingWork.push(built.work), {
        name: inlineObject.name,
        title: inlineObject.title,
        fields: built.holes
      };
    })
  };
  return pushInOrder(conversion, pendingWork), drain(conversion), result;
}
function pushAncestor(conversion, name) {
  const count = conversion.ancestors.get(name) ?? 0;
  count === 0 && conversion.distinctAncestorCount++, conversion.ancestors.set(name, count + 1);
}
function popAncestor(conversion, name) {
  const count = conversion.ancestors.get(name) ?? 0;
  if (count <= 1) {
    conversion.ancestors.delete(name), count === 1 && conversion.distinctAncestorCount--;
    return;
  }
  conversion.ancestors.set(name, count - 1);
}
function hasAncestor(conversion, name) {
  return (conversion.ancestors.get(name) ?? 0) > 0;
}
function drain(conversion) {
  let steps = 0;
  for (; conversion.work.length > 0; )
    if (conversion.work.pop()(), ++steps > 1e8)
      throw new Error(
        `Portable Text schema conversion exceeded ${steps} work items; the schema type graph likely contains a cycle the conversion cannot cut (in-flight ancestors: [${[...conversion.ancestors.keys()].join(", ")}])`
      );
}
function buildFields(conversion, fields, seedName, inFlightKey) {
  const holes = [];
  return { holes, work: () => {
    if (inFlightKey !== void 0) {
      const entrySizes = conversion.inFlight.get(inFlightKey), ancestorCount = conversion.distinctAncestorCount;
      if (entrySizes !== void 0 && entrySizes.length > 0 && ancestorCount <= entrySizes[entrySizes.length - 1])
        return;
      entrySizes === void 0 ? conversion.inFlight.set(inFlightKey, [ancestorCount]) : entrySizes.push(ancestorCount), conversion.work.push(() => {
        const sizes = conversion.inFlight.get(inFlightKey);
        sizes?.pop(), sizes !== void 0 && sizes.length === 0 && conversion.inFlight.delete(inFlightKey);
      });
    }
    seedName !== void 0 && (pushAncestor(conversion, seedName), conversion.work.push(() => popAncestor(conversion, seedName))), holes.length = fields.length;
    for (let index = fields.length - 1; index >= 0; index--) {
      const field = fields[index];
      conversion.work.push(() => scheduleField(conversion, field, holes, index));
    }
  } };
}
function pushInOrder(conversion, works) {
  for (let index = works.length - 1; index >= 0; index--)
    conversion.work.push(works[index]);
}
function scheduleField(conversion, field, target, index) {
  if (field.type.jsonType !== "array") {
    target[index] = {
      name: field.name,
      type: field.type.jsonType,
      ...field.type.title ? { title: field.type.title } : {}
    };
    return;
  }
  const ofMembers = safeGetOf(field.type), of = ofMembers ? new Array(ofMembers.length) : [];
  if (target[index] = {
    name: field.name,
    type: "array",
    ...field.type.title ? { title: field.type.title } : {},
    of
  }, ofMembers)
    for (let memberIndex = ofMembers.length - 1; memberIndex >= 0; memberIndex--) {
      const member = ofMembers[memberIndex];
      conversion.work.push(
        () => scheduleOfMember(conversion, member, of, memberIndex)
      );
    }
}
function scheduleOfMember(conversion, memberType, target, index) {
  if (findBlockType(memberType)) {
    scheduleBlockOfMember(
      conversion,
      memberType,
      target,
      index
    );
    return;
  }
  if (!(memberType.jsonType === "object" && "fields" in memberType && Array.isArray(memberType.fields)) || hasAncestor(conversion, memberType.name) || conversion.rootBlockObjects.has(memberType)) {
    target[index] = {
      type: memberType.name,
      ...memberType.title ? { title: memberType.title } : {}
    };
    return;
  }
  const memoized = conversion.memo.get(memberType);
  if (memoized) {
    target[index] = memoized;
    return;
  }
  const fields = memberType.fields, holes = new Array(fields.length), definition = {
    type: "object",
    name: memberType.name,
    ...memberType.title ? { title: memberType.title } : {},
    fields: holes
  };
  conversion.memo.set(memberType, definition), target[index] = definition, pushAncestor(conversion, memberType.name), conversion.work.push(() => popAncestor(conversion, memberType.name));
  for (let fieldIndex = fields.length - 1; fieldIndex >= 0; fieldIndex--) {
    const field = fields[fieldIndex];
    conversion.work.push(
      () => scheduleField(conversion, field, holes, fieldIndex)
    );
  }
}
function scheduleBlockOfMember(conversion, blockType, target, index) {
  const styleList = blockType.fields?.find((field) => field.name === "style")?.type.options?.list, listItemList = blockType.fields?.find(
    (field) => field.name === "listItem"
  )?.type.options?.list, childrenOf = blockType.fields?.find((field) => field.name === "children")?.type.of, spanType = childrenOf?.find(
    (memberType) => memberType.name === "span"
  ), inlineObjectTypes = (Array.isArray(childrenOf) ? childrenOf : []).filter((memberType) => memberType.name !== "span"), spanDecorators = spanType?.decorators, spanAnnotations = spanType?.annotations, pendingWork = [], definition = {
    type: "block",
    styles: (Array.isArray(styleList) ? styleList : []).filter((style) => style.value).map((style) => ({
      name: style.value,
      title: style.title,
      value: style.value
    })),
    lists: (Array.isArray(listItemList) ? listItemList : []).filter((list) => list.value).map((list) => ({
      name: list.value,
      title: list.title,
      value: list.value
    })),
    decorators: (Array.isArray(spanDecorators) ? spanDecorators : []).map(
      (decorator) => ({
        name: decorator.value,
        title: decorator.title,
        value: decorator.value
      })
    ),
    annotations: (Array.isArray(spanAnnotations) ? spanAnnotations : []).map(
      (annotation) => {
        const built = buildFields(
          conversion,
          annotation.fields,
          void 0,
          annotation
        );
        return pendingWork.push(built.work), {
          name: annotation.name,
          title: annotation.title,
          fields: built.holes
        };
      }
    ),
    inlineObjects: inlineObjectTypes.map((inlineObject) => {
      const built = buildFields(
        conversion,
        inlineObject.fields ?? [],
        inlineObject.name,
        inlineObject
      );
      return pendingWork.push(built.work), {
        name: inlineObject.name,
        title: inlineObject.title,
        fields: built.holes
      };
    })
  };
  target[index] = definition, pushInOrder(conversion, pendingWork);
}
function safeGetOf(schemaType) {
  try {
    if (schemaType.jsonType === "array") {
      const arrayOf = schemaType.of;
      return Array.isArray(arrayOf) ? arrayOf : void 0;
    }
  } catch {
  }
}
function resolveEnabledStyles(blockType) {
  const styleField = blockType.fields?.find(
    (btField) => btField.name === "style"
  );
  if (!styleField)
    throw new Error(
      "A field with name 'style' is not defined in the block type (required)."
    );
  const textStyles = styleField.type.options?.list && styleField.type.options.list?.filter(
    (style) => style.value
  );
  if (!textStyles || textStyles.length === 0)
    throw new Error(
      "The style fields need at least one style defined. I.e: {title: 'Normal', value: 'normal'}."
    );
  return textStyles;
}
function resolveEnabledDecorators(spanType) {
  return spanType.decorators;
}
function resolveEnabledListItems(blockType) {
  const listField = blockType.fields?.find(
    (btField) => btField.name === "listItem"
  );
  if (!listField)
    throw new Error(
      "A field with name 'listItem' is not defined in the block type (required)."
    );
  const listItems = listField.type.options?.list && listField.type.options.list.filter((list) => list.value);
  if (!listItems)
    throw new Error("The list field need at least to be an empty array");
  return listItems;
}
function findBlockType(type) {
  return type.type ? findBlockType(type.type) : type.name === "block" ? type : null;
}
function compileType(rawType) {
  return Schema.compile({
    name: "blockTypeSchema",
    types: [rawType, ...builtinTypes]
  }).get(rawType.name);
}
export {
  createPortableTextMemberSchemaTypes,
  getSanitySubSchema,
  sanitySchemaToPortableTextSchema
};
