// sanity/lib/_chunks-es/index8.js
import { jsx, jsxs, Fragment } from "react/jsx-runtime";
import { c } from "react/compiler-runtime";
import { isPortableTextTextBlock, isArrayOfBlocksSchemaType } from "@sanity/types";
import { PortableTextEditor } from "@portabletext/editor";
import { useClickOutsideEvent, Stack, usePortal, BoundaryElementProvider } from "@sanity/ui";
import * as PathUtils from "@sanity/util/paths";
import { uuid } from "@sanity/uuid";
import debounce from "lodash-es/debounce.js";
import isEqual from "lodash-es/isEqual.js";
import { motion, AnimatePresence } from "motion/react";
import { useState, useEffect, useRef, memo, useLayoutEffect, startTransition } from "react";
import { Popover, Button } from "./index.js";
import "react-i18next";
import "sanity/_singletons";
import { useTranslation, commentsLocaleNamespace, CommentDisabledIcon, hasCommentMessageValue, CommentInput, useCommentsEnabled, useFieldActions, useCurrentUser, useComments, useAddonDataset, useCommentsSelectedPath, useCommentsScroll, useFormValue, isTextSelectionComment, parseCommentFieldPath, getCommentFieldPath, buildTextSelectionFromFragment, buildRangeDecorationSelectionsFromComments, buildCommentRangeDecorations, CommentInlineHighlightSpan } from "./index2.js";
import "history";
import "@sanity/icons/ArrowLeft";
import "@sanity/icons/Launch";
import { styled } from "styled-components";
import "react-fast-compare";
import "rxjs/operators";
import "react-rx";
import "rxjs";
import "@sanity/client/csm";
import "quick-lru";
import "lodash-es/throttle.js";
import "./createAgentBundlesStore.js";
import "@portabletext/toolkit";
import "@sanity/icons/WarningOutline";
import "@sanity/icons/ChevronRight";
import "@sanity/icons/Revert";
import "@sanity/telemetry/react";
import "@sanity/icons/Document";
import "@sanity/icons/Image";
import "@sanity/color";
import "color2k";
import "lodash-es/isString.js";
import "@sanity/icons/Copy";
import "@sanity/icons/Clipboard";
import "@sanity/icons/ErrorOutline";
import "@sanity/icons/InfoOutline";
import "@portabletext/sanity-bridge";
import "@sanity/icons/Edit";
import "lodash-es/pick.js";
import "@sanity/util/legacyDateFormat";
import "./version.js";
import "lodash-es/get.js";
import "@algorithm.ts/lcs";
import "lodash-es/intersection.js";
import "json-stable-stringify";
import "mendoza";
import "shallow-equals";
import "scroll-into-view-if-needed";
import "@sanity/schema/_internal";
import "lodash-es/groupBy.js";
import "lodash-es/sortBy.js";
import "lodash-es/compact.js";
import "lodash-es/keyBy.js";
import "lodash-es/partition.js";
import "lodash-es/toLower.js";
import "lodash-es/union.js";
import "lodash-es/uniq.js";
import "lodash-es/words.js";
import "lodash-es/isFinite.js";
import "lodash-es/flatten.js";
import "@sanity/icons/Add";
import "@sanity/id-utils";
import "sanity/router";
import "@sanity/icons/ChevronDown";
import "date-fns/addHours";
import "date-fns/startOfHour";
import "@sanity/icons/Bolt";
import "@sanity/icons/Clock";
import "@sanity/icons/Dot";
import "@sanity/icons/EarthGlobe";
import "date-fns/format";
import "date-fns/isValid";
import "date-fns/parse";
import "date-fns/getMinutes";
import "date-fns/parseISO";
import "date-fns/setMinutes";
import "@sanity/icons/EarthAmericas";
import "lodash-es/noop.js";
import "@sanity/icons/Search";
import "@sanity/ui/theme";
import "observable-callback";
import "@sanity/util/content";
import "@sanity/icons/BulbOutline";
import "@sanity/icons/Unknown";
import "@sanity/client/stega";
import "@sanity/icons/ToggleArrowRight";
import "lodash-es/isPlainObject.js";
import "@sanity/icons/ThLarge";
import "@sanity/icons/Ulist";
import "lodash-es/startCase.js";
import "react-is";
import "@sanity/icons/Trash";
import "@dnd-kit/sortable";
import "@sanity/icons/DragHandle";
import "@sanity/icons/AddDocument";
import "@sanity/icons/InsertAbove";
import "@sanity/icons/InsertBelow";
import "@tanstack/react-virtual";
import "speakingurl";
import "@sanity/icons/Reset";
import "@sanity/icons/Sync";
import "@sanity/asset-utils";
import "@sanity/icons/AccessDenied";
import "@sanity/icons/HelpCircle";
import "@sanity/image-url";
import "@sanity/icons/Close";
import "@portabletext/react";
import "@sanity/icons";
import "@sanity/icons/Link";
import "lodash-es/pickBy.js";
import "@sanity/client";
import "swr";
import "@sanity/icons/Download";
import "@sanity/icons/Upload";
import "@sanity/icons/BinaryDocument";
import "@sanity/icons/Lock";
import "@sanity/diff";
import "lodash-es/castArray.js";
import "nanoid";
import "use-hot-module-reload";
import "@sanity/icons/EllipsisHorizontal";
import "@sanity/icons/EyeOpen";
import "react-focus-lock";
import "lodash-es/memoize.js";
import "lodash-es/sortedIndex.js";
import "lodash-es/difference.js";
import "lodash-es/identity.js";
import "lodash-es/values.js";
import "@sanity/comlink";
import "@sanity/mutator";
import "lodash-es/omit.js";
import "@isaacs/ttlcache";
import "@sanity/schema";
import "@sanity/util/client";
import "@sanity/util/concurrency-limiter";
import "lodash-es/uniqBy.js";
import "json-reduce";
import "rxjs-exhaustmap-with-trailing";
import "dataloader";
import "raf";
import "groq-js";
import "use-sync-external-store/with-selector";
import "date-fns/differenceInDays";
import "date-fns/differenceInHours";
import "date-fns/differenceInMinutes";
import "date-fns/differenceInMonths";
import "date-fns/differenceInSeconds";
import "date-fns/differenceInWeeks";
import "date-fns/differenceInYears";
import "@sanity/icons/AddCircle";
import "@sanity/icons/Calendar";
import "@sanity/icons/Circle";
import "@sanity/icons/Publish";
import "@sanity/icons/Timeline";
import "@sanity/icons/Unpublish";
import "@sanity/icons/ArrowDown";
import "@sanity/icons/ArrowRight";
import "@sanity/diff-patch";
import "clsx";
import "@sanity/icons/LinkRemoved";
import "lodash-es/findIndex.js";
import { useCommentsUpsell } from "./useCommentsUpsell.js";
import "@sanity/icons/ChevronUp";
import "@sanity/icons/Feedback";
import "@xstate/react";
import "xstate";
import "@sanity/icons/FaceHappy";
import "@sanity/icons/FaceIndifferent";
import "@sanity/icons/FaceSad";
import "@sentry/react";
import "rxjs-mergemap-array";
import "date-fns/isBefore";
import "date-fns/startOfMinute";
import "@sanity/icons/Documents";
import "@sanity/icons/User";
import "@sanity/icons/CheckmarkCircle";
import "lodash-es/deburr.js";
import "@sanity/icons/ChevronLeft";
import "@sanity/icons/Checkmark";
import "@sanity/logos";
import "is-network-error";
import "lodash-es/escapeRegExp.js";
import "@sanity/icons/Desktop";
import "@sanity/icons/Moon";
import "@sanity/icons/Sun";
import "lodash-es/isEmpty.js";
import "react-dom/server";
import "@sanity/icons/Controls";
import "@sanity/icons/Sort";
import "lodash-es/capitalize.js";
import "@sanity/icons/Undo";
import "@sanity/icons/Menu";
import "@sanity/icons/AddUser";
import "@sanity/icons/Cog";
import "@sanity/icons/Users";
import "semver";
import "@sanity/icons/Github";
import "@sanity/icons/Refresh";
import "@sanity/icons/Leave";
import "lodash-es/last.js";
import "react-dom/client";
import "date-fns/addWeeks";
import "date-fns/isAfter";
import { AddCommentIcon } from "@sanity/icons/AddComment";
import "@sanity/diff-match-patch";
function createDomRectFromElements(elements) {
  if (!elements || !elements.length) return null;
  const rects = elements.map((el) => el.getBoundingClientRect()), minX = Math.min(...rects.map((r) => r.x)) || 0, minY = Math.min(...rects.map((r) => r.y)) || 0, maxRight = Math.max(...rects.map((r) => r.right)) || 0, maxBottom = Math.max(...rects.map((r) => r.bottom)) || 0;
  return {
    x: minX,
    y: minY,
    width: maxRight - minX,
    height: maxBottom - minY,
    top: minY,
    right: maxRight,
    bottom: maxBottom,
    left: minX
  };
}
function useRectFromElements(props) {
  const $ = c(5), {
    scrollElement,
    disabled,
    selector
  } = props, [rect, setRect] = useState(null);
  let t0, t1;
  return $[0] !== disabled || $[1] !== scrollElement || $[2] !== selector ? (t0 = () => {
    if (disabled)
      return;
    const handleSetRect = () => {
      const elements = document?.querySelectorAll(selector);
      if (!elements)
        return;
      const nextRect = createDomRectFromElements(Array.from(elements));
      setRect(nextRect);
    };
    if (handleSetRect(), !!scrollElement)
      return scrollElement.addEventListener("wheel", handleSetRect), () => {
        scrollElement.removeEventListener("wheel", handleSetRect);
      };
  }, t1 = [scrollElement, disabled, selector], $[0] = disabled, $[1] = scrollElement, $[2] = selector, $[3] = t0, $[4] = t1) : (t0 = $[3], t1 = $[4]), useEffect(t0, t1), rect;
}
function useAuthoringReferenceElement(props) {
  const $ = c(6), {
    scrollElement,
    disabled,
    selector
  } = props;
  let t0;
  $[0] !== disabled || $[1] !== scrollElement || $[2] !== selector ? (t0 = {
    scrollElement,
    disabled,
    selector
  }, $[0] = disabled, $[1] = scrollElement, $[2] = selector, $[3] = t0) : t0 = $[3];
  const rect = useRectFromElements(t0);
  let t1;
  bb0: {
    if (!rect) {
      t1 = null;
      break bb0;
    }
    let t2;
    $[4] !== rect ? (t2 = {
      getBoundingClientRect: () => rect
    }, $[4] = rect, $[5] = t2) : t2 = $[5], t1 = t2;
  }
  return t1;
}
function getSelectionBoundingRect() {
  const selection = window.getSelection();
  let range = null;
  return selection && selection.rangeCount > 0 && (range = selection.getRangeAt(0)), range?.getBoundingClientRect() || null;
}
const MotionPopover = /* @__PURE__ */ styled(motion.create(Popover)).withConfig({
  displayName: "MotionPopover",
  componentId: "sc-wmroq-0"
})(["user-select:none;"]), POPOVER_FALLBACK_PLACEMENTS$1 = ["bottom", "top"], VARIANTS = {
  hidden: {
    opacity: 0,
    y: -4
  },
  visible: {
    opacity: 1,
    y: 0
  }
};
function FloatingButtonPopover(props) {
  const $ = c(13), {
    disabled,
    onClick,
    onClickOutside,
    referenceElement
  } = props, buttonRef = useRef(null), {
    t
  } = useTranslation(commentsLocaleNamespace);
  let t0;
  $[0] === /* @__PURE__ */ Symbol.for("react.memo_cache_sentinel") ? (t0 = () => [buttonRef.current], $[0] = t0) : t0 = $[0], useClickOutsideEvent(onClickOutside, t0);
  let t1;
  $[1] !== t ? (t1 = t("inline-add-comment-button.disabled-overlap-title"), $[1] = t, $[2] = t1) : t1 = $[2];
  const disabledText = t1;
  let t2;
  $[3] !== t ? (t2 = t("inline-add-comment-button.title"), $[3] = t, $[4] = t2) : t2 = $[4];
  const text = disabled ? disabledText : t2, t3 = disabled ? CommentDisabledIcon : AddCommentIcon;
  let t4;
  $[5] !== disabled || $[6] !== onClick || $[7] !== t3 || $[8] !== text ? (t4 = /* @__PURE__ */ jsx(Button, { "data-testid": "inline-comment-button", disabled, icon: t3, mode: "bleed", onClick, ref: buttonRef, text }), $[5] = disabled, $[6] = onClick, $[7] = t3, $[8] = text, $[9] = t4) : t4 = $[9];
  const content = t4;
  let t5;
  return $[10] !== content || $[11] !== referenceElement ? (t5 = /* @__PURE__ */ jsx(MotionPopover, { animate: "visible", content, contentEditable: !1, fallbackPlacements: POPOVER_FALLBACK_PLACEMENTS$1, initial: "hidden", open: !0, padding: 1, placement: "bottom", portal: !0, referenceElement, variants: VARIANTS }), $[10] = content, $[11] = referenceElement, $[12] = t5) : t5 = $[12], t5;
}
const POPOVER_FALLBACK_PLACEMENTS = ["bottom", "top"], RootStack = /* @__PURE__ */ styled(Stack).withConfig({
  displayName: "RootStack",
  componentId: "sc-w2wgl6-0"
})(["width:250px;"]);
function InlineCommentInputPopover(props) {
  const $ = c(17), {
    currentUser,
    mentionOptions,
    onChange,
    onClickOutside,
    onDiscardConfirm,
    onSubmit,
    referenceElement,
    value
  } = props, commentInputRef = useRef(null), contentElementRef = useRef(null);
  let t0;
  $[0] !== onDiscardConfirm ? (t0 = () => {
    commentInputRef.current?.discardDialogController.close(), onDiscardConfirm();
  }, $[0] = onDiscardConfirm, $[1] = t0) : t0 = $[1];
  const handleDiscardConfirm = t0;
  let t1;
  $[2] === /* @__PURE__ */ Symbol.for("react.memo_cache_sentinel") ? (t1 = () => {
    commentInputRef.current?.discardDialogController.close();
  }, $[2] = t1) : t1 = $[2];
  const handleDiscardCancel = t1;
  let t2;
  $[3] !== onClickOutside || $[4] !== value ? (t2 = () => {
    if (hasCommentMessageValue(value)) {
      commentInputRef.current?.discardDialogController.open();
      return;
    }
    onClickOutside();
  }, $[3] = onClickOutside, $[4] = value, $[5] = t2) : t2 = $[5];
  let t3;
  $[6] === /* @__PURE__ */ Symbol.for("react.memo_cache_sentinel") ? (t3 = () => [contentElementRef.current], $[6] = t3) : t3 = $[6], useClickOutsideEvent(t2, t3);
  let t4;
  $[7] !== currentUser || $[8] !== handleDiscardConfirm || $[9] !== mentionOptions || $[10] !== onChange || $[11] !== onSubmit || $[12] !== value ? (t4 = /* @__PURE__ */ jsx(RootStack, { padding: 2, ref: contentElementRef, children: /* @__PURE__ */ jsx(CommentInput, { currentUser, focusLock: !0, focusOnMount: !0, mentionOptions, onChange, onDiscardCancel: handleDiscardCancel, onDiscardConfirm: handleDiscardConfirm, onSubmit, ref: commentInputRef, value }) }), $[7] = currentUser, $[8] = handleDiscardConfirm, $[9] = mentionOptions, $[10] = onChange, $[11] = onSubmit, $[12] = value, $[13] = t4) : t4 = $[13];
  const content = t4;
  let t5;
  return $[14] !== content || $[15] !== referenceElement ? (t5 = /* @__PURE__ */ jsx(Popover, { content, "data-ui": "InlineCommentInputPopover", fallbackPlacements: POPOVER_FALLBACK_PLACEMENTS, open: !0, placement: "bottom", portal: !0, referenceElement }), $[14] = content, $[15] = referenceElement, $[16] = t5) : t5 = $[16], t5;
}
const EMPTY_ARRAY = [], AI_ASSIST_TYPE = "sanity.assist.instruction.prompt";
function CommentsPortableTextInput(props) {
  const $ = c(5), {
    enabled,
    mode
  } = useCommentsEnabled(), fieldActions = useFieldActions(), isAiAssist = props.schemaType.name === AI_ASSIST_TYPE, isCommentsEnabledInField = !!fieldActions.__internal_comments;
  if (!enabled || isAiAssist || !isCommentsEnabledInField) {
    let t02;
    return $[0] !== props ? (t02 = props.renderDefault(props), $[0] = props, $[1] = t02) : t02 = $[1], t02;
  }
  let t0;
  return $[2] !== mode || $[3] !== props ? (t0 = /* @__PURE__ */ jsx(CommentsPortableTextInputInner, { ...props, mode }), $[2] = mode, $[3] = props, $[4] = t0) : t0 = $[4], t0;
}
const CommentsPortableTextInputInner = memo(function(props) {
  const $ = c(101), {
    mode
  } = props, currentUser = useCurrentUser(), portal = usePortal(), {
    comments,
    getComment,
    mentionOptions,
    onCommentsOpen,
    operation,
    setStatus,
    status
  } = useComments(), {
    error: addonDatasetError
  } = useAddonDataset(), {
    setSelectedPath,
    selectedPath
  } = useCommentsSelectedPath(), {
    scrollToComment,
    scrollToGroup
  } = useCommentsScroll(), {
    handleOpenDialog
  } = useCommentsUpsell(), [mousePressed, setMousePressed] = useState(!1), editorRef = useRef(null);
  let t0;
  $[0] === /* @__PURE__ */ Symbol.for("react.memo_cache_sentinel") ? (t0 = [], $[0] = t0) : t0 = $[0];
  const documentValue = useFormValue(t0), documentValueRef = useRef(documentValue);
  let t1, t2;
  $[1] !== documentValue ? (t1 = () => {
    documentValueRef.current = documentValue;
  }, t2 = [documentValue], $[1] = documentValue, $[2] = t1, $[3] = t2) : (t1 = $[2], t2 = $[3]), useLayoutEffect(t1, t2);
  const [authoringDecorationElement, setAuthoringDecorationElement] = useState(null), [nextCommentValue, setNextCommentValue] = useState(null), [nextCommentSelection, setNextCommentSelection] = useState(null), [currentSelection, setCurrentSelection] = useState(null), [currentSelectionRect, setCurrenSelectionRect] = useState(null), [currentHoveredCommentId, setCurrentHoveredCommentId] = useState(null), [canSubmit, setCanSubmit] = useState(!1), [rootElement, setRootElement] = useState(null), [isFullScreen, setIsFullScreen] = useState(!1), [addedCommentsDecorations, setAddedCommentsDecorations] = useState(EMPTY_ARRAY), [fragment, setFragment] = useState(null);
  let t3;
  $[4] === /* @__PURE__ */ Symbol.for("react.memo_cache_sentinel") ? (t3 = () => editorRef.current ? PortableTextEditor.getFragment(editorRef.current) : EMPTY_ARRAY, $[4] = t3) : t3 = $[4];
  const getFragment = t3;
  let t4;
  $[5] === /* @__PURE__ */ Symbol.for("react.memo_cache_sentinel") ? (t4 = () => {
    const rect = getSelectionBoundingRect();
    setCurrenSelectionRect(rect);
  }, $[5] = t4) : t4 = $[5];
  const handleSetCurrentSelectionRect = t4;
  let t5;
  $[6] === /* @__PURE__ */ Symbol.for("react.memo_cache_sentinel") ? (t5 = () => {
    setCurrentSelection(null), setCurrenSelectionRect(null), setNextCommentSelection(null), setNextCommentValue(null), setCanSubmit(!1), setAuthoringDecorationElement(null), setFragment(null);
  }, $[6] = t5) : t5 = $[6];
  const resetStates = t5;
  let t6;
  $[7] !== currentSelection || $[8] !== handleOpenDialog || $[9] !== mode ? (t6 = () => {
    if (mode === "upsell") {
      handleOpenDialog("pte");
      return;
    }
    const currentFragment = getFragment() || null;
    setFragment(currentFragment), setNextCommentSelection(currentSelection);
  }, $[7] = currentSelection, $[8] = handleOpenDialog, $[9] = mode, $[10] = t6) : t6 = $[10];
  const handleSelectCurrentSelection = t6;
  let t7;
  $[11] === /* @__PURE__ */ Symbol.for("react.memo_cache_sentinel") ? (t7 = () => {
    resetStates();
  }, $[11] = t7) : t7 = $[11];
  const handleCommentDiscardConfirm = t7;
  let t8;
  if ($[12] !== comments.data.open || $[13] !== props.path) {
    let t92;
    $[15] !== props.path ? (t92 = (c_0) => {
      const fieldPath = parseCommentFieldPath(c_0.fieldPath);
      return fieldPath ? PathUtils.startsWith(props.path, fieldPath) : !1;
    }, $[15] = props.path, $[16] = t92) : t92 = $[16], t8 = comments.data.open.filter(_temp).filter(t92).map(_temp2), $[12] = comments.data.open, $[13] = props.path, $[14] = t8;
  } else
    t8 = $[14];
  const textComments = t8;
  let t9;
  $[17] !== fragment || $[18] !== nextCommentSelection || $[19] !== onCommentsOpen || $[20] !== operation || $[21] !== props.path || $[22] !== scrollToGroup || $[23] !== setSelectedPath || $[24] !== setStatus || $[25] !== status ? (t9 = (nextValue) => {
    if (!nextCommentSelection || !editorRef.current)
      return;
    const normalizedSelection = nextCommentSelection.backward ? {
      backward: !1,
      anchor: nextCommentSelection.focus,
      focus: nextCommentSelection.anchor
    } : nextCommentSelection, fieldPath_0 = getCommentFieldPath(documentValueRef.current, props.path, normalizedSelection.focus.path);
    if (!fieldPath_0)
      return;
    const textSelection = buildTextSelectionFromFragment({
      fragment: fragment || EMPTY_ARRAY,
      selection: normalizedSelection,
      documentValue: documentValueRef.current,
      basePath: props.path
    }), threadId = uuid();
    operation.create({
      type: "field",
      contentSnapshot: fragment,
      fieldPath: fieldPath_0,
      message: nextValue,
      parentCommentId: void 0,
      reactions: EMPTY_ARRAY,
      selection: textSelection,
      status: "open",
      threadId
    }), onCommentsOpen?.(), status === "resolved" && setStatus("open"), setSelectedPath({
      fieldPath: fieldPath_0,
      threadId,
      origin: "form"
    }), scrollToGroup(threadId), resetStates();
  }, $[17] = fragment, $[18] = nextCommentSelection, $[19] = onCommentsOpen, $[20] = operation, $[21] = props.path, $[22] = scrollToGroup, $[23] = setSelectedPath, $[24] = setStatus, $[25] = status, $[26] = t9) : t9 = $[26];
  const handleSubmit = t9;
  let t10;
  $[27] !== getComment || $[28] !== onCommentsOpen || $[29] !== scrollToComment || $[30] !== setSelectedPath ? (t10 = (commentId) => {
    const comment = getComment(commentId);
    comment && (setSelectedPath({
      fieldPath: comment.target.path?.field || "",
      threadId: comment.threadId,
      origin: "form"
    }), onCommentsOpen?.(), scrollToComment(comment._id));
  }, $[27] = getComment, $[28] = onCommentsOpen, $[29] = scrollToComment, $[30] = setSelectedPath, $[31] = t10) : t10 = $[31];
  const handleDecoratorClick = t10, blurred = useRef(!1);
  let t11;
  $[32] !== mousePressed ? (t11 = (selection) => {
    const isRangeSelected = selection?.anchor.offset !== selection?.focus.offset;
    if (!getFragment()?.every(isPortableTextTextBlock) || !isRangeSelected) {
      setCanSubmit(!1);
      return;
    }
    if (blurred.current) {
      blurred.current = !1;
      return;
    }
    mousePressed || handleSetCurrentSelectionRect(), setCurrentSelection(selection), setCanSubmit(!0);
  }, $[32] = mousePressed, $[33] = t11) : t11 = $[33];
  const debounceSelectionChange = useDebounceSelectionChange(t11);
  let t12;
  $[34] === /* @__PURE__ */ Symbol.for("react.memo_cache_sentinel") ? (t12 = () => {
    startTransition(() => setMousePressed(!0));
  }, $[34] = t12) : t12 = $[34];
  const handleMouseDown = t12;
  let t13;
  $[35] === /* @__PURE__ */ Symbol.for("react.memo_cache_sentinel") ? (t13 = () => {
    startTransition(() => setMousePressed(!1)), handleSetCurrentSelectionRect();
  }, $[35] = t13) : t13 = $[35];
  const handleMouseUp = t13;
  let t14;
  $[36] === /* @__PURE__ */ Symbol.for("react.memo_cache_sentinel") ? (t14 = (details) => {
    const {
      rangeDecoration,
      newSelection
    } = details, commentId_0 = rangeDecoration.payload?.commentId;
    setAddedCommentsDecorations((prev) => prev.map((p) => p.payload?.commentId === commentId_0 ? {
      ...rangeDecoration,
      selection: newSelection,
      payload: {
        ...rangeDecoration.payload,
        dirty: !0
      }
    } : p));
  }, $[36] = t14) : t14 = $[36];
  const handleRangeDecorationMoved = t14;
  let t15;
  $[37] !== addedCommentsDecorations || $[38] !== getComment || $[39] !== operation || $[40] !== props.path ? (t15 = () => {
    const decoratorsToUpdate = addedCommentsDecorations.filter(_temp3);
    decoratorsToUpdate.length !== 0 && (decoratorsToUpdate.forEach((decorator_0) => {
      const commentId_1 = decorator_0.payload?.commentId, comment_0 = getComment(commentId_1 || "");
      if (!comment_0 || !editorRef.current)
        return;
      const editorValue = PortableTextEditor.getValue(editorRef.current) || EMPTY_ARRAY, [updatedDecoration] = buildRangeDecorationSelectionsFromComments({
        comments: [comment_0],
        value: editorValue,
        documentValue: documentValueRef.current,
        basePath: props.path
      }), nextRange = updatedDecoration?.range ? [updatedDecoration.range] : EMPTY_ARRAY, nextValue_0 = updatedDecoration ? [...comment_0.target.path?.selection?.value.filter((r) => r._key !== nextRange[0]?._key).concat(nextRange).flat().sort(_temp4) || EMPTY_ARRAY] : EMPTY_ARRAY, nextComment = {
        target: {
          ...comment_0.target,
          path: {
            ...comment_0.target?.path,
            field: comment_0.target.path?.field || "",
            selection: {
              type: "text",
              value: nextValue_0
            }
          }
        }
      };
      !isEqual(comment_0.target, nextComment.target) && operation.update(comment_0._id, nextComment);
    }), setAddedCommentsDecorations((prev_0) => prev_0.map((p_0) => decoratorsToUpdate.find((d) => d.payload?.commentId === p_0.payload?.commentId)?.payload?.dirty ? {
      ...p_0,
      payload: {
        ...p_0.payload,
        dirty: !1
      }
    } : p_0).filter(_temp5)));
  }, $[37] = addedCommentsDecorations, $[38] = getComment, $[39] = operation, $[40] = props.path, $[41] = t15) : t15 = $[41];
  const updateCommentRange = t15;
  let t16;
  $[42] !== currentHoveredCommentId || $[43] !== handleDecoratorClick || $[44] !== props.path || $[45] !== selectedPath?.threadId ? (t16 = (commentsToDecorate) => {
    if (!editorRef.current)
      return EMPTY_ARRAY;
    const editorValue_0 = PortableTextEditor.getValue(editorRef.current) || EMPTY_ARRAY;
    return buildCommentRangeDecorations({
      comments: commentsToDecorate,
      currentHoveredCommentId,
      onDecorationClick: handleDecoratorClick,
      onDecorationHoverEnd: setCurrentHoveredCommentId,
      onDecorationHoverStart: setCurrentHoveredCommentId,
      onDecorationMoved: handleRangeDecorationMoved,
      selectedThreadId: selectedPath?.threadId || null,
      value: editorValue_0,
      documentValue: documentValueRef.current,
      basePath: props.path
    });
  }, $[42] = currentHoveredCommentId, $[43] = handleDecoratorClick, $[44] = props.path, $[45] = selectedPath?.threadId, $[46] = t16) : t16 = $[46], selectedPath?.threadId;
  const handleBuildRangeDecorations = t16;
  let t17;
  $[47] !== debounceSelectionChange || $[48] !== updateCommentRange ? (t17 = (change) => {
    change.type === "mutation" && updateCommentRange(), change.type === "blur" && (blurred.current = !0), change.type === "selection" && debounceSelectionChange(change.selection);
  }, $[47] = debounceSelectionChange, $[48] = updateCommentRange, $[49] = t17) : t17 = $[49];
  const onEditorChange = t17;
  let t18;
  bb0: {
    if (!nextCommentSelection) {
      t18 = null;
      break bb0;
    }
    let t192;
    $[50] === /* @__PURE__ */ Symbol.for("react.memo_cache_sentinel") ? (t192 = (t203) => {
      const {
        children
      } = t203;
      return /* @__PURE__ */ jsx(CommentInlineHighlightSpan, { isAuthoring: !0, ref: setAuthoringDecorationElement, children });
    }, $[50] = t192) : t192 = $[50];
    let t202;
    $[51] !== nextCommentSelection ? (t202 = {
      component: t192,
      selection: nextCommentSelection
    }, $[51] = nextCommentSelection, $[52] = t202) : t202 = $[52], t18 = t202;
  }
  const authoringDecoration = t18, t19 = props?.rangeDecorations || EMPTY_ARRAY;
  let t20;
  $[53] !== addedCommentsDecorations || $[54] !== authoringDecoration || $[55] !== t19 ? (t20 = [...t19, ...authoringDecoration ? [authoringDecoration] : EMPTY_ARRAY, ...addedCommentsDecorations], $[53] = addedCommentsDecorations, $[54] = authoringDecoration, $[55] = t19, $[56] = t20) : t20 = $[56];
  const rangeDecorations = t20, [currentSelectionIsOverlapping, setCurrentSelectionIsOverlapping] = useState(!1);
  let t21, t22;
  $[57] !== addedCommentsDecorations || $[58] !== currentSelection ? (t21 = () => {
    startTransition(() => setCurrentSelectionIsOverlapping(() => !currentSelection || addedCommentsDecorations.length === 0 ? !1 : addedCommentsDecorations.some((d_0) => {
      if (!editorRef.current)
        return !1;
      const testA = PortableTextEditor.isSelectionsOverlapping(editorRef.current, currentSelection, d_0.selection), testB = PortableTextEditor.isSelectionsOverlapping(editorRef.current, d_0.selection, currentSelection);
      return testA || testB;
    })));
  }, t22 = [addedCommentsDecorations, currentSelection], $[57] = addedCommentsDecorations, $[58] = currentSelection, $[59] = t21, $[60] = t22) : (t21 = $[59], t22 = $[60]), useEffect(t21, t22);
  const scrollElement = isFullScreen ? document.body : portal.elements?.documentScrollElement || document.body, boundaryElement = isFullScreen ? portal.elements?.documentScrollElement || document.body : rootElement, t23 = !nextCommentSelection || !authoringDecorationElement;
  let t24;
  $[61] !== scrollElement || $[62] !== t23 ? (t24 = {
    scrollElement,
    disabled: t23,
    selector: '[data-inline-comment-state="authoring"]'
  }, $[61] = scrollElement, $[62] = t23, $[63] = t24) : t24 = $[63];
  const popoverAuthoringReferenceElement = useAuthoringReferenceElement(t24);
  let t25;
  bb1: {
    if (!currentSelectionRect) {
      t25 = null;
      break bb1;
    }
    let t262;
    $[64] !== currentSelectionRect ? (t262 = {
      getBoundingClientRect: () => currentSelectionRect
    }, $[64] = currentSelectionRect, $[65] = t262) : t262 = $[65], t25 = t262;
  }
  const selectionReferenceElement = t25;
  let t26, t27;
  $[66] !== currentSelection || $[67] !== scrollElement ? (t26 = () => {
    if (currentSelection)
      return scrollElement?.addEventListener("wheel", handleSetCurrentSelectionRect), () => {
        scrollElement?.removeEventListener("wheel", handleSetCurrentSelectionRect);
      };
  }, t27 = [currentSelection, scrollElement, handleSetCurrentSelectionRect], $[66] = currentSelection, $[67] = scrollElement, $[68] = t26, $[69] = t27) : (t26 = $[68], t27 = $[69]), useEffect(t26, t27);
  let t28, t29;
  $[70] !== handleBuildRangeDecorations || $[71] !== textComments ? (t28 = () => {
    const nextDecorations = handleBuildRangeDecorations(textComments);
    startTransition(() => setAddedCommentsDecorations((current) => nextDecorations.map((nextDecoration_1) => {
      const prevDecoration = current.find((p_2) => p_2.payload?.commentId === nextDecoration_1.payload?.commentId);
      return prevDecoration?.payload?.dirty ? {
        ...nextDecoration_1,
        payload: {
          ...nextDecoration_1.payload,
          dirty: prevDecoration.payload.dirty
        }
      } : nextDecoration_1;
    })));
  }, t29 = [handleBuildRangeDecorations, textComments], $[70] = handleBuildRangeDecorations, $[71] = textComments, $[72] = t28, $[73] = t29) : (t28 = $[72], t29 = $[73]), useEffect(t28, t29);
  const showFloatingButton = !!(currentSelection && canSubmit && selectionReferenceElement && !mousePressed), showFloatingInput = !!(nextCommentSelection && popoverAuthoringReferenceElement);
  let t30;
  $[74] !== currentUser || $[75] !== handleSubmit || $[76] !== mentionOptions || $[77] !== nextCommentValue || $[78] !== popoverAuthoringReferenceElement || $[79] !== showFloatingInput ? (t30 = showFloatingInput && currentUser && /* @__PURE__ */ jsx(InlineCommentInputPopover, { currentUser, mentionOptions, onChange: setNextCommentValue, onClickOutside: resetStates, onDiscardConfirm: handleCommentDiscardConfirm, onSubmit: handleSubmit, referenceElement: popoverAuthoringReferenceElement, value: nextCommentValue }), $[74] = currentUser, $[75] = handleSubmit, $[76] = mentionOptions, $[77] = nextCommentValue, $[78] = popoverAuthoringReferenceElement, $[79] = showFloatingInput, $[80] = t30) : t30 = $[80];
  let t31;
  $[81] !== addonDatasetError || $[82] !== currentSelectionIsOverlapping || $[83] !== handleSelectCurrentSelection || $[84] !== selectionReferenceElement || $[85] !== showFloatingButton || $[86] !== showFloatingInput ? (t31 = showFloatingButton && !showFloatingInput && /* @__PURE__ */ jsx(FloatingButtonPopover, { disabled: currentSelectionIsOverlapping || !!addonDatasetError, onClick: handleSelectCurrentSelection, onClickOutside: resetStates, referenceElement: selectionReferenceElement }), $[81] = addonDatasetError, $[82] = currentSelectionIsOverlapping, $[83] = handleSelectCurrentSelection, $[84] = selectionReferenceElement, $[85] = showFloatingButton, $[86] = showFloatingInput, $[87] = t31) : t31 = $[87];
  let t32;
  $[88] !== t31 ? (t32 = /* @__PURE__ */ jsx(AnimatePresence, { children: t31 }), $[88] = t31, $[89] = t32) : t32 = $[89];
  let t33;
  $[90] !== boundaryElement || $[91] !== t30 || $[92] !== t32 ? (t33 = /* @__PURE__ */ jsxs(BoundaryElementProvider, { element: boundaryElement, children: [
    t30,
    t32
  ] }), $[90] = boundaryElement, $[91] = t30, $[92] = t32, $[93] = t33) : t33 = $[93];
  let t34;
  $[94] !== onEditorChange || $[95] !== props || $[96] !== rangeDecorations ? (t34 = /* @__PURE__ */ jsx(Stack, { ref: setRootElement, onMouseDown: handleMouseDown, onMouseUp: handleMouseUp, children: /* @__PURE__ */ jsx(RenderDefaultCommentsPortableTextInputInner, { ...props, onEditorChange, editorRef, rangeDecorations, onFullScreenChange: setIsFullScreen }) }), $[94] = onEditorChange, $[95] = props, $[96] = rangeDecorations, $[97] = t34) : t34 = $[97];
  let t35;
  return $[98] !== t33 || $[99] !== t34 ? (t35 = /* @__PURE__ */ jsxs(Fragment, { children: [
    t33,
    t34
  ] }), $[98] = t33, $[99] = t34, $[100] = t35) : t35 = $[100], t35;
});
function useDebounceSelectionChange(handleSelectionChange) {
  const $ = c(2);
  let t0;
  return $[0] !== handleSelectionChange ? (t0 = debounce(handleSelectionChange, 200), $[0] = handleSelectionChange, $[1] = t0) : t0 = $[1], t0;
}
function RenderDefaultCommentsPortableTextInputInner(props) {
  return props.renderDefault(props);
}
function _temp(c2) {
  return isTextSelectionComment(c2.parentComment);
}
function _temp2(c_1) {
  return c_1.parentComment;
}
function _temp3(decorator) {
  return decorator.payload?.dirty;
}
function _temp4(a, b) {
  return a._key.localeCompare(b._key);
}
function _temp5(p_1) {
  return p_1.selection !== null;
}
function isPortableTextInputProps(inputProps) {
  return isArrayOfBlocksSchemaType(inputProps.schemaType);
}
function CommentsInput(props) {
  const $ = c(4);
  if (isPortableTextInputProps(props)) {
    let t02;
    return $[0] !== props ? (t02 = /* @__PURE__ */ jsx(CommentsPortableTextInput, { ...props }), $[0] = props, $[1] = t02) : t02 = $[1], t02;
  }
  let t0;
  return $[2] !== props ? (t0 = props.renderDefault(props), $[2] = props, $[3] = t0) : t0 = $[3], t0;
}
export {
  CommentsInput
};
