// sanity/lib/_chunks-es/TasksStudioNavbar.js
import { jsx } from "react/jsx-runtime";
import { c } from "react/compiler-runtime";
import { CheckmarkCircleIcon } from "@sanity/icons/CheckmarkCircle";
import { Button } from "./index.js";
import "@sanity/ui";
import "react";
import "react-i18next";
import "sanity/_singletons";
import "styled-components";
import "@sanity/util/paths";
import "react-fast-compare";
import { useTasksEnabled, useTasksNavigation, useTranslation, tasksLocaleNamespace } from "./index2.js";
import "history";
import "@sanity/icons/ArrowLeft";
import "@sanity/icons/Launch";
import "@sanity/logos";
import "@sanity/icons/ChevronRight";
import "react-rx";
import "@sanity/client";
import "rxjs";
import "rxjs/operators";
import "./version.js";
import "@sanity/icons/Checkmark";
import "react-is";
import "@sanity/client/csm";
import "quick-lru";
import "lodash-es/throttle.js";
import "lodash-es/isEqual.js";
import "./createAgentBundlesStore.js";
import "@portabletext/toolkit";
import "@sanity/icons/WarningOutline";
import "@sanity/icons/Document";
import "@sanity/icons/Image";
import "@sanity/uuid";
import "@sanity/color";
import "color2k";
import "lodash-es/isString.js";
import "@sanity/icons/Copy";
import "@sanity/icons/Clipboard";
import "@sanity/telemetry/react";
import "@portabletext/editor";
import "@sanity/types";
import "@sanity/icons/Revert";
import "@sanity/ui/theme";
import "date-fns/differenceInDays";
import "date-fns/differenceInHours";
import "date-fns/differenceInMinutes";
import "date-fns/differenceInMonths";
import "date-fns/differenceInSeconds";
import "date-fns/differenceInWeeks";
import "date-fns/differenceInYears";
import "@sanity/icons/AddCircle";
import "@sanity/icons/Calendar";
import "@sanity/icons/Circle";
import "@sanity/icons/Close";
import "@sanity/icons/Edit";
import "@sanity/icons/Publish";
import "@sanity/icons/Timeline";
import "@sanity/icons/Trash";
import "@sanity/icons/Unpublish";
import "@sanity/icons/ArrowDown";
import "@sanity/icons/ArrowRight";
import "@sanity/util/legacyDateFormat";
import "@sanity/diff-patch";
import "@sanity/icons/ErrorOutline";
import "@sanity/diff";
import "mendoza";
import "@sanity/icons/InfoOutline";
import "@portabletext/sanity-bridge";
import "@sanity/id-utils";
import "sanity/router";
import "@sanity/icons/ChevronDown";
import "date-fns/addHours";
import "date-fns/startOfHour";
import "@sanity/icons/Bolt";
import "@sanity/icons/Clock";
import "@sanity/icons/Dot";
import "@sanity/icons/EarthGlobe";
import "date-fns/format";
import "date-fns/isValid";
import "date-fns/parse";
import "lodash-es/flatten.js";
import "date-fns/getMinutes";
import "date-fns/parseISO";
import "date-fns/setMinutes";
import "@sanity/icons/EarthAmericas";
import "lodash-es/noop.js";
import "@sanity/icons/Search";
import "motion/react";
import "lodash-es/pick.js";
import "lodash-es/get.js";
import "@algorithm.ts/lcs";
import "lodash-es/intersection.js";
import "json-stable-stringify";
import "shallow-equals";
import "scroll-into-view-if-needed";
import "@sanity/schema/_internal";
import "lodash-es/groupBy.js";
import "lodash-es/sortBy.js";
import "lodash-es/compact.js";
import "lodash-es/keyBy.js";
import "lodash-es/partition.js";
import "lodash-es/toLower.js";
import "lodash-es/union.js";
import "lodash-es/uniq.js";
import "lodash-es/words.js";
import "lodash-es/isFinite.js";
import "@sanity/icons/Add";
import "observable-callback";
import "@sanity/util/content";
import "@sanity/icons/BulbOutline";
import "@sanity/icons/Unknown";
import "@sanity/client/stega";
import "@sanity/icons/ToggleArrowRight";
import "lodash-es/isPlainObject.js";
import "@sanity/icons/ThLarge";
import "@sanity/icons/Ulist";
import "lodash-es/startCase.js";
import "@dnd-kit/sortable";
import "@sanity/icons/DragHandle";
import "@sanity/icons/AddDocument";
import "@sanity/icons/InsertAbove";
import "@sanity/icons/InsertBelow";
import "@tanstack/react-virtual";
import "speakingurl";
import "@sanity/icons/Reset";
import "@sanity/icons/Sync";
import "@sanity/asset-utils";
import "@sanity/icons/AccessDenied";
import "@sanity/icons/HelpCircle";
import "@sanity/image-url";
import "@portabletext/react";
import "@sanity/icons";
import "@sanity/icons/Link";
import "lodash-es/pickBy.js";
import "swr";
import "@sanity/icons/Download";
import "@sanity/icons/Upload";
import "@sanity/icons/BinaryDocument";
import "@sanity/icons/Lock";
import "lodash-es/castArray.js";
import "nanoid";
import "use-hot-module-reload";
import "@sanity/icons/EllipsisHorizontal";
import "@sanity/icons/EyeOpen";
import "rxjs-mergemap-array";
import "@sanity/mutator";
import "react-focus-lock";
import "lodash-es/memoize.js";
import "lodash-es/sortedIndex.js";
import "lodash-es/difference.js";
import "lodash-es/identity.js";
import "lodash-es/values.js";
import "@sanity/comlink";
import "lodash-es/omit.js";
import "@isaacs/ttlcache";
import "@sanity/schema";
import "@sanity/util/client";
import "@sanity/util/concurrency-limiter";
import "lodash-es/uniqBy.js";
import "json-reduce";
import "rxjs-exhaustmap-with-trailing";
import "dataloader";
import "raf";
import "groq-js";
import "use-sync-external-store/with-selector";
import "is-network-error";
import "lodash-es/escapeRegExp.js";
import "@sanity/icons/Desktop";
import "@sanity/icons/Moon";
import "@sanity/icons/Sun";
import "lodash-es/isEmpty.js";
import "react-dom/server";
import "@sanity/icons/Controls";
import "@sanity/icons/Sort";
import "lodash-es/capitalize.js";
import "@sanity/icons/Undo";
import "@sanity/icons/Documents";
import "@sanity/icons/Menu";
import "@sanity/icons/AddUser";
import "@sanity/icons/Cog";
import "@sanity/icons/Users";
import "semver";
import "@sanity/icons/FaceHappy";
import "@sanity/icons/FaceIndifferent";
import "@sanity/icons/FaceSad";
import "@sentry/react";
import "@sanity/icons/Feedback";
import "@sanity/icons/Github";
import "@sanity/icons/Refresh";
import "@sanity/icons/Leave";
import "lodash-es/last.js";
import "react-dom/client";
import "date-fns/addWeeks";
import "date-fns/isAfter";
import "date-fns/isBefore";
import "lodash-es/debounce.js";
import "clsx";
import "@sanity/icons/LinkRemoved";
import "lodash-es/findIndex.js";
import "@sanity/diff-match-patch";
import "@sanity/icons/ChevronUp";
import "@xstate/react";
import "xstate";
import "date-fns/startOfMinute";
import "@sanity/icons/User";
import "lodash-es/deburr.js";
import "@sanity/icons/ChevronLeft";
const EMPTY_ARRAY = [], TasksToolbar = (t0) => {
  const $ = c(8), {
    onClick,
    isOpen
  } = t0, {
    t
  } = useTranslation(tasksLocaleNamespace);
  let t1;
  $[0] !== t ? (t1 = t("toolbar.tooltip"), $[0] = t, $[1] = t1) : t1 = $[1];
  let t2;
  $[2] !== t1 ? (t2 = {
    content: t1
  }, $[2] = t1, $[3] = t2) : t2 = $[3];
  let t3;
  return $[4] !== isOpen || $[5] !== onClick || $[6] !== t2 ? (t3 = /* @__PURE__ */ jsx(Button, { tooltipProps: t2, icon: CheckmarkCircleIcon, mode: "bleed", onClick, selected: isOpen, "data-testid": "tasks-toolbar" }), $[4] = isOpen, $[5] = onClick, $[6] = t2, $[7] = t3) : t3 = $[7], t3;
};
function TasksStudioNavbarInner(props) {
  const $ = c(22), {
    handleOpenTasks,
    handleCloseTasks,
    state: t0
  } = useTasksNavigation(), {
    isOpen
  } = t0, {
    t
  } = useTranslation(tasksLocaleNamespace);
  let t1;
  $[0] !== handleCloseTasks || $[1] !== handleOpenTasks || $[2] !== isOpen ? (t1 = () => {
    isOpen ? handleCloseTasks() : handleOpenTasks();
  }, $[0] = handleCloseTasks, $[1] = handleOpenTasks, $[2] = isOpen, $[3] = t1) : t1 = $[3];
  const handleClick = t1;
  let t2;
  $[4] !== handleClick || $[5] !== isOpen ? (t2 = () => /* @__PURE__ */ jsx(TasksToolbar, { onClick: handleClick, isOpen }), $[4] = handleClick, $[5] = isOpen, $[6] = t2) : t2 = $[6];
  const renderTasksNav = t2;
  props?.__internal_actions;
  const t3 = props?.__internal_actions || EMPTY_ARRAY;
  let t4;
  $[7] !== renderTasksNav ? (t4 = {
    location: "topbar",
    name: "tasks-topbar",
    render: renderTasksNav
  }, $[7] = renderTasksNav, $[8] = t4) : t4 = $[8];
  let t5;
  $[9] !== t ? (t5 = t("actions.open.text"), $[9] = t, $[10] = t5) : t5 = $[10];
  let t6;
  $[11] !== handleClick || $[12] !== isOpen || $[13] !== t5 ? (t6 = {
    icon: CheckmarkCircleIcon,
    location: "sidebar",
    name: "tasks-sidebar",
    onAction: handleClick,
    selected: isOpen,
    title: t5
  }, $[11] = handleClick, $[12] = isOpen, $[13] = t5, $[14] = t6) : t6 = $[14];
  let t7;
  $[15] !== t3 || $[16] !== t4 || $[17] !== t6 ? (t7 = [...t3, t4, t6], $[15] = t3, $[16] = t4, $[17] = t6, $[18] = t7) : t7 = $[18];
  const actions = t7;
  let t8;
  return $[19] !== actions || $[20] !== props ? (t8 = props.renderDefault({
    ...props,
    __internal_actions: actions
  }), $[19] = actions, $[20] = props, $[21] = t8) : t8 = $[21], t8;
}
function TasksStudioNavbar(props) {
  const $ = c(4), {
    enabled
  } = useTasksEnabled();
  if (!enabled) {
    let t02;
    return $[0] !== props ? (t02 = props.renderDefault(props), $[0] = props, $[1] = t02) : t02 = $[1], t02;
  }
  let t0;
  return $[2] !== props ? (t0 = /* @__PURE__ */ jsx(TasksStudioNavbarInner, { ...props }), $[2] = props, $[3] = t0) : t0 = $[3], t0;
}
export {
  TasksStudioNavbar
};
