// sanity/lib/_chunks-es/useCommentsUpsell.js
import { c } from "react/compiler-runtime";
import { useContext } from "react";
import { CommentsUpsellContext } from "sanity/_singletons";
function useCommentsUpsell() {
  const $ = c(1), value = useContext(CommentsUpsellContext);
  if (!value) {
    let t0;
    return $[0] === /* @__PURE__ */ Symbol.for("react.memo_cache_sentinel") ? (t0 = {
      upsellData: null,
      handleOpenDialog: _temp,
      handleClose: _temp2,
      upsellDialogOpen: !1,
      telemetryLogs: {
        dialogSecondaryClicked: _temp3,
        dialogPrimaryClicked: _temp4,
        panelViewed: _temp5,
        panelDismissed: _temp6,
        panelPrimaryClicked: _temp7,
        panelSecondaryClicked: _temp8
      }
    }, $[0] = t0) : t0 = $[0], t0;
  }
  return value;
}
function _temp8() {
  return null;
}
function _temp7() {
  return null;
}
function _temp6() {
  return null;
}
function _temp5() {
  return null;
}
function _temp4() {
  return null;
}
function _temp3() {
  return null;
}
function _temp2() {
  return null;
}
function _temp() {
  return null;
}
export {
  useCommentsUpsell
};
