// sanity/lib/_chunks-es/VariantsTool.js
import { jsx, jsxs, Fragment } from "react/jsx-runtime";
import { c } from "react/compiler-runtime";
import { useRouter, IntentLink, StateLink } from "sanity/router";
import { Autocomplete, Stack, Flex, Box, Text, TextInput, TextArea, useToast, Card, Menu, Badge, Container, rem, Skeleton } from "@sanity/ui";
import { useId, useState, useMemo, useCallback, forwardRef, memo } from "react";
import { Button, Dialog, MenuItem, MenuButton, Tooltip } from "./index.js";
import { getVariantId, variantsClient, useClient, VARIANTS_STUDIO_CLIENT_OPTIONS, VARIANT_DOCUMENTS_PATH, useTranslation, variantsLocaleNamespace, useAllVariants, TextWithTone, DOCUMENT_SYSTEM_FIELD, useBundleDocuments, usePerspective, useSetVariant, getVariantTitle, isPublishedBundleId, ContextMenuButton, RelativeTime, isReleaseBundle, getReleaseDocumentIdFromReleaseId, getReleaseIdFromReleaseDocumentId, ReleaseTitle, RELEASES_INTENT, useDocumentPresence, DocumentPreviewPresence, useDocumentPreviewValues, SanityDefaultPreview, useSchema, ToneIcon, useActiveReleases, decodeVariantIdFromRoute, LoadingBlock, getVariantDescription, getVariantConditionsText, useVariantsStore, listenQuery, filterVariantsForSearch } from "./index2.js";
import "history";
import "sanity/_singletons";
import "@sanity/icons/ArrowLeft";
import "@sanity/icons/Launch";
import "styled-components";
import "react-i18next";
import { map, distinctUntilChanged, switchMap, scan, startWith, catchError } from "rxjs/operators";
import { useObservable } from "react-rx";
import { of } from "rxjs";
import { getPublishedId } from "@sanity/client/csm";
import "quick-lru";
import "lodash-es/throttle.js";
import "lodash-es/isEqual.js";
import "./createAgentBundlesStore.js";
import "@sanity/logos";
import "@sanity/icons/ChevronRight";
import "@sanity/client";
import "./version.js";
import "@sanity/icons/Checkmark";
import "react-is";
import "@sanity/icons/Document";
import "@sanity/icons/Image";
import "@sanity/uuid";
import "@sanity/color";
import "color2k";
import "lodash-es/isString.js";
import "@sanity/icons/Copy";
import "@sanity/icons/Clipboard";
import "@sanity/telemetry/react";
import "@portabletext/editor";
import "@sanity/util/paths";
import "@sanity/types";
import "@sanity/icons/Revert";
import "@sanity/ui/theme";
import "date-fns/differenceInDays";
import "date-fns/differenceInHours";
import "date-fns/differenceInMinutes";
import "date-fns/differenceInMonths";
import "date-fns/differenceInSeconds";
import "date-fns/differenceInWeeks";
import "date-fns/differenceInYears";
import "@sanity/icons/AddCircle";
import "@sanity/icons/Calendar";
import "@sanity/icons/Circle";
import "@sanity/icons/Close";
import "@sanity/icons/Edit";
import "@sanity/icons/Publish";
import "@sanity/icons/Timeline";
import { TrashIcon } from "@sanity/icons/Trash";
import "@sanity/icons/Unpublish";
import "@sanity/icons/ArrowDown";
import "@sanity/icons/ArrowRight";
import "@sanity/util/legacyDateFormat";
import "@sanity/diff-patch";
import { ErrorOutlineIcon } from "@sanity/icons/ErrorOutline";
import "@sanity/diff";
import "mendoza";
import "@sanity/icons/InfoOutline";
import "@sanity/icons/WarningOutline";
import "@portabletext/sanity-bridge";
import "@sanity/id-utils";
import "@sanity/icons/ChevronDown";
import "date-fns/addHours";
import "date-fns/startOfHour";
import "@sanity/icons/Bolt";
import "@sanity/icons/Clock";
import "@sanity/icons/Dot";
import "@sanity/icons/EarthGlobe";
import "date-fns/format";
import "date-fns/isValid";
import "date-fns/parse";
import "lodash-es/flatten.js";
import "date-fns/getMinutes";
import "date-fns/parseISO";
import "date-fns/setMinutes";
import "react-fast-compare";
import "clsx";
import "@sanity/icons/EarthAmericas";
import "lodash-es/noop.js";
import { SearchIcon } from "@sanity/icons/Search";
import "motion/react";
import "lodash-es/pick.js";
import "lodash-es/get.js";
import "@algorithm.ts/lcs";
import "lodash-es/intersection.js";
import "json-stable-stringify";
import "shallow-equals";
import "scroll-into-view-if-needed";
import "@sanity/schema/_internal";
import "lodash-es/groupBy.js";
import "lodash-es/sortBy.js";
import "lodash-es/compact.js";
import "lodash-es/keyBy.js";
import "lodash-es/partition.js";
import "lodash-es/toLower.js";
import "lodash-es/union.js";
import "lodash-es/uniq.js";
import "lodash-es/words.js";
import "lodash-es/isFinite.js";
import { AddIcon } from "@sanity/icons/Add";
import "observable-callback";
import { randomKey } from "@sanity/util/content";
import "@sanity/icons/BulbOutline";
import "@sanity/icons/Unknown";
import "@sanity/client/stega";
import "@sanity/icons/ToggleArrowRight";
import "lodash-es/isPlainObject.js";
import "@sanity/icons/ThLarge";
import "@sanity/icons/Ulist";
import "lodash-es/startCase.js";
import "@dnd-kit/sortable";
import "@sanity/icons/DragHandle";
import "@sanity/icons/AddDocument";
import "@sanity/icons/InsertAbove";
import "@sanity/icons/InsertBelow";
import "@tanstack/react-virtual";
import "speakingurl";
import "@portabletext/react";
import "@sanity/icons/LinkRemoved";
import "@sanity/asset-utils";
import "@sanity/icons/AccessDenied";
import "@sanity/icons/HelpCircle";
import "@sanity/image-url";
import { customAlphabet } from "nanoid";
import "lodash-es/castArray.js";
import "rxjs-mergemap-array";
import "@sanity/mutator";
import "lodash-es/findIndex.js";
import "@sanity/diff-match-patch";
import "@sanity/icons";
import "@sanity/icons/Link";
import "@sanity/icons/ChevronUp";
import "@sanity/icons/EyeOpen";
import "@sanity/icons/Feedback";
import "@xstate/react";
import "xstate";
import "@sanity/icons/FaceHappy";
import "@sanity/icons/FaceIndifferent";
import "@sanity/icons/FaceSad";
import "@sentry/react";
import "@sanity/schema";
import "@sanity/util/client";
import "@sanity/util/concurrency-limiter";
import "lodash-es/uniqBy.js";
import "json-reduce";
import "rxjs-exhaustmap-with-trailing";
import "date-fns/isBefore";
import "date-fns/startOfMinute";
import "@sanity/icons/Documents";
import "lodash-es/debounce.js";
import "@sanity/icons/User";
import "@sanity/icons/CheckmarkCircle";
import "lodash-es/deburr.js";
import "react-focus-lock";
import "@sanity/icons/ChevronLeft";
import "@sanity/icons/Reset";
import "@sanity/icons/Sync";
import "lodash-es/pickBy.js";
import "swr";
import "@sanity/icons/Download";
import "@sanity/icons/Upload";
import "@sanity/icons/BinaryDocument";
import "@sanity/icons/Lock";
import "use-hot-module-reload";
import "@sanity/icons/EllipsisHorizontal";
import "lodash-es/memoize.js";
import "lodash-es/sortedIndex.js";
import "lodash-es/difference.js";
import "lodash-es/identity.js";
import "lodash-es/values.js";
import "@sanity/comlink";
import "lodash-es/omit.js";
import "@isaacs/ttlcache";
import "dataloader";
import "raf";
import "groq-js";
import "use-sync-external-store/with-selector";
import "is-network-error";
import "lodash-es/escapeRegExp.js";
import "@sanity/icons/Desktop";
import "@sanity/icons/Moon";
import "@sanity/icons/Sun";
import "lodash-es/isEmpty.js";
import "react-dom/server";
import "@sanity/icons/Controls";
import "@sanity/icons/Sort";
import "lodash-es/capitalize.js";
import "@sanity/icons/Undo";
import "@sanity/icons/Menu";
import "@sanity/icons/AddUser";
import "@sanity/icons/Cog";
import "@sanity/icons/Users";
import "semver";
import "@sanity/icons/Github";
import "@sanity/icons/Refresh";
import "@sanity/icons/Leave";
import "lodash-es/last.js";
import "react-dom/client";
import "date-fns/addWeeks";
import "date-fns/isAfter";
import { isPortableTextBlock, toPlainText } from "@portabletext/toolkit";
import { at, set } from "@sanity/mutate";
import { applyPatches } from "@sanity/mutate/_unstable_apply";
import { PinIcon } from "@sanity/icons/Pin";
import { PinFilledIcon } from "@sanity/icons/PinFilled";
import { DiamondIcon } from "@sanity/icons/Diamond";
import { Headers, Table, searchDocumentRelease } from "./searchDocumentRelease.js";
function createVariantOperationsStore(options) {
  const client = variantsClient(options.client);
  return {
    createVariant: async (variant) => {
      const action = {
        actionType: "sanity.action.variant.definition.create",
        variantId: getVariantId(variant._id),
        conditions: variant.conditions,
        priority: variant.priority,
        ...variant.metadata ? {
          metadata: variant.metadata
        } : {}
      };
      return await client.action(action, {
        tag: "variants.create"
      });
    },
    updateVariant: async (variant) => {
      const variantId = getVariantId(variant._id), setPayload = {
        conditions: variant.conditions,
        priority: variant.priority
      };
      variant.metadata && (setPayload.metadata = variant.metadata);
      const patch = {
        set: setPayload
      };
      variant.metadata || (patch.unset = ["metadata"]);
      const action = {
        actionType: "sanity.action.variant.definition.edit",
        variantId,
        patch
      };
      return await client.action(action, {
        tag: "variants.edit"
      });
    },
    deleteVariant: async (variantIdOrDocumentId) => {
      const action = {
        actionType: "sanity.action.variant.definition.delete",
        variantId: getVariantId(variantIdOrDocumentId)
      };
      return await client.action(action, {
        tag: "variants.delete"
      });
    }
  };
}
function useVariantOperations() {
  const $ = c(2), studioClient = useClient(VARIANTS_STUDIO_CLIENT_OPTIONS);
  let t0;
  return $[0] !== studioClient ? (t0 = createVariantOperationsStore({
    client: studioClient
  }), $[0] = studioClient, $[1] = t0) : t0 = $[1], t0;
}
function getVariantTitleValue(variant) {
  const title = variant.metadata?.title;
  return typeof title == "string" ? title.trim() : "";
}
function getIsVariantInvalid(variant) {
  return !getVariantTitleValue(variant) || !getHasValidConditions(variant.conditions);
}
function getHasValidConditions(conditions) {
  const entries = Object.entries(conditions);
  return entries.length > 0 && entries.every(([key, value]) => !!(key.trim() && value.trim()));
}
const CONDITION_KEY_REGEX = /^[a-z][a-z0-9_-]{0,63}$/, RESERVED_KEY_PREFIXES = ["_", "$"];
function getConditionKeyValidationError(key) {
  const trimmedKey = key.trim();
  if (trimmedKey) {
    for (const prefix of RESERVED_KEY_PREFIXES)
      if (trimmedKey.startsWith(prefix))
        return "reserved";
    if (trimmedKey.includes(":") || !CONDITION_KEY_REGEX.test(trimmedKey))
      return "invalid";
  }
}
function getConditionValueValidationError(value) {
  if (!value.trim())
    return "empty";
  if (value.includes(":"))
    return "invalid";
}
const createVariantIdSuffix = customAlphabet("abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789", 8);
function createVariantId() {
  return `${VARIANT_DOCUMENTS_PATH}.${createVariantIdSuffix()}`;
}
function getVariantDefaults() {
  return {
    _id: createVariantId(),
    _type: "system.variant",
    conditions: {},
    priority: 0,
    metadata: {
      title: "",
      description: []
    }
  };
}
function createPortableTextDescription(text) {
  return text.trim() ? [{
    _key: randomKey(12),
    _type: "block",
    children: [{
      _key: randomKey(12),
      _type: "span",
      marks: [],
      text
    }],
    markDefs: [],
    style: "normal"
  }] : [];
}
function sortValues(values) {
  return Array.from(values).filter(Boolean).toSorted((a, b) => a.localeCompare(b));
}
function buildConditionSuggestionIndex(variants) {
  const keys = /* @__PURE__ */ new Set(), valuesByKey = /* @__PURE__ */ new Map();
  return variants.forEach((variant) => {
    Object.entries(variant.conditions).forEach(([rawKey, rawValue]) => {
      const key = rawKey.trim();
      if (!key)
        return;
      keys.add(key);
      const value = rawValue?.trim();
      if (!value)
        return;
      let values = valuesByKey.get(key);
      values || (values = /* @__PURE__ */ new Set(), valuesByKey.set(key, values)), values.add(value);
    });
  }), {
    keys: sortValues(keys),
    valuesByKey: new Map(Array.from(valuesByKey, ([key, values]) => [key, sortValues(values)]))
  };
}
function getConditionKeyOptions(index, rows, currentRowIndex) {
  const usedKeys = new Set(rows.filter((_, rowIndex) => rowIndex !== currentRowIndex).map((row) => row.key.trim()).filter(Boolean));
  return index.keys.filter((key) => !usedKeys.has(key)).map((value) => ({
    value
  }));
}
function getConditionValueOptions(index, conditionKey) {
  const key = conditionKey.trim();
  return key ? index.valuesByKey.get(key)?.map((value) => ({
    value
  })) ?? [] : [];
}
function filterConditionOption(query, option) {
  return option.value.toLowerCase().includes(query.trim().toLowerCase());
}
function ConditionAutocompleteInput(props) {
  const $ = c(16), {
    ariaLabel,
    autoFocus,
    customValidity,
    invalid,
    onChange,
    options,
    placeholder,
    testId,
    value
  } = props, id = useId(), [focused, setFocused] = useState(!1);
  let t0;
  $[0] !== onChange ? (t0 = (nextValue) => {
    nextValue !== null && onChange(nextValue);
  }, $[0] = onChange, $[1] = t0) : t0 = $[1];
  const handleQueryChange = t0, t1 = invalid ? "true" : void 0;
  let t2;
  $[2] === /* @__PURE__ */ Symbol.for("react.memo_cache_sentinel") ? (t2 = () => setFocused(!1), $[2] = t2) : t2 = $[2];
  let t3;
  $[3] === /* @__PURE__ */ Symbol.for("react.memo_cache_sentinel") ? (t3 = () => setFocused(!0), $[3] = t3) : t3 = $[3];
  const t4 = focused ? void 0 : value;
  let t5;
  return $[4] !== ariaLabel || $[5] !== autoFocus || $[6] !== customValidity || $[7] !== handleQueryChange || $[8] !== id || $[9] !== onChange || $[10] !== options || $[11] !== placeholder || $[12] !== t1 || $[13] !== t4 || $[14] !== testId ? (t5 = /* @__PURE__ */ jsx(Autocomplete, { id, "aria-invalid": t1, "aria-label": ariaLabel, autoFocus, customValidity, "data-testid": testId, filterOption: filterConditionOption, fontSize: 1, onBlur: t2, onChange, onFocus: t3, onQueryChange: handleQueryChange, openButton: !0, options, placeholder, value: t4 }), $[4] = ariaLabel, $[5] = autoFocus, $[6] = customValidity, $[7] = handleQueryChange, $[8] = id, $[9] = onChange, $[10] = options, $[11] = placeholder, $[12] = t1, $[13] = t4, $[14] = testId, $[15] = t5) : t5 = $[15], t5;
}
function getConditionRows(conditions) {
  const rows = Object.entries(conditions).map(([key, value]) => ({
    id: randomKey(12),
    key,
    value
  }));
  return rows.length ? rows : [{
    id: randomKey(12),
    key: "",
    value: ""
  }];
}
function getConditionsFromRows(rows) {
  return rows.reduce((conditions, row) => {
    const key = row.key.trim(), value = row.value.trim();
    return key && value && (conditions[key] = value), conditions;
  }, {});
}
function isConditionRowEmpty(row) {
  return !row.key.trim() && !row.value.trim();
}
function getEmptyConditionRowValidation() {
  return {
    key: null,
    value: null
  };
}
function getConditionRowsValidation(rows) {
  const conditionRowsValidation = /* @__PURE__ */ new Map(), seenKeys = /* @__PURE__ */ new Set();
  return rows.forEach((row, index) => {
    const validation = {
      key: null,
      value: null
    }, key = row.key.trim(), value = row.value.trim();
    if (key) {
      seenKeys.has(key) ? validation.key = "dialog.create.condition-key.duplicate" : seenKeys.add(key);
      const keyError = getConditionKeyValidationError(row.key);
      !validation.key && keyError === "reserved" ? validation.key = "dialog.create.condition-key.reserved" : !validation.key && keyError === "invalid" && (validation.key = "dialog.create.condition-key.invalid");
    } else value && (validation.key = "dialog.create.condition-key.required");
    const valueError = getConditionValueValidationError(row.value);
    (key || isConditionRowEmpty(row)) && valueError === "empty" ? validation.value = "dialog.create.condition-value.required" : valueError === "invalid" && (validation.value = "dialog.create.condition-value.invalid"), conditionRowsValidation.set(index, validation);
  }), conditionRowsValidation;
}
function hasConditionRowsValidationErrors(validation) {
  return Array.from(validation.values()).some(({
    key,
    value
  }) => key || value);
}
function getPortableTextDescriptionValue(description) {
  return !Array.isArray(description) || !description.every(isPortableTextBlock) ? "" : toPlainText(description);
}
function VariantForm(props) {
  const $ = c(108), {
    onChange,
    onConditionValidityChange,
    showValidation: t0,
    value
  } = props, showValidation = t0 === void 0 ? !1 : t0, {
    t
  } = useTranslation(variantsLocaleNamespace), {
    data: variants
  } = useAllVariants();
  let t1;
  $[0] !== variants ? (t1 = buildConditionSuggestionIndex(variants), $[0] = variants, $[1] = t1) : t1 = $[1];
  const suggestionIndex = t1, titleId = useId(), descriptionId = useId();
  let t2;
  $[2] !== value.conditions ? (t2 = () => getConditionRows(value.conditions), $[2] = value.conditions, $[3] = t2) : t2 = $[3];
  const [conditionRows, setConditionRows] = useState(t2);
  let t3;
  $[4] !== conditionRows ? (t3 = getConditionRowsValidation(conditionRows), $[4] = conditionRows, $[5] = t3) : t3 = $[5];
  const conditionsValidation = t3;
  let t4;
  $[6] !== value ? (t4 = getVariantTitleValue(value), $[6] = value, $[7] = t4) : t4 = $[7];
  const showTitleError = showValidation && !t4, lastConditionRow = conditionRows[conditionRows.length - 1];
  let t5;
  $[8] !== conditionsValidation || $[9] !== lastConditionRow ? (t5 = lastConditionRow && !hasConditionRowsValidationErrors(conditionsValidation), $[8] = conditionsValidation, $[9] = lastConditionRow, $[10] = t5) : t5 = $[10];
  const canAddCondition = !!t5;
  let t6;
  $[11] !== onChange || $[12] !== onConditionValidityChange ? (t6 = (nextRows) => {
    const rows = nextRows.length ? nextRows : getConditionRows({});
    setConditionRows(rows);
    const nextRowsValidation = getConditionRowsValidation(rows), nextRowsInvalid = hasConditionRowsValidationErrors(nextRowsValidation);
    onConditionValidityChange?.(nextRowsInvalid), (nextRows.length === 0 || !nextRowsInvalid) && onChange(["conditions"], getConditionsFromRows(nextRows));
  }, $[11] = onChange, $[12] = onConditionValidityChange, $[13] = t6) : t6 = $[13];
  const updateConditionRows = t6;
  let t7;
  $[14] !== onChange ? (t7 = (event) => {
    onChange(["metadata", "title"], event.currentTarget.value);
  }, $[14] = onChange, $[15] = t7) : t7 = $[15];
  const handleTitleChange = t7;
  let t8;
  $[16] !== onChange ? (t8 = (event_0) => {
    onChange(["metadata", "description"], createPortableTextDescription(event_0.currentTarget.value));
  }, $[16] = onChange, $[17] = t8) : t8 = $[17];
  const handleDescriptionChange = t8;
  let t9;
  $[18] !== conditionRows || $[19] !== updateConditionRows ? (t9 = (index, field, nextValue) => {
    const nextRows_0 = conditionRows.map((row, rowIndex) => rowIndex === index ? {
      ...row,
      [field]: nextValue
    } : row);
    updateConditionRows(nextRows_0);
  }, $[18] = conditionRows, $[19] = updateConditionRows, $[20] = t9) : t9 = $[20];
  const handleConditionChange = t9;
  let t10;
  $[21] !== canAddCondition || $[22] !== conditionRows || $[23] !== updateConditionRows ? (t10 = () => {
    canAddCondition && updateConditionRows([...conditionRows, {
      id: randomKey(12),
      key: "",
      value: ""
    }]);
  }, $[21] = canAddCondition, $[22] = conditionRows, $[23] = updateConditionRows, $[24] = t10) : t10 = $[24];
  const handleAddCondition = t10;
  let t11;
  $[25] !== conditionRows || $[26] !== updateConditionRows ? (t11 = (index_0) => {
    const nextRows_1 = conditionRows.filter((_, rowIndex_0) => rowIndex_0 !== index_0);
    updateConditionRows(nextRows_1);
  }, $[25] = conditionRows, $[26] = updateConditionRows, $[27] = t11) : t11 = $[27];
  const handleRemoveCondition = t11;
  let t12;
  $[28] !== t ? (t12 = t("dialog.create.variant-title.label"), $[28] = t, $[29] = t12) : t12 = $[29];
  let t13;
  $[30] !== t12 || $[31] !== titleId ? (t13 = /* @__PURE__ */ jsx(Text, { as: "label", htmlFor: titleId, size: 1, weight: "medium", children: t12 }), $[30] = t12, $[31] = titleId, $[32] = t13) : t13 = $[32];
  const t14 = showTitleError ? "true" : void 0;
  let t15;
  $[33] !== showTitleError || $[34] !== t ? (t15 = showTitleError ? t("dialog.create.variant-title.required") : void 0, $[33] = showTitleError, $[34] = t, $[35] = t15) : t15 = $[35];
  let t16;
  $[36] !== t ? (t16 = t("dialog.create.variant-title.placeholder"), $[36] = t, $[37] = t16) : t16 = $[37];
  const t17 = typeof value.metadata?.title == "string" ? value.metadata.title : "";
  let t18;
  $[38] !== handleTitleChange || $[39] !== t14 || $[40] !== t15 || $[41] !== t16 || $[42] !== t17 || $[43] !== titleId ? (t18 = /* @__PURE__ */ jsx(TextInput, { autoFocus: !0, "aria-invalid": t14, customValidity: t15, "data-testid": "variant-form-title", fontSize: 2, id: titleId, onChange: handleTitleChange, placeholder: t16, value: t17 }), $[38] = handleTitleChange, $[39] = t14, $[40] = t15, $[41] = t16, $[42] = t17, $[43] = titleId, $[44] = t18) : t18 = $[44];
  let t19;
  $[45] !== showTitleError || $[46] !== t ? (t19 = showTitleError && /* @__PURE__ */ jsx(TextWithTone, { "data-testid": "variant-form-title-error", size: 1, tone: "critical", children: t("dialog.create.variant-title.required") }), $[45] = showTitleError, $[46] = t, $[47] = t19) : t19 = $[47];
  let t20;
  $[48] !== t13 || $[49] !== t18 || $[50] !== t19 ? (t20 = /* @__PURE__ */ jsxs(Stack, { space: 3, children: [
    t13,
    t18,
    t19
  ] }), $[48] = t13, $[49] = t18, $[50] = t19, $[51] = t20) : t20 = $[51];
  let t21;
  $[52] !== t ? (t21 = t("dialog.create.description.label"), $[52] = t, $[53] = t21) : t21 = $[53];
  let t22;
  $[54] !== descriptionId || $[55] !== t21 ? (t22 = /* @__PURE__ */ jsx(Text, { as: "label", htmlFor: descriptionId, size: 1, weight: "medium", children: t21 }), $[54] = descriptionId, $[55] = t21, $[56] = t22) : t22 = $[56];
  let t23;
  $[57] !== t ? (t23 = t("dialog.create.description.placeholder"), $[57] = t, $[58] = t23) : t23 = $[58];
  const t24 = value.metadata?.description;
  let t25;
  $[59] !== t24 ? (t25 = getPortableTextDescriptionValue(t24), $[59] = t24, $[60] = t25) : t25 = $[60];
  let t26;
  $[61] !== descriptionId || $[62] !== handleDescriptionChange || $[63] !== t23 || $[64] !== t25 ? (t26 = /* @__PURE__ */ jsx(TextArea, { "data-testid": "variant-form-description", fontSize: 1, id: descriptionId, onChange: handleDescriptionChange, placeholder: t23, rows: 3, value: t25 }), $[61] = descriptionId, $[62] = handleDescriptionChange, $[63] = t23, $[64] = t25, $[65] = t26) : t26 = $[65];
  let t27;
  $[66] !== t22 || $[67] !== t26 ? (t27 = /* @__PURE__ */ jsxs(Stack, { space: 3, children: [
    t22,
    t26
  ] }), $[66] = t22, $[67] = t26, $[68] = t27) : t27 = $[68];
  let t28;
  $[69] !== t ? (t28 = t("dialog.create.conditions.title"), $[69] = t, $[70] = t28) : t28 = $[70];
  let t29;
  $[71] !== t28 ? (t29 = /* @__PURE__ */ jsx(Text, { size: 1, weight: "medium", children: t28 }), $[71] = t28, $[72] = t29) : t29 = $[72];
  let t30;
  $[73] !== t ? (t30 = t("dialog.create.conditions.description"), $[73] = t, $[74] = t30) : t30 = $[74];
  let t31;
  $[75] !== t30 ? (t31 = /* @__PURE__ */ jsx(Text, { muted: !0, size: 1, children: t30 }), $[75] = t30, $[76] = t31) : t31 = $[76];
  let t32;
  $[77] !== t29 || $[78] !== t31 ? (t32 = /* @__PURE__ */ jsxs(Stack, { space: 2, children: [
    t29,
    t31
  ] }), $[77] = t29, $[78] = t31, $[79] = t32) : t32 = $[79];
  let t33;
  $[80] !== conditionRows || $[81] !== conditionsValidation || $[82] !== handleConditionChange || $[83] !== handleRemoveCondition || $[84] !== showValidation || $[85] !== suggestionIndex || $[86] !== t ? (t33 = conditionRows.map((row_0, index_1) => {
    const validation = conditionsValidation.get(index_1) ?? getEmptyConditionRowValidation(), valueValidation = showValidation ? validation.value : null, conditionValidationError = validation.key || valueValidation;
    return /* @__PURE__ */ jsxs(Stack, { space: 2, children: [
      /* @__PURE__ */ jsxs(Flex, { align: "center", gap: 2, children: [
        /* @__PURE__ */ jsx(Box, { flex: 1, children: /* @__PURE__ */ jsx(ConditionAutocompleteInput, { autoFocus: index_1 > 0, ariaLabel: t("dialog.create.condition-key.label"), customValidity: validation.key ? t(validation.key) : void 0, invalid: !!validation.key, onChange: (nextValue_0) => handleConditionChange(index_1, "key", nextValue_0), options: getConditionKeyOptions(suggestionIndex, conditionRows, index_1), placeholder: t("dialog.create.condition-key.placeholder"), testId: "variant-form-condition-key", value: row_0.key }) }),
        /* @__PURE__ */ jsx(Box, { flex: 1, children: /* @__PURE__ */ jsx(ConditionAutocompleteInput, { ariaLabel: t("dialog.create.condition-value.label"), customValidity: valueValidation ? t(valueValidation) : void 0, invalid: !!valueValidation, onChange: (nextValue_1) => handleConditionChange(index_1, "value", nextValue_1), options: getConditionValueOptions(suggestionIndex, row_0.key), placeholder: t("dialog.create.condition-value.placeholder"), testId: "variant-form-condition-value", value: row_0.value }) }),
        /* @__PURE__ */ jsx(Button, { disabled: isConditionRowEmpty(row_0) && conditionRows.length === 1, icon: TrashIcon, mode: "bleed", onClick: () => handleRemoveCondition(index_1), tone: "critical", tooltipProps: {
          content: t("dialog.create.remove-condition")
        }, type: "button" })
      ] }),
      conditionValidationError && /* @__PURE__ */ jsx(TextWithTone, { "data-testid": validation.key ? "variant-form-condition-key-error" : "variant-form-condition-value-error", size: 1, tone: "critical", children: t(conditionValidationError) })
    ] }, row_0.id);
  }), $[80] = conditionRows, $[81] = conditionsValidation, $[82] = handleConditionChange, $[83] = handleRemoveCondition, $[84] = showValidation, $[85] = suggestionIndex, $[86] = t, $[87] = t33) : t33 = $[87];
  let t34;
  $[88] !== t33 ? (t34 = /* @__PURE__ */ jsx(Stack, { space: 2, children: t33 }), $[88] = t33, $[89] = t34) : t34 = $[89];
  const t35 = !canAddCondition;
  let t36;
  $[90] !== t ? (t36 = t("dialog.create.action.add-condition"), $[90] = t, $[91] = t36) : t36 = $[91];
  let t37;
  $[92] !== canAddCondition || $[93] !== t ? (t37 = canAddCondition ? null : {
    content: t("dialog.create.action.add-condition.disabled-hint")
  }, $[92] = canAddCondition, $[93] = t, $[94] = t37) : t37 = $[94];
  let t38;
  $[95] !== handleAddCondition || $[96] !== t35 || $[97] !== t36 || $[98] !== t37 ? (t38 = /* @__PURE__ */ jsx(Flex, { children: /* @__PURE__ */ jsx(Button, { disabled: t35, icon: AddIcon, mode: "ghost", onClick: handleAddCondition, text: t36, tooltipProps: t37, type: "button" }) }), $[95] = handleAddCondition, $[96] = t35, $[97] = t36, $[98] = t37, $[99] = t38) : t38 = $[99];
  let t39;
  $[100] !== t32 || $[101] !== t34 || $[102] !== t38 ? (t39 = /* @__PURE__ */ jsxs(Stack, { space: 3, children: [
    t32,
    t34,
    t38
  ] }), $[100] = t32, $[101] = t34, $[102] = t38, $[103] = t39) : t39 = $[103];
  let t40;
  return $[104] !== t20 || $[105] !== t27 || $[106] !== t39 ? (t40 = /* @__PURE__ */ jsxs(Stack, { space: 5, children: [
    t20,
    t27,
    t39
  ] }), $[104] = t20, $[105] = t27, $[106] = t39, $[107] = t40) : t40 = $[107], t40;
}
function VariantDialog(props) {
  const $ = c(35), {
    confirmDataTestId,
    confirmText,
    errorTitle,
    header,
    id,
    initialValue,
    onCancel,
    onSubmit,
    renderCancelButton: t0
  } = props, renderCancelButton = t0 === void 0 ? !1 : t0, toast = useToast(), {
    t
  } = useTranslation(variantsLocaleNamespace), [variant, setVariant] = useState(initialValue), [isSubmitting, setIsSubmitting] = useState(!1), [showValidation, setShowValidation] = useState(!1), [conditionsInvalid, setConditionsInvalid] = useState(!1);
  let t1;
  $[0] !== conditionsInvalid || $[1] !== variant ? (t1 = getIsVariantInvalid(variant) || conditionsInvalid, $[0] = conditionsInvalid, $[1] = variant, $[2] = t1) : t1 = $[2];
  const invalid = t1;
  let t2;
  $[3] === /* @__PURE__ */ Symbol.for("react.memo_cache_sentinel") ? (t2 = (path, nextValue) => {
    setVariant((currentVariant) => applyPatches([at(path, set(nextValue))], currentVariant));
  }, $[3] = t2) : t2 = $[3];
  const handleVariantChange = t2;
  let t3;
  $[4] !== errorTitle || $[5] !== invalid || $[6] !== onSubmit || $[7] !== toast || $[8] !== variant ? (t3 = async (event) => {
    if (event.preventDefault(), invalid) {
      setShowValidation(!0);
      return;
    }
    setIsSubmitting(!0);
    try {
      await onSubmit(variant), setIsSubmitting(!1);
    } catch (t42) {
      const error = t42;
      setIsSubmitting(!1), console.error(error), toast.push({
        closable: !0,
        status: "error",
        title: errorTitle
      });
    }
  }, $[4] = errorTitle, $[5] = invalid, $[6] = onSubmit, $[7] = toast, $[8] = variant, $[9] = t3) : t3 = $[9];
  const handleSubmit = t3, t4 = isSubmitting ? void 0 : onCancel, t5 = isSubmitting ? void 0 : onCancel;
  let t6;
  $[10] !== showValidation || $[11] !== variant ? (t6 = /* @__PURE__ */ jsx(Box, { paddingBottom: 4, children: /* @__PURE__ */ jsx(VariantForm, { onChange: handleVariantChange, onConditionValidityChange: setConditionsInvalid, showValidation, value: variant }) }), $[10] = showValidation, $[11] = variant, $[12] = t6) : t6 = $[12];
  let t7;
  $[13] !== isSubmitting || $[14] !== onCancel || $[15] !== renderCancelButton || $[16] !== t ? (t7 = renderCancelButton && /* @__PURE__ */ jsx(Button, { disabled: isSubmitting, mode: "ghost", onClick: onCancel, text: t("dialog.edit.action.cancel"), type: "button" }), $[13] = isSubmitting, $[14] = onCancel, $[15] = renderCancelButton, $[16] = t, $[17] = t7) : t7 = $[17];
  let t8;
  $[18] !== confirmDataTestId || $[19] !== confirmText || $[20] !== isSubmitting ? (t8 = /* @__PURE__ */ jsx(Button, { "data-testid": confirmDataTestId, disabled: isSubmitting, loading: isSubmitting, size: "large", text: confirmText, type: "submit" }), $[18] = confirmDataTestId, $[19] = confirmText, $[20] = isSubmitting, $[21] = t8) : t8 = $[21];
  let t9;
  $[22] !== t7 || $[23] !== t8 ? (t9 = /* @__PURE__ */ jsxs(Flex, { gap: 2, justify: "flex-end", paddingTop: 5, children: [
    t7,
    t8
  ] }), $[22] = t7, $[23] = t8, $[24] = t9) : t9 = $[24];
  let t10;
  $[25] !== handleSubmit || $[26] !== t6 || $[27] !== t9 ? (t10 = /* @__PURE__ */ jsx(Card, { borderTop: !0, padding: 4, children: /* @__PURE__ */ jsxs("form", { onSubmit: handleSubmit, children: [
    t6,
    t9
  ] }) }), $[25] = handleSubmit, $[26] = t6, $[27] = t9, $[28] = t10) : t10 = $[28];
  let t11;
  return $[29] !== header || $[30] !== id || $[31] !== t10 || $[32] !== t4 || $[33] !== t5 ? (t11 = /* @__PURE__ */ jsx(Dialog, { __unstable_autoFocus: !1, header, id, onClickOutside: t4, onClose: t5, padding: !1, width: 1, children: t10 }), $[29] = header, $[30] = id, $[31] = t10, $[32] = t4, $[33] = t5, $[34] = t11) : t11 = $[34], t11;
}
function toEditableVariant(variant) {
  return {
    _id: variant._id,
    _type: variant._type,
    conditions: variant.conditions,
    priority: variant.priority,
    metadata: variant.metadata
  };
}
function EditVariantDialog(props) {
  const $ = c(18), {
    onCancel,
    onSubmit,
    variant: initialVariant
  } = props, {
    t
  } = useTranslation(variantsLocaleNamespace), {
    updateVariant
  } = useVariantOperations();
  let t0;
  $[0] !== initialVariant ? (t0 = toEditableVariant(initialVariant), $[0] = initialVariant, $[1] = t0) : t0 = $[1];
  const initialValue = t0;
  let t1;
  $[2] !== onSubmit || $[3] !== updateVariant ? (t1 = async (variant) => {
    await updateVariant(variant), onSubmit();
  }, $[2] = onSubmit, $[3] = updateVariant, $[4] = t1) : t1 = $[4];
  const handleSubmit = t1;
  let t2;
  $[5] !== t ? (t2 = t("dialog.edit.action.confirm"), $[5] = t, $[6] = t2) : t2 = $[6];
  let t3;
  $[7] !== t ? (t3 = t("dialog.edit.error.title"), $[7] = t, $[8] = t3) : t3 = $[8];
  let t4;
  $[9] !== t ? (t4 = t("dialog.edit.title"), $[9] = t, $[10] = t4) : t4 = $[10];
  let t5;
  return $[11] !== handleSubmit || $[12] !== initialValue || $[13] !== onCancel || $[14] !== t2 || $[15] !== t3 || $[16] !== t4 ? (t5 = /* @__PURE__ */ jsx(VariantDialog, { confirmDataTestId: "save-variant-button", confirmText: t2, errorTitle: t3, header: t4, id: "edit-variant-dialog", initialValue, onCancel, onSubmit: handleSubmit, renderCancelButton: !0 }), $[11] = handleSubmit, $[12] = initialValue, $[13] = onCancel, $[14] = t2, $[15] = t3, $[16] = t4, $[17] = t5) : t5 = $[17], t5;
}
function toVariantDocumentVersion(document) {
  const system = document[DOCUMENT_SYSTEM_FIELD];
  return system?.group?._ref ? {
    documentId: document._id,
    bundleId: system.bundleId ?? void 0,
    releaseRef: system.release?._ref ?? null,
    updatedAt: document._updatedAt ?? ""
  } : null;
}
function useVariantDocuments(variantId) {
  const $ = c(15), enabled = !!variantId;
  let t0;
  $[0] !== variantId ? (t0 = variantId ? getVariantId(variantId) : "", $[0] = variantId, $[1] = t0) : t0 = $[1];
  let t1;
  $[2] !== t0 ? (t1 = {
    variantId: t0
  }, $[2] = t0, $[3] = t1) : t1 = $[3];
  const params = t1, t2 = enabled ? `variant-${variantId}` : "variant-disabled";
  let t3;
  $[4] !== enabled || $[5] !== params || $[6] !== t2 ? (t3 = {
    groqFilter: "sanity::partOfVariant($variantId)",
    params,
    cacheKey: t2,
    enabled
  }, $[4] = enabled, $[5] = params, $[6] = t2, $[7] = t3) : t3 = $[7];
  const {
    loading,
    results,
    error
  } = useBundleDocuments(t3);
  let t4;
  $[8] !== enabled || $[9] !== results ? (t4 = enabled ? toDocumentInVariants(results) : [], $[8] = enabled, $[9] = results, $[10] = t4) : t4 = $[10];
  const variantResults = t4, t5 = enabled && loading, t6 = enabled ? error : null;
  let t7;
  return $[11] !== t5 || $[12] !== t6 || $[13] !== variantResults ? (t7 = {
    loading: t5,
    results: variantResults,
    error: t6
  }, $[11] = t5, $[12] = t6, $[13] = variantResults, $[14] = t7) : t7 = $[14], t7;
}
function toDocumentInVariants(results) {
  return results.flatMap((result) => {
    const version = toVariantDocumentVersion(result.document);
    return version ? [{
      ...result,
      version
    }] : [];
  });
}
function VariantPinButton(t0) {
  const $ = c(22), {
    variant,
    "data-testid": t1
  } = t0, dataTestId = t1 === void 0 ? "pin-variant-button" : t1, {
    t
  } = useTranslation(variantsLocaleNamespace), {
    selectedVariant
  } = usePerspective(), setVariant = useSetVariant(), isPinned = selectedVariant?._id === variant._id;
  let t2;
  $[0] !== variant ? (t2 = getVariantTitle(variant), $[0] = variant, $[1] = t2) : t2 = $[1];
  const variantTitle = t2;
  let t3;
  $[2] !== isPinned || $[3] !== setVariant || $[4] !== variant._id ? (t3 = () => {
    setVariant(isPinned ? {
      variantId: void 0
    } : {
      variantId: variant._id
    });
  }, $[2] = isPinned, $[3] = setVariant, $[4] = variant._id, $[5] = t3) : t3 = $[5];
  const handlePinVariant = t3;
  let t4;
  $[6] !== isPinned || $[7] !== t || $[8] !== variantTitle ? (t4 = isPinned ? `${t("detail.unpin-variant")}: "${variantTitle}"` : `${t("detail.pin-variant")}: "${variantTitle}"`, $[6] = isPinned, $[7] = t, $[8] = variantTitle, $[9] = t4) : t4 = $[9];
  const t5 = isPinned ? PinFilledIcon : PinIcon;
  let t6;
  $[10] !== isPinned || $[11] !== t ? (t6 = t(isPinned ? "detail.unpin-variant" : "detail.pin-variant"), $[10] = isPinned, $[11] = t, $[12] = t6) : t6 = $[12];
  let t7;
  $[13] !== t6 ? (t7 = {
    content: t6
  }, $[13] = t6, $[14] = t7) : t7 = $[14];
  let t8;
  return $[15] !== dataTestId || $[16] !== handlePinVariant || $[17] !== isPinned || $[18] !== t4 || $[19] !== t5 || $[20] !== t7 ? (t8 = /* @__PURE__ */ jsx(Button, { "aria-label": t4, "aria-live": "assertive", "data-testid": dataTestId, icon: t5, mode: "bleed", onClick: handlePinVariant, radius: "full", selected: isPinned, tooltipProps: t7 }), $[15] = dataTestId, $[16] = handlePinVariant, $[17] = isPinned, $[18] = t4, $[19] = t5, $[20] = t7, $[21] = t8) : t8 = $[21], t8;
}
function getBundleSortOrder(bundleId) {
  return isPublishedBundleId(bundleId) ? 0 : bundleId === "drafts" ? 1 : 2;
}
function compareVariantDocumentVersions(left, right) {
  const leftOrder = getBundleSortOrder(left.bundleId), rightOrder = getBundleSortOrder(right.bundleId);
  return leftOrder !== rightOrder ? leftOrder - rightOrder : left.documentId.localeCompare(right.documentId);
}
function pickRepresentativeDocument(documents) {
  return documents.reduce((latest, current) => {
    const latestUpdatedAt = latest.document._updatedAt ?? "";
    return (current.document._updatedAt ?? "") > latestUpdatedAt ? current : latest;
  });
}
function aggregateValidation(documents) {
  const representative = pickRepresentativeDocument(documents);
  return {
    isValidating: documents.some((document) => document.validation.isValidating),
    validation: documents.flatMap((document) => document.validation.validation),
    revision: representative.validation.revision,
    hasError: documents.some((document) => document.validation.hasError)
  };
}
function groupVariantDocumentsByGroup(documents) {
  const groups = /* @__PURE__ */ new Map();
  for (const document of documents) {
    const groupId = document.document[DOCUMENT_SYSTEM_FIELD]?.group?._ref;
    if (!groupId)
      continue;
    const existingGroup = groups.get(groupId);
    existingGroup ? existingGroup.push(document) : groups.set(groupId, [document]);
  }
  return Array.from(groups.entries()).map(([groupId, groupDocuments]) => {
    const representative = pickRepresentativeDocument(groupDocuments), versions = groupDocuments.map((item) => item.version).toSorted(compareVariantDocumentVersions);
    return {
      ...representative,
      groupId,
      versions,
      validation: aggregateValidation(groupDocuments)
    };
  });
}
function DeleteVariantDialog(t0) {
  const $ = c(23), {
    isDeleting,
    onClose,
    onConfirm,
    variantTitle
  } = t0, {
    t
  } = useTranslation(variantsLocaleNamespace);
  let t1;
  $[0] !== isDeleting ? (t1 = {
    disabled: isDeleting
  }, $[0] = isDeleting, $[1] = t1) : t1 = $[1];
  let t2;
  $[2] !== t ? (t2 = t("dialog.delete.action.confirm"), $[2] = t, $[3] = t2) : t2 = $[3];
  let t3;
  $[4] !== isDeleting || $[5] !== onConfirm || $[6] !== t2 ? (t3 = {
    disabled: isDeleting,
    loading: isDeleting,
    onClick: onConfirm,
    text: t2,
    tone: "critical"
  }, $[4] = isDeleting, $[5] = onConfirm, $[6] = t2, $[7] = t3) : t3 = $[7];
  let t4;
  $[8] !== t1 || $[9] !== t3 ? (t4 = {
    cancelButton: t1,
    confirmButton: t3
  }, $[8] = t1, $[9] = t3, $[10] = t4) : t4 = $[10];
  let t5;
  $[11] !== t ? (t5 = t("dialog.delete.title"), $[11] = t, $[12] = t5) : t5 = $[12];
  const t6 = isDeleting ? void 0 : onClose;
  let t7;
  $[13] !== t || $[14] !== variantTitle ? (t7 = t("dialog.delete.description", {
    title: variantTitle
  }), $[13] = t, $[14] = variantTitle, $[15] = t7) : t7 = $[15];
  let t8;
  $[16] !== t7 ? (t8 = /* @__PURE__ */ jsx(Text, { muted: !0, size: 1, children: t7 }), $[16] = t7, $[17] = t8) : t8 = $[17];
  let t9;
  return $[18] !== t4 || $[19] !== t5 || $[20] !== t6 || $[21] !== t8 ? (t9 = /* @__PURE__ */ jsx(Dialog, { "data-testid": "delete-variant-dialog", footer: t4, header: t5, id: "delete-variant-dialog", onClose: t6, children: t8 }), $[18] = t4, $[19] = t5, $[20] = t6, $[21] = t8, $[22] = t9) : t9 = $[22], t9;
}
function useVariantDeleteAction(variantId, options) {
  const {
    documentCount,
    documentsLoading = !1,
    onDeleted,
    variantTitle = ""
  } = options ?? {}, {
    t
  } = useTranslation(variantsLocaleNamespace), toast = useToast(), {
    deleteVariant
  } = useVariantOperations(), [isDeleting, setIsDeleting] = useState(!1), [isDeleteDialogOpen, setIsDeleteDialogOpen] = useState(!1), hasDocuments = typeof documentCount == "number" && documentCount > 0, countUnknown = documentsLoading || documentCount === void 0 || documentCount === null, deleteDisabled = isDeleting || countUnknown || hasDocuments, deleteDisabledTooltip = useMemo(() => {
    if (!(!hasDocuments || typeof documentCount != "number"))
      return t(documentCount === 1 ? "overview.action.delete-variant.disabled-hint_one" : "overview.action.delete-variant.disabled-hint_other", {
        count: documentCount
      });
  }, [documentCount, hasDocuments, t]), handleDelete = useCallback(() => {
    countUnknown || hasDocuments || setIsDeleteDialogOpen(!0);
  }, [countUnknown, hasDocuments]), handleCloseDeleteDialog = useCallback(() => {
    isDeleting || setIsDeleteDialogOpen(!1);
  }, [isDeleting]), handleConfirmDelete = useCallback(async () => {
    if (!(countUnknown || hasDocuments)) {
      setIsDeleting(!0);
      try {
        await deleteVariant(variantId), setIsDeleting(!1), setIsDeleteDialogOpen(!1), onDeleted?.();
      } catch (error) {
        console.error(error), toast.push({
          closable: !0,
          status: "error",
          title: t("overview.action.delete-variant.error.title")
        }), setIsDeleting(!1);
      }
    }
  }, [countUnknown, deleteVariant, hasDocuments, onDeleted, t, toast, variantId]);
  return {
    deleteDisabled,
    deleteDisabledTooltip,
    handleCloseDeleteDialog,
    handleConfirmDelete,
    handleDelete,
    isDeleteDialogOpen,
    isDeleting,
    variantTitle
  };
}
function VariantDetailMenuButton(t0) {
  const $ = c(36), {
    documentCount,
    documentsLoading: t1,
    variant
  } = t0, documentsLoading = t1 === void 0 ? !1 : t1, {
    t
  } = useTranslation(variantsLocaleNamespace), router = useRouter();
  let t2;
  $[0] !== variant ? (t2 = getVariantTitle(variant), $[0] = variant, $[1] = t2) : t2 = $[1];
  const variantTitle = t2;
  let t3;
  $[2] !== router ? (t3 = () => router.navigate({}), $[2] = router, $[3] = t3) : t3 = $[3];
  let t4;
  $[4] !== documentCount || $[5] !== documentsLoading || $[6] !== t3 || $[7] !== variantTitle ? (t4 = {
    documentCount,
    documentsLoading,
    onDeleted: t3,
    variantTitle
  }, $[4] = documentCount, $[5] = documentsLoading, $[6] = t3, $[7] = variantTitle, $[8] = t4) : t4 = $[8];
  const {
    deleteDisabled,
    deleteDisabledTooltip,
    handleCloseDeleteDialog,
    handleConfirmDelete,
    handleDelete,
    isDeleteDialogOpen,
    isDeleting,
    variantTitle: dialogVariantTitle
  } = useVariantDeleteAction(variant._id, t4);
  let t5;
  $[9] !== isDeleting ? (t5 = /* @__PURE__ */ jsx(ContextMenuButton, { disabled: isDeleting, loading: isDeleting }), $[9] = isDeleting, $[10] = t5) : t5 = $[10];
  let t6;
  $[11] !== variant._id ? (t6 = getVariantId(variant._id), $[11] = variant._id, $[12] = t6) : t6 = $[12];
  const t7 = `variant-detail-actions-${t6}`;
  let t8;
  $[13] !== t ? (t8 = t("overview.action.delete-variant"), $[13] = t, $[14] = t8) : t8 = $[14];
  let t9;
  $[15] !== deleteDisabledTooltip ? (t9 = deleteDisabledTooltip ? {
    content: deleteDisabledTooltip
  } : void 0, $[15] = deleteDisabledTooltip, $[16] = t9) : t9 = $[16];
  let t10;
  $[17] !== deleteDisabled || $[18] !== handleDelete || $[19] !== t8 || $[20] !== t9 ? (t10 = /* @__PURE__ */ jsx(Menu, { children: /* @__PURE__ */ jsx(MenuItem, { disabled: deleteDisabled, onClick: handleDelete, text: t8, tone: "critical", tooltipProps: t9 }) }), $[17] = deleteDisabled, $[18] = handleDelete, $[19] = t8, $[20] = t9, $[21] = t10) : t10 = $[21];
  let t11;
  $[22] === /* @__PURE__ */ Symbol.for("react.memo_cache_sentinel") ? (t11 = {
    placement: "bottom-end",
    portal: !0
  }, $[22] = t11) : t11 = $[22];
  let t12;
  $[23] !== t10 || $[24] !== t5 || $[25] !== t7 ? (t12 = /* @__PURE__ */ jsx(MenuButton, { button: t5, id: t7, menu: t10, popover: t11 }), $[23] = t10, $[24] = t5, $[25] = t7, $[26] = t12) : t12 = $[26];
  let t13;
  $[27] !== dialogVariantTitle || $[28] !== handleCloseDeleteDialog || $[29] !== handleConfirmDelete || $[30] !== isDeleteDialogOpen || $[31] !== isDeleting ? (t13 = isDeleteDialogOpen && /* @__PURE__ */ jsx(DeleteVariantDialog, { isDeleting, onClose: handleCloseDeleteDialog, onConfirm: handleConfirmDelete, variantTitle: dialogVariantTitle }), $[27] = dialogVariantTitle, $[28] = handleCloseDeleteDialog, $[29] = handleConfirmDelete, $[30] = isDeleteDialogOpen, $[31] = isDeleting, $[32] = t13) : t13 = $[32];
  let t14;
  return $[33] !== t12 || $[34] !== t13 ? (t14 = /* @__PURE__ */ jsxs(Fragment, { children: [
    t12,
    t13
  ] }), $[33] = t12, $[34] = t13, $[35] = t14) : t14 = $[35], t14;
}
function VariantDetailFooter(t0) {
  const $ = c(24), {
    openEditDialog,
    documentCount,
    documentsLoading: t1,
    variant
  } = t0, documentsLoading = t1 === void 0 ? !1 : t1, {
    t
  } = useTranslation(variantsLocaleNamespace);
  let t2;
  $[0] === /* @__PURE__ */ Symbol.for("react.memo_cache_sentinel") ? (t2 = /* @__PURE__ */ jsx(Card, { borderTop: !0, marginX: 2, style: {
    opacity: 0.6
  } }), $[0] = t2) : t2 = $[0];
  let t3;
  $[1] === /* @__PURE__ */ Symbol.for("react.memo_cache_sentinel") ? (t3 = /* @__PURE__ */ jsx(Text, { size: 1, children: /* @__PURE__ */ jsx(DiamondIcon, {}) }), $[1] = t3) : t3 = $[1];
  let t4;
  $[2] !== t ? (t4 = t("detail.footer.created"), $[2] = t, $[3] = t4) : t4 = $[3];
  let t5;
  $[4] !== variant._createdAt ? (t5 = /* @__PURE__ */ jsx(RelativeTime, { time: variant._createdAt, useTemporalPhrase: !0, minimal: !0 }), $[4] = variant._createdAt, $[5] = t5) : t5 = $[5];
  let t6;
  $[6] !== t4 || $[7] !== t5 ? (t6 = /* @__PURE__ */ jsx(Flex, { flex: 1, gap: 1, children: /* @__PURE__ */ jsx(Card, { children: /* @__PURE__ */ jsxs(Flex, { align: "center", gap: 2, children: [
    t3,
    /* @__PURE__ */ jsxs(Text, { muted: !0, size: 1, children: [
      t4,
      " ",
      t5
    ] })
  ] }) }) }), $[6] = t4, $[7] = t5, $[8] = t6) : t6 = $[8];
  let t7;
  $[9] !== t ? (t7 = t("detail.action.edit-variant"), $[9] = t, $[10] = t7) : t7 = $[10];
  let t8;
  $[11] !== openEditDialog || $[12] !== t7 ? (t8 = /* @__PURE__ */ jsx(Button, { onClick: openEditDialog, text: t7 }), $[11] = openEditDialog, $[12] = t7, $[13] = t8) : t8 = $[13];
  let t9;
  $[14] !== documentCount || $[15] !== documentsLoading || $[16] !== variant ? (t9 = /* @__PURE__ */ jsx(VariantDetailMenuButton, { documentCount, documentsLoading, variant }), $[14] = documentCount, $[15] = documentsLoading, $[16] = variant, $[17] = t9) : t9 = $[17];
  let t10;
  $[18] !== t8 || $[19] !== t9 ? (t10 = /* @__PURE__ */ jsxs(Flex, { flex: "none", gap: 1, "data-testid": "variant-detail-footer-actions", children: [
    t8,
    t9
  ] }), $[18] = t8, $[19] = t9, $[20] = t10) : t10 = $[20];
  let t11;
  return $[21] !== t10 || $[22] !== t6 ? (t11 = /* @__PURE__ */ jsxs(Card, { flex: "none", children: [
    t2,
    /* @__PURE__ */ jsxs(Flex, { padding: 3, children: [
      t6,
      t10
    ] })
  ] }), $[21] = t10, $[22] = t6, $[23] = t11) : t11 = $[23], t11;
}
function getDocumentPreviewTitle(document) {
  const title = document.title || document.name;
  return typeof title == "string" && title.trim() ? title : document._id;
}
function getReleaseDocumentForVersion(version, releasesById) {
  if (version.releaseRef)
    return releasesById.get(version.releaseRef);
  if (isReleaseBundle(version.bundleId)) {
    const releaseDocumentId = getReleaseDocumentIdFromReleaseId(version.bundleId);
    return releasesById.get(releaseDocumentId);
  }
}
function ReleaseBundleChip(t0) {
  const $ = c(10), {
    release
  } = t0, {
    t
  } = useTranslation();
  let t1;
  $[0] !== release._id ? (t1 = getReleaseIdFromReleaseDocumentId(release._id), $[0] = release._id, $[1] = t1) : t1 = $[1];
  const releaseId = t1, t2 = release.metadata?.title;
  let t3;
  $[2] !== t ? (t3 = t("release.placeholder-untitled-release"), $[2] = t, $[3] = t3) : t3 = $[3];
  let t4;
  $[4] !== releaseId ? (t4 = (t52) => {
    const {
      displayTitle
    } = t52;
    return /* @__PURE__ */ jsx(IntentLink, { intent: RELEASES_INTENT, params: {
      id: releaseId
    }, children: /* @__PURE__ */ jsx(Badge, { radius: 2, tone: "primary", children: displayTitle }) });
  }, $[4] = releaseId, $[5] = t4) : t4 = $[5];
  let t5;
  return $[6] !== t2 || $[7] !== t3 || $[8] !== t4 ? (t5 = /* @__PURE__ */ jsx(ReleaseTitle, { title: t2, fallback: t3, children: t4 }), $[6] = t2, $[7] = t3, $[8] = t4, $[9] = t5) : t5 = $[9], t5;
}
function StaticBundleChip(t0) {
  const $ = c(3), {
    label,
    tone
  } = t0;
  let t1;
  return $[0] !== label || $[1] !== tone ? (t1 = /* @__PURE__ */ jsx(Badge, { radius: 2, tone, children: label }), $[0] = label, $[1] = tone, $[2] = t1) : t1 = $[2], t1;
}
function VariantDocumentBundleChips(t0) {
  const $ = c(11), {
    versions,
    releasesById
  } = t0, {
    t
  } = useTranslation();
  let t1;
  $[0] === /* @__PURE__ */ Symbol.for("react.memo_cache_sentinel") ? (t1 = {
    minWidth: 0,
    width: "100%",
    overflowX: "auto",
    overflowY: "hidden"
  }, $[0] = t1) : t1 = $[0];
  let t2;
  $[1] === /* @__PURE__ */ Symbol.for("react.memo_cache_sentinel") ? (t2 = {
    width: "max-content"
  }, $[1] = t2) : t2 = $[1];
  let t3;
  if ($[2] !== releasesById || $[3] !== t || $[4] !== versions) {
    let t42;
    $[6] !== releasesById || $[7] !== t ? (t42 = (version) => {
      if (isPublishedBundleId(version.bundleId))
        return /* @__PURE__ */ jsx(StaticBundleChip, { label: t("release.chip.published"), tone: "positive" }, version.documentId);
      if (version.bundleId === "drafts")
        return /* @__PURE__ */ jsx(StaticBundleChip, { label: t("release.chip.draft"), tone: "default" }, version.documentId);
      const release = getReleaseDocumentForVersion(version, releasesById);
      return release ? /* @__PURE__ */ jsx(ReleaseBundleChip, { release }, version.documentId) : /* @__PURE__ */ jsx(StaticBundleChip, { label: t("release.placeholder-untitled-release"), tone: "default" }, version.documentId);
    }, $[6] = releasesById, $[7] = t, $[8] = t42) : t42 = $[8], t3 = versions.map(t42), $[2] = releasesById, $[3] = t, $[4] = versions, $[5] = t3;
  } else
    t3 = $[5];
  let t4;
  return $[9] !== t3 ? (t4 = /* @__PURE__ */ jsx(Box, { style: t1, children: /* @__PURE__ */ jsx(Flex, { align: "center", gap: 1, wrap: "nowrap", style: t2, children: t3 }) }), $[9] = t3, $[10] = t4) : t4 = $[10], t4;
}
const VariantDocumentPreviewLink = forwardRef(function(t0, ref) {
  const $ = c(13);
  let documentType, linkProps, publishedId, searchParams;
  $[0] !== t0 ? ({
    publishedId,
    documentType,
    searchParams,
    ...linkProps
  } = t0, $[0] = t0, $[1] = documentType, $[2] = linkProps, $[3] = publishedId, $[4] = searchParams) : (documentType = $[1], linkProps = $[2], publishedId = $[3], searchParams = $[4]);
  let t1;
  $[5] !== documentType || $[6] !== publishedId ? (t1 = {
    id: publishedId,
    type: documentType
  }, $[5] = documentType, $[6] = publishedId, $[7] = t1) : t1 = $[7];
  let t2;
  return $[8] !== linkProps || $[9] !== ref || $[10] !== searchParams || $[11] !== t1 ? (t2 = /* @__PURE__ */ jsx(IntentLink, { ...linkProps, intent: "edit", params: t1, searchParams, ref }), $[8] = linkProps, $[9] = ref, $[10] = searchParams, $[11] = t1, $[12] = t2) : t2 = $[12], t2;
});
function getVersionPerspectiveNavigation(bundleId, variantId) {
  const searchParams = variantId ? [["variant", variantId]] : [];
  return isPublishedBundleId(bundleId) ? {
    perspectiveStack: ["published"],
    searchParams: [...searchParams, ["perspective", "published"]]
  } : bundleId === "drafts" ? {
    perspectiveStack: ["drafts"],
    searchParams: searchParams.length > 0 ? searchParams : void 0
  } : {
    perspectiveStack: [bundleId ?? "", "drafts"],
    searchParams: [...searchParams, ["perspective", bundleId ?? ""]]
  };
}
function VariantDocumentPreview(t0) {
  const $ = c(23), {
    row,
    layout,
    variantId
  } = t0;
  let t1;
  $[0] !== row.groupId ? (t1 = getPublishedId(row.groupId), $[0] = row.groupId, $[1] = t1) : t1 = $[1];
  const publishedId = t1, documentPresence = useDocumentPresence(row.document._id);
  let t2;
  $[2] !== row.version.bundleId || $[3] !== variantId ? (t2 = getVersionPerspectiveNavigation(row.version.bundleId, variantId), $[2] = row.version.bundleId, $[3] = variantId, $[4] = t2) : t2 = $[4];
  const {
    perspectiveStack,
    searchParams
  } = t2;
  let t3;
  $[5] !== documentPresence ? (t3 = documentPresence?.length > 0 && /* @__PURE__ */ jsx(DocumentPreviewPresence, { presence: documentPresence }), $[5] = documentPresence, $[6] = t3) : t3 = $[6];
  const previewPresence = t3;
  let t4;
  $[7] !== perspectiveStack || $[8] !== row.document._id || $[9] !== row.document._type ? (t4 = {
    documentId: row.document._id,
    documentType: row.document._type,
    perspectiveStack
  }, $[7] = perspectiveStack, $[8] = row.document._id, $[9] = row.document._type, $[10] = t4) : t4 = $[10];
  const {
    isLoading: previewLoading,
    value: resolvedPreview
  } = useDocumentPreviewValues(t4);
  let t5;
  $[11] !== resolvedPreview ? (t5 = resolvedPreview || {}, $[11] = resolvedPreview, $[12] = t5) : t5 = $[12];
  let t6;
  $[13] !== layout || $[14] !== previewLoading || $[15] !== previewPresence || $[16] !== t5 ? (t6 = /* @__PURE__ */ jsx(SanityDefaultPreview, { ...t5, status: previewPresence, isPlaceholder: previewLoading, layout }), $[13] = layout, $[14] = previewLoading, $[15] = previewPresence, $[16] = t5, $[17] = t6) : t6 = $[17];
  let t7;
  return $[18] !== publishedId || $[19] !== row.document._type || $[20] !== searchParams || $[21] !== t6 ? (t7 = /* @__PURE__ */ jsx(Card, { tone: "inherit", as: VariantDocumentPreviewLink, publishedId, documentType: row.document._type, searchParams, radius: 2, "data-as": "a", children: t6 }), $[18] = publishedId, $[19] = row.document._type, $[20] = searchParams, $[21] = t6, $[22] = t7) : t7 = $[22], t7;
}
const MemoVariantDocumentPreview = memo(function(t0) {
  const $ = c(3), {
    row,
    variantId
  } = t0;
  let t1;
  return $[0] !== row || $[1] !== variantId ? (t1 = /* @__PURE__ */ jsx(VariantDocumentPreview, { row, variantId }), $[0] = row, $[1] = variantId, $[2] = t1) : t1 = $[2], t1;
}, (prev, next) => prev.row.memoKey === next.row.memoKey && prev.variantId === next.variantId), MemoDocumentType = memo(function(t0) {
  const $ = c(5), {
    type
  } = t0, schema = useSchema();
  let t1;
  $[0] !== schema || $[1] !== type ? (t1 = schema.get(type), $[0] = schema, $[1] = type, $[2] = t1) : t1 = $[2];
  const t2 = t1?.title || type;
  let t3;
  return $[3] !== t2 ? (t3 = /* @__PURE__ */ jsx(Text, { size: 1, children: t2 }), $[3] = t2, $[4] = t3) : t3 = $[4], t3;
}, (prev, next) => prev.type === next.type), getVariantDocumentTableColumnDefs = (t, variantId, releasesById) => [{
  id: "documentGroup",
  hidden: !0,
  width: null,
  sorting: !0,
  sortTransform: (row) => row.groupId
}, {
  id: "bundle",
  width: 140,
  style: {
    minWidth: 100,
    maxWidth: 140
  },
  sorting: !1,
  header: (props) => /* @__PURE__ */ jsx(Flex, { ...props.headerProps, paddingY: 3, sizing: "border", children: /* @__PURE__ */ jsx(Headers.BasicHeader, { text: t("detail.documents.table.bundle") }) }),
  cell: ({
    cellProps,
    datum
  }) => /* @__PURE__ */ jsx(Flex, { align: "center", ...cellProps, style: {
    ...cellProps.style,
    flex: "0 1 auto",
    minWidth: 0,
    overflow: "hidden"
  }, children: /* @__PURE__ */ jsx(Box, { flex: 1, paddingX: 2, style: {
    minWidth: 0,
    overflow: "hidden"
  }, children: !datum.isLoading && /* @__PURE__ */ jsx(VariantDocumentBundleChips, { versions: datum.versions, releasesById }) }) })
}, {
  id: "document._type",
  width: null,
  style: {
    minWidth: 100
  },
  sorting: !0,
  header: (props) => /* @__PURE__ */ jsx(Flex, { ...props.headerProps, paddingY: 3, sizing: "border", children: /* @__PURE__ */ jsx(Headers.SortHeaderButton, { text: t("detail.documents.table.type"), ...props }) }),
  cell: ({
    cellProps,
    datum
  }) => /* @__PURE__ */ jsx(Flex, { align: "center", ...cellProps, children: /* @__PURE__ */ jsx(Box, { paddingX: 2, children: !datum.isLoading && /* @__PURE__ */ jsx(MemoDocumentType, { type: datum.document._type }) }) })
}, {
  id: "search",
  width: null,
  style: {
    minWidth: "min(50%, calc(100vw - 80px))",
    maxWidth: "min(50%, calc(100vw - 80px))"
  },
  sortTransform: ({
    document
  }) => getDocumentPreviewTitle(document).toLowerCase(),
  header: (props) => /* @__PURE__ */ jsx(Headers.TableHeaderSearch, { ...props, placeholder: t("detail.documents.table.search-placeholder") }),
  cell: ({
    cellProps,
    datum
  }) => /* @__PURE__ */ jsx(Box, { ...cellProps, flex: 1, padding: 1, paddingRight: 2, sizing: "border", children: datum.isLoading ? /* @__PURE__ */ jsx(SanityDefaultPreview, { isPlaceholder: !0 }) : /* @__PURE__ */ jsx(MemoVariantDocumentPreview, { row: datum, variantId }) })
}, {
  id: "document._updatedAt",
  sorting: !0,
  width: 130,
  header: (props) => /* @__PURE__ */ jsx(Flex, { ...props.headerProps, paddingY: 3, sizing: "border", children: /* @__PURE__ */ jsx(Headers.SortHeaderButton, { text: t("detail.documents.table.edited"), ...props }) }),
  cell: ({
    cellProps,
    datum
  }) => /* @__PURE__ */ jsx(Flex, { ...cellProps, align: "center", paddingX: 2, paddingY: 3, style: {
    minWidth: 130
  }, children: !datum.isLoading && datum.document._updatedAt && /* @__PURE__ */ jsx(Text, { muted: !0, size: 1, children: /* @__PURE__ */ jsx(RelativeTime, { time: datum.document._updatedAt, useTemporalPhrase: !0, minimal: !0 }) }) })
}, {
  id: "validation",
  sorting: !1,
  width: 50,
  header: ({
    headerProps
  }) => /* @__PURE__ */ jsx(Flex, { ...headerProps, paddingY: 3, sizing: "border", children: /* @__PURE__ */ jsx(Headers.BasicHeader, { text: "" }) }),
  cell: ({
    cellProps,
    datum
  }) => {
    if (datum.isLoading) return null;
    const validationErrorCount = datum.validation.validation.filter((validation) => validation.level === "error").length;
    return /* @__PURE__ */ jsx(Flex, { ...cellProps, flex: 1, padding: 1, justify: "center", align: "center", sizing: "border", children: datum.validation.hasError && /* @__PURE__ */ jsx(Tooltip, { portal: !0, placement: "bottom-end", content: /* @__PURE__ */ jsx(Text, { muted: !0, size: 1, children: /* @__PURE__ */ jsxs(Flex, { align: "center", gap: 3, padding: 1, children: [
      /* @__PURE__ */ jsx(ToneIcon, { icon: ErrorOutlineIcon, tone: "critical" }),
      t(validationErrorCount === 1 ? "detail.documents.table.validation.error_one" : "detail.documents.table.validation.error_other", {
        count: validationErrorCount
      })
    ] }) }), children: /* @__PURE__ */ jsx(Text, { size: 1, children: /* @__PURE__ */ jsx(ToneIcon, { icon: ErrorOutlineIcon, tone: "critical" }) }) }) });
  }
}], TABLE_CARD_STYLE = {
  height: "100%",
  overflow: "auto"
};
function filterDocuments(rows, searchTerm) {
  return searchTerm.trim() ? rows.filter(({
    document
  }) => searchDocumentRelease(document, searchTerm)) : rows;
}
function VariantDocumentsTable(t0) {
  const $ = c(15), {
    rows,
    loading: t1,
    variantId
  } = t0, loading = t1 === void 0 ? !1 : t1, {
    t
  } = useTranslation(variantsLocaleNamespace), [scrollContainerRef, setScrollContainerRef] = useState(null), {
    data: releases
  } = useActiveReleases();
  let t2;
  $[0] !== releases ? (t2 = new Map(releases.map(_temp$2)), $[0] = releases, $[1] = t2) : t2 = $[1];
  const releasesById = t2;
  let t3;
  $[2] !== releasesById || $[3] !== t || $[4] !== variantId ? (t3 = getVariantDocumentTableColumnDefs(t, variantId, releasesById), $[2] = releasesById, $[3] = t, $[4] = variantId, $[5] = t3) : t3 = $[5];
  const columnDefs = t3;
  let t4;
  $[6] === /* @__PURE__ */ Symbol.for("react.memo_cache_sentinel") ? (t4 = {
    column: "documentGroup",
    direction: "asc"
  }, $[6] = t4) : t4 = $[6];
  let t5;
  $[7] !== t ? (t5 = t("detail.documents.no-documents"), $[7] = t, $[8] = t5) : t5 = $[8];
  let t6;
  return $[9] !== columnDefs || $[10] !== loading || $[11] !== rows || $[12] !== scrollContainerRef || $[13] !== t5 ? (t6 = /* @__PURE__ */ jsx(Card, { flex: 1, ref: setScrollContainerRef, style: TABLE_CARD_STYLE, children: /* @__PURE__ */ jsx(Table, { columnDefs, data: rows, defaultSort: t4, emptyState: t5, loading, rowId: "groupId", scrollContainerRef, searchFilter: filterDocuments }) }), $[9] = columnDefs, $[10] = loading, $[11] = rows, $[12] = scrollContainerRef, $[13] = t5, $[14] = t6) : t6 = $[14], t6;
}
function _temp$2(release) {
  return [release._id, release];
}
function VariantDetail() {
  const $ = c(90), router = useRouter(), {
    t
  } = useTranslation(variantsLocaleNamespace), [editDialogOpen, setEditDialogOpen] = useState(!1), variantIdRaw = typeof router.state.variantId == "string" ? router.state.variantId : void 0, variantId = decodeVariantIdFromRoute(variantIdRaw), {
    byId,
    loading
  } = useAllVariants(), variant = variantId ? byId.get(variantId) : void 0, {
    loading: documentsLoading,
    results: variantDocuments,
    error: variantDocumentsError
  } = useVariantDocuments(variant?._id);
  let t0;
  $[0] !== variantDocuments ? (t0 = groupVariantDocumentsByGroup(variantDocuments), $[0] = variantDocuments, $[1] = t0) : t0 = $[1];
  const tableRows = t0;
  if (loading) {
    let t110;
    $[2] !== t ? (t110 = t("detail.loading"), $[2] = t, $[3] = t110) : t110 = $[3];
    let t25;
    return $[4] !== t110 ? (t25 = /* @__PURE__ */ jsx(LoadingBlock, { fill: !0, title: t110 }), $[4] = t110, $[5] = t25) : t25 = $[5], t25;
  }
  if (!variant) {
    let t110;
    $[6] !== router ? (t110 = () => router.navigate({}), $[6] = router, $[7] = t110) : t110 = $[7];
    let t25;
    $[8] !== t ? (t25 = t("detail.back"), $[8] = t, $[9] = t25) : t25 = $[9];
    let t32;
    $[10] !== t110 || $[11] !== t25 ? (t32 = /* @__PURE__ */ jsx(Card, { borderBottom: !0, flex: "none", padding: 3, children: /* @__PURE__ */ jsx(Button, { mode: "bleed", onClick: t110, text: t25 }) }), $[10] = t110, $[11] = t25, $[12] = t32) : t32 = $[12];
    let t42;
    $[13] !== t ? (t42 = t("detail.not-found.title"), $[13] = t, $[14] = t42) : t42 = $[14];
    let t52;
    $[15] !== t42 ? (t52 = /* @__PURE__ */ jsx(Text, { size: 2, weight: "semibold", children: t42 }), $[15] = t42, $[16] = t52) : t52 = $[16];
    let t62;
    $[17] !== t ? (t62 = t("detail.not-found.description"), $[17] = t, $[18] = t62) : t62 = $[18];
    let t72;
    $[19] !== t62 ? (t72 = /* @__PURE__ */ jsx(Text, { muted: !0, size: 1, children: t62 }), $[19] = t62, $[20] = t72) : t72 = $[20];
    let t82;
    $[21] !== t52 || $[22] !== t72 ? (t82 = /* @__PURE__ */ jsx(Box, { padding: 4, children: /* @__PURE__ */ jsx(Card, { border: !0, padding: 4, radius: 3, children: /* @__PURE__ */ jsxs(Stack, { space: 3, children: [
      t52,
      t72
    ] }) }) }), $[21] = t52, $[22] = t72, $[23] = t82) : t82 = $[23];
    let t92;
    return $[24] !== t32 || $[25] !== t82 ? (t92 = /* @__PURE__ */ jsxs(Flex, { direction: "column", flex: 1, height: "fill", children: [
      t32,
      t82
    ] }), $[24] = t32, $[25] = t82, $[26] = t92) : t92 = $[26], t92;
  }
  const description = getVariantDescription(variant), conditionsText = getVariantConditionsText(variant.conditions);
  let t1;
  $[27] !== router ? (t1 = () => router.navigate({}), $[27] = router, $[28] = t1) : t1 = $[28];
  let t2;
  $[29] !== t ? (t2 = t("detail.back"), $[29] = t, $[30] = t2) : t2 = $[30];
  let t3;
  $[31] !== t1 || $[32] !== t2 ? (t3 = /* @__PURE__ */ jsx(Card, { borderBottom: !0, flex: "none", padding: 3, children: /* @__PURE__ */ jsx(Button, { mode: "bleed", onClick: t1, text: t2 }) }), $[31] = t1, $[32] = t2, $[33] = t3) : t3 = $[33];
  let t4;
  $[34] === /* @__PURE__ */ Symbol.for("react.memo_cache_sentinel") ? (t4 = {
    minHeight: 0
  }, $[34] = t4) : t4 = $[34];
  let t5;
  $[35] !== variant ? (t5 = /* @__PURE__ */ jsx(VariantPinButton, { variant }), $[35] = variant, $[36] = t5) : t5 = $[36];
  let t6;
  $[37] !== variant ? (t6 = getVariantTitle(variant), $[37] = variant, $[38] = t6) : t6 = $[38];
  let t7;
  $[39] !== t6 ? (t7 = /* @__PURE__ */ jsx(Text, { as: "h1", size: 4, weight: "bold", children: t6 }), $[39] = t6, $[40] = t7) : t7 = $[40];
  let t8;
  $[41] !== t5 || $[42] !== t7 ? (t8 = /* @__PURE__ */ jsxs(Flex, { align: "center", gap: 1, children: [
    t5,
    t7
  ] }), $[41] = t5, $[42] = t7, $[43] = t8) : t8 = $[43];
  let t9;
  $[44] !== description || $[45] !== t ? (t9 = description || t("detail.no-description"), $[44] = description, $[45] = t, $[46] = t9) : t9 = $[46];
  let t10;
  $[47] !== t9 ? (t10 = /* @__PURE__ */ jsx(Text, { muted: !0, size: 1, children: t9 }), $[47] = t9, $[48] = t10) : t10 = $[48];
  let t11;
  $[49] !== t10 || $[50] !== t8 ? (t11 = /* @__PURE__ */ jsx(Flex, { align: "flex-start", gap: 4, justify: "space-between", children: /* @__PURE__ */ jsxs(Stack, { space: 3, children: [
    t8,
    t10
  ] }) }), $[49] = t10, $[50] = t8, $[51] = t11) : t11 = $[51];
  let t12;
  $[52] !== t ? (t12 = t("dialog.create.conditions.title"), $[52] = t, $[53] = t12) : t12 = $[53];
  let t13;
  $[54] !== t12 ? (t13 = /* @__PURE__ */ jsx(Text, { size: 1, weight: "medium", children: t12 }), $[54] = t12, $[55] = t13) : t13 = $[55];
  let t14;
  $[56] !== conditionsText || $[57] !== t ? (t14 = conditionsText || t("overview.table.no-conditions"), $[56] = conditionsText, $[57] = t, $[58] = t14) : t14 = $[58];
  let t15;
  $[59] !== t14 ? (t15 = /* @__PURE__ */ jsx(Text, { muted: !0, size: 1, children: t14 }), $[59] = t14, $[60] = t15) : t15 = $[60];
  let t16;
  $[61] !== t13 || $[62] !== t15 ? (t16 = /* @__PURE__ */ jsx(Box, { paddingTop: 4, children: /* @__PURE__ */ jsxs(Stack, { space: 2, children: [
    t13,
    t15
  ] }) }), $[61] = t13, $[62] = t15, $[63] = t16) : t16 = $[63];
  let t17;
  $[64] !== t11 || $[65] !== t16 ? (t17 = /* @__PURE__ */ jsx(Container, { flex: "none", width: 3, children: /* @__PURE__ */ jsx(Flex, { direction: "column", paddingX: 3, children: /* @__PURE__ */ jsxs(Card, { paddingY: 5, children: [
    t11,
    t16
  ] }) }) }), $[64] = t11, $[65] = t16, $[66] = t17) : t17 = $[66];
  let t18;
  $[67] === /* @__PURE__ */ Symbol.for("react.memo_cache_sentinel") ? (t18 = {
    minHeight: 0
  }, $[67] = t18) : t18 = $[67];
  let t19;
  $[68] !== documentsLoading || $[69] !== t || $[70] !== tableRows || $[71] !== variant || $[72] !== variantDocumentsError ? (t19 = /* @__PURE__ */ jsx(Flex, { direction: "column", flex: 1, overflow: "hidden", style: t18, children: variantDocumentsError ? /* @__PURE__ */ jsx(Box, { padding: 4, children: /* @__PURE__ */ jsx(Text, { muted: !0, size: 1, children: t("detail.documents.error") }) }) : /* @__PURE__ */ jsx(VariantDocumentsTable, { loading: documentsLoading, rows: tableRows, variantId: getVariantId(variant._id) }) }), $[68] = documentsLoading, $[69] = t, $[70] = tableRows, $[71] = variant, $[72] = variantDocumentsError, $[73] = t19) : t19 = $[73];
  let t20;
  $[74] !== t17 || $[75] !== t19 ? (t20 = /* @__PURE__ */ jsxs(Flex, { direction: "column", flex: 1, height: "fill", overflow: "hidden", style: t4, children: [
    t17,
    t19
  ] }), $[74] = t17, $[75] = t19, $[76] = t20) : t20 = $[76];
  let t21;
  $[77] === /* @__PURE__ */ Symbol.for("react.memo_cache_sentinel") ? (t21 = () => setEditDialogOpen(!0), $[77] = t21) : t21 = $[77];
  let t22;
  $[78] !== documentsLoading || $[79] !== tableRows.length || $[80] !== variant ? (t22 = /* @__PURE__ */ jsx(VariantDetailFooter, { openEditDialog: t21, documentCount: tableRows.length, documentsLoading, variant }), $[78] = documentsLoading, $[79] = tableRows.length, $[80] = variant, $[81] = t22) : t22 = $[81];
  let t23;
  $[82] !== editDialogOpen || $[83] !== variant ? (t23 = editDialogOpen && /* @__PURE__ */ jsx(EditVariantDialog, { onCancel: () => setEditDialogOpen(!1), onSubmit: () => setEditDialogOpen(!1), variant }), $[82] = editDialogOpen, $[83] = variant, $[84] = t23) : t23 = $[84];
  let t24;
  return $[85] !== t20 || $[86] !== t22 || $[87] !== t23 || $[88] !== t3 ? (t24 = /* @__PURE__ */ jsxs(Flex, { direction: "column", flex: 1, height: "fill", overflow: "hidden", children: [
    t3,
    t20,
    t22,
    t23
  ] }), $[85] = t20, $[86] = t22, $[87] = t23, $[88] = t3, $[89] = t24) : t24 = $[89], t24;
}
function CreateVariantDialog(props) {
  const $ = c(16), {
    onCancel,
    onSubmit
  } = props, {
    t
  } = useTranslation(variantsLocaleNamespace), {
    createVariant
  } = useVariantOperations();
  let t0;
  $[0] === /* @__PURE__ */ Symbol.for("react.memo_cache_sentinel") ? (t0 = getVariantDefaults(), $[0] = t0) : t0 = $[0];
  const initialValue = t0;
  let t1;
  $[1] !== createVariant || $[2] !== onSubmit ? (t1 = async (variant) => {
    await createVariant(variant), onSubmit(variant._id);
  }, $[1] = createVariant, $[2] = onSubmit, $[3] = t1) : t1 = $[3];
  const handleSubmit = t1;
  let t2;
  $[4] !== t ? (t2 = t("dialog.create.action.confirm"), $[4] = t, $[5] = t2) : t2 = $[5];
  let t3;
  $[6] !== t ? (t3 = t("dialog.create.error.title"), $[6] = t, $[7] = t3) : t3 = $[7];
  let t4;
  $[8] !== t ? (t4 = t("dialog.create.title"), $[8] = t, $[9] = t4) : t4 = $[9];
  let t5;
  return $[10] !== handleSubmit || $[11] !== onCancel || $[12] !== t2 || $[13] !== t3 || $[14] !== t4 ? (t5 = /* @__PURE__ */ jsx(VariantDialog, { confirmDataTestId: "submit-variant-button", confirmText: t2, errorTitle: t3, header: t4, id: "create-variant-dialog", initialValue, onCancel, onSubmit: handleSubmit }), $[10] = handleSubmit, $[11] = onCancel, $[12] = t2, $[13] = t3, $[14] = t4, $[15] = t5) : t5 = $[15], t5;
}
function buildVariantsDocumentCountsQuery(variantDocumentIds) {
  const projections = [], filters = [];
  for (const variantDocumentId of variantDocumentIds) {
    const membership = `sanity::partOfVariant(${JSON.stringify(getVariantId(variantDocumentId))})`;
    projections.push(`${JSON.stringify(variantDocumentId)}: count(array::unique(*[${membership}]._system.group._ref))`), filters.push(membership);
  }
  return {
    fetch: `{${projections.join(",")}}`,
    listen: `*[${filters.join(" || ")}]`
  };
}
const INITIAL_STATE = {
  data: null,
  loading: !0,
  error: null
}, EMPTY_STATE = {
  data: {},
  loading: !1,
  error: null
};
function listenVariantsDocumentCounts(client, variantDocumentIds) {
  if (variantDocumentIds.length === 0)
    return of(EMPTY_STATE);
  const {
    fetch,
    listen
  } = buildVariantsDocumentCountsQuery(variantDocumentIds);
  return listenQuery(
    client,
    {
      fetch,
      listen
    },
    {},
    // Variant-scoped documents are version documents, which are only visible to queries
    // using the raw perspective.
    {
      tag: "variants-doc-counts.listen",
      perspective: "raw"
    }
  ).pipe(
    map((response) => ({
      data: Object.fromEntries(variantDocumentIds.map((variantDocumentId) => [variantDocumentId, response?.[variantDocumentId] ?? 0])),
      loading: !1,
      error: null
    })),
    startWith(INITIAL_STATE),
    // catchError stays on this inner observable so the outer stream in
    // `getVariantsDocumentCounts` keeps reacting to variant list changes after a failure.
    catchError((error) => of({
      data: null,
      loading: !1,
      error
    }))
  );
}
function getVariantsDocumentCounts(client, variantsState$) {
  return variantsState$.pipe(
    // Document ids cannot contain commas, so a sorted joined string is a stable identity
    // key that lets `distinctUntilChanged` ignore store emissions (e.g. variant metadata
    // edits) that don't change the set of variants.
    map((state) => Array.from(state.variants.keys()).toSorted().join(",")),
    distinctUntilChanged(),
    switchMap((idsKey) => listenVariantsDocumentCounts(client, idsKey ? idsKey.split(",") : [])),
    scan((previous, next) => ({
      // Keep the previous counts while the next fetch is loading (or errored), so counts
      // for unchanged variants remain visible across variant list changes.
      data: next.data ? {
        ...previous.data,
        ...next.data
      } : previous.data,
      loading: next.loading && previous.data === null,
      error: next.error
    }), INITIAL_STATE)
  );
}
function useVariantsDocumentCounts() {
  const $ = c(3), client = useClient(VARIANTS_STUDIO_CLIENT_OPTIONS), {
    state$
  } = useVariantsStore();
  let t0;
  return $[0] !== client || $[1] !== state$ ? (t0 = getVariantsDocumentCounts(client, state$), $[0] = client, $[1] = state$, $[2] = t0) : t0 = $[2], useObservable(t0, INITIAL_STATE);
}
function VariantMenuButton(t0) {
  const $ = c(32), {
    documentCount,
    variant
  } = t0, {
    t
  } = useTranslation(variantsLocaleNamespace);
  let t1;
  $[0] !== variant ? (t1 = getVariantTitle(variant), $[0] = variant, $[1] = t1) : t1 = $[1];
  const variantTitle = t1;
  let t2;
  $[2] !== documentCount || $[3] !== variantTitle ? (t2 = {
    documentCount,
    variantTitle
  }, $[2] = documentCount, $[3] = variantTitle, $[4] = t2) : t2 = $[4];
  const {
    deleteDisabled,
    deleteDisabledTooltip,
    handleCloseDeleteDialog,
    handleConfirmDelete,
    handleDelete,
    isDeleteDialogOpen,
    isDeleting,
    variantTitle: dialogVariantTitle
  } = useVariantDeleteAction(variant._id, t2);
  let t3;
  $[5] !== isDeleting ? (t3 = /* @__PURE__ */ jsx(ContextMenuButton, { disabled: isDeleting, loading: isDeleting }), $[5] = isDeleting, $[6] = t3) : t3 = $[6];
  let t4;
  $[7] !== variant._id ? (t4 = getVariantId(variant._id), $[7] = variant._id, $[8] = t4) : t4 = $[8];
  const t5 = `variant-actions-${t4}`;
  let t6;
  $[9] !== t ? (t6 = t("overview.action.delete-variant"), $[9] = t, $[10] = t6) : t6 = $[10];
  let t7;
  $[11] !== deleteDisabledTooltip ? (t7 = deleteDisabledTooltip ? {
    content: deleteDisabledTooltip
  } : void 0, $[11] = deleteDisabledTooltip, $[12] = t7) : t7 = $[12];
  let t8;
  $[13] !== deleteDisabled || $[14] !== handleDelete || $[15] !== t6 || $[16] !== t7 ? (t8 = /* @__PURE__ */ jsx(Menu, { children: /* @__PURE__ */ jsx(MenuItem, { disabled: deleteDisabled, onClick: handleDelete, text: t6, tone: "critical", tooltipProps: t7 }) }), $[13] = deleteDisabled, $[14] = handleDelete, $[15] = t6, $[16] = t7, $[17] = t8) : t8 = $[17];
  let t9;
  $[18] === /* @__PURE__ */ Symbol.for("react.memo_cache_sentinel") ? (t9 = {
    placement: "bottom-end",
    portal: !0
  }, $[18] = t9) : t9 = $[18];
  let t10;
  $[19] !== t3 || $[20] !== t5 || $[21] !== t8 ? (t10 = /* @__PURE__ */ jsx(MenuButton, { button: t3, id: t5, menu: t8, popover: t9 }), $[19] = t3, $[20] = t5, $[21] = t8, $[22] = t10) : t10 = $[22];
  let t11;
  $[23] !== dialogVariantTitle || $[24] !== handleCloseDeleteDialog || $[25] !== handleConfirmDelete || $[26] !== isDeleteDialogOpen || $[27] !== isDeleting ? (t11 = isDeleteDialogOpen && /* @__PURE__ */ jsx(DeleteVariantDialog, { isDeleting, onClose: handleCloseDeleteDialog, onConfirm: handleConfirmDelete, variantTitle: dialogVariantTitle }), $[23] = dialogVariantTitle, $[24] = handleCloseDeleteDialog, $[25] = handleConfirmDelete, $[26] = isDeleteDialogOpen, $[27] = isDeleting, $[28] = t11) : t11 = $[28];
  let t12;
  return $[29] !== t10 || $[30] !== t11 ? (t12 = /* @__PURE__ */ jsxs(Fragment, { children: [
    t10,
    t11
  ] }), $[29] = t10, $[30] = t11, $[31] = t12) : t12 = $[31], t12;
}
const BASE_PANEL = "M64 150.174V81.8262C64 79.1748 65.498 76.751 67.8695 75.5653L103.87 57.5652C108.524 55.2381 114 58.6226 114 63.8262V132.174C114 134.825 112.502 137.249 110.131 138.435L74.1305 156.435C69.4762 158.762 64 155.377 64 150.174Z", LINES_FULL = ["M72 86 L106 86", "M72 96 L102 96", "M72 106 L104 106", "M72 116 L98 116"], LINES_MEDIUM = ["M72 88 L100 88", "M72 98 L96 98", "M72 108 L102 108"], LINES_SHORT = ["M72 90 L94 90", "M72 100 L92 100"], VariantIllustration = () => {
  const $ = c(6);
  let t0;
  $[0] === /* @__PURE__ */ Symbol.for("react.memo_cache_sentinel") ? (t0 = /* @__PURE__ */ jsxs("g", { opacity: 0.12, children: [
    /* @__PURE__ */ jsx("path", { d: BASE_PANEL, fill: "var(--card-muted-bg-color)", stroke: "var(--card-muted-fg-color)", strokeWidth: 1.2, transform: "translate(12 10)" }),
    /* @__PURE__ */ jsx("path", { d: BASE_PANEL, fill: "var(--card-muted-bg-color)", stroke: "var(--card-muted-fg-color)", strokeWidth: 1.2, transform: "translate(6 6)" })
  ] }), $[0] = t0) : t0 = $[0];
  let t1;
  $[1] === /* @__PURE__ */ Symbol.for("react.memo_cache_sentinel") ? (t1 = /* @__PURE__ */ jsx("g", { opacity: 0.28, children: /* @__PURE__ */ jsxs("g", { transform: "translate(118 6) scale(0.5)", children: [
    /* @__PURE__ */ jsx("path", { d: BASE_PANEL, fill: "var(--card-muted-bg-color)", stroke: "var(--card-muted-fg-color)", strokeWidth: 1.2 }),
    LINES_MEDIUM.map(_temp$1)
  ] }) }), $[1] = t1) : t1 = $[1];
  let t2;
  $[2] === /* @__PURE__ */ Symbol.for("react.memo_cache_sentinel") ? (t2 = /* @__PURE__ */ jsx("g", { opacity: 0.34, children: /* @__PURE__ */ jsxs("g", { transform: "translate(126 52) scale(0.52)", children: [
    /* @__PURE__ */ jsx("path", { d: BASE_PANEL, fill: "var(--card-muted-bg-color)", stroke: "var(--card-muted-fg-color)", strokeWidth: 1.2 }),
    LINES_FULL.slice(0, 3).map(_temp2)
  ] }) }), $[2] = t2) : t2 = $[2];
  let t3;
  $[3] === /* @__PURE__ */ Symbol.for("react.memo_cache_sentinel") ? (t3 = /* @__PURE__ */ jsx("g", { opacity: 0.26, children: /* @__PURE__ */ jsxs("g", { transform: "translate(106 102) scale(0.48)", children: [
    /* @__PURE__ */ jsx("path", { d: BASE_PANEL, fill: "var(--card-muted-bg-color)", stroke: "var(--card-muted-fg-color)", strokeWidth: 1.2 }),
    LINES_SHORT.map(_temp3)
  ] }) }), $[3] = t3) : t3 = $[3];
  let t4;
  $[4] === /* @__PURE__ */ Symbol.for("react.memo_cache_sentinel") ? (t4 = /* @__PURE__ */ jsxs("g", { opacity: 0.14, stroke: "var(--card-muted-fg-color)", strokeWidth: 1.2, strokeLinecap: "round", children: [
    /* @__PURE__ */ jsx("path", { d: "M118 112 Q 152 72 168 58", fill: "none" }),
    /* @__PURE__ */ jsx("path", { d: "M124 124 Q 168 108 188 92", fill: "none" }),
    /* @__PURE__ */ jsx("path", { d: "M118 138 Q 150 152 162 168", fill: "none" })
  ] }), $[4] = t4) : t4 = $[4];
  let t5;
  return $[5] === /* @__PURE__ */ Symbol.for("react.memo_cache_sentinel") ? (t5 = /* @__PURE__ */ jsxs("svg", { "aria-hidden": "true", "data-testid": "variant-illustration", width: "248", height: "201", viewBox: "0 0 248 201", fill: "none", xmlns: "http://www.w3.org/2000/svg", children: [
    t0,
    t1,
    t2,
    t3,
    t4,
    /* @__PURE__ */ jsxs("g", { transform: "translate(26 20) scale(1.06)", children: [
      /* @__PURE__ */ jsx("path", { d: BASE_PANEL, fill: "var(--card-muted-bg-color)", stroke: "var(--card-muted-fg-color)", strokeWidth: 1.2, opacity: 0.92 }),
      LINES_FULL.map(_temp4)
    ] })
  ] }), $[5] = t5) : t5 = $[5], t5;
};
function _temp$1(d) {
  return /* @__PURE__ */ jsx("path", { d, stroke: "var(--card-muted-fg-color)", strokeWidth: 1.2, strokeLinecap: "round", opacity: 0.55 }, d);
}
function _temp2(d_0) {
  return /* @__PURE__ */ jsx("path", { d: d_0, stroke: "var(--card-muted-fg-color)", strokeWidth: 1.2, strokeLinecap: "round", opacity: 0.5 }, d_0);
}
function _temp3(d_1) {
  return /* @__PURE__ */ jsx("path", { d: d_1, stroke: "var(--card-muted-fg-color)", strokeWidth: 1.2, strokeLinecap: "round", opacity: 0.55 }, d_1);
}
function _temp4(d_2, index) {
  return /* @__PURE__ */ jsx("path", { d: d_2, stroke: "var(--card-muted-fg-color)", strokeWidth: 1.2, strokeLinecap: "round", opacity: 0.42 + index * 0.06 }, d_2);
}
const VARIANTS_DOCUMENTATION_URL = "https://www.sanity.io/docs/content-variants", VariantsEmptyState = (t0) => {
  const $ = c(22), {
    createVariantButton
  } = t0, {
    t
  } = useTranslation(variantsLocaleNamespace);
  let t1;
  $[0] === /* @__PURE__ */ Symbol.for("react.memo_cache_sentinel") ? (t1 = {
    maxWidth: rem(300)
  }, $[0] = t1) : t1 = $[0];
  let t2;
  $[1] === /* @__PURE__ */ Symbol.for("react.memo_cache_sentinel") ? (t2 = /* @__PURE__ */ jsx(VariantIllustration, {}), $[1] = t2) : t2 = $[1];
  let t3;
  $[2] !== t ? (t3 = t("overview.title"), $[2] = t, $[3] = t3) : t3 = $[3];
  let t4;
  $[4] !== t3 ? (t4 = /* @__PURE__ */ jsx(Text, { as: "h1", size: 1, weight: "semibold", "data-testid": "no-variants-info-text", children: t3 }), $[4] = t3, $[5] = t4) : t4 = $[5];
  let t5;
  $[6] === /* @__PURE__ */ Symbol.for("react.memo_cache_sentinel") ? (t5 = {
    textAlign: "center"
  }, $[6] = t5) : t5 = $[6];
  let t6;
  $[7] !== t ? (t6 = t("overview.empty-state.description"), $[7] = t, $[8] = t6) : t6 = $[8];
  let t7;
  $[9] !== t6 ? (t7 = /* @__PURE__ */ jsx(Text, { size: 1, muted: !0, style: t5, children: t6 }), $[9] = t6, $[10] = t7) : t7 = $[10];
  let t8;
  $[11] !== t ? (t8 = t("overview.action.documentation"), $[11] = t, $[12] = t8) : t8 = $[12];
  let t9;
  $[13] !== t8 ? (t9 = /* @__PURE__ */ jsx(Button, { as: "a", href: VARIANTS_DOCUMENTATION_URL, target: "_blank", mode: "ghost", text: t8 }), $[13] = t8, $[14] = t9) : t9 = $[14];
  let t10;
  $[15] !== createVariantButton || $[16] !== t9 ? (t10 = /* @__PURE__ */ jsxs(Flex, { gap: 2, children: [
    createVariantButton,
    t9
  ] }), $[15] = createVariantButton, $[16] = t9, $[17] = t10) : t10 = $[17];
  let t11;
  return $[18] !== t10 || $[19] !== t4 || $[20] !== t7 ? (t11 = /* @__PURE__ */ jsxs(Flex, { align: "center", "data-testid": "variants-empty-state", flex: 1, direction: "column", gap: 3, style: t1, paddingBottom: 5, children: [
    t2,
    t4,
    t7,
    t10
  ] }), $[18] = t10, $[19] = t4, $[20] = t7, $[21] = t11) : t11 = $[21], t11;
}, VariantDocumentsCell = (t0) => {
  const $ = c(8), {
    cellProps,
    datum
  } = t0;
  if (datum.isLoading || datum.documentCount === void 0) {
    let t12;
    $[0] === /* @__PURE__ */ Symbol.for("react.memo_cache_sentinel") ? (t12 = /* @__PURE__ */ jsx(Text, { size: 1, children: /* @__PURE__ */ jsx(Skeleton, { animated: !0, radius: 1, style: {
      width: "4ch"
    } }) }), $[0] = t12) : t12 = $[0];
    let t22;
    return $[1] !== cellProps ? (t22 = /* @__PURE__ */ jsx(Flex, { ...cellProps, align: "center", paddingX: 2, paddingY: 3, sizing: "border", children: t12 }), $[1] = cellProps, $[2] = t22) : t22 = $[2], t22;
  }
  const t1 = datum.documentCount === null ? "-" : datum.documentCount;
  let t2;
  $[3] !== t1 ? (t2 = /* @__PURE__ */ jsx(Text, { muted: !0, size: 1, children: t1 }), $[3] = t1, $[4] = t2) : t2 = $[4];
  let t3;
  return $[5] !== cellProps || $[6] !== t2 ? (t3 = /* @__PURE__ */ jsx(Flex, { ...cellProps, align: "center", paddingX: 2, paddingY: 3, sizing: "border", children: t2 }), $[5] = cellProps, $[6] = t2, $[7] = t3) : t3 = $[7], t3;
}, VariantTitleCell = (t0) => {
  const $ = c(33), {
    cellProps,
    datum: variant
  } = t0, {
    t
  } = useTranslation(variantsLocaleNamespace);
  let t1;
  $[0] !== variant._id ? (t1 = getVariantId(variant._id), $[0] = variant._id, $[1] = t1) : t1 = $[1];
  const encodedVariantId = t1;
  let t2;
  $[2] !== encodedVariantId ? (t2 = forwardRef(function(linkProps, ref) {
    return /* @__PURE__ */ jsx(StateLink, { ...linkProps, ref, state: {
      variantId: encodedVariantId
    } });
  }), $[2] = encodedVariantId, $[3] = t2) : t2 = $[3];
  const VariantLink = t2;
  if (variant.isLoading) {
    let t32;
    $[4] === /* @__PURE__ */ Symbol.for("react.memo_cache_sentinel") ? (t32 = /* @__PURE__ */ jsx(Text, { size: 1, weight: "medium", children: /* @__PURE__ */ jsx(Skeleton, { animated: !0, radius: 1, style: {
      width: "16ch"
    } }) }), $[4] = t32) : t32 = $[4];
    let t42;
    $[5] === /* @__PURE__ */ Symbol.for("react.memo_cache_sentinel") ? (t42 = /* @__PURE__ */ jsxs(Stack, { space: 2, children: [
      t32,
      /* @__PURE__ */ jsx(Text, { muted: !0, size: 1, children: /* @__PURE__ */ jsx(Skeleton, { animated: !0, radius: 1, style: {
        width: "32ch"
      } }) })
    ] }), $[5] = t42) : t42 = $[5];
    let t52;
    return $[6] !== cellProps ? (t52 = /* @__PURE__ */ jsx(Box, { ...cellProps, flex: 1, paddingLeft: 3, paddingRight: 2, paddingY: 2, sizing: "border", children: t42 }), $[6] = cellProps, $[7] = t52) : t52 = $[7], t52;
  }
  let t3;
  $[8] !== variant.conditions ? (t3 = getVariantConditionsText(variant.conditions), $[8] = variant.conditions, $[9] = t3) : t3 = $[9];
  const conditionsText = t3;
  let t4;
  $[10] !== variant ? (t4 = /* @__PURE__ */ jsx(VariantPinButton, { variant }), $[10] = variant, $[11] = t4) : t4 = $[11];
  let t5;
  $[12] !== variant ? (t5 = getVariantTitle(variant), $[12] = variant, $[13] = t5) : t5 = $[13];
  let t6;
  $[14] !== t5 ? (t6 = /* @__PURE__ */ jsx(Text, { size: 1, weight: "medium", children: t5 }), $[14] = t5, $[15] = t6) : t6 = $[15];
  let t7;
  $[16] !== conditionsText || $[17] !== t ? (t7 = conditionsText || t("overview.table.no-conditions"), $[16] = conditionsText, $[17] = t, $[18] = t7) : t7 = $[18];
  let t8;
  $[19] !== t7 ? (t8 = /* @__PURE__ */ jsx(Text, { muted: !0, size: 1, children: t7 }), $[19] = t7, $[20] = t8) : t8 = $[20];
  let t9;
  $[21] !== t6 || $[22] !== t8 ? (t9 = /* @__PURE__ */ jsx(Flex, { align: "center", gap: 3, children: /* @__PURE__ */ jsxs(Stack, { flex: 1, space: 2, children: [
    t6,
    t8
  ] }) }), $[21] = t6, $[22] = t8, $[23] = t9) : t9 = $[23];
  let t10;
  $[24] !== VariantLink || $[25] !== t9 ? (t10 = /* @__PURE__ */ jsx(Card, { as: VariantLink, "data-as": "a", flex: 1, padding: 2, radius: 2, tone: "inherit", children: t9 }), $[24] = VariantLink, $[25] = t9, $[26] = t10) : t10 = $[26];
  let t11;
  $[27] !== t10 || $[28] !== t4 ? (t11 = /* @__PURE__ */ jsxs(Flex, { align: "center", gap: 3, children: [
    t4,
    t10
  ] }), $[27] = t10, $[28] = t4, $[29] = t11) : t11 = $[29];
  let t12;
  return $[30] !== cellProps || $[31] !== t11 ? (t12 = /* @__PURE__ */ jsx(Box, { ...cellProps, flex: 1, paddingLeft: 3, paddingRight: 2, paddingY: 1, sizing: "border", children: t11 }), $[30] = cellProps, $[31] = t11, $[32] = t12) : t12 = $[32], t12;
};
function variantsOverviewColumnDefs(t) {
  return [{
    id: "metadata.title",
    sorting: !0,
    width: null,
    style: {
      minWidth: "min(50%, calc(100vw - 80px))"
    },
    header: (props) => /* @__PURE__ */ jsx(Flex, { ...props.headerProps, flex: 1, paddingLeft: 3, paddingRight: 2, paddingY: 3, sizing: "border", children: /* @__PURE__ */ jsx(Headers.SortHeaderButton, { ...props, text: t("overview.table.variant") }) }),
    cell: VariantTitleCell,
    sortTransform: (variant) => getVariantTitle(variant)
  }, {
    id: "documentCount",
    sorting: !1,
    width: 120,
    header: ({
      headerProps
    }) => /* @__PURE__ */ jsx(Flex, { ...headerProps, paddingY: 3, sizing: "border", children: /* @__PURE__ */ jsx(Headers.BasicHeader, { text: t("overview.table.documents") }) }),
    cell: VariantDocumentsCell
  }];
}
const VARIANT_TABLE_ROW_ID = "_id";
function VariantsOverview() {
  const $ = c(67), {
    t
  } = useTranslation(variantsLocaleNamespace), router = useRouter(), {
    data: variants,
    error,
    loading
  } = useAllVariants(), [scrollContainerRef, setScrollContainerRef] = useState(null), [searchQuery, setSearchQuery] = useState(""), [isCreateVariantDialogOpen, setIsCreateVariantDialogOpen] = useState(!1);
  let t0;
  $[0] === /* @__PURE__ */ Symbol.for("react.memo_cache_sentinel") ? (t0 = () => {
    setIsCreateVariantDialogOpen(!0);
  }, $[0] = t0) : t0 = $[0];
  const handleCreateVariant = t0;
  let t1;
  $[1] !== router ? (t1 = (createdVariantId) => {
    setIsCreateVariantDialogOpen(!1), router.navigate({
      variantId: getVariantId(createdVariantId)
    });
  }, $[1] = router, $[2] = t1) : t1 = $[2];
  const handleOnCreateVariant = t1;
  let t2;
  $[3] !== t ? (t2 = variantsOverviewColumnDefs(t), $[3] = t, $[4] = t2) : t2 = $[4];
  const columnDefs = t2, renderRowActions = _temp;
  let t3;
  $[5] !== variants ? (t3 = variants ?? [], $[5] = variants, $[6] = t3) : t3 = $[6];
  const variantsList = t3, {
    data: documentCounts,
    error: documentCountsError
  } = useVariantsDocumentCounts();
  let t4;
  if ($[7] !== documentCounts || $[8] !== documentCountsError || $[9] !== searchQuery || $[10] !== variantsList) {
    let t52;
    $[12] !== documentCounts || $[13] !== documentCountsError ? (t52 = (variant) => ({
      ...variant,
      documentCount: documentCounts?.[variant._id] ?? (documentCountsError ? null : void 0)
    }), $[12] = documentCounts, $[13] = documentCountsError, $[14] = t52) : t52 = $[14], t4 = filterVariantsForSearch(variantsList, searchQuery).map(t52), $[7] = documentCounts, $[8] = documentCountsError, $[9] = searchQuery, $[10] = variantsList, $[11] = t4;
  } else
    t4 = $[11];
  const filteredVariants = t4, hasVariants = variantsList.length > 0;
  let t5;
  $[15] !== t ? (t5 = t("overview.action.create-variant"), $[15] = t, $[16] = t5) : t5 = $[16];
  let t6;
  $[17] !== isCreateVariantDialogOpen || $[18] !== t5 ? (t6 = /* @__PURE__ */ jsx(Button, { disabled: isCreateVariantDialogOpen, icon: AddIcon, onClick: handleCreateVariant, text: t5 }), $[17] = isCreateVariantDialogOpen, $[18] = t5, $[19] = t6) : t6 = $[19];
  const createVariantButton = t6;
  let t7;
  $[20] !== createVariantButton || $[21] !== error || $[22] !== hasVariants || $[23] !== t ? (t7 = () => error && !hasVariants ? /* @__PURE__ */ jsx(Flex, { align: "center", direction: "column", gap: 3, justify: "center", padding: 4, children: /* @__PURE__ */ jsx(Text, { muted: !0, size: 1, children: t("overview.error") }) }) : /* @__PURE__ */ jsx(VariantsEmptyState, { createVariantButton }), $[20] = createVariantButton, $[21] = error, $[22] = hasVariants, $[23] = t, $[24] = t7) : t7 = $[24];
  const tableEmptyState = t7;
  let t8;
  $[25] !== t ? (t8 = t("overview.title"), $[25] = t, $[26] = t8) : t8 = $[26];
  let t9;
  $[27] !== t8 ? (t9 = /* @__PURE__ */ jsx(Text, { as: "h1", size: 4, weight: "bold", children: t8 }), $[27] = t8, $[28] = t9) : t9 = $[28];
  let t10;
  $[29] !== t ? (t10 = t("overview.description"), $[29] = t, $[30] = t10) : t10 = $[30];
  let t11;
  $[31] !== t10 ? (t11 = /* @__PURE__ */ jsx(Text, { muted: !0, size: 1, children: t10 }), $[31] = t10, $[32] = t11) : t11 = $[32];
  let t12;
  $[33] !== t11 || $[34] !== t9 ? (t12 = /* @__PURE__ */ jsxs(Stack, { space: 3, children: [
    t9,
    t11
  ] }), $[33] = t11, $[34] = t9, $[35] = t12) : t12 = $[35];
  let t13;
  $[36] !== createVariantButton || $[37] !== t12 ? (t13 = /* @__PURE__ */ jsx(Card, { flex: "none", paddingY: 5, children: /* @__PURE__ */ jsxs(Flex, { align: "flex-start", gap: 4, justify: "space-between", children: [
    t12,
    createVariantButton
  ] }) }), $[36] = createVariantButton, $[37] = t12, $[38] = t13) : t13 = $[38];
  const t14 = !!searchQuery;
  let t15, t16;
  $[39] === /* @__PURE__ */ Symbol.for("react.memo_cache_sentinel") ? (t15 = (event) => setSearchQuery(event.currentTarget.value), t16 = () => setSearchQuery(""), $[39] = t15, $[40] = t16) : (t15 = $[39], t16 = $[40]);
  let t17;
  $[41] !== t ? (t17 = t("overview.search.placeholder"), $[41] = t, $[42] = t17) : t17 = $[42];
  let t18;
  $[43] !== searchQuery || $[44] !== t14 || $[45] !== t17 ? (t18 = /* @__PURE__ */ jsx(Box, { flex: "none", paddingBottom: 4, paddingTop: 2, children: /* @__PURE__ */ jsx(TextInput, { clearButton: t14, fontSize: 1, icon: SearchIcon, onChange: t15, onClear: t16, placeholder: t17, radius: 3, value: searchQuery }) }), $[43] = searchQuery, $[44] = t14, $[45] = t17, $[46] = t18) : t18 = $[46];
  let t19;
  $[47] !== error || $[48] !== t ? (t19 = error && /* @__PURE__ */ jsx(Card, { flex: "none", padding: 3, tone: "critical", children: /* @__PURE__ */ jsx(Text, { size: 1, children: t("overview.error") }) }), $[47] = error, $[48] = t, $[49] = t19) : t19 = $[49];
  let t20;
  $[50] !== t13 || $[51] !== t18 || $[52] !== t19 ? (t20 = /* @__PURE__ */ jsx(Container, { flex: "none", width: 3, children: /* @__PURE__ */ jsxs(Flex, { direction: "column", paddingX: 3, children: [
    t13,
    t18,
    t19
  ] }) }), $[50] = t13, $[51] = t18, $[52] = t19, $[53] = t20) : t20 = $[53];
  let t21;
  $[54] !== columnDefs || $[55] !== filteredVariants || $[56] !== loading || $[57] !== scrollContainerRef || $[58] !== tableEmptyState ? (t21 = /* @__PURE__ */ jsx(Box, { flex: 1, overflow: "auto", ref: setScrollContainerRef, children: /* @__PURE__ */ jsx(Table, { columnDefs, data: filteredVariants, emptyState: tableEmptyState, loading, rowId: VARIANT_TABLE_ROW_ID, rowActions: renderRowActions, scrollContainerRef }) }), $[54] = columnDefs, $[55] = filteredVariants, $[56] = loading, $[57] = scrollContainerRef, $[58] = tableEmptyState, $[59] = t21) : t21 = $[59];
  let t22;
  $[60] !== handleOnCreateVariant || $[61] !== isCreateVariantDialogOpen ? (t22 = isCreateVariantDialogOpen && /* @__PURE__ */ jsx(CreateVariantDialog, { onCancel: () => setIsCreateVariantDialogOpen(!1), onSubmit: handleOnCreateVariant }), $[60] = handleOnCreateVariant, $[61] = isCreateVariantDialogOpen, $[62] = t22) : t22 = $[62];
  let t23;
  return $[63] !== t20 || $[64] !== t21 || $[65] !== t22 ? (t23 = /* @__PURE__ */ jsxs(Flex, { direction: "column", flex: 1, height: "fill", children: [
    t20,
    t21,
    t22
  ] }), $[63] = t20, $[64] = t21, $[65] = t22, $[66] = t23) : t23 = $[66], t23;
}
function _temp(t0) {
  const {
    datum
  } = t0;
  return /* @__PURE__ */ jsx(VariantMenuButton, { documentCount: datum.documentCount, variant: datum });
}
function VariantsTool() {
  const $ = c(3), router = useRouter(), variantId = typeof router.state.variantId == "string" ? router.state.variantId : void 0;
  if (variantId) {
    let t02;
    return $[0] !== variantId ? (t02 = /* @__PURE__ */ jsx(VariantDetail, {}, variantId), $[0] = variantId, $[1] = t02) : t02 = $[1], t02;
  }
  let t0;
  return $[2] === /* @__PURE__ */ Symbol.for("react.memo_cache_sentinel") ? (t0 = /* @__PURE__ */ jsx(VariantsOverview, {}), $[2] = t0) : t0 = $[2], t0;
}
export {
  VariantsTool
};
