// @portabletext/plugin-table/dist/index.js
import { jsxs, Fragment, jsx } from "react/jsx-runtime";
import { c } from "react/compiler-runtime";
import { defineContainer } from "@portabletext/editor";
import { NodePlugin, BehaviorPlugin } from "@portabletext/editor/plugins";
import { defineBehavior, raise } from "@portabletext/editor/behaviors";
import { resolveTableSelection, tableRows, rowCells, memberCells, cellValue, resolveCell, createTableGuards, getTableSelection, registerTableConfig } from "./_chunks-es/get-table-selection.js";
import { isActiveDecorator, isActiveAnnotation, isActiveStyle, isActiveListItem, getSelectedTextBlocks, getFocusTextBlock, getFocusBlockObject, isSelectionCollapsed } from "@portabletext/editor/selectors";
import { getFirstChild, getBlock, getLastChild, getPathSubSchema, getEnclosingBlock, getParent, getChildren, getSibling, comparePoints, getLeaf, pathContains } from "@portabletext/editor/traversal";
import { getBlockStartPoint, getBlockEndPoint, isEqualSelectionPoints, isEqualPaths } from "@portabletext/editor/utils";
import { createKeyboardShortcut } from "@portabletext/keyboard-shortcuts";
function createDeleteBehaviors(config) {
  return [defineBehavior({
    on: "delete",
    guard: ({
      snapshot,
      event
    }) => {
      if (event.direction === void 0)
        return !1;
      const resolved = resolveTableSelection(snapshot, config);
      if (!resolved)
        return !1;
      const [rowStart, rowEnd] = resolved.tableSelection.rowRange, [colStart, colEnd] = resolved.tableSelection.colRange, rows = tableRows(config, resolved.table.node), columnCount = Math.max(...rows.map((row) => rowCells(config, row).length));
      return rowStart === 0 && colStart === 0 && rowEnd === rows.length - 1 && colEnd === columnCount - 1 ? {
        tablePath: resolved.table.path
      } : !1;
    },
    actions: [(_, {
      tablePath
    }) => [raise({
      type: "unset",
      at: tablePath
    })]]
  }), defineBehavior({
    on: "delete",
    guard: ({
      snapshot,
      event
    }) => event.at ? !1 : resolveTableSelection(snapshot, config) ?? !1,
    actions: [(_, resolved) => clearCellsAndCollapse(config, resolved)]
  }), defineBehavior({
    on: "split",
    guard: ({
      snapshot
    }) => resolveTableSelection(snapshot, config) ?? !1,
    actions: [(_, resolved) => clearCellsAndCollapse(config, resolved)]
  })];
}
function clearCellsAndCollapse(config, {
  tableSelection,
  table
}) {
  const actions = [];
  let topLeftCellPath = null;
  for (const cell of memberCells(tableSelection, table.node, config)) {
    topLeftCellPath === null && (topLeftCellPath = cell.path);
    for (const block of cellValue(config, cell.node))
      actions.push(raise({
        type: "unset",
        at: [...cell.path, config.valueField, {
          _key: block._key
        }]
      }));
  }
  return topLeftCellPath && actions.push(raise({
    type: "select.block",
    at: topLeftCellPath,
    select: "start"
  })), actions;
}
function cellStartPoint(snapshot, cellPath2) {
  const firstChild = getFirstChild(snapshot, cellPath2), firstBlock = firstChild && getBlock(snapshot, firstChild.path);
  if (firstBlock)
    return getBlockStartPoint({
      context: snapshot.context,
      block: firstBlock
    });
}
function cellEndPoint(snapshot, cellPath2) {
  const lastChild = getLastChild(snapshot, cellPath2), lastBlock = lastChild && getBlock(snapshot, lastChild.path);
  if (lastBlock)
    return getBlockEndPoint({
      context: snapshot.context,
      block: lastBlock
    });
}
function createFormatBehaviors(config) {
  return [
    defineBehavior({
      on: "decorator.toggle",
      guard: ({
        snapshot,
        event
      }) => {
        if (event.at)
          return !1;
        const ranges = spanRanges(config, snapshot);
        return ranges ? resolveMembers(snapshot, ranges, isActiveDecorator(event.decorator)) : !1;
      },
      actions: [({
        event
      }, {
        members,
        active
      }) => (
        // Toggling each member cell separately would checkerboard a mixed
        // rectangle. Aggregate first: if any member cell misses the
        // decorator, add it where missing, otherwise remove it everywhere.
        members.filter((member) => active ? !0 : !member.active).map((member) => raise({
          type: active ? "decorator.remove" : "decorator.add",
          decorator: event.decorator,
          at: member.range
        }))
      )]
    }),
    defineBehavior({
      on: "annotation.toggle",
      guard: ({
        snapshot,
        event
      }) => {
        if (event.at)
          return !1;
        const ranges = spanRanges(config, snapshot);
        return ranges ? resolveMembers(snapshot, ranges, isActiveAnnotation(event.annotation.name)) : !1;
      },
      actions: [({
        event
      }, {
        members,
        active
      }) => (
        // Unlike decorator marks, annotations are not set-semantic: every
        // `annotation.add` mints a new markDef. The add branch must skip
        // members that are already active, or a mixed rectangle would
        // stack a second annotation onto them.
        members.filter((member) => active ? !0 : !member.active).map((member) => active ? raise({
          type: "annotation.remove",
          annotation: {
            name: event.annotation.name
          },
          at: member.range
        }) : raise({
          type: "annotation.add",
          annotation: event.annotation,
          at: member.range
        }))
      )]
    }),
    fanOutOverRectangle(config, "decorator.add"),
    fanOutOverRectangle(config, "decorator.remove"),
    fanOutOverRectangle(config, "annotation.add"),
    fanOutOverRectangle(config, "annotation.remove"),
    defineBehavior({
      on: "style.toggle",
      guard: ({
        snapshot,
        event
      }) => {
        const ranges = rectangleRanges(config, snapshot);
        return ranges ? resolveMembers(snapshot, ranges, isActiveStyle(event.style)) : !1;
      },
      actions: [({
        event
      }, {
        active
      }) => [
        // The re-raise is address-less and re-enters the interceptors
        // below, which decompose it per member cell.
        raise({
          type: active ? "style.remove" : "style.add",
          style: event.style
        })
      ]]
    }),
    defineBehavior({
      on: "list item.toggle",
      guard: ({
        snapshot,
        event
      }) => {
        const ranges = rectangleRanges(config, snapshot);
        return ranges ? resolveMembers(snapshot, ranges, isActiveListItem(event.listItem)) : !1;
      },
      actions: [({
        event
      }, {
        active
      }) => [raise({
        type: active ? "list item.remove" : "list item.add",
        listItem: event.listItem
      })]]
    }),
    // Keep in sync with core's `behavior.abstract.style.ts`, which decomposes
    // the selection-scoped style events into `block.set`/`block.unset` per
    // selected text block. Here they decompose over the member cells' ranges
    // instead of the linear selection.
    defineBehavior({
      on: "style.add",
      guard: ({
        snapshot
      }) => rectangleTextBlocks(config, snapshot),
      actions: [({
        event
      }, blocks) => blocks.map((block) => raise({
        type: "block.set",
        at: block.path,
        props: {
          style: event.style
        }
      }))]
    }),
    defineBehavior({
      on: "style.remove",
      guard: ({
        snapshot
      }) => rectangleTextBlocks(config, snapshot),
      actions: [(_, blocks) => blocks.map((block) => raise({
        type: "block.unset",
        at: block.path,
        props: ["style"]
      }))]
    }),
    // Keep in sync with core's `behavior.abstract.list-item.ts`: adding only
    // touches blocks whose sub-schema declares the list, removing unsets both
    // `level` and `listItem`.
    defineBehavior({
      on: "list item.add",
      guard: ({
        snapshot,
        event
      }) => {
        const blocks = rectangleTextBlocks(config, snapshot);
        if (!blocks)
          return !1;
        const listBlocks = blocks.filter((block) => getPathSubSchema(snapshot, block.path).lists.some((list) => list.name === event.listItem));
        return listBlocks.length > 0 ? listBlocks : !1;
      },
      actions: [({
        event
      }, blocks) => blocks.map((block) => raise({
        type: "block.set",
        at: block.path,
        props: {
          level: block.node.level ?? 1,
          listItem: event.listItem
        }
      }))]
    }),
    defineBehavior({
      on: "list item.remove",
      guard: ({
        snapshot
      }) => rectangleTextBlocks(config, snapshot),
      actions: [(_, blocks) => blocks.map((block) => raise({
        type: "block.unset",
        at: block.path,
        props: ["level", "listItem"]
      }))]
    })
  ];
}
function fanOutOverRectangle(config, on) {
  return defineBehavior({
    on,
    guard: ({
      snapshot,
      event
    }) => event.at ? !1 : spanRanges(config, snapshot),
    actions: [({
      event
    }, ranges) => ranges.map((range) => raise({
      ...event,
      at: range
    }))]
  });
}
function resolveMembers(snapshot, ranges, isActive) {
  const members = ranges.map((range) => ({
    range,
    active: isActive({
      ...snapshot,
      context: {
        ...snapshot.context,
        selection: range
      }
    })
  }));
  return {
    members,
    active: members.every((member) => member.active)
  };
}
function rectangleTextBlocks(config, snapshot) {
  const ranges = rectangleRanges(config, snapshot);
  if (!ranges)
    return !1;
  const blocks = ranges.flatMap((range) => getSelectedTextBlocks({
    ...snapshot,
    context: {
      ...snapshot.context,
      selection: range
    }
  }));
  return blocks.length > 0 ? blocks : !1;
}
function spanRanges(config, snapshot) {
  const ranges = rectangleRanges(config, snapshot);
  if (!ranges)
    return !1;
  const inhabited = ranges.filter((range) => !isEqualSelectionPoints(range.anchor, range.focus));
  return inhabited.length > 0 ? inhabited : !1;
}
function rectangleRanges(config, snapshot) {
  const resolved = resolveTableSelection(snapshot, config);
  if (!resolved)
    return !1;
  const ranges = [];
  for (const cell of memberCells(resolved.tableSelection, resolved.table.node, config)) {
    const anchor = cellStartPoint(snapshot, cell.path), focus = cellEndPoint(snapshot, cell.path);
    !anchor || !focus || ranges.push({
      anchor,
      focus
    });
  }
  return ranges.length > 0 ? ranges : !1;
}
function createInsertBehaviors(config) {
  const {
    isRow,
    isTable
  } = createTableGuards(config);
  return [defineBehavior({
    on: "custom.insert.row",
    guard: ({
      snapshot,
      event
    }) => {
      const enclosingRow = getEnclosingBlock(snapshot, event.at, {
        match: isRow
      });
      if (enclosingRow)
        return {
          row: enclosingRow.node,
          rowPath: enclosingRow.path
        };
      const enclosingTable = getEnclosingBlock(snapshot, event.at, {
        match: isTable
      });
      if (!enclosingTable)
        return !1;
      const edgeRow = event.position === "before" ? tableRows(config, enclosingTable.node).at(0) : tableRows(config, enclosingTable.node).at(-1);
      return edgeRow ? {
        row: edgeRow,
        rowPath: [...enclosingTable.path, config.rowsField, {
          _key: edgeRow._key
        }]
      } : !1;
    },
    actions: [({
      event
    }, {
      row,
      rowPath
    }) => {
      const point = {
        path: rowPath,
        offset: 0
      };
      return [raise({
        type: "insert.block",
        block: {
          _type: config.rowType,
          [config.cellsField]: rowCells(config, row).map(() => ({
            _type: config.cellType
          }))
        },
        placement: event.position,
        select: "none",
        at: {
          anchor: point,
          focus: point
        }
      })];
    }]
  }), defineBehavior({
    on: "custom.insert.column",
    guard: ({
      snapshot,
      event
    }) => {
      const resolved = resolveCell(snapshot, event.at, config);
      if (resolved) {
        const columnIndex2 = rowCells(config, resolved.row.node).findIndex((cell) => cell._key === resolved.cell.node._key);
        return columnIndex2 === -1 ? !1 : {
          table: resolved.table.node,
          tablePath: resolved.table.path,
          columnIndex: columnIndex2
        };
      }
      const enclosingTable = getEnclosingBlock(snapshot, event.at, {
        match: isTable
      });
      if (!enclosingTable)
        return !1;
      const firstRow = tableRows(config, enclosingTable.node).at(0);
      if (!firstRow)
        return !1;
      const columnIndex = event.position === "before" ? 0 : rowCells(config, firstRow).length - 1;
      return {
        table: enclosingTable.node,
        tablePath: enclosingTable.path,
        columnIndex
      };
    },
    actions: [({
      event
    }, {
      table,
      tablePath,
      columnIndex
    }) => tableRows(config, table).flatMap((row) => {
      const cellAtColumn = rowCells(config, row).at(columnIndex);
      if (!cellAtColumn)
        return [];
      const point = {
        path: [...tablePath, config.rowsField, {
          _key: row._key
        }, config.cellsField, {
          _key: cellAtColumn._key
        }],
        offset: 0
      };
      return [raise({
        type: "insert.block",
        block: {
          _type: config.cellType
        },
        placement: event.position,
        select: "none",
        at: {
          anchor: point,
          focus: point
        }
      })];
    })]
  })];
}
function createMoveBehaviors(config) {
  const {
    isCell,
    isRow,
    isTable
  } = createTableGuards(config);
  return [defineBehavior({
    on: "custom.move.row",
    guard: ({
      snapshot,
      event
    }) => {
      const originRow = getEnclosingBlock(snapshot, event.at, {
        match: isRow
      });
      if (!originRow)
        return !1;
      const destinationRow = getEnclosingBlock(snapshot, event.to, {
        match: isRow
      });
      if (!destinationRow || originRow.node._key === destinationRow.node._key)
        return !1;
      const originTable = getParent(snapshot, originRow.path, {
        match: isTable
      }), destinationTable = getParent(snapshot, destinationRow.path, {
        match: isTable
      });
      return !originTable || !destinationTable || originTable.node._key !== destinationTable.node._key ? !1 : {
        originRowPath: originRow.path,
        destinationRowPath: destinationRow.path,
        savedSelection: snapshot.context.selection
      };
    },
    actions: [(_, {
      originRowPath,
      destinationRowPath,
      savedSelection
    }) => {
      const actions = [raise({
        type: "move.block",
        at: originRowPath,
        to: destinationRowPath
      })];
      return savedSelection && actions.push(raise({
        type: "select",
        at: savedSelection
      })), actions;
    }]
  }), defineBehavior({
    on: "custom.move.column",
    guard: ({
      snapshot,
      event
    }) => {
      const originCell = getEnclosingBlock(snapshot, event.at, {
        match: isCell
      });
      if (!originCell)
        return !1;
      const destinationCell = getEnclosingBlock(snapshot, event.to, {
        match: isCell
      });
      if (!destinationCell)
        return !1;
      const originRow = getParent(snapshot, originCell.path, {
        match: isRow
      }), destinationRow = getParent(snapshot, destinationCell.path, {
        match: isRow
      });
      if (!originRow || !destinationRow)
        return !1;
      const originTable = getParent(snapshot, originRow.path, {
        match: isTable
      }), destinationTable = getParent(snapshot, destinationRow.path, {
        match: isTable
      });
      if (!originTable || !destinationTable || originTable.node._key !== destinationTable.node._key)
        return !1;
      const originIndex = rowCells(config, originRow.node).findIndex((cell) => cell._key === originCell.node._key), destinationIndex = rowCells(config, destinationRow.node).findIndex((cell) => cell._key === destinationCell.node._key);
      return originIndex === -1 || destinationIndex === -1 || originIndex === destinationIndex ? !1 : {
        table: originTable.node,
        tablePath: originTable.path,
        originIndex,
        destinationIndex,
        savedSelection: snapshot.context.selection
      };
    },
    actions: [(_, {
      table,
      tablePath,
      originIndex,
      destinationIndex,
      savedSelection
    }) => {
      const actions = tableRows(config, table).flatMap((row) => {
        const originCell = rowCells(config, row).at(originIndex), destinationCell = rowCells(config, row).at(destinationIndex);
        if (!originCell || !destinationCell)
          return [];
        const rowPath = [...tablePath, config.rowsField, {
          _key: row._key
        }];
        return [raise({
          type: "move.block",
          at: [...rowPath, config.cellsField, {
            _key: originCell._key
          }],
          to: [...rowPath, config.cellsField, {
            _key: destinationCell._key
          }]
        })];
      });
      return savedSelection && actions.push(raise({
        type: "select",
        at: savedSelection
      })), actions;
    }]
  })];
}
const tab = createKeyboardShortcut({
  default: [{
    key: "Tab",
    alt: !1,
    ctrl: !1,
    meta: !1,
    shift: !1
  }]
}), shiftTab = createKeyboardShortcut({
  default: [{
    key: "Tab",
    alt: !1,
    ctrl: !1,
    meta: !1,
    shift: !0
  }]
}), arrowDown = createKeyboardShortcut({
  default: [{
    key: "ArrowDown",
    alt: !1,
    ctrl: !1,
    meta: !1,
    shift: !1
  }]
}), arrowUp = createKeyboardShortcut({
  default: [{
    key: "ArrowUp",
    alt: !1,
    ctrl: !1,
    meta: !1,
    shift: !1
  }]
});
function createNavBehaviors(config) {
  return [
    // Tab: move to the start of the next cell, wrapping to the first cell of the
    // next row. At the last cell the guard fails, so Tab falls through to the
    // editor default and is not overwritten.
    defineBehavior({
      on: "keyboard.keydown",
      guard: ({
        snapshot,
        event
      }) => {
        if (!tab.guard(event.originEvent) || getFocusTextBlock(snapshot)?.node.listItem !== void 0)
          return !1;
        const position = resolveFocusCell(config, snapshot);
        if (!position)
          return !1;
        const cells = getChildren(snapshot, position.row.path), nextRow = getSibling(snapshot, position.row.path, {
          direction: "next"
        }), target = cells[indexOf(cells, position.cell) + 1] ?? (nextRow && getFirstChild(snapshot, nextRow.path)), at = target && cellStart(snapshot, target.path);
        return at ? {
          at
        } : !1;
      },
      actions: [(_, {
        at
      }) => [raise({
        type: "select",
        at
      })]]
    }),
    // Shift+Tab: move to the start of the previous cell, wrapping to the last
    // cell of the previous row. At the first cell it falls through.
    defineBehavior({
      on: "keyboard.keydown",
      guard: ({
        snapshot,
        event
      }) => {
        if (!shiftTab.guard(event.originEvent) || getFocusTextBlock(snapshot)?.node.listItem !== void 0)
          return !1;
        const position = resolveFocusCell(config, snapshot);
        if (!position)
          return !1;
        const cells = getChildren(snapshot, position.row.path), columnIndex = indexOf(cells, position.cell), previousRow = getSibling(snapshot, position.row.path, {
          direction: "previous"
        }), target = (columnIndex > 0 ? cells[columnIndex - 1] : void 0) ?? (previousRow && getLastChild(snapshot, previousRow.path)), at = target && cellStart(snapshot, target.path);
        return at ? {
          at
        } : !1;
      },
      actions: [(_, {
        at
      }) => [raise({
        type: "select",
        at
      })]]
    }),
    // ArrowDown: when the caret is in the last block of a cell, move to the cell
    // directly below (same column). Otherwise fall through so the caret moves
    // within the cell.
    defineBehavior({
      on: "keyboard.keydown",
      guard: ({
        snapshot,
        event,
        dom
      }) => {
        if (!arrowDown.guard(event.originEvent) || getFocusBlockObject(snapshot))
          return !1;
        const position = resolveFocusCell(config, snapshot);
        if (!position || !focusAtCellEdge(snapshot, position.cell.path, "last") || !focusOnVisualEdge(snapshot, dom, "last"))
          return !1;
        const rowBelow = getSibling(snapshot, position.row.path, {
          direction: "next"
        });
        if (!rowBelow) {
          const siblingBelow = getSibling(snapshot, position.table.path, {
            direction: "next"
          });
          if (siblingBelow) {
            const at2 = blockEntrySelection(snapshot, dom, siblingBelow, "first");
            return at2 ? {
              at: at2
            } : !1;
          }
          return {
            escapeTablePath: position.table.path
          };
        }
        const target = sameColumnCell(snapshot, position.cell, position.row, rowBelow), at = target && cellEntrySelection(config, snapshot, dom, target.path, "first");
        return at ? {
          at
        } : !1;
      },
      actions: [({
        snapshot
      }, result) => "at" in result ? [raise({
        type: "select",
        at: result.at
      })] : [escapeTableAction(snapshot, result.escapeTablePath, "after")]]
    }),
    // ArrowUp: when the caret is in the first block of a cell, move to the cell
    // directly above (same column). Otherwise fall through.
    defineBehavior({
      on: "keyboard.keydown",
      guard: ({
        snapshot,
        event,
        dom
      }) => {
        if (!arrowUp.guard(event.originEvent) || getFocusBlockObject(snapshot))
          return !1;
        const position = resolveFocusCell(config, snapshot);
        if (!position || !focusAtCellEdge(snapshot, position.cell.path, "first") || !focusOnVisualEdge(snapshot, dom, "first"))
          return !1;
        const rowAbove = getSibling(snapshot, position.row.path, {
          direction: "previous"
        });
        if (!rowAbove) {
          const siblingAbove = getSibling(snapshot, position.table.path, {
            direction: "previous"
          });
          if (siblingAbove) {
            const at2 = blockEntrySelection(snapshot, dom, siblingAbove, "last");
            return at2 ? {
              at: at2
            } : !1;
          }
          return {
            escapeTablePath: position.table.path
          };
        }
        const target = sameColumnCell(snapshot, position.cell, position.row, rowAbove), at = target && cellEntrySelection(config, snapshot, dom, target.path, "last");
        return at ? {
          at
        } : !1;
      },
      actions: [({
        snapshot
      }, result) => "at" in result ? [raise({
        type: "select",
        at: result.at
      })] : [escapeTableAction(snapshot, result.escapeTablePath, "before")]]
    })
  ];
}
function escapeTableAction(snapshot, tablePath, placement) {
  return raise({
    type: "insert.block",
    block: {
      _type: snapshot.context.schema.block.name
    },
    placement,
    at: {
      anchor: {
        path: tablePath,
        offset: 0
      },
      focus: {
        path: tablePath,
        offset: 0
      }
    },
    select: "start"
  });
}
function resolveFocusCell(config, snapshot) {
  const selection = snapshot.context.selection;
  if (!(!selection || !isSelectionCollapsed(snapshot)))
    return resolveCell(snapshot, selection.focus.path, config);
}
function cellStart(snapshot, cellPath2) {
  const point = cellStartPoint(snapshot, cellPath2);
  return point && {
    anchor: point,
    focus: point
  };
}
function cellEnd(snapshot, cellPath2) {
  const point = cellEndPoint(snapshot, cellPath2);
  return point && {
    anchor: point,
    focus: point
  };
}
function cellEntrySelection(config, snapshot, dom, cellPath2, edge) {
  const fallback = edge === "first" ? cellStart(snapshot, cellPath2) : cellEnd(snapshot, cellPath2);
  if (!fallback)
    return;
  const caretRect = dom.getSelectionRect(snapshot), entryLineRect = dom.getSelectionRect({
    ...snapshot,
    context: {
      ...snapshot.context,
      selection: fallback
    }
  });
  if (!caretRect || !entryLineRect)
    return fallback;
  const point = dom.getPointAtCoordinates({
    x: caretRect.left,
    y: entryLineRect.top + entryLineRect.height / 2
  }), resolved = point && resolveCell(snapshot, point.path, config);
  return !resolved || !isEqualPaths(resolved.cell.path, cellPath2) ? fallback : {
    anchor: point,
    focus: point
  };
}
function blockEntrySelection(snapshot, dom, sibling, edge) {
  const block = {
    // A root-level sibling is a block by the data model; the traversal
    // node union cannot prove it.
    node: sibling.node,
    path: sibling.path
  }, edgePoint = edge === "first" ? getBlockStartPoint({
    context: snapshot.context,
    block
  }) : getBlockEndPoint({
    context: snapshot.context,
    block
  }), fallback = {
    anchor: edgePoint,
    focus: edgePoint
  }, caretRect = dom.getSelectionRect(snapshot), entryLineRect = dom.getSelectionRect({
    ...snapshot,
    context: {
      ...snapshot.context,
      selection: fallback
    }
  });
  if (!caretRect || !entryLineRect)
    return fallback;
  const point = dom.getPointAtCoordinates({
    x: caretRect.left,
    y: entryLineRect.top + entryLineRect.height / 2
  }), pointRootSegment = point?.path[0], blockRootSegment = block.path[0];
  return !point || pointRootSegment === void 0 || blockRootSegment === void 0 || !isEqualPaths([pointRootSegment], [blockRootSegment]) ? fallback : {
    anchor: point,
    focus: point
  };
}
function focusAtCellEdge(snapshot, cellPath2, edge) {
  const selection = snapshot.context.selection;
  if (!selection)
    return !1;
  const focusBlock = getEnclosingBlock(snapshot, selection.focus.path), edgeBlock = edge === "first" ? getFirstChild(snapshot, cellPath2) : getLastChild(snapshot, cellPath2);
  return !!focusBlock && !!edgeBlock && isEqualPaths(focusBlock.path, edgeBlock.path);
}
function focusOnVisualEdge(snapshot, dom, edge) {
  const selection = snapshot.context.selection;
  if (!selection)
    return !1;
  const focusBlock = getEnclosingBlock(snapshot, selection.focus.path);
  if (!focusBlock)
    return !1;
  const edgePoint = edge === "first" ? getBlockStartPoint({
    context: snapshot.context,
    block: focusBlock
  }) : getBlockEndPoint({
    context: snapshot.context,
    block: focusBlock
  }), caretRect = dom.getSelectionRect(snapshot), edgeRect = dom.getSelectionRect({
    ...snapshot,
    context: {
      ...snapshot.context,
      selection: {
        anchor: edgePoint,
        focus: edgePoint
      }
    }
  });
  return !!caretRect && !!edgeRect && sameLine(caretRect, edgeRect);
}
function sameColumnCell(snapshot, cell, row, neighborRow) {
  const columnIndex = indexOf(getChildren(snapshot, row.path), cell), cells = getChildren(snapshot, neighborRow.path);
  return cells[Math.min(columnIndex, cells.length - 1)];
}
function sameLine(first, second) {
  return withinBand(first, second) && withinBand(second, first);
}
function withinBand(rect, compareRect) {
  const middle = (compareRect.top + compareRect.bottom) / 2;
  return rect.top <= middle && rect.bottom >= middle;
}
function indexOf(entries, entry) {
  return entries.findIndex((candidate) => isEqualPaths(candidate.path, entry.path));
}
function createPasteBehaviors(config) {
  return [defineBehavior({
    on: "clipboard.paste",
    guard: ({
      snapshot,
      event
    }) => {
      const fragment = tableFragment(config, snapshot, event.originEvent.dataTransfer);
      if (!fragment)
        return !1;
      const anchor = resolveAnchorCell(config, snapshot);
      return anchor ? planDistribution(config, snapshot, fragment, anchor) : !1;
    },
    actions: [(_, {
      replacements,
      cellAppends,
      rowAppends,
      selection
    }) => [...cellAppends.map((append) => raise({
      type: "insert",
      at: append.afterCellPath,
      value: append.cell,
      position: "after"
    })), ...rowAppends.map((append) => raise({
      type: "insert",
      at: append.afterRowPath,
      value: append.row,
      position: "after"
    })), ...replacements.flatMap((replacement) => {
      const lastOriginalKey = replacement.originalBlockKeys[replacement.originalBlockKeys.length - 1];
      return lastOriginalKey === void 0 ? [] : [...replacement.blocks.map((block, index) => raise({
        type: "insert",
        at: [...replacement.cellPath, config.valueField, {
          _key: index === 0 ? lastOriginalKey : replacement.blocks[index - 1]?._key
        }],
        value: block,
        position: "after"
      })), ...replacement.originalBlockKeys.map((blockKey) => raise({
        type: "unset",
        at: [...replacement.cellPath, config.valueField, {
          _key: blockKey
        }]
      }))];
    }), raise({
      type: "select",
      at: selection
    })]]
  })];
}
function planDistribution(config, snapshot, fragment, anchor) {
  const keyGenerator = snapshot.context.keyGenerator, fragmentRows = tableRows(config, fragment), anchorRows = tableRows(config, anchor.table), fragmentColCount = Math.max(...fragmentRows.map((row) => rowCells(config, row).length)), tableColCount = Math.max(...anchorRows.map((row) => rowCells(config, row).length)), grownColCount = Math.max(tableColCount, anchor.colIndex + fragmentColCount), grownRowCount = Math.max(anchorRows.length, anchor.rowIndex + fragmentRows.length), firstRow = anchorRows[0], firstCell = firstRow ? rowCells(config, firstRow)[0] : void 0;
  if (!firstRow || !firstCell)
    return !1;
  const cellSchema = getPathSubSchema(snapshot, cellPathFor(config, anchor.tablePath, firstRow._key, firstCell._key)), allowedTypes = /* @__PURE__ */ new Set([cellSchema.block.name, ...cellSchema.blockObjects.map((blockObject) => blockObject.name)]), contentAt = (rowIndex, colIndex) => {
    const fragmentRow = fragmentRows[rowIndex - anchor.rowIndex], fragmentCell = fragmentRow ? rowCells(config, fragmentRow)[colIndex - anchor.colIndex] : void 0;
    if (!fragmentCell)
      return;
    const blocks = cellValue(config, fragmentCell).filter((block) => allowedTypes.has(block._type));
    return blocks.length > 0 ? {
      ...fragmentCell,
      [config.valueField]: blocks
    } : void 0;
  }, replacements = [], cellAppends = [], rowAppends = [], contentTargets = [], buildCell = (fragmentCell) => ({
    _type: config.cellType,
    _key: keyGenerator(),
    [config.valueField]: fragmentCell ? rekeyBlocks(keyGenerator, cellValue(config, fragmentCell)) : [emptyBlock(keyGenerator)]
  });
  let previousRowKey = anchorRows[anchorRows.length - 1]?._key;
  if (previousRowKey === void 0)
    return !1;
  for (let rowIndex = 0; rowIndex < grownRowCount; rowIndex++) {
    const existingRow = anchorRows[rowIndex];
    if (!existingRow) {
      const newCells = [], row = {
        _type: config.rowType,
        _key: keyGenerator(),
        [config.cellsField]: newCells
      };
      for (let colIndex = 0; colIndex < grownColCount; colIndex++) {
        const fragmentCell = contentAt(rowIndex, colIndex), cellNode = buildCell(fragmentCell);
        newCells.push(cellNode), fragmentCell && contentTargets.push({
          cellPath: cellPathFor(config, anchor.tablePath, row._key, cellNode._key),
          blocks: cellValue(config, cellNode)
        });
      }
      rowAppends.push({
        afterRowPath: [...anchor.tablePath, config.rowsField, {
          _key: previousRowKey
        }],
        row
      }), previousRowKey = row._key;
      continue;
    }
    const existingCells = rowCells(config, existingRow);
    let previousCellKey = existingCells[existingCells.length - 1]?._key;
    for (let colIndex = 0; colIndex < grownColCount; colIndex++) {
      const fragmentCell = contentAt(rowIndex, colIndex), existingCell = existingCells[colIndex];
      if (existingCell) {
        if (fragmentCell) {
          const blocks = rekeyBlocks(keyGenerator, cellValue(config, fragmentCell)), cellPath2 = cellPathFor(config, anchor.tablePath, existingRow._key, existingCell._key);
          replacements.push({
            cellPath: cellPath2,
            originalBlockKeys: cellValue(config, existingCell).map((block) => block._key),
            blocks
          }), contentTargets.push({
            cellPath: cellPath2,
            blocks
          });
        }
        continue;
      }
      if (previousCellKey === void 0)
        return !1;
      const cellNode = buildCell(fragmentCell);
      cellAppends.push({
        afterCellPath: [...anchor.tablePath, config.rowsField, {
          _key: existingRow._key
        }, config.cellsField, {
          _key: previousCellKey
        }],
        cell: cellNode
      }), previousCellKey = cellNode._key, fragmentCell && contentTargets.push({
        cellPath: cellPathFor(config, anchor.tablePath, existingRow._key, cellNode._key),
        blocks: cellValue(config, cellNode)
      });
    }
  }
  const selection = pastedSelection(config, snapshot.context, contentTargets);
  return selection ? {
    replacements,
    cellAppends,
    rowAppends,
    selection
  } : !1;
}
function cellPathFor(config, tablePath, rowKey, cellKey) {
  return [...tablePath, config.rowsField, {
    _key: rowKey
  }, config.cellsField, {
    _key: cellKey
  }];
}
function emptyBlock(keyGenerator) {
  return {
    _type: "block",
    _key: keyGenerator(),
    style: "normal",
    markDefs: [],
    children: [{
      _type: "span",
      _key: keyGenerator(),
      text: "",
      marks: []
    }]
  };
}
function tableFragment(config, snapshot, dataTransfer) {
  const data = dataTransfer.getData("application/x-portable-text");
  if (!data)
    return;
  const converter = snapshot.context.converters.find((candidate) => candidate.mimeType === "application/x-portable-text");
  if (!converter)
    return;
  const result = converter.deserialize({
    snapshot,
    event: {
      type: "deserialize",
      data
    }
  });
  if (result.type !== "deserialization.success")
    return;
  const blocks = result.data;
  if (!Array.isArray(blocks) || blocks.length !== 1)
    return;
  const block = blocks[0], {
    isTable
  } = createTableGuards(config);
  if (!isTable(block))
    return;
  const blockRows = tableRows(config, block);
  return blockRows.length > 0 && blockRows.every((row) => {
    const cells = rowCells(config, row);
    return cells.length > 0 && cells.every((cellNode) => Array.isArray(cellValue(config, cellNode)));
  }) ? block : void 0;
}
function resolveAnchorCell(config, snapshot) {
  const tableSelection = getTableSelection(snapshot, config);
  if (tableSelection) {
    const anchorCell = resolveCell(snapshot, snapshot.context.selection?.anchor.path ?? [], config);
    return anchorCell ? {
      table: anchorCell.table.node,
      tablePath: anchorCell.table.path,
      rowIndex: tableSelection.rowRange[0],
      colIndex: tableSelection.colRange[0]
    } : void 0;
  }
  if (!isSelectionCollapsed(snapshot))
    return;
  const caretCell = resolveCell(snapshot, snapshot.context.selection?.focus.path ?? [], config);
  if (!caretCell)
    return;
  const rowIndex = tableRows(config, caretCell.table.node).findIndex((row) => row._key === caretCell.row.node._key), colIndex = rowCells(config, caretCell.row.node).findIndex((cellNode) => cellNode._key === caretCell.cell.node._key);
  if (!(rowIndex === -1 || colIndex === -1))
    return {
      table: caretCell.table.node,
      tablePath: caretCell.table.path,
      rowIndex,
      colIndex
    };
}
function rekeyBlocks(keyGenerator, blocks) {
  return blocks.map((block) => {
    const blockKey = keyGenerator();
    if (!Array.isArray(block.children))
      return {
        ...block,
        _key: blockKey
      };
    const markDefKeys = /* @__PURE__ */ new Map(), markDefs = (Array.isArray(block.markDefs) ? block.markDefs : []).map((markDef) => {
      const markDefKey = keyGenerator();
      return markDefKeys.set(markDef._key, markDefKey), {
        ...markDef,
        _key: markDefKey
      };
    }), children = block.children.map((child) => ({
      ...child,
      _key: keyGenerator(),
      ...Array.isArray(child.marks) ? {
        marks: child.marks.map((mark) => markDefKeys.get(mark) ?? mark)
      } : {}
    }));
    return {
      ...block,
      _key: blockKey,
      markDefs,
      children
    };
  });
}
function pastedSelection(config, context, targets) {
  const first = targets.find((target) => target.blocks.length > 0), last = [...targets].reverse().find((target) => target.blocks.length > 0);
  if (!first || !last)
    return;
  const firstBlock = first.blocks[0], lastBlock = last.blocks[last.blocks.length - 1];
  return {
    anchor: getBlockStartPoint({
      context,
      block: {
        node: firstBlock,
        path: [...first.cellPath, config.valueField, {
          _key: firstBlock._key
        }]
      }
    }),
    focus: getBlockEndPoint({
      context,
      block: {
        node: lastBlock,
        path: [...last.cellPath, config.valueField, {
          _key: lastBlock._key
        }]
      }
    })
  };
}
function createSelectBehaviors(config) {
  const {
    isCell,
    isTable
  } = createTableGuards(config);
  return [defineBehavior({
    on: "select",
    guard: ({
      snapshot,
      event
    }) => {
      if (!event.at)
        return !1;
      const anchorCell = getEnclosingBlock(snapshot, event.at.anchor.path, {
        match: isCell
      });
      if (!anchorCell || getEnclosingBlock(snapshot, event.at.focus.path, {
        match: isCell
      }))
        return !1;
      const focusTable = getEnclosingBlock(snapshot, event.at.focus.path, {
        match: isTable
      }), anchorTable = getEnclosingBlock(snapshot, anchorCell.path, {
        match: isTable
      });
      if (!focusTable || !anchorTable || !isEqualPaths(focusTable.path, anchorTable.path))
        return !1;
      const previousFocus = snapshot.context.selection?.focus, previousFocusCell = previousFocus ? getEnclosingBlock(snapshot, previousFocus.path, {
        match: isCell
      }) : void 0, previousFocusTable = previousFocusCell ? getEnclosingBlock(snapshot, previousFocusCell.path, {
        match: isTable
      }) : void 0, clampedFocus = previousFocus && previousFocusTable && isEqualPaths(previousFocusTable.path, anchorTable.path) ? previousFocus : event.at.backward ? cellStartPoint(snapshot, anchorCell.path) : cellEndPoint(snapshot, anchorCell.path);
      return clampedFocus ? {
        anchor: event.at.anchor,
        focus: clampedFocus,
        backward: comparePoints(snapshot, clampedFocus, event.at.anchor) === -1
      } : !1;
    },
    actions: [(_, at) => [raise({
      type: "select",
      at
    })]]
  })];
}
function createSerializeBehaviors(config) {
  return [defineBehavior({
    on: "serialize.data",
    guard: ({
      snapshot,
      event
    }) => {
      if (event.originEvent.type !== "clipboard.copy" && event.originEvent.type !== "clipboard.cut")
        return !1;
      const resolved = resolveTableSelection(snapshot, config);
      if (!resolved)
        return !1;
      const converter = snapshot.context.converters.find((candidate) => candidate.mimeType === event.mimeType);
      if (!converter)
        return !1;
      const rectangle = sliceTable(config, resolved.table.node, resolved.tableSelection);
      if (event.mimeType === "text/plain")
        return {
          type: "serialization.success",
          data: tableToTsv(config, rectangle),
          mimeType: "text/plain",
          originEvent: event.originEvent.type
        };
      const rectangleRows = tableRows(config, rectangle), firstRow = rectangleRows[0], lastRow = rectangleRows[rectangleRows.length - 1], firstCell = firstRow ? rowCells(config, firstRow)[0] : void 0, lastCells = lastRow ? rowCells(config, lastRow) : [], lastCell = lastCells[lastCells.length - 1];
      if (!firstRow || !lastRow || !firstCell || !lastCell)
        return !1;
      const doctoredSnapshot = {
        ...snapshot,
        context: {
          ...snapshot.context,
          value: [rectangle]
        }
      }, anchor = cellStartPoint(doctoredSnapshot, cellPath(config, rectangle, firstRow, firstCell)), focus = cellEndPoint(doctoredSnapshot, cellPath(config, rectangle, lastRow, lastCell));
      return !anchor || !focus ? !1 : converter.serialize({
        snapshot: {
          ...doctoredSnapshot,
          context: {
            ...doctoredSnapshot.context,
            selection: {
              anchor,
              focus
            }
          }
        },
        event: {
          type: "serialize",
          originEvent: event.originEvent.type
        }
      });
    },
    actions: [({
      event
    }, serialization) => [raise({
      ...serialization,
      originEvent: event.originEvent
    })]]
  })];
}
function sliceTable(config, table, tableSelection) {
  const [rowStart, rowEnd] = tableSelection.rowRange, [colStart, colEnd] = tableSelection.colRange, slicedRows = tableRows(config, table).slice(rowStart, rowEnd + 1).map((row) => ({
    ...row,
    [config.cellsField]: rowCells(config, row).slice(colStart, colEnd + 1)
  })), sliced = {
    ...table,
    [config.rowsField]: slicedRows
  };
  return typeof table.headerRows == "number" && (sliced.headerRows = rowStart === 0 ? Math.min(table.headerRows, slicedRows.length) : 0), sliced;
}
function tableToTsv(config, table) {
  return tableRows(config, table).map((row) => rowCells(config, row).map((cell) => cellText(config, cell)).join("	")).join(`
`);
}
function cellText(config, cell) {
  return cellValue(config, cell).map((block) => "children" in block && Array.isArray(block.children) ? block.children.map((child) => typeof child.text == "string" ? child.text : "").join("") : "").join(" ").replace(/[\t\n]/g, " ");
}
function cellPath(config, table, row, cell) {
  return [{
    _key: table._key
  }, config.rowsField, {
    _key: row._key
  }, config.cellsField, {
    _key: cell._key
  }];
}
function selectionTouches(selection, paths) {
  return selection ? paths.some((path) => pathContains(path, selection.anchor.path) || pathContains(path, selection.focus.path)) : !1;
}
function createUnsetBehaviors(config) {
  const {
    isCell,
    isRow,
    isTable
  } = createTableGuards(config);
  return [defineBehavior({
    on: "custom.unset.row",
    guard: ({
      snapshot,
      event
    }) => {
      const enclosingRow = getEnclosingBlock(snapshot, event.at, {
        match: isRow
      });
      if (!enclosingRow)
        return !1;
      const table = getParent(snapshot, enclosingRow.path, {
        match: isTable
      });
      if (!table)
        return !1;
      const rows = tableRows(config, table.node), rowIndex = rows.findIndex((row) => row._key === enclosingRow.node._key);
      if (rowIndex === -1)
        return !1;
      const neighborRow = rows[rowIndex + 1] ?? rows[rowIndex - 1];
      if (!neighborRow)
        return !1;
      const enclosingCell = getEnclosingBlock(snapshot, event.at, {
        match: isCell
      }), columnIndex = enclosingCell ? rowCells(config, enclosingRow.node).findIndex((cell) => cell._key === enclosingCell.node._key) : 0, safeColumnIndex = Math.min(Math.max(columnIndex, 0), rowCells(config, neighborRow).length - 1), neighborCell = rowCells(config, neighborRow).at(safeColumnIndex);
      if (!neighborCell)
        return !1;
      const neighborCellPath = [...table.path, config.rowsField, {
        _key: neighborRow._key
      }, config.cellsField, {
        _key: neighborCell._key
      }];
      return {
        rowPath: enclosingRow.path,
        columnIndex: safeColumnIndex,
        neighborCellPath
      };
    },
    actions: [({
      snapshot
    }, {
      rowPath,
      neighborCellPath
    }) => {
      const actions = [raise({
        type: "unset",
        at: rowPath
      })];
      if (selectionTouches(snapshot.context.selection, [rowPath])) {
        const leaf = getLeaf(snapshot, neighborCellPath, {
          edge: "start"
        });
        if (leaf) {
          const point = {
            path: leaf.path,
            offset: 0
          };
          actions.push(raise({
            type: "select",
            at: {
              anchor: point,
              focus: point
            }
          }));
        }
      }
      return actions;
    }]
  }), defineBehavior({
    on: "custom.unset.column",
    guard: ({
      snapshot,
      event
    }) => {
      const enclosingCell = getEnclosingBlock(snapshot, event.at, {
        match: isCell
      });
      if (!enclosingCell)
        return !1;
      const enclosingRow = getParent(snapshot, enclosingCell.path, {
        match: isRow
      });
      if (!enclosingRow)
        return !1;
      const columnIndex = rowCells(config, enclosingRow.node).findIndex((cell) => cell._key === enclosingCell.node._key);
      if (columnIndex === -1)
        return !1;
      const table = getParent(snapshot, enclosingRow.path, {
        match: isTable
      });
      if (!table || rowCells(config, enclosingRow.node).length <= 1)
        return !1;
      const neighborColumnIndex = columnIndex + 1 < rowCells(config, enclosingRow.node).length ? columnIndex + 1 : columnIndex - 1, neighborCell = rowCells(config, enclosingRow.node).at(neighborColumnIndex);
      if (!neighborCell)
        return !1;
      const neighborCellPath = [...enclosingRow.path, config.cellsField, {
        _key: neighborCell._key
      }];
      return {
        cellPaths: tableRows(config, table.node).flatMap((row) => {
          const cellAtColumn = rowCells(config, row).at(columnIndex);
          return cellAtColumn ? [[...table.path, config.rowsField, {
            _key: row._key
          }, config.cellsField, {
            _key: cellAtColumn._key
          }]] : [];
        }),
        neighborCellPath
      };
    },
    actions: [({
      snapshot
    }, {
      cellPaths,
      neighborCellPath
    }) => {
      const actions = cellPaths.map((cellPath2) => raise({
        type: "unset",
        at: cellPath2
      }));
      if (selectionTouches(snapshot.context.selection, cellPaths)) {
        const leaf = getLeaf(snapshot, neighborCellPath, {
          edge: "start"
        });
        if (leaf) {
          const point = {
            path: leaf.path,
            offset: 0
          };
          actions.push(raise({
            type: "select",
            at: {
              anchor: point,
              focus: point
            }
          }));
        }
      }
      return actions;
    }]
  }), defineBehavior({
    on: "custom.unset.table",
    guard: ({
      snapshot,
      event
    }) => {
      const enclosingTable = getEnclosingBlock(snapshot, event.at, {
        match: isTable
      });
      if (!enclosingTable)
        return !1;
      const nextSibling = getSibling(snapshot, enclosingTable.path, {
        direction: "next"
      }), previousSibling = getSibling(snapshot, enclosingTable.path, {
        direction: "previous"
      });
      return {
        tablePath: enclosingTable.path,
        neighborBlockPath: nextSibling?.path ?? previousSibling?.path
      };
    },
    actions: [({
      snapshot
    }, {
      tablePath,
      neighborBlockPath
    }) => {
      const actions = [raise({
        type: "unset",
        at: tablePath
      })];
      if (neighborBlockPath && selectionTouches(snapshot.context.selection, [tablePath])) {
        const leaf = getLeaf(snapshot, neighborBlockPath, {
          edge: "start"
        });
        if (leaf) {
          const point = {
            path: leaf.path,
            offset: 0
          };
          actions.push(raise({
            type: "select",
            at: {
              anchor: point,
              focus: point
            }
          }));
        }
      }
      return actions;
    }]
  })];
}
function defineTable(options = {}) {
  const containers = options.containers ?? {}, table = withFallbackRender(containers.table ?? canonicalTableContainer, bareTableRender), row = withFallbackRender(containers.row ?? canonicalRowContainer, bareRowRender), cell = withFallbackRender(containers.cell ?? canonicalCellContainer, bareCellRender);
  for (const [role, definition] of [["table", table], ["row", row]])
    definition.of && console.warn(`[@portabletext/plugin-table] The ${role} definition ('${definition.type}') declares an \`of\`, which \`defineTable\` owns: the nesting is grafted as table.of \u2192 row.of \u2192 cell. The declared \`of\` is ignored. Cell-scoped node definitions belong on the cell definition's \`of\`.`);
  const config = {
    tableType: table.type,
    rowsField: table.arrayField,
    rowType: row.type,
    cellsField: row.arrayField,
    cellType: cell.type,
    valueField: cell.arrayField
  };
  registerTableConfig(config);
  const graftedTable = {
    ...table,
    of: [{
      ...row,
      of: [cell]
    }]
  }, behaviors = createTableBehaviors(config);
  function Plugin() {
    const $ = c(1);
    let t0;
    return $[0] === /* @__PURE__ */ Symbol.for("react.memo_cache_sentinel") ? (t0 = /* @__PURE__ */ jsxs(Fragment, { children: [
      /* @__PURE__ */ jsx(NodePlugin, { nodes: [graftedTable] }),
      /* @__PURE__ */ jsx(BehaviorPlugin, { behaviors })
    ] }), $[0] = t0) : t0 = $[0], t0;
  }
  return {
    Plugin,
    behaviors,
    createBlock: ({
      rows = 3,
      columns = 3,
      headerRows
    } = {}) => ({
      _type: config.tableType,
      ...headerRows === void 0 ? {} : {
        headerRows
      },
      [config.rowsField]: Array.from({
        length: rows
      }, () => ({
        _type: config.rowType,
        [config.cellsField]: Array.from({
          length: columns
        }, () => ({
          _type: config.cellType,
          [config.valueField]: [{
            _type: "block",
            style: "normal",
            markDefs: [],
            children: [{
              _type: "span",
              text: "",
              marks: []
            }]
          }]
        }))
      }))
    }),
    getTableSelection: (snapshot) => getTableSelection(snapshot, config),
    ...createTableGuards(config)
  };
}
function createTableBehaviors(config) {
  return [...createInsertBehaviors(config), ...createUnsetBehaviors(config), ...createMoveBehaviors(config), ...createDeleteBehaviors(config), ...createNavBehaviors(config), ...createSelectBehaviors(config), ...createFormatBehaviors(config), ...createSerializeBehaviors(config), ...createPasteBehaviors(config)];
}
const bareTableRender = ({
  attributes,
  children
}) => /* @__PURE__ */ jsx("table", { ...attributes, children: /* @__PURE__ */ jsx("tbody", { children }) }), bareRowRender = ({
  attributes,
  children
}) => /* @__PURE__ */ jsx("tr", { ...attributes, children }), bareCellRender = ({
  attributes,
  children
}) => /* @__PURE__ */ jsx("td", { ...attributes, children }), canonicalTableContainer = defineContainer({
  type: "table",
  arrayField: "rows"
}), canonicalRowContainer = defineContainer({
  type: "row",
  arrayField: "cells"
}), canonicalCellContainer = defineContainer({
  type: "cell",
  arrayField: "value"
});
function withFallbackRender(definition, fallback) {
  return definition.render ? definition : {
    ...definition,
    render: fallback
  };
}
export {
  defineTable
};
