// @sanity/workbench-cli/dist/actions/build/vite/plugins/plugin-sanity-environment.js
import { FEDERATION_DIR_NAME } from '../constants.js';
export function sanityEnvironmentPlugin(options) {
    const { clientInput, input } = options;
    return {
        config () {
            return {
                builder: {
                    async buildApp (builder) {
                        // `emptyOutDir` is false on both environments and the CLI clears
                        // `dist` once up-front, so the SPA and federation outputs coexist
                        // without either build wiping the other's files.
                        if (clientInput) {
                            await builder.build(builder.environments.client);
                        }
                        await builder.build(builder.environments[FEDERATION_DIR_NAME]);
                    }
                },
                environments: {
                    ...clientInput ? {
                        client: {
                            build: {
                                assetsDir: 'static',
                                copyPublicDir: false,
                                emptyOutDir: false,
                                outDir: `dist`,
                                rolldownOptions: {
                                    input: {
                                        sanity: clientInput
                                    }
                                }
                            },
                            consumer: 'client'
                        }
                    } : {},
                    [FEDERATION_DIR_NAME]: {
                        build: {
                            copyPublicDir: false,
                            emptyOutDir: false,
                            outDir: `dist`,
                            rolldownOptions: {
                                input
                            }
                        },
                        consumer: 'client'
                    }
                }
            };
        },
        name: 'sanity/environment'
    };
}
