// @sanity/runtime-cli/dist/utils/logger.js
import { format } from 'node:util';
import ora from 'ora';
import { clearProgress, showBusy, showError, showProgress } from './progress.js';
var LogLevel;
(function (LogLevel) {
    LogLevel[LogLevel["TRACE"] = 0] = "TRACE";
    LogLevel[LogLevel["VERBOSE"] = 1] = "VERBOSE";
    LogLevel[LogLevel["INFO"] = 2] = "INFO";
    LogLevel[LogLevel["WARN"] = 3] = "WARN";
    LogLevel[LogLevel["ERROR"] = 4] = "ERROR";
})(LogLevel || (LogLevel = {}));
export function Logger(log, flags = {}) {
    const logger = (msg) => {
        log(msg);
    };
    const level = flags.verbose ? LogLevel.VERBOSE : flags.trace ? LogLevel.TRACE : LogLevel.INFO;
    logger.isTraceEnabled = level <= LogLevel.TRACE;
    logger.trace = (formatter, ...args) => level <= LogLevel.TRACE && logger(format(formatter, ...args));
    logger.verbose = (formatter, ...args) => level <= LogLevel.VERBOSE && logger(format(formatter, ...args));
    logger.info = (formatter, ...args) => level <= LogLevel.INFO && logger(format(formatter, ...args));
    logger.warn = (formatter, ...args) => level <= LogLevel.WARN && logger(format(formatter, ...args));
    logger.error = (formatter, ...args) => level <= LogLevel.ERROR && logger(format(formatter, ...args));
    logger.showProgress = (percent) => {
        if (flags.json)
            return;
        if (percent === undefined)
            showBusy();
        else
            showProgress(percent);
    };
    logger.clearProgress = () => {
        if (flags.json)
            return;
        clearProgress();
    };
    const oraWrapper = (opts) => {
        if (flags.json)
            return createOraLineLoggingWrapper(opts, logger);
        if (level >= LogLevel.INFO)
            return withTerminalProgress(ora(opts));
        return createOraLineLoggingWrapper(opts, logger);
    };
    logger.ora = oraWrapper;
    return logger;
}
function withTerminalProgress(spinner) {
    const wrap = (name, effect) => {
        if (typeof spinner[name] !== 'function')
            return;
        const original = spinner[name].bind(spinner);
        spinner[name] = (...args) => {
            const result = original(...args);
            effect();
            return result;
        };
    };
    wrap('start', showBusy);
    wrap('fail', showError);
    for (const name of ['stop', 'succeed', 'warn', 'info', 'stopAndPersist']) {
        wrap(name, clearProgress);
    }
    return spinner;
}
/**
 * An ora-like wrapper that uses standard line output logging instead of rendering a spinner; used in verbose and trace modes.
 */
function createOraLineLoggingWrapper(opts, logger) {
    const text = typeof opts === 'string' ? opts : (typeof opts === 'object' ? opts.text : '') || '';
    const wrapper = {
        clear() {
            logger('');
            return wrapper;
        },
        fail(subtext) {
            logger.error(subtext);
            return wrapper;
        },
        info(subtext) {
            logger(subtext || wrapper.text);
            return wrapper;
        },
        start(subtext) {
            logger(subtext || wrapper.text);
            return wrapper;
        },
        stop() {
            return wrapper;
        },
        succeed(subtext) {
            logger(subtext || wrapper.text);
            return wrapper;
        },
        text,
    };
    return wrapper;
}
