// @sanity/runtime-cli/dist/utils/progress.js
import { stderr } from 'node:process';
import { onExit } from 'signal-exit';
const ST = '\x1b\\';
const STATES = {
    clear: 0,
    normal: 1,
    error: 2,
    paused: 4,
    busy: 3,
};
let exitHookRegistered = false;
function write(state, value = 0) {
    if (!stderr.isTTY)
        return;
    if (!exitHookRegistered) {
        exitHookRegistered = true;
        function clear() {
            stderr.write(`\x1b]9;4;0;0${ST}`);
        }
        // arc/utils SIGINT handler clobbers our signal-exit; luckily it then calls process.exit()
        process.once('exit', clear); // triggered by arc/utils SIGINT handler
        onExit(clear); // triggered by all other exit paths
    }
    stderr.write(`\x1b]9;4;${STATES[state]};${Math.round(value)}${ST}`);
}
export function showProgress(percent) {
    write('normal', Math.max(0, Math.min(100, percent)));
}
export function showBusy() {
    write('busy');
}
export function showError() {
    write('error', 100);
}
export function showPaused(percent) {
    write('paused', Math.max(0, Math.min(100, percent)));
}
export function clearProgress() {
    write('clear');
}
