// @sanity/runtime-cli/dist/config.js
import { readFileSync } from 'node:fs';
import { env } from 'node:process';
import { fileURLToPath } from 'node:url';
import * as pkg from 'empathic/package';
import getToken from './utils/get-token.js';
import isTrueish from './utils/isTrueish.js';
export const BLUEPRINT_CONFIG_VERSION = 'v2025-05-08';
// Default API version for Blueprints service calls.
// Overridable globally via env, or per-endpoint via the `version` arg to `blueprintsUrl`.
export const BLUEPRINTS_API_VERSION = env.SANITY_BLUEPRINTS_API_VERSION ?? 'v2026-06-15';
export const BLUEPRINT_CONFIG_DIR = '.sanity';
export const BLUEPRINT_CONFIG_FILE = 'blueprint.config.json';
export const RUNTIME_CLI_NAME = '@sanity/runtime-cli';
export let RUNTIME_CLI_VERSION;
try {
    const packageJsonPath = pkg.up({ cwd: fileURLToPath(new URL('.', import.meta.url)) });
    if (packageJsonPath) {
        const packageJson = JSON.parse(readFileSync(packageJsonPath, 'utf8'));
        if (packageJson.name === RUNTIME_CLI_NAME) {
            RUNTIME_CLI_VERSION = packageJson.version;
        }
    }
}
catch { }
export const RUNTIME_CLI_USER_AGENT = RUNTIME_CLI_VERSION
    ? `${RUNTIME_CLI_NAME}@${RUNTIME_CLI_VERSION}`
    : RUNTIME_CLI_NAME;
const nodeEnv = env.NODE_ENV?.toLowerCase() ?? 'production';
const isTest = nodeEnv === 'test';
const sanityEnv = env.SANITY_INTERNAL_ENV?.toLowerCase() ?? 'production';
const sanityProd = sanityEnv === 'production';
const isLive = sanityProd || sanityEnv === 'staging';
const isPublishing = isTrueish(env?.PUBLISHING);
// https://github.com/sanity-io/cli/blob/adf90ea6691ee6c36849e426de9729533706f47f/packages/%40sanity/cli-core/src/util/isCi.ts
const isCi = 
// Travis CI, CircleCI, Gitlab CI, Appveyor, CodeShip, Travis CI, Jenkins, TeamCity
isTrueish(env?.CI) || isTrueish(env?.CONTINUOUS_INTEGRATION) || isTrueish(env?.BUILD_NUMBER);
const apiUrls = {
    production: 'https://api.sanity.io/',
    staging: 'https://api.sanity.work/',
    default: 'http://api.sanity/',
};
const apiUrl = sanityEnv ? (apiUrls[sanityEnv] ?? sanityEnv) : apiUrls.default;
// allow override of populus URL
const populusApiUrl = env.POPULUS_API_URL ?? apiUrl;
/**
 * Build a Blueprints API URL for the given path.
 *
 * Uses `BLUEPRINTS_API_VERSION` by default (itself overridable via the
 * `SANITY_BLUEPRINTS_API_VERSION` env var); pass `version` to override per-endpoint.
 */
export function blueprintsUrl(path, version = BLUEPRINTS_API_VERSION) {
    return `${apiUrl}${version}/blueprints/${path}`;
}
let _token;
function loadToken() {
    let token = null;
    if (isTest || isPublishing) {
        token = 'token';
    }
    else {
        try {
            token = getToken({ prod: sanityProd });
        }
        catch (error) {
            if (error instanceof Error && 'code' in error && error.code === 'ENOENT') {
                // do nothing, validateToken and commands handle missing config file
            }
            else {
                // rethrow other errors like JSON.parse failure
                throw error;
            }
        }
    }
    return token;
}
export default {
    isTest,
    isLive,
    apiUrl,
    populusApiUrl,
    isCi,
    get token() {
        return _token === undefined ? loadToken() : _token;
    },
    set token(value) {
        _token = value;
    },
};
