// @sanity/cli/dist/commands/organizations/list.js
import { exitCodes, SanityCommand, subdebug } from '@sanity/cli-core';
import { getErrorMessage } from '@sanity/cli-core/errors';
import { Table } from 'console-table-printer';
import { listOrganizations } from '../../services/organizations.js';
import { organizationAliases } from '../../util/organizationAliases.js';
const listOrgsDebug = subdebug('organizations:list');
export class ListOrganizationsCommand extends SanityCommand {
    static description = 'List organizations you are a member of';
    static examples = [
        {
            command: '<%= config.bin %> <%= command.id %>',
            description: 'List all your organizations'
        }
    ];
    static hiddenAliases = organizationAliases('list');
    async run() {
        let organizations;
        try {
            organizations = await listOrganizations();
        } catch (error) {
            listOrgsDebug('Error listing organizations', error);
            this.error(`Failed to list organizations: ${getErrorMessage(error)}`, {
                exit: exitCodes.RUNTIME_ERROR
            });
        }
        if (organizations.length === 0) {
            this.log('No organizations found');
            return;
        }
        const table = new Table({
            columns: [
                {
                    alignment: 'left',
                    name: 'id',
                    title: 'ID'
                },
                {
                    alignment: 'left',
                    name: 'name',
                    title: 'Name'
                },
                {
                    alignment: 'left',
                    name: 'slug',
                    title: 'Slug'
                }
            ]
        });
        for (const { id, name, slug } of organizations){
            table.addRow({
                id,
                name,
                slug: slug ?? '-'
            });
        }
        table.printTable();
    }
}
