// @portabletext/editor/lib/_chunks-es/get-path-sub-schema.js
import { isKeyedSegment, serializePath, getNodeChildren, getNode, isTypedObject, getChildren, parentPath, getParent } from "./get-parent.js";
import { getSubSchema } from "@portabletext/schema";
function getAncestors(snapshot, path) {
  const keyedIndices = [];
  for (let i = 0; i < path.length; i++)
    isKeyedSegment(path[i]) && keyedIndices.push(i);
  if (keyedIndices.length <= 1)
    return [];
  const {
    context,
    blockIndexMap
  } = snapshot;
  let currentChildren = context.value, currentParent;
  const ancestorsByDepth = [], resolvedPath = [], targetKeyedIndex = keyedIndices[keyedIndices.length - 1];
  let segmentIndex = 0;
  for (; segmentIndex < targetKeyedIndex; ) {
    const segment = path[segmentIndex];
    if (typeof segment == "string") {
      resolvedPath.push(segment), segmentIndex++;
      continue;
    }
    let node;
    if (isKeyedSegment(segment)) {
      resolvedPath.push(segment);
      const index = blockIndexMap.get(serializePath(resolvedPath));
      index !== void 0 && currentChildren[index]?._key === segment._key ? node = currentChildren[index] : (node = currentChildren.find((child) => child._key === segment._key), node && node._key !== void 0 && (resolvedPath[resolvedPath.length - 1] = {
        _key: node._key
      }));
    } else if (typeof segment == "number")
      node = currentChildren.at(segment), node && resolvedPath.push({
        _key: node._key
      });
    else
      return [];
    if (!node)
      return [];
    const next = getNodeChildren(context, node, currentParent);
    if (!next)
      return [];
    ancestorsByDepth.push({
      node,
      path: resolvedPath.slice()
    }), currentChildren = next.children, currentParent = next.parent, segmentIndex++;
  }
  return ancestorsByDepth.reverse();
}
function hasNode(snapshot, path) {
  return getNode(snapshot, path) !== void 0;
}
function resolveContainerAt(containers, value, path) {
  const keyedIndices = [];
  for (let index = 0; index < path.length; index++)
    isKeyedSegment(path[index]) && keyedIndices.push(index);
  if (keyedIndices.length === 0)
    return;
  let currentChildren = value, parent, resolved;
  const targetKeyedIndex = keyedIndices[keyedIndices.length - 1];
  let segmentIndex = 0;
  for (; segmentIndex <= targetKeyedIndex; ) {
    const segment = path[segmentIndex];
    if (typeof segment == "string") {
      segmentIndex++;
      continue;
    }
    let node;
    if (isKeyedSegment(segment))
      node = currentChildren.find((child) => child._key === segment._key);
    else if (typeof segment == "number")
      node = currentChildren.at(segment);
    else
      return;
    if (!node || (resolved = resolveNodeEntry(containers, parent, node), !resolved))
      return;
    if (segmentIndex < targetKeyedIndex) {
      if (!("field" in resolved))
        return;
      const fieldValue = node[resolved.field.name];
      if (!Array.isArray(fieldValue))
        return;
      parent = resolved, currentChildren = fieldValue;
    }
    segmentIndex++;
  }
  return resolved;
}
function resolveNodeEntry(containers, parent, node) {
  if (parent?.of) {
    for (const entry of parent.of)
      if (entry.type === node._type)
        return entry;
  }
  return containers.get(node._type);
}
function isEditableContainer(snapshot, _node, path) {
  if (snapshot.context.containers.size === 0)
    return !1;
  const resolved = resolveContainerAt(snapshot.context.containers, snapshot.context.value, path);
  return !!(resolved && "field" in resolved);
}
function isObject(snapshot, node) {
  return isTypedObject(node) && node._type !== snapshot.context.schema.block.name && node._type !== snapshot.context.schema.span.name;
}
function comparePaths(path, another, root) {
  const min = Math.min(path.length, another.length);
  let currentChildren = root?.value, currentNode;
  for (let i = 0; i < min; i++) {
    const segment = path[i], otherSegment = another[i];
    if (isKeyedSegment(segment) && isKeyedSegment(otherSegment)) {
      if (segment._key === otherSegment._key) {
        currentChildren && (currentNode = currentChildren.find((c) => c._key === segment._key), currentChildren = void 0);
        continue;
      }
      if (currentChildren) {
        const segmentIndex = currentChildren.findIndex((c) => c._key === segment._key), otherSegmentIndex = currentChildren.findIndex((c) => c._key === otherSegment._key);
        if (segmentIndex !== -1 && otherSegmentIndex !== -1)
          return segmentIndex < otherSegmentIndex ? -1 : 1;
      }
      if (segment._key < otherSegment._key)
        return -1;
      if (segment._key > otherSegment._key)
        return 1;
      continue;
    }
    if (typeof segment == "string" && typeof otherSegment == "string") {
      if (segment === otherSegment) {
        if (currentNode) {
          const fieldValue = currentNode[segment];
          currentChildren = Array.isArray(fieldValue) ? fieldValue : void 0, currentNode = void 0;
        }
        continue;
      }
      if (segment < otherSegment)
        return -1;
      if (segment > otherSegment)
        return 1;
      continue;
    }
    if (typeof segment == "number" && typeof otherSegment == "number") {
      if (segment < otherSegment)
        return -1;
      if (segment > otherSegment)
        return 1;
      continue;
    }
    break;
  }
  return 0;
}
function comparePoints$1(point, another, root) {
  const result = comparePaths(point.path, another.path, root);
  return result === 0 ? point.offset < another.offset ? -1 : point.offset > another.offset ? 1 : 0 : result;
}
function isAfterPoint(point, another, root) {
  return comparePoints$1(point, another, root) === 1;
}
function isBackwardRange(range, root) {
  const {
    anchor,
    focus
  } = range;
  return isAfterPoint(anchor, focus, root);
}
function rangeEdges(range, root) {
  const {
    anchor,
    focus
  } = range;
  return isBackwardRange(range, root) ? [focus, anchor] : [anchor, focus];
}
function isTextBlockNode(context, node) {
  return isTypedObject(node) && node._type === context.schema.block.name;
}
function isAncestorPath(path, another) {
  if (path.length >= another.length)
    return !1;
  for (let i = 0; i < path.length; i++) {
    const segment = path[i], otherSegment = another[i];
    if (isKeyedSegment(segment) && isKeyedSegment(otherSegment)) {
      if (segment._key !== otherSegment._key)
        return !1;
    } else if (segment !== otherSegment)
      return !1;
  }
  return !0;
}
function resolveChildEntryIndex(blockIndexMap, children, childPath) {
  const lastSegment = childPath[childPath.length - 1];
  if (typeof lastSegment == "number")
    return lastSegment >= 0 && lastSegment < children.length ? lastSegment : -1;
  if (!isKeyedSegment(lastSegment))
    return -1;
  let fullyKeyed = !0;
  for (let segmentIndex = 0; segmentIndex < childPath.length - 1; segmentIndex++) {
    const segment = childPath[segmentIndex];
    if (typeof segment != "string" && !isKeyedSegment(segment)) {
      fullyKeyed = !1;
      break;
    }
  }
  if (fullyKeyed) {
    const mappedIndex = blockIndexMap.get(serializePath(childPath));
    if (mappedIndex !== void 0 && children[mappedIndex]?.node._key === lastSegment._key)
      return mappedIndex;
  }
  return children.findIndex((child) => child.node._key === lastSegment._key);
}
function* getNodes(snapshot, options = {}) {
  const {
    at = [],
    from,
    to,
    match,
    reverse = !1
  } = options;
  if (from === void 0 && to === void 0) {
    yield* getNodesSimple(snapshot, at, {
      match,
      reverse
    });
    return;
  }
  yield* getNodesInRange(snapshot, at, {
    from,
    to,
    match,
    reverse
  });
}
function* getNodesSimple(snapshot, path, options) {
  const {
    match,
    reverse = !1
  } = options, children = getChildren(snapshot, path), entries = reverse ? [...children].reverse() : children;
  for (const entry of entries)
    (!match || match(entry.node, entry.path)) && (yield entry), yield* getNodesSimple(snapshot, entry.path, options);
}
function comparePathsInTree(snapshot, pathA, pathB) {
  const minLength = Math.min(pathA.length, pathB.length), prefix = [];
  for (let i = 0; i < minLength; i++) {
    const segA = pathA[i], segB = pathB[i];
    if (typeof segA == "string" || typeof segB == "string" || typeof segA == "number" || typeof segB == "number" || Array.isArray(segA) || Array.isArray(segB)) {
      prefix.push(segA);
      continue;
    }
    if (segA._key === segB._key) {
      prefix.push(segA);
      continue;
    }
    const indexA = snapshot.blockIndexMap.get(serializePath([...prefix, segA])) ?? -1, indexB = snapshot.blockIndexMap.get(serializePath([...prefix, segB])) ?? -1;
    return indexA < indexB ? -1 : indexA > indexB ? 1 : 0;
  }
  return pathA.length < pathB.length ? -1 : pathA.length > pathB.length ? 1 : 0;
}
function* getNodesInRange(snapshot, path, options) {
  const {
    from,
    to,
    match,
    reverse = !1
  } = options, children = getChildren(snapshot, path), startIndex = from === void 0 ? 0 : boundaryChildIndex(snapshot, children, from) ?? 0, endIndex = to === void 0 ? children.length - 1 : boundaryChildIndex(snapshot, children, to) ?? children.length - 1, window = children.slice(startIndex, endIndex + 1), entries = reverse ? window.reverse() : window;
  for (const entry of entries) {
    if (canStopTraversal(snapshot, entry.path, from, to, reverse))
      return;
    couldContainInRangeNodes(snapshot, entry.path, from, to) && (isInRange(snapshot, entry.path, from, to) && (!match || match(entry.node, entry.path)) && (yield entry), yield* getNodesInRange(snapshot, entry.path, options));
  }
}
function boundaryChildIndex(snapshot, children, boundary) {
  const firstChild = children[0];
  if (!firstChild)
    return;
  const childPathLength = firstChild.path.length;
  if (boundary.length < childPathLength)
    return;
  const sharedPrefix = firstChild.path.slice(0, -1);
  if (sharedPrefix.length > 0 && !isAncestorPath(sharedPrefix, boundary))
    return;
  const resolvedIndex = resolveChildEntryIndex(snapshot.blockIndexMap, children, boundary.slice(0, childPathLength));
  return resolvedIndex === -1 ? void 0 : resolvedIndex;
}
function isInRange(snapshot, nodePath, from, to) {
  return !(from !== void 0 && comparePathsInTree(snapshot, nodePath, from) === -1 && !isAncestorPath(nodePath, from) || to !== void 0 && comparePathsInTree(snapshot, nodePath, to) === 1 && !isAncestorPath(nodePath, to));
}
function couldContainInRangeNodes(snapshot, nodePath, from, to) {
  return !!(isInRange(snapshot, nodePath, from, to) || from !== void 0 && isAncestorPath(nodePath, from) || to !== void 0 && isAncestorPath(nodePath, to));
}
function canStopTraversal(snapshot, nodePath, from, to, reverse) {
  return reverse ? from === void 0 ? !1 : comparePathsInTree(snapshot, nodePath, from) === -1 && !isAncestorPath(nodePath, from) : to === void 0 ? !1 : comparePathsInTree(snapshot, nodePath, to) === 1;
}
function getSibling(snapshot, path, options) {
  const {
    direction,
    match
  } = options;
  if (path.length === 0)
    return;
  const parent = parentPath(path), children = getChildren(snapshot, parent), currentIndex = resolveChildEntryIndex(snapshot.blockIndexMap, children, path);
  if (currentIndex !== -1) {
    if (!match) {
      const siblingIndex = direction === "next" ? currentIndex + 1 : currentIndex - 1;
      return siblingIndex < 0 || siblingIndex >= children.length ? void 0 : children[siblingIndex];
    }
    return (direction === "next" ? children.slice(currentIndex + 1) : children.slice(0, currentIndex).reverse()).find((child) => match(child.node, child.path));
  }
}
function isSpanNode(context, node) {
  return isTypedObject(node) && node._type === context.schema.span.name;
}
function isBlock(snapshot, path) {
  const parent = getParent(snapshot, path);
  return parent ? !isTextBlockNode({
    schema: snapshot.context.schema
  }, parent.node) : !0;
}
function getBlock(snapshot, path) {
  const entry = getNode(snapshot, path);
  if (entry && isBlock(snapshot, path) && !isSpanNode({
    schema: snapshot.context.schema
  }, entry.node))
    return {
      node: entry.node,
      path: entry.path
    };
}
function isInline(snapshot, path) {
  return !isBlock(snapshot, path);
}
function getEnclosingBlock(snapshot, path, options) {
  const match = options?.match;
  if ((options?.mode ?? "lowest") === "highest") {
    const ancestors = getAncestors(snapshot, path);
    for (const ancestor of [...ancestors].reverse())
      if (!match || match(ancestor.node, ancestor.path))
        return ancestor;
    const direct2 = getBlock(snapshot, path);
    return direct2 && (!match || match(direct2.node, direct2.path)) ? direct2 : void 0;
  }
  const direct = getBlock(snapshot, path);
  if (direct && (!match || match(direct.node, direct.path)))
    return direct;
  for (const ancestor of getAncestors(snapshot, path))
    if (!match || match(ancestor.node, ancestor.path))
      return ancestor;
}
function comparePoints(snapshot, pointA, pointB) {
  const pathComparison = comparePaths(pointA.path, pointB.path, {
    value: snapshot.context.value
  });
  return pathComparison !== 0 ? pathComparison : pointA.offset < pointB.offset ? -1 : pointA.offset > pointB.offset ? 1 : 0;
}
function descendToParent(snapshot, path) {
  const ancestors = getAncestors(snapshot, path);
  for (const ancestor of ancestors) {
    if (!isObject(snapshot, ancestor.node))
      continue;
    const resolved = resolveContainerAt(snapshot.context.containers, snapshot.context.value, ancestor.path);
    return !resolved || !("field" in resolved) ? void 0 : {
      parent: resolved,
      parentPath: ancestor.path
    };
  }
}
function getEnclosingContainer(snapshot, path) {
  const descent = descendToParent(snapshot, path);
  if (descent)
    return {
      of: descent.parent.field.of,
      path: descent.parentPath
    };
}
function getPathSubSchema(snapshot, path) {
  const enclosing = getEnclosingContainer(snapshot, path);
  return enclosing ? getSubSchema(snapshot.context.schema, enclosing.of) : snapshot.context.schema;
}
export {
  comparePaths,
  comparePoints,
  comparePoints$1,
  getAncestors,
  getBlock,
  getEnclosingBlock,
  getEnclosingContainer,
  getNodes,
  getPathSubSchema,
  getSibling,
  hasNode,
  isAfterPoint,
  isAncestorPath,
  isBackwardRange,
  isBlock,
  isEditableContainer,
  isInline,
  isObject,
  isSpanNode,
  isTextBlockNode,
  rangeEdges,
  resolveChildEntryIndex,
  resolveContainerAt
};
