// @sanity/workbench-cli/dist/actions/preview/startWorkbenchPreview.js
import { readFile } from 'node:fs/promises';
import path from 'node:path';
import { styleText } from 'node:util';
import { findProjectRoot } from '@sanity/cli-core';
import { buildAppId, SANITY_APP_ID_FILE } from '../../appId.js';
import { resolveWorkbenchApp } from '../../resolveWorkbenchApp.js';
import { createServerLifecycle, toDisplayHost } from '../../util/serverOrchestration.js';
import { deriveConfigs, deriveInterfaces } from '../dev/deriveInterfaces.js';
import { registerDevServer } from '../dev/registry.js';
import { startWorkbenchDevServer } from '../dev/startWorkbenchDevServer.js';
import { serveBuiltApplication } from './serveBuiltApplication.js';
/**
 * `sanity start` for a workbench app: serve a production build the way dev serves
 * a live one. The same singleton workbench shell renders it and the same registry
 * advertises it — only the remote differs, static files from the build output
 * instead of a live Vite dev server. There's no config watcher or rebuild: a
 * build is fixed, so nothing re-syncs.
 *
 * A running workbench claims the configured port, so the built remote binds the
 * next one. Without one the remote takes the configured port and announces its
 * own URL.
 */ export async function startWorkbenchPreview(options) {
    const { cacheDir, checkForDeprecatedAppId, cliConfig, extractManifest, httpHost, httpPort, isApp, outDir, output, reactStrictMode, workDir } = options;
    const { close, closers, installSignalHandlers } = createServerLifecycle();
    const workbench = await startWorkbenchDevServer({
        cacheDir,
        cliConfig,
        httpHost,
        httpPort,
        mode: 'preview',
        output,
        reactStrictMode,
        workDir
    });
    closers.push(workbench.close);
    const remotePort = workbench.workbenchAvailable ? workbench.workbenchPort + 1 : httpPort;
    const remote = await serveBuiltApplication({
        cacheDir,
        httpHost,
        httpPort: remotePort,
        outDir,
        workDir
    }).catch(async (err)=>{
        await close();
        throw err;
    });
    closers.push(remote.close);
    try {
        // Callers provide CLI-only validation and manifest extraction to keep them
        // out of workbench-cli.
        checkForDeprecatedAppId();
        const configPath = (await findProjectRoot(workDir)).path;
        const workbench = resolveWorkbenchApp(cliConfig);
        // Read the id the build inlined so start matches it even for a deploy build
        // (which carries the API id, not the shape hash); fall back for older builds.
        const inlinedId = await readInlinedAppId(outDir);
        const configs = await deriveConfigs(cliConfig.app);
        // `start` serves a build, so it advertises the build's inlined id (matching
        // the bundle's `__SANITY_APP_ID__`), not the dev host-port.
        const id = workbench ? inlinedId ?? await buildAppId(workbench) : `${remote.host}-${remote.port}`;
        const registration = registerDevServer({
            configs,
            host: remote.host,
            id,
            interfaces: deriveInterfaces(cliConfig.app, {
                isApp
            }),
            manifest: await extractManifest({
                configPath,
                workDir
            }),
            manifestUpdatedAt: new Date().toISOString(),
            port: remote.port,
            projectId: cliConfig?.api?.projectId,
            type: isApp ? 'coreApp' : 'studio',
            workDir
        });
        closers.push(async ()=>registration.release());
    } catch (err) {
        await close();
        throw err;
    }
    if (workbench.workbenchAvailable) {
        const workbenchUrl = `http://${toDisplayHost(workbench.httpHost)}:${workbench.workbenchPort}`;
        output.log(`Workbench preview server started at ${styleText([
            'blue',
            'underline'
        ], workbenchUrl)} (serving build on port ${remote.port})`);
    } else {
        const remoteUrl = `http://${toDisplayHost(remote.host)}:${remote.port}`;
        output.log(`Serving build at ${styleText([
            'blue',
            'underline'
        ], remoteUrl)}`);
    }
    installSignalHandlers();
    return {
        close
    };
}
/** The id the build inlined into its bundle, or undefined when absent. */ async function readInlinedAppId(outDir) {
    try {
        return (await readFile(path.join(outDir, SANITY_APP_ID_FILE), 'utf8')).trim() || undefined;
    } catch  {
        return undefined;
    }
}
