// @sanity/runtime-cli/dist/utils/functions/invoke/local.js
import { spawn } from 'node:child_process';
import { performance } from 'node:perf_hooks';
import { cwd } from 'node:process';
import { setTimeout } from 'node:timers';
import { fileURLToPath } from 'node:url';
import * as groq from 'groq-js';
import config from '../../../config.js';
import { groupBlueprintResources } from '../../blueprints/group-resources.js';
import { isGroqContextOptions, } from '../../types.js';
import { acquireFunctionBuild } from './prepare.js';
function getChildProcessWrapperPath() {
    return fileURLToPath(new URL('../child-process-wrapper.js', import.meta.url));
}
export function sanitizeLogs(logs) {
    return logs.replace(/([a-zA-Z0-9]{10})[a-zA-Z0-9]{65,}/g, '$1**********');
}
export async function applyGroqRule(resource, payload, projectId, dataset) {
    const { before, after, payload: data = null } = payload;
    // If there is no rule set return everything
    if (!resource.event)
        return data;
    // default groq rule is: gimme full doc content. otherwise, parse + eval custom rule
    // applying the GROQ filter may result in a slimmer set of documents
    if (resource.event.filter || resource.event.projection) {
        const query = `*[${resource.event.filter || ''}]${resource.event.projection || ''}`;
        try {
            const rule = groq.parse(query, { mode: 'delta' });
            const sanity = projectId && dataset ? { projectId, dataset } : undefined;
            const queryResults = await groq.evaluate(rule, { dataset: [data], before, after, sanity });
            const currentFunctionDocumentSet = await queryResults.get();
            return currentFunctionDocumentSet[0];
        }
        catch (groqErr) {
            const detail = groqErr instanceof Error && `; ${groqErr.message}`;
            throw new Error(`Invalid rule configuration: ${query}${detail}`, { cause: groqErr });
        }
    }
    // default groq rule so just return the data
    return data;
}
export default async function invoke(resource, payload, context, options) {
    if (!resource.src) {
        throw new Error(`Function resource "${resource.name}" is missing the 'src' property.`);
    }
    const { forceColor = true, timeout = 10, onNestedInvoke, depth = 0, callStack = [] } = options;
    // Apply the GROQ rule before preparing a build so a filtered-out event skips
    // the (potentially expensive) transpile entirely.
    let filteredData = payload?.payload?.data || {};
    if (resource.type === 'sanity.function.document' ||
        resource.type === 'sanity.function.media-library.asset') {
        if (!isGroqContextOptions(context)) {
            throw new Error('GROQ-based functions require a context with clientOptions');
        }
        const { projectId, dataset } = context.clientOptions;
        if ('before' in payload) {
            filteredData = await applyGroqRule(resource, payload, projectId, dataset);
            if (typeof filteredData === 'undefined') {
                return {
                    logs: `⚠️ Filter "${resource.event?.filter}" returned an empty result. Skipping invoke.`,
                    error: undefined,
                    json: undefined,
                };
            }
        }
    }
    // Acquire a shared, reference-counted build. Concurrent invokes of the same
    // function reuse one transpile + temp package.json instead of racing on the
    // per-function build paths; the build is torn down once every invoke releases.
    const build = acquireFunctionBuild(resource, context.blueprintFileDir);
    let functionPath;
    let bundleTimings;
    try {
        const prepared = await build.ready;
        functionPath = prepared.functionPath;
        bundleTimings = prepared.bundleTimings;
    }
    catch (err) {
        await build.release();
        throw err;
    }
    return new Promise((resolve, reject) => {
        let child;
        let timer;
        let executionStart;
        // The protocol is no longer one-shot: the child stays alive across nested
        // `context.invoke` calls, so several handlers (result/exit/error/timeout)
        // can race to settle this Promise. `settled` ensures exactly one wins and
        // that shutdown/cleanup only runs once.
        let settled = false;
        // Timeout budget is paused while nested invokes run so a function's own
        // time limit isn't consumed by the functions it calls.
        let pendingNested = 0;
        let timerStart = 0;
        let remaining = timeout * 1000;
        const nestedTimings = {};
        // Output the child writes directly to its stdio pipes (anything that escapes
        // the in-process console capture, e.g. a native crash). Hoisted to the
        // Promise scope so both the timeout and exit handlers can surface it.
        let stdout = '';
        let stderr = '';
        function startTimer() {
            timerStart = performance.now();
            timer = setTimeout(() => {
                if (settled)
                    return;
                settled = true;
                const executionTimeMs = performance.now() - executionStart;
                shutdown();
                // Resolve with error to retain all logs
                resolve({
                    json: undefined,
                    logs: sanitizeLogs(stdout),
                    error: {
                        name: 'Error',
                        message: `Timeout: Function exceeded the ${timeout}-second time limit.`,
                        stack: '',
                    },
                    timings: {
                        ...bundleTimings,
                        ...nestedTimings,
                        execute: executionTimeMs,
                    },
                });
            }, Math.max(0, remaining));
        }
        function pauseTimer() {
            if (pendingNested === 0) {
                clearTimeout(timer);
                remaining -= performance.now() - timerStart;
            }
            pendingNested += 1;
        }
        function resumeTimer() {
            pendingNested -= 1;
            if (pendingNested === 0 && !settled) {
                startTimer();
            }
        }
        // Resolve a nested `context.invoke(name, payload)` by re-entering the invoke
        // pipeline via the supplied callback, then hand the result back to the child.
        async function handleNestedInvoke(msg) {
            const { id, name, payload: nestedPayload } = msg;
            pauseTimer();
            const nestedStart = performance.now();
            try {
                if (!onNestedInvoke) {
                    throw new Error('Nested context.invoke is not supported in this context');
                }
                const nested = await onNestedInvoke(name, nestedPayload ?? {}, context, depth + 1, callStack);
                nestedTimings[`invoke:${name}`] =
                    (nestedTimings[`invoke:${name}`] ?? 0) + (performance.now() - nestedStart);
                // Fire-and-forget contract: in production `context.invoke` returns once the
                // message is enqueued (Lambda InvocationType:'Event', SNS/SQS), so a
                // downstream handler error never reaches the caller. Mirror that here —
                // the target ran (it was "dispatched"), so resolve the caller regardless
                // of `nested.error`. The error isn't lost: it travels back in the returned
                // result and is logged/streamed to the UI by the caller (app/test core).
                if (child && !child.killed) {
                    const response = {
                        type: 'invoke-response',
                        id,
                        ok: true,
                        result: { json: nested.json },
                    };
                    child.send(JSON.stringify(response));
                }
            }
            catch (error) {
                // The invoke could not be dispatched at all (unknown function, recursion
                // guard, unsupported context). That is an enqueue-time failure, which in
                // production rejects synchronously, so propagate it to the caller.
                if (child && !child.killed) {
                    const response = {
                        type: 'invoke-response',
                        id,
                        ok: false,
                        error: {
                            message: error instanceof Error ? error.message : String(error),
                            stack: error instanceof Error ? (error.stack ?? '') : '',
                            name: error instanceof Error ? error.name : 'Error',
                        },
                    };
                    child.send(JSON.stringify(response));
                }
            }
            finally {
                resumeTimer();
            }
        }
        function start() {
            executionStart = performance.now();
            child = spawn('node', ['--enable-source-maps', getChildProcessWrapperPath()], {
                cwd: cwd(),
                stdio: ['pipe', 'pipe', 'pipe', 'ipc'], // changed: use 'pipe' so we can capture output
                env: { ...process.env, ...resource.env, FORCE_COLOR: forceColor ? '1' : '0' },
            });
            child.stdout?.on('data', (data) => {
                stdout += data.toString();
            });
            child.stderr?.on('data', (data) => {
                stderr += data.toString();
            });
            child.on('message', (data) => {
                let msg;
                try {
                    msg = JSON.parse(data.toString());
                }
                catch {
                    return;
                }
                // Nested invoke request — handle without tearing down the child.
                if (msg?.type === 'invoke-request') {
                    handleNestedInvoke({ id: msg.id, name: msg.name, payload: msg.payload });
                    return;
                }
                // Terminal result message.
                if (settled)
                    return;
                settled = true;
                const executionTimeMs = performance.now() - executionStart;
                shutdown();
                resolve({
                    json: msg.json,
                    logs: sanitizeLogs(msg.logs ?? ''),
                    error: msg.error,
                    timings: {
                        ...bundleTimings,
                        ...nestedTimings,
                        execute: executionTimeMs,
                    },
                });
            });
            child.on('error', (error) => {
                if (settled)
                    return;
                settled = true;
                shutdown();
                reject(new Error(`encountered error ${error.message}`));
            });
            // Resolve on 'close' to drain the pipes
            child.on('close', (code) => {
                if (settled)
                    return;
                settled = true;
                const executionTimeMs = performance.now() - executionStart;
                shutdown();
                if (code !== 0) {
                    // use captured stderr if available
                    const message = stderr.trim().length > 0
                        ? `Process failed with code ${code}\n${stderr}`
                        : `exited with code ${code}`;
                    // Resolve with error to retain all logs
                    resolve({
                        json: undefined,
                        logs: sanitizeLogs(stdout),
                        error: { name: 'Error', message, stack: '' },
                        timings: {
                            ...bundleTimings,
                            ...nestedTimings,
                            execute: executionTimeMs,
                        },
                    });
                }
                else {
                    resolve({
                        json: undefined,
                        logs: sanitizeLogs(stdout),
                        error: undefined,
                        timings: {
                            ...bundleTimings,
                            ...nestedTimings,
                            execute: executionTimeMs,
                        },
                    });
                }
            });
            startTimer();
            const payload = {
                data: { ...filteredData },
                context: {
                    ...context,
                    local: true,
                    resources: groupBlueprintResources(context.resources ?? []),
                    ...(isGroqContextOptions(context) && {
                        clientOptions: {
                            ...context.clientOptions,
                            apiHost: config.apiUrl,
                        },
                    }),
                },
            };
            const startMessage = { type: 'start', srcPath: functionPath, payload };
            child.send(JSON.stringify(startMessage));
        }
        function shutdown() {
            clearTimeout(timer);
            if (child && !child.killed) {
                child.kill();
            }
            // Release our reference to the shared build; its temp package.json and
            // bundle are cleaned up once the last concurrent invoke releases.
            build.release().catch((err) => console.warn('Build cleanup failed:', err));
        }
        start();
    });
}
