// @sanity/runtime-cli/dist/utils/functions/invoke/prepare.js
import path from 'node:path';
import { findFunctionEntryPoint } from '../find-entry-point.js';
import { cleanupTempPackageJson, createTempPackageJson, doesPackageJsonExists, } from '../package-json.js';
import { resolveResourceDependencies } from '../resolve-dependencies.js';
import { shouldAutoResolveDependencies } from '../should-auto-resolve-deps.js';
import { shouldTranspileFunction } from '../should-transpile.js';
import { transpileFunction } from '../transpile/function.js';
/**
 * In-progress builds keyed by function identity. Concurrent invokes of the same
 * function (e.g. `Promise.all([context.invoke('f'), context.invoke('f')])`) share
 * a single build and a single temp package.json, so they don't race on the
 * deterministic per-function paths (`.build/function-<name>`, the temp
 * package.json). The entry is removed once the last concurrent invoke releases
 * it, so a later, non-overlapping invoke rebuilds and picks up source changes.
 */
const builds = new Map();
function buildKey(resource) {
    return `${resource.name}:${resource.src}`;
}
/**
 * Acquire a shared, reference-counted build for `resource`. The first caller runs
 * the transpile + temp package.json setup; concurrent callers reuse it. Every
 * caller MUST call `release()` exactly once when its child has exited, so the
 * shared artifacts are torn down only after the last in-flight invoke finishes.
 */
export function acquireFunctionBuild(resource, blueprintFileDir) {
    const key = buildKey(resource);
    let entry = builds.get(key);
    if (!entry) {
        let cleanupBundle = async () => { };
        const resolvedFunctionPath = path.resolve(blueprintFileDir, resource.src);
        let functionPath = resolvedFunctionPath;
        let createdTempPackageJson = false;
        // If no package.json exists we create a temporary one to avoid the
        // "module type" warning from node.
        const existingPackageJson = doesPackageJsonExists(functionPath);
        const ready = (async () => {
            if (!resource.src) {
                throw new Error(`Function resource "${resource.name}" is missing the 'src' property.`);
            }
            let bundleTimings;
            const shouldTranspile = await shouldTranspileFunction(resource, functionPath);
            if (shouldTranspile.transpile) {
                const transpileResult = await transpileFunction({
                    resource,
                    functionPath,
                    entryPoint: shouldTranspile.entryPoint,
                });
                functionPath = await findFunctionEntryPoint(transpileResult.outputDir);
                bundleTimings = transpileResult.timings;
                cleanupBundle = transpileResult.cleanup;
            }
            else {
                functionPath = await findFunctionEntryPoint(functionPath, resource.displayName ?? resource.name);
            }
            const shouldResolveDependencies = await shouldAutoResolveDependencies(resource);
            if (shouldResolveDependencies && existingPackageJson) {
                await resolveResourceDependencies({
                    resource,
                    functionPath: resolvedFunctionPath,
                    blueprintFileDir,
                    transpiled: shouldTranspile.transpile,
                });
            }
            if (!existingPackageJson) {
                createTempPackageJson(functionPath);
                createdTempPackageJson = true;
            }
            return { functionPath, bundleTimings };
        })();
        entry = {
            refs: 0,
            ready,
            teardown: async () => {
                // rmSync is synchronous, so the temp package.json is gone before this
                // function yields — a later invoke of the same function can't observe a
                // half-removed state.
                if (createdTempPackageJson) {
                    cleanupTempPackageJson(functionPath);
                }
                await cleanupBundle().catch((err) => console.warn('Bundle cleanup failed:', err));
            },
        };
        builds.set(key, entry);
    }
    const activeEntry = entry;
    activeEntry.refs += 1;
    let released = false;
    const release = async () => {
        if (released)
            return;
        released = true;
        activeEntry.refs -= 1;
        if (activeEntry.refs === 0) {
            if (builds.get(key) === activeEntry)
                builds.delete(key);
            await activeEntry.teardown();
        }
    };
    return { ready: activeEntry.ready, release };
}
