// @portabletext/plugin-character-pair-decorator/dist/index.js
import { jsx } from "react/jsx-runtime";
import { c } from "react/compiler-runtime";
import { defineInputRule, InputRulePlugin } from "@portabletext/plugin-input-rule";
import "react";
import { raise } from "@portabletext/editor/behaviors";
import { getPathSubSchema } from "@portabletext/editor/traversal";
function createCharacterPairRegex(char, amount) {
  const prePrefix = `(?<!\\${char})`, prefix = `\\${char}`.repeat(Math.max(amount, 1)), postPrefix = "(?!\\s)", content = `(?<content>[^${char}\\n]+?)`, preSuffix = "(?<!\\s)", suffix = `\\${char}`.repeat(Math.max(amount, 1)), postSuffix = `(?!\\${char})`;
  return `${prePrefix}${prefix}${postPrefix}${content}${preSuffix}${suffix}${postSuffix}`;
}
function createCharacterPairRule(config) {
  config.pair.amount < 1 && console.warn("The amount of characters in the pair should be greater than 0");
  const pairRegex = createCharacterPairRegex(config.pair.char, config.pair.amount);
  return defineInputRule({
    // `$`-anchored: the pair must close at the insertion point, so at most
    // one match fires per insert.
    on: new RegExp(`${pairRegex}$`),
    // A pair only deletes its markers, so a match whose CONTENT spans an
    // inline object is harmless and must decorate (`**bo`, insert an
    // object, `ld**`). An inline object anywhere else in the match, between
    // the marker characters, means this is not a pair, and the machinery
    // drops the match.
    inlineObjects: {
      allow: ["content"]
    },
    guard: ({
      snapshot,
      event
    }) => {
      if (config.pair.amount < 1)
        return !1;
      const subSchema = getPathSubSchema(snapshot, event.focusBlock.path), decorator = config.decorator({
        context: {
          schema: subSchema
        },
        schema: subSchema
      });
      if (decorator === void 0)
        return !1;
      const match = event.matches.at(0);
      if (!match)
        return !1;
      const content = match.groups.content;
      if (!content)
        return !1;
      const prefixOffsets = {
        anchor: match.targetOffsets.anchor,
        focus: content.targetOffsets.anchor
      }, suffixOffsets = {
        anchor: content.targetOffsets.focus,
        focus: match.targetOffsets.focus
      };
      return {
        decorator,
        contentOffsets: {
          anchor: content.targetOffsets.anchor,
          focus: content.targetOffsets.focus
        },
        prefixOffsets,
        suffixOffsets
      };
    },
    actions: [(_, {
      decorator,
      contentOffsets,
      prefixOffsets,
      suffixOffsets
    }) => [
      // Decorate the text between the prefix and suffix
      raise({
        type: "decorator.add",
        decorator,
        at: contentOffsets
      }),
      // Delete the suffix, then the prefix (suffix first so the prefix
      // offsets stay valid)
      raise({
        type: "delete.text",
        at: suffixOffsets
      }),
      raise({
        type: "delete.text",
        at: prefixOffsets
      }),
      // Toggle the decorator off so the next inserted text isn't decorated
      raise({
        type: "decorator.remove",
        decorator
      })
    ]]
  });
}
function CharacterPairDecoratorPlugin(props) {
  const $ = c(7);
  let t0;
  $[0] !== props.decorator || $[1] !== props.pair ? (t0 = createCharacterPairRule({
    decorator: props.decorator,
    pair: props.pair
  }), $[0] = props.decorator, $[1] = props.pair, $[2] = t0) : t0 = $[2];
  let t1;
  $[3] !== t0 ? (t1 = [t0], $[3] = t0, $[4] = t1) : t1 = $[4];
  const rules = t1;
  let t2;
  return $[5] !== rules ? (t2 = /* @__PURE__ */ jsx(InputRulePlugin, { rules }), $[5] = rules, $[6] = t2) : t2 = $[6], t2;
}
export {
  CharacterPairDecoratorPlugin
};
