// @sanity/cli/dist/actions/deploy/createUserApplication.js
import { CLIError } from '@oclif/core/errors';
import { input, spinner } from '@sanity/cli-core/ux';
import { customAlphabet } from 'nanoid';
import { createUserApplication as createUserApplicationRequest } from '../../services/userApplications.js';
import { NO_ORGANIZATION_ID } from '../../util/errorMessages.js';
import { deployDebug } from './deployDebug.js';
import { normalizeUrl, validateUrl } from './urlUtils.js';
const LETTERS = 'abcdefghijklmnopqrstuvwxyz';
// appHosts have some restrictions (no uppercase, must start with a letter)
export function generateAppSlug() {
    const firstChar = customAlphabet(LETTERS, 1)();
    const rest = customAlphabet(`${LETTERS}0123456789`, 11)();
    return `${firstChar}${rest}`;
}
// TODO: replace with `Promise.withResolvers()` once it lands in node 22
function promiseWithResolvers() {
    let resolve;
    let reject;
    const promise = new Promise((res, rej)=>{
        resolve = res;
        reject = rej;
    });
    return {
        promise,
        reject,
        resolve
    };
}
export async function createStudioUserApplication(options) {
    const { projectId, title, urlType = 'internal' } = options;
    const { promise, resolve } = promiseWithResolvers();
    const isExternal = urlType === 'external';
    await input({
        message: isExternal ? 'Studio URL (https://...):' : 'Studio hostname (<value>.sanity.studio):',
        // if a string is returned here, it is relayed to the user and prompt allows
        // the user to try again until this function returns true
        validate: async (inp)=>{
            let appHost;
            if (isExternal) {
                const normalized = normalizeUrl(inp);
                const validation = validateUrl(normalized);
                if (validation !== true) {
                    return validation;
                }
                appHost = normalized;
            } else {
                appHost = inp.replace(/\.sanity\.studio$/i, '');
            }
            try {
                const response = await createUserApplicationRequest({
                    appType: 'studio',
                    body: {
                        appHost,
                        title,
                        type: 'studio',
                        urlType
                    },
                    projectId
                });
                resolve(response);
                return true;
            } catch (e) {
                // if the name is taken, it should return a 409 so we relay to the user
                if ([
                    402,
                    409
                ].includes(e?.statusCode)) {
                    return e?.response?.body?.message || 'Bad request' // just in case
                    ;
                }
                deployDebug('Error creating user application', e);
                // otherwise, it's a fatal error
                throw new CLIError('Error creating user application', {
                    exit: 1
                });
            }
        }
    });
    return await promise;
}
export async function createUserApplication(organizationId, title, visibility) {
    if (!organizationId) {
        throw new Error(NO_ORGANIZATION_ID);
    }
    const resolvedTitle = title ?? await input({
        message: 'Enter a title for your application:',
        validate: (value)=>value.length > 0 || 'Title is required'
    });
    return tryCreateApp(resolvedTitle, organizationId, visibility);
}
const tryCreateApp = async (title, organizationId, visibility)=>{
    // we will likely prepend this with an org ID or other parameter in the future
    const appHost = generateAppSlug();
    const spin = spinner('Creating application').start();
    try {
        const response = await createUserApplicationRequest({
            appType: 'coreApp',
            body: {
                appHost,
                title,
                type: 'coreApp',
                urlType: 'internal',
                ...visibility ? {
                    dashboardStatus: visibility
                } : {}
            },
            organizationId
        });
        spin.succeed();
        return response;
    } catch (e) {
        // if the name is taken, generate a new one and try again
        if ([
            402,
            409
        ].includes(e?.statusCode)) {
            deployDebug('App host taken, retrying with new host');
            return tryCreateApp(title, organizationId, visibility);
        }
        spin.fail();
        deployDebug('Error creating core application', e);
        throw e;
    }
};
