// @sanity/cli/dist/actions/deploy/deployRunner.js
import { CLIError } from '@oclif/core/errors';
import { getErrorMessage } from '@sanity/cli-core/errors';
import { createCollectingReporter, createFailFastReporter } from '../../util/checks.js';
import { toStderrOutput } from '../../util/toStderrOutput.js';
import { deployDebug } from './deployDebug.js';
import { deploymentPlanToJson, isDeployable, renderDeploymentPlan } from './deploymentPlan.js';
/**
 * Runs a deploy in the mode the flags select: a real deploy fails fast and
 * mutates, `--dry-run` drives the same `run` sequence read-only and renders a
 * plan, and `--json` emits the same information as machine-readable JSON.
 */ export async function runDeploy(options, spec) {
    const { output } = options;
    const json = !!options.flags.json;
    const emitJson = (payload)=>output.log(JSON.stringify(payload, null, 2));
    // The JSON payload owns stdout, so the run's progress logs go to stderr; only
    // the final JSON.stringify writes to stdout.
    const runOptions = json ? {
        ...options,
        output: toStderrOutput(output)
    } : options;
    try {
        if (options.flags['dry-run']) {
            const plan = await collectPlan(runOptions, spec);
            if (json) emitJson(deploymentPlanToJson(plan));
            else renderDeploymentPlan(plan, output);
            exitIfBlocked(plan, output);
            return;
        }
        const result = await spec.run(runOptions, createFailFastReporter(runOptions.output));
        if (json && result) emitJson({
            deployed: true,
            ...result
        });
    } catch (error) {
        const failure = normalizeFailure(error, spec.type);
        // A blocked dry run reaches this catch too (its exit throws) and already
        // printed its plan, so only a real deploy adds the {deployed: false} envelope.
        if (json && !options.flags['dry-run']) {
            emitJson({
                deployed: false,
                error: {
                    message: failure.message
                }
            });
        }
        output.error(failure.message, {
            exit: failure.exit
        });
    }
}
/** Runs the step sequence read-only and gathers the plan a dry run reports. */ async function collectPlan(options, spec) {
    const reporter = createCollectingReporter();
    await spec.run(options, reporter);
    const plan = {
        checks: reporter.results,
        config: reporter.results.find((check)=>check.config)?.config ?? null,
        exposes: reporter.results.find((check)=>check.exposes)?.exposes ?? [],
        files: [],
        isSingleton: reporter.results.find((check)=>check.isSingleton !== undefined)?.isSingleton,
        target: reporter.results.find((check)=>check.target)?.target ?? null,
        type: spec.type,
        version: reporter.results.find((check)=>check.version)?.version ?? null
    };
    // A blocked deploy uploads nothing, so only enumerate files for a deployable plan.
    if (isDeployable(plan)) plan.files = await spec.listFiles(options);
    return plan;
}
/** Exits like a real (fail-fast) deploy would, on the first failing check's exit code. */ function exitIfBlocked(plan, output) {
    if (isDeployable(plan)) return;
    const failed = plan.checks.find((check)=>check.status === 'fail');
    output.error('Deploy blocked by failing checks.', {
        exit: failed?.exitCode ?? 1
    });
}
/** The one failure diagnosis both the stderr message and the `--json` envelope read. */ function normalizeFailure(error, type) {
    // Ctrl+C on an interactive prompt isn't a real failure
    if (error instanceof Error && error.name === 'ExitPromptError') {
        return {
            exit: 1,
            message: 'Deployment cancelled by user'
        };
    }
    // A failed check already carries its own message and exit code
    if (error instanceof CLIError) {
        return {
            exit: error.oclif?.exit ?? 1,
            message: error.message
        };
    }
    deployDebug(`Error deploying ${type === 'coreApp' ? 'application' : 'studio'}`, error);
    return {
        exit: 1,
        message: `Error deploying ${type === 'coreApp' ? 'application' : 'studio'}: ${getErrorMessage(error)}`
    };
}
