// @sanity/cli/dist/actions/deploy/resolveDeployTarget.js
import { getApplication, getApplicationUrl } from '@sanity/workbench-cli/deploy';
import { getUserApplication, getUserApplications } from '../../services/userApplications.js';
import { APP_ID_NOT_FOUND_IN_ORGANIZATION } from '../../util/errorMessages.js';
import { normalizeUrl, validateUrl } from './urlUtils.js';
/**
 * Owns the studio deploy-target rules: the --url flag over studioHost config,
 * appId over appHost precedence, external URL normalization and validation.
 * Both the real deploy and the dry run consume these verdicts.
 *
 * @internal
 */ export async function resolveStudioDeployTarget(options) {
    const { appId, isExternal, projectId, studioHost, urlFlag } = options;
    // appId wins over host config (and undeploy resolves it the same way): a
    // configured appId deploys even when studioHost is stale or invalid, so it's
    // resolved before any host validation.
    if (appId) {
        if (!projectId) {
            return {
                message: 'api.projectId is missing',
                type: 'blocked'
            };
        }
        const application = await getUserApplication({
            appId,
            isSdkApp: false,
            projectId
        });
        if (application) {
            return {
                application,
                type: 'found'
            };
        }
        return {
            message: `Cannot find app with app ID ${appId}`,
            reason: 'app-not-found',
            type: 'invalid'
        };
    }
    const { error: hostError, host: resolvedHost } = resolveAppHost({
        isExternal,
        studioHost,
        url: urlFlag
    });
    if (hostError) {
        return {
            message: hostError,
            reason: 'invalid-host',
            type: 'invalid'
        };
    }
    // A host from config hasn't passed through the --url validation yet
    let appHost = resolvedHost;
    if (appHost && isExternal) {
        appHost = normalizeUrl(appHost);
        const validation = validateUrl(appHost);
        if (validation !== true) {
            return {
                message: validation,
                reason: 'invalid-host',
                type: 'invalid'
            };
        }
    }
    if (appHost) {
        if (!projectId) {
            return {
                message: 'api.projectId is missing',
                type: 'blocked'
            };
        }
        const application = await getUserApplication({
            appHost,
            isSdkApp: false,
            projectId
        });
        if (application) {
            return {
                application,
                type: 'found'
            };
        }
        return {
            appHost,
            type: 'would-create'
        };
    }
    // Neither appId nor host configured — a deploy would prompt.
    // Without a project there is nothing to list; a deploy would still prompt.
    const existing = projectId ? await listStudioApplications(projectId, isExternal) : [];
    return {
        existing,
        type: 'needs-input'
    };
}
/**
 * Owns the app deploy-target rules: appId lookup, falling back to listing the
 * organization's applications. Both the real deploy and the dry run consume
 * these verdicts.
 *
 * @internal
 */ export async function resolveAppDeployTarget(options) {
    const { appId, organizationId } = options;
    if (appId) {
        const application = await getUserApplication({
            appId,
            isSdkApp: true
        });
        if (application) {
            return {
                application,
                type: 'found'
            };
        }
        return {
            message: `Cannot find app with app ID ${appId}`,
            reason: 'app-not-found',
            type: 'invalid'
        };
    }
    if (!organizationId) {
        return {
            message: 'app.organizationId is missing',
            type: 'blocked'
        };
    }
    const existing = await getUserApplications({
        appType: 'coreApp',
        organizationId
    });
    if (existing?.length) {
        return {
            existing,
            type: 'needs-input'
        };
    }
    return {
        type: 'would-create'
    };
}
/**
 * The dry-run counterpart to a workbench app's create-on-deploy: a configured
 * `appId` is looked up read-only, otherwise a coreApp would be created.
 * @internal
 */ export async function resolveWorkbenchApp({ appId }) {
    return appId ? resolveAppById(appId) : {
        type: 'would-create'
    };
}
/**
 * The studio counterpart to {@link resolveWorkbenchApp}: a redeploy targets
 * `deployment.appId`, a first deploy creates the studio at its `slug`.
 *
 * @internal
 */ export async function resolveWorkbenchStudio({ appId, slug }) {
    if (appId) return resolveAppById(appId);
    return {
        appHost: slug,
        type: 'would-create'
    };
}
async function resolveAppById(appId) {
    const application = await getApplication(appId);
    return application ? {
        application: {
            appHost: application.slug ?? '',
            id: application.id,
            organizationId: application.organizationId,
            title: application.title,
            url: getApplicationUrl(application)
        },
        type: 'found'
    } : {
        message: APP_ID_NOT_FOUND_IN_ORGANIZATION,
        reason: 'app-not-found',
        type: 'invalid'
    };
}
async function listStudioApplications(projectId, isExternal) {
    const urlType = isExternal ? 'external' : 'internal';
    const applications = await getUserApplications({
        appType: 'studio',
        projectId
    });
    // External deploys should only see external studios and vice versa
    return applications?.filter((application)=>application.urlType === urlType) ?? [];
}
function resolveAppHost({ isExternal, studioHost, url }) {
    if (!url) {
        return {
            host: studioHost
        };
    }
    if (isExternal) {
        const normalized = normalizeUrl(url);
        const validation = validateUrl(normalized);
        if (validation !== true) {
            return {
                error: validation
            };
        }
        return {
            host: normalized
        };
    }
    // For internal deploys, strip protocol prefix and .sanity.studio suffix if present
    const hostname = url.replace(/^https?:\/\//i, '').replace(/\.sanity\.studio\/?$/i, '');
    // If the result still looks like a URL (contains dots), the user likely meant --external
    if (hostname.includes('.')) {
        return {
            error: `"${hostname}" does not look like a sanity.studio hostname. Did you mean to use --external?`
        };
    }
    // Validate hostname characters (alphanumeric and hyphens only)
    if (!/^[a-z0-9]([a-z0-9-]*[a-z0-9])?$/i.test(hostname)) {
        return {
            error: `Invalid studio hostname "${hostname}". Hostnames can only contain letters, numbers, and hyphens.`
        };
    }
    return {
        host: hostname
    };
}
