// sanity/lib/index.js
import "sanity/bundle.css";
import { ALL_FIELDS_GROUP, ActiveWorkspaceMatcher, AddonDatasetProvider, ArrayOfObjectOptionsInput, ArrayOfObjectsFunctions, ArrayOfObjectsInput, ArrayOfObjectsInputMember, ArrayOfObjectsInputMembers, ArrayOfObjectsItem, ArrayOfOptionsInput, ArrayOfPrimitiveOptionsInput, ArrayOfPrimitivesFunctions, ArrayOfPrimitivesInput, ArrayOfPrimitivesItem, AutoCollapseMenu, AvatarSkeleton, BetaBadge, PortableTextInput, BlockImagePreview, BlockPreview, BooleanInput, COMMENTS_INSPECTOR_NAME, CONNECTING, CapabilityGate, ChangeBreadcrumb, ChangeConnectorRoot, ChangeFieldWrapper, ChangeIndicator, ChangeIndicatorsTracker, ChangeList, ChangeResolver, ChangeTitleSegment, ChangesError, Chip, CircularProgress, CollapseMenu, CollapseMenuButton, ColorSchemeCustomProvider, ColorSchemeLocalStorageProvider, ColorSchemeProvider, CommandList, CommentDeleteDialog, CommentDisabledIcon, CommentInlineHighlightSpan, CommentInput, CommentsAuthoringPathProvider, CommentsEnabledProvider, CommentsIntentProvider, CommentsList, CommentsProvider, CommentsSelectedPathProvider, CompactPreview, Rule, ConfigErrorGate, ConfigPropertyError, ConfigResolutionError, ContextMenuButton, CopyPasteProvider, CorsOriginErrorScreen, CreateButton, StudioCrossDatasetReferenceInput, CrossDatasetReferencePreview, DEFAULT_MAX_RECURSION_DEPTH, DEFAULT_STUDIO_CLIENT_OPTIONS, DateInput, DateTimeInput$1, DefaultPreview, DetailPreview, DiffCard, DiffErrorBoundary, DiffFromTo, DiffInspectWrapper, DiffString, DiffStringSegment, DiffTooltip, DivergencesProvider, DocumentGroupInventory, DocumentGroupInventoryAction, DocumentPreviewPresence, DocumentStatus, DocumentStatusIndicator, EMPTY_ARRAY, EMPTY_OBJECT, EditPortal, EditScheduleForm, EmailInput, EnhancedObjectDialog, ErrorActions, ErrorMessage, Event, EventsProvider, FallbackDiff, FeedbackDialog, FieldActionMenu, FieldActionsProvider, FieldActionsResolver, FieldChange, FieldPresence, FieldPresenceInner, FieldPresenceWithOverlay, StudioFileInput, Filters, FormBuilder, FormCallbacksProvider, FormCell, FormContainer, FormField, FormFieldHeaderText, FormFieldSet, FormFieldStatus, FormFieldValidationStatus, FormInput, FormProvider, FormRow, FormValueProvider, FromTo, FromToArrow, GetFormValueProvider, GetHookCollectionState, GroupChange, HoveredFieldProvider, StudioImageInput, ImperativeToast, InlinePreview, InsufficientPermissionsMessage, IntentButton, IsLastPaneProvider, LATEST, LegacyLayerProvider, LinearProgress, LiveManifestRegisterProvider, LoadingBlock, LocaleProvider, LocaleProviderBase, MediaPreview, MemberField, MemberFieldError, MemberFieldSet, MemberItemError, MetaInfo, NoChanges, NumberInput, ObjectInput, ObjectInputMember, ObjectInputMembers, ObjectMembers, PUBLISHED, ParseErrorsProvider, PatchEvent, PerspectiveProvider, PopoverDialog, PortableTextInput as PortableTextInput2, PresenceOverlay, PresenceScope, Preview, PreviewCard, PreviewLoader, RELEASES_INTENT, RELEASES_STUDIO_CLIENT_OPTIONS, RESOLVE_INITIAL_VALUE_TIMEOUT_MS, ReferenceAutocomplete, StudioReferenceInput, ReferenceInputOptionsProvider, ReferenceInputPreviewCard, RelativeTime, ReleaseAvatar, ReleaseAvatarIcon, ReleaseTitle, ReleasesNav, RequestErrorDialog, Resizable, ResourceCacheProvider, RevertChangesButton, RevertChangesConfirmDialog, SANITY_PATCH_TYPE, SESSION_ID, SanityDefaultPreview, useScheduleAction, ScheduledBadge, SchemaError, ScrollContainer, SearchButton, SearchDialog, SearchHeader, SearchPopover, SearchProvider, SearchResultItemPreview, SelectInput, SingleDocReleaseProvider, SlugInput, SourceProvider, StatusButton, StringInput, Studio, StudioAnnouncementsCard, StudioAnnouncementsDialog, StudioFeedbackDialog, StudioFeedbackProvider, StudioLayout, StudioLayoutComponent, StudioLogo, StudioNavbar, StudioProvider, StudioToolMenu, TIMELINE_ITEM_I18N_KEY_MAPPING, TagsArrayInput, TelephoneInput, TemplatePreview, TextInput, TextWithTone, Timeline, TimelineController, ToolLink, TooltipOfDisabled, TransformPatches, Translate, UniversalArrayInput, UpdateReadOnlyPlugin, UpsellDescriptionSerializer, UpsellDialogDismissed, UpsellDialogLearnMoreCtaClicked, UpsellDialogUpgradeCtaClicked, UpsellDialogViewed, UrlInput, UserAvatar, UserColorManagerProvider, VARIANTS_STUDIO_CLIENT_OPTIONS, ValueError, VersionChip, VersionInlineBadge, VirtualizerScrollInstanceProvider, VisibleWorkspacesProvider, WithReferringDocuments, WorkspaceLoaderBoundary, WorkspaceProvider, WorkspacesProvider, ZIndexProvider, _createAuthStore, _isCustomDocumentTypeDefinition, _isSanityDocumentTypeDefinition, _isType, asLoadable, buildCommentRangeDecorations, buildLegacyTheme, buildRangeDecorationSelectionsFromComments, buildTextSelectionFromFragment, catchWithCount, checkoutPair, classifyConfigError, classifyRequestError, collate, compileFieldPath, connectionState, createAuthStore, createBufferedDocument, createConfig, createConnectionStatusStore, createDefaultIcon, createDocumentPreviewStore, createDocumentStore, createDraftFrom, createGrantsStore, createHistoryStore, createHookFromObservableFactory, createKeyValueStore, createMockAuthStore, createObservableBufferedDocument, createPatchChannel, createPlugin, createPresenceStore, createProjectStore, createPublishedFrom, createRequestErrorChannel, createSWR, createSanityMediaLibraryFileSource, createSanityMediaLibraryImageSource, createSchema, createSearch, createSourceFromConfig, createUserColorManager, createUserStore, createWorkspaceFromConfig, dec, decodePath, defaultLocale, defaultRenderAnnotation, defaultRenderBlock, defaultRenderField, defaultRenderInlineBlock, defaultRenderInput, defaultRenderItem, defaultRenderPreview, defaultTemplateForType, defaultTemplatesForSchema, defaultTheme, defineConfig, defineDocumentFieldAction, defineDocumentInspector, defineLocale, defineLocaleResourceBundle, defineLocalesResources, definePlugin, defineSearchFilter, defineSearchFilterOperators, defineSearchOperator, deriveDocumentSyncState, diffMatchPatch, documentFieldActionsReducer, documentIdEquals, editState, emitOperation, emptyValuesByType, encodePath, escapeField, evaluateWorkspaceHidden, fieldNeedsEscape, findIndex, flattenConfig, formatRelativeLocale, formatRelativeLocalePublishDate, fromMutationPatches, generateStudioManifest, getAnnotationAtPath, getAnnotationColor, getApiErrorCode, getCalendarLabels, getConfigContextFromSource, getDiffAtPath, getDocumentIdForCanvasLink, getDocumentIsInPerspective, getDocumentPairPermissions, getDocumentValuePermissions, getDocumentVariantType, getErrorMessage, getExpandOperations, getIdPair, getInitialValueStream, getItemKey, getItemKeySegment, getNamelessWorkspaceIdentifier, getPairListener, getPairTarget, getPreviewPaths, getPreviewStateObservable, getPreviewValueWithFallback, getProjectGrants, getProviderTitle, getReferencePaths, getReleaseIdFromReleaseDocumentId, getReleaseTone, getSchemaTypeTitle, getSearchableTypes, getSelectedVariant, getTargetDocument, getTargetScopeId, getTemplatePermissions, getValueAtPath, getValueError, getVariantPublishedSibling, getVariantTitle, getVersionInlineBadge, getWorkspaceIdentifier, globalScope, grantsPermissionOn, hasCommentMessageValue, idMatchesPerspective, inc, initialDocumentFieldActions, insert, isAddedItemDiff, isArray, isArrayOfBlocksInputProps, isArrayOfObjectsInputProps, isArrayOfPrimitivesInputProps, isAuthStore, isBooleanInputProps, isBuilder, isCardinalityOnePerspective, isCardinalityOneRelease, isClientRequestError, isCookielessCompatibleLoginMethod, isCreateDocumentVersionEvent, isCreateLiveDocumentEvent, isDeleteDocumentGroupEvent, isDeleteDocumentVersionEvent, isDev, isDocumentInSelectedVariant, isDocumentLimitError, isDraft, isDraftPerspective, isDraftVersion, isEditDocumentVersionEvent, isEmptyObject, isFieldChange, isGoingToUnpublish, isGroupChange, isInvalidSessionError, isNetworkError, isNewDocument, isNonNullable, isNumberInputProps, isObjectInputProps, isObjectItemProps, isPausedCardinalityOneRelease, isPerspectiveRaw, isPerspectiveWriteable, isProd, isPublishDocumentVersionEvent, isPublishedPerspective, isPublishedVersion, isRecord, isReleaseDocument, isReleasePerspective, isReleaseScheduledOrScheduling, isReleaseVersion, isRemovedItemDiff, isSanityDefinedAction, isScheduleDocumentVersionEvent, isString, isStringInputProps, isSystemBundle, isSystemBundleName, isTextSelectionComment, isTimeoutError, isTruthy, isUnauthorizedError, isUnchangedDiff, isUnpublishDocumentEvent, isUnscheduleDocumentVersionEvent, isUpdateLiveDocumentEvent, isValidAnnouncementAudience, isValidAnnouncementRole, isVariantVersion, joinPath, listenQuery, matchWorkspace, measureFirstEmission, measureFirstMatch, mergeParseErrors, newDraftFrom, noop, normalizeIndexSegment, normalizeIndexTupleSegment, normalizeKeySegment, normalizePathSegment, onRetry, operationEvents, operatorDefinitions, parseRetryAfter, passthroughErrorHandler, pathToString, pathsAreEqual, prefixPath, prepareConfig, prepareForPreview, prepareTemplates, readVersionType, remoteSnapshots, removeDupes, removeMissingReferences, removeUndefinedLocaleResources, renderStudio, resolveConditionalProperty, resolveConfig, resolveDiffComponent, resolveInitialObjectValue, resolveInitialValue, resolveInitialValueForType, resolveSchemaTypes, selectUpstreamVersion, serializeError, set, setAtPath, setIfMissing, sliceString, snapshotPair, sortReleases, stringToPath, supportsTouch, systemBundles, toMutationPatches, truncateString, uncaughtErrorHandler, unset, unstable_useObserveDocument, unstable_useValuePreview, uploadSchema, usEnglishLocale, useActiveReleases, useActiveWorkspace, useAddonDataset, useAgentVersionDisplay, useAllVariants, useAnnotationColor, useArchivedReleases, useCanvasCompanionDoc, useChangeIndicatorsReportedValues, useChangeIndicatorsReporter, useClient, useColorScheme, useColorSchemeInternalValue, useColorSchemeOptions, useColorSchemeSetValue, useColorSchemeValue, useComlinkStore, useComments, useCommentsEnabled, useCommentsSelectedPath, useCommentsTelemetry, useConditionalToast, useConfigContextFromSource, useConnectionState, useConnectionStatusStore, useCopyErrorDetails, useCopyPaste, useCurrentLocale, useCurrentUser, useDataset, useDateTimeFormat, useDialogStack, useDidUpdate, useDiffAnnotationColor, useDivergenceNavigator, useDocumentChange, useDocumentDivergences, useDocumentForm, useDocumentIdStack, useDocumentLimitsUpsellContext, useDocumentOperation, useDocumentOperationEvent, useDocumentPairPermissions, useDocumentPairPermissionsFromHookFactory, useDocumentPresence, useDocumentPreviewStore, useDocumentPreviewValues, useDocumentStore, useDocumentSyncState, useDocumentType, useDocumentValuePermissions, useDocumentValues, useDocumentVersionInfo, useDocumentVersionTypeSortedList, useDocumentVersions, useEditState, useEnhancedObjectDialog, useEvents, useEventsStore, useExcludedPerspective, useFeatureEnabled, useFeedback, useFieldActions, useFilteredReleases, useFormBuilder, useFormCallbacks, useFormState, useFormValue, useFormatRelativeLocalePublishDate, useFormattedDuration, useGetDefaultPerspective, useGetFormValue, useGetI18nText, useGlobalCopyPasteElementHandler, useGlobalPresence, useGrantsStore, useHistoryStore, useHoveredField, useI18nText, useInStudioFeedback, useInitialValue, useInitialValueResolverContext, useIsReleaseActive, useKeyValueStore, useListFormat, useLoadable, useLocale, useManageFavorite, useMiddlewareComponents, useNavigateToCanvasDoc, useNumberFormat, useOnScroll, useOnlyHasVersions, useParseErrorForPath, useParseErrors, usePausedScheduledDraft, usePerspective, usePresenceStore, usePreviewCard, useProject, useProjectDatasets, useProjectId, useProjectStore, useReconnectingToast, useReferenceInputOptions, useReferringDocuments, useRelativeTime, useReleasesIds, useRenderingContextStore, useReportParseError, useResolveInitialValueForType, useResourceCache, useRetryCountdown, useReviewChanges, useRovingFocus, useScheduledDraftDocument, useScheduledDraftsEnabled, useSchema, useSearchMaxFieldDepth, useSearchState, useSetPerspective, useSetVariant, useSingleDocRelease, useSource, useStudioErrorHandler, useStudioFeedbackTags, useStudioUrl, useSyncState, useTargetDocumentState, useTelemetryConsent, useTemplatePermissions, useTemplatePermissionsFromHookFactory, useTemplates, useThrottledCallback, useTimeAgo, useTimelineSelector, useTimelineStore, useTools, useTrackerStore, useTrackerStoreReporter, useTranslation, useUnique, useUnitFormatter, useUnstableObserveDocument, useUser, useUserColor, useUserColorManager, useUserListWithPermissions, useUserStore, useValidationStatus, useValuePreview, useVariantDocumentOperations, useVersionOperations, useVersionRelease, useVirtualizerScrollInstance, useVisibleWorkspaces, useWorkspace, useWorkspaceLoader, useWorkspaceSchemaId, useWorkspaces, useZIndex, userHasRole, validateBasePaths, validateDocument, validateNames, validateWorkspaces, validation, visitDiff } from "./_chunks-es/index2.js";
import "react/jsx-runtime";
import "react/compiler-runtime";
import "@sanity/icons/ChevronRight";
import "@sanity/ui";
import "react";
import { Hotkeys } from "./_chunks-es/index.js";
import "react-i18next";
import { FeedbackContext, SchedulesContext } from "sanity/_singletons";
import "@portabletext/react";
import "@sanity/color";
import "@sanity/icons/LinkRemoved";
export * from "@sanity/types";
import "styled-components";
import "@portabletext/editor";
import "@sanity/util/paths";
import "lodash-es/findIndex.js";
import { isAgentBundleName } from "./_chunks-es/createAgentBundlesStore.js";
import "@sanity/ui/theme";
import "@sanity/icons/Launch";
import "@sanity/telemetry/react";
import "@sanity/icons/Close";
import "@sanity/uuid";
import "sanity/router";
import "react-rx";
import "rxjs";
import "lodash-es/throttle.js";
import "scroll-into-view-if-needed";
import "@sanity/id-utils";
import "@sanity/icons/ChevronDown";
import "@sanity/icons/InfoOutline";
import "date-fns/addHours";
import "date-fns/startOfHour";
import "@sanity/icons/EarthGlobe";
import "date-fns/format";
import "date-fns/isValid";
import "date-fns/parse";
import "@sanity/icons/Search";
import "rxjs/operators";
import "motion/react";
import "lodash-es/debounce.js";
import "@sanity/icons/User";
import "@sanity/icons/CheckmarkCircle";
import "@sanity/icons/Circle";
import "lodash-es/deburr.js";
import "@sanity/icons/Calendar";
import "@sanity/util/legacyDateFormat";
import "react-focus-lock";
import "@sanity/icons/Document";
import "@sanity/icons/Trash";
import "mendoza";
import "@sanity/icons/Copy";
import "@sanity/icons/Link";
import "@sanity/asset-utils";
import "@sanity/image-url";
import "react-fast-compare";
import "@sanity/icons/Add";
import "@sanity/icons/ChevronLeft";
import "@sanity/icons/Checkmark";
import "json-reduce";
import "rxjs-exhaustmap-with-trailing";
import "shallow-equals";
import { SANITY_VERSION } from "./_chunks-es/version.js";
import { DEFAULT_ANNOTATIONS, DEFAULT_DECORATORS } from "@sanity/schema";
import { DRAFTS_FOLDER, VERSION_FOLDER, getDraftId, getPublishedId, getVersionFromId, getVersionId, isDraftId, isPublishedId, isVersionId } from "@sanity/client/csm";
export {
  ALL_FIELDS_GROUP,
  ActiveWorkspaceMatcher,
  AddonDatasetProvider,
  ArrayOfObjectOptionsInput,
  ArrayOfObjectsFunctions,
  ArrayOfObjectsInput,
  ArrayOfObjectsInputMember,
  ArrayOfObjectsInputMembers,
  ArrayOfObjectsItem,
  ArrayOfOptionsInput,
  ArrayOfPrimitiveOptionsInput,
  ArrayOfPrimitivesFunctions,
  ArrayOfPrimitivesInput,
  ArrayOfPrimitivesItem,
  AutoCollapseMenu,
  AvatarSkeleton,
  BetaBadge,
  PortableTextInput as BlockEditor,
  BlockImagePreview,
  BlockPreview,
  BooleanInput,
  COMMENTS_INSPECTOR_NAME,
  CONNECTING,
  CapabilityGate,
  ChangeBreadcrumb,
  ChangeConnectorRoot,
  ChangeFieldWrapper,
  ChangeIndicator,
  ChangeIndicatorsTracker,
  ChangeList,
  ChangeResolver,
  ChangeTitleSegment,
  ChangesError,
  Chip,
  CircularProgress,
  CollapseMenu,
  CollapseMenuButton,
  ColorSchemeCustomProvider,
  ColorSchemeLocalStorageProvider,
  ColorSchemeProvider,
  CommandList,
  CommentDeleteDialog,
  CommentDisabledIcon,
  CommentInlineHighlightSpan,
  CommentInput,
  CommentsAuthoringPathProvider,
  CommentsEnabledProvider,
  CommentsIntentProvider,
  CommentsList,
  CommentsProvider,
  CommentsSelectedPathProvider,
  CompactPreview,
  Rule as ConcreteRuleClass,
  ConfigErrorGate,
  ConfigPropertyError,
  ConfigResolutionError,
  ContextMenuButton,
  CopyPasteProvider,
  CorsOriginErrorScreen,
  CreateButton as CreateReferenceButton,
  StudioCrossDatasetReferenceInput as CrossDatasetReferenceInput,
  CrossDatasetReferencePreview,
  DEFAULT_ANNOTATIONS,
  DEFAULT_DECORATORS,
  DEFAULT_MAX_RECURSION_DEPTH,
  DEFAULT_STUDIO_CLIENT_OPTIONS,
  DRAFTS_FOLDER,
  DateInput,
  DateTimeInput$1 as DateTimeInput,
  DefaultPreview,
  DetailPreview,
  DiffCard,
  DiffErrorBoundary,
  DiffFromTo,
  DiffInspectWrapper,
  DiffString,
  DiffStringSegment,
  DiffTooltip,
  DivergencesProvider,
  DocumentGroupInventory,
  DocumentGroupInventoryAction,
  DocumentPreviewPresence,
  DocumentStatus,
  DocumentStatusIndicator,
  EMPTY_ARRAY,
  EMPTY_OBJECT,
  EditPortal,
  EditScheduleForm,
  EmailInput,
  EnhancedObjectDialog,
  ErrorActions,
  ErrorMessage,
  Event,
  EventsProvider,
  FallbackDiff,
  FeedbackContext,
  FeedbackDialog,
  FieldActionMenu,
  FieldActionsProvider,
  FieldActionsResolver,
  FieldChange,
  FieldPresence,
  FieldPresenceInner,
  FieldPresenceWithOverlay,
  StudioFileInput as FileInput,
  Filters,
  FormBuilder,
  FormCallbacksProvider,
  FormCell,
  FormContainer,
  FormField,
  FormFieldHeaderText,
  FormFieldSet,
  FormFieldStatus,
  FormFieldValidationStatus,
  FormInput,
  FormProvider,
  FormRow,
  FormValueProvider,
  FromTo,
  FromToArrow,
  GetFormValueProvider,
  GetHookCollectionState,
  GroupChange,
  Hotkeys,
  HoveredFieldProvider,
  StudioImageInput as ImageInput,
  ImperativeToast,
  InlinePreview,
  InsufficientPermissionsMessage,
  IntentButton,
  IsLastPaneProvider,
  LATEST,
  LegacyLayerProvider,
  LinearProgress,
  LiveManifestRegisterProvider,
  LoadingBlock,
  LocaleProvider,
  LocaleProviderBase,
  MediaPreview,
  MemberField,
  MemberFieldError,
  MemberFieldSet,
  MemberItemError,
  MetaInfo,
  NoChanges,
  NumberInput,
  ObjectInput,
  ObjectInputMember,
  ObjectInputMembers,
  ObjectMembers,
  PUBLISHED,
  ParseErrorsProvider,
  PatchEvent,
  PerspectiveProvider,
  PopoverDialog,
  PortableTextInput2 as PortableTextInput,
  PresenceOverlay,
  PresenceScope,
  Preview,
  PreviewCard,
  PreviewLoader,
  RELEASES_INTENT,
  RELEASES_STUDIO_CLIENT_OPTIONS,
  RESOLVE_INITIAL_VALUE_TIMEOUT_MS,
  ReferenceAutocomplete,
  StudioReferenceInput as ReferenceInput,
  ReferenceInputOptionsProvider,
  ReferenceInputPreviewCard,
  RelativeTime,
  ReleaseAvatar,
  ReleaseAvatarIcon,
  ReleaseTitle,
  ReleasesNav,
  RequestErrorDialog,
  Resizable,
  ResourceCacheProvider,
  RevertChangesButton,
  RevertChangesConfirmDialog,
  SANITY_PATCH_TYPE,
  SANITY_VERSION,
  SESSION_ID,
  SanityDefaultPreview,
  useScheduleAction as ScheduleAction,
  ScheduledBadge,
  SchedulesContext,
  SchemaError,
  ScrollContainer,
  SearchButton,
  SearchDialog,
  SearchHeader,
  SearchPopover,
  SearchProvider,
  SearchResultItemPreview,
  SelectInput,
  SingleDocReleaseProvider,
  SlugInput,
  SourceProvider,
  StatusButton,
  StringInput,
  Studio,
  StudioAnnouncementsCard,
  StudioAnnouncementsDialog,
  StudioFeedbackDialog,
  StudioFeedbackProvider,
  StudioLayout,
  StudioLayoutComponent,
  StudioLogo,
  StudioNavbar,
  StudioProvider,
  StudioToolMenu,
  TIMELINE_ITEM_I18N_KEY_MAPPING,
  TagsArrayInput,
  TelephoneInput,
  TemplatePreview,
  TextInput,
  TextWithTone,
  Timeline,
  TimelineController,
  ToolLink,
  TooltipOfDisabled,
  TransformPatches,
  Translate,
  UniversalArrayInput,
  UpdateReadOnlyPlugin,
  UpsellDescriptionSerializer,
  UpsellDialogDismissed,
  UpsellDialogLearnMoreCtaClicked,
  UpsellDialogUpgradeCtaClicked,
  UpsellDialogViewed,
  UrlInput,
  UserAvatar,
  UserColorManagerProvider,
  VARIANTS_STUDIO_CLIENT_OPTIONS,
  VERSION_FOLDER,
  ValueError,
  VersionChip,
  VersionInlineBadge,
  VirtualizerScrollInstanceProvider,
  VisibleWorkspacesProvider,
  WithReferringDocuments,
  WorkspaceLoaderBoundary as WorkspaceLoader,
  WorkspaceProvider,
  WorkspacesProvider,
  ZIndexProvider,
  _createAuthStore,
  _isCustomDocumentTypeDefinition,
  _isSanityDocumentTypeDefinition,
  _isType,
  asLoadable,
  buildCommentRangeDecorations,
  buildLegacyTheme,
  buildRangeDecorationSelectionsFromComments,
  buildTextSelectionFromFragment,
  catchWithCount,
  checkoutPair,
  classifyConfigError,
  classifyRequestError,
  collate,
  compileFieldPath,
  connectionState,
  createAuthStore,
  createBufferedDocument,
  createConfig,
  createConnectionStatusStore,
  createDefaultIcon,
  createDocumentPreviewStore,
  createDocumentStore,
  createDraftFrom,
  createGrantsStore,
  createHistoryStore,
  createHookFromObservableFactory,
  createKeyValueStore,
  createMockAuthStore,
  createObservableBufferedDocument,
  createPatchChannel,
  createPlugin,
  createPresenceStore,
  createProjectStore,
  createPublishedFrom,
  createRequestErrorChannel,
  createSWR,
  createSanityMediaLibraryFileSource,
  createSanityMediaLibraryImageSource,
  createSchema,
  createSearch,
  createSourceFromConfig,
  createUserColorManager,
  createUserStore,
  createWorkspaceFromConfig,
  dec,
  decodePath,
  defaultLocale,
  defaultRenderAnnotation,
  defaultRenderBlock,
  defaultRenderField,
  defaultRenderInlineBlock,
  defaultRenderInput,
  defaultRenderItem,
  defaultRenderPreview,
  defaultTemplateForType,
  defaultTemplatesForSchema,
  defaultTheme,
  defineConfig,
  defineDocumentFieldAction,
  defineDocumentInspector,
  defineLocale,
  defineLocaleResourceBundle,
  defineLocalesResources,
  definePlugin,
  defineSearchFilter,
  defineSearchFilterOperators,
  defineSearchOperator,
  deriveDocumentSyncState,
  diffMatchPatch,
  documentFieldActionsReducer,
  documentIdEquals,
  editState,
  emitOperation,
  emptyValuesByType,
  encodePath,
  escapeField,
  evaluateWorkspaceHidden,
  fieldNeedsEscape,
  findIndex,
  flattenConfig,
  formatRelativeLocale,
  formatRelativeLocalePublishDate,
  fromMutationPatches,
  generateStudioManifest,
  getAnnotationAtPath,
  getAnnotationColor,
  getApiErrorCode,
  getCalendarLabels,
  getConfigContextFromSource,
  getDiffAtPath,
  getDocumentIdForCanvasLink,
  getDocumentIsInPerspective,
  getDocumentPairPermissions,
  getDocumentValuePermissions,
  getDocumentVariantType,
  getDraftId,
  getErrorMessage,
  getExpandOperations,
  getIdPair,
  getInitialValueStream,
  getItemKey,
  getItemKeySegment,
  getNamelessWorkspaceIdentifier,
  getPairListener,
  getPairTarget,
  getPreviewPaths,
  getPreviewStateObservable,
  getPreviewValueWithFallback,
  getProjectGrants,
  getProviderTitle,
  getPublishedId,
  getReferencePaths,
  getReleaseIdFromReleaseDocumentId,
  getReleaseTone,
  getSchemaTypeTitle,
  getSearchableTypes,
  getSelectedVariant,
  getTargetDocument,
  getTargetScopeId,
  getTemplatePermissions,
  getValueAtPath,
  getValueError,
  getVariantPublishedSibling,
  getVariantTitle,
  getVersionFromId,
  getVersionId,
  getVersionInlineBadge,
  getWorkspaceIdentifier,
  globalScope,
  grantsPermissionOn,
  hasCommentMessageValue,
  idMatchesPerspective,
  inc,
  initialDocumentFieldActions,
  insert,
  isAddedItemDiff,
  isAgentBundleName,
  isArray,
  isArrayOfBlocksInputProps,
  isArrayOfObjectsInputProps,
  isArrayOfPrimitivesInputProps,
  isAuthStore,
  isBooleanInputProps,
  isBuilder,
  isCardinalityOnePerspective,
  isCardinalityOneRelease,
  isClientRequestError,
  isCookielessCompatibleLoginMethod,
  isCreateDocumentVersionEvent,
  isCreateLiveDocumentEvent,
  isDeleteDocumentGroupEvent,
  isDeleteDocumentVersionEvent,
  isDev,
  isDocumentInSelectedVariant,
  isDocumentLimitError,
  isDraft,
  isDraftId,
  isDraftPerspective,
  isDraftVersion,
  isEditDocumentVersionEvent,
  isEmptyObject,
  isFieldChange,
  isGoingToUnpublish,
  isGroupChange,
  isInvalidSessionError,
  isNetworkError,
  isNewDocument,
  isNonNullable,
  isNumberInputProps,
  isObjectInputProps,
  isObjectItemProps,
  isPausedCardinalityOneRelease,
  isPerspectiveRaw,
  isPerspectiveWriteable,
  isProd,
  isPublishDocumentVersionEvent,
  isPublishedId,
  isPublishedPerspective,
  isPublishedVersion,
  isRecord,
  isReleaseDocument,
  isReleasePerspective,
  isReleaseScheduledOrScheduling,
  isReleaseVersion,
  isRemovedItemDiff,
  isSanityDefinedAction,
  isScheduleDocumentVersionEvent,
  isString,
  isStringInputProps,
  isSystemBundle,
  isSystemBundleName,
  isTextSelectionComment,
  isTimeoutError,
  isTruthy,
  isUnauthorizedError,
  isUnchangedDiff,
  isUnpublishDocumentEvent,
  isUnscheduleDocumentVersionEvent,
  isUpdateLiveDocumentEvent,
  isValidAnnouncementAudience,
  isValidAnnouncementRole,
  isVariantVersion,
  isVersionId,
  joinPath,
  listenQuery,
  matchWorkspace,
  measureFirstEmission,
  measureFirstMatch,
  mergeParseErrors,
  newDraftFrom,
  noop,
  normalizeIndexSegment,
  normalizeIndexTupleSegment,
  normalizeKeySegment,
  normalizePathSegment,
  onRetry,
  operationEvents,
  operatorDefinitions,
  parseRetryAfter,
  passthroughErrorHandler,
  pathToString,
  pathsAreEqual,
  prefixPath,
  prepareConfig,
  prepareForPreview,
  prepareTemplates,
  readVersionType,
  remoteSnapshots,
  removeDupes,
  removeMissingReferences,
  removeUndefinedLocaleResources,
  renderStudio,
  resolveConditionalProperty,
  resolveConfig,
  resolveDiffComponent,
  resolveInitialObjectValue,
  resolveInitialValue,
  resolveInitialValueForType,
  resolveSchemaTypes,
  selectUpstreamVersion,
  serializeError,
  set,
  setAtPath,
  setIfMissing,
  sliceString,
  snapshotPair,
  sortReleases,
  stringToPath,
  supportsTouch,
  systemBundles,
  toMutationPatches,
  truncateString,
  uncaughtErrorHandler,
  unset,
  unstable_useObserveDocument,
  unstable_useValuePreview,
  uploadSchema,
  usEnglishLocale,
  useActiveReleases,
  useActiveWorkspace,
  useAddonDataset,
  useAgentVersionDisplay,
  useAllVariants,
  useAnnotationColor,
  useArchivedReleases,
  useCanvasCompanionDoc,
  useChangeIndicatorsReportedValues,
  useChangeIndicatorsReporter,
  useClient,
  useColorScheme,
  useColorSchemeInternalValue,
  useColorSchemeOptions,
  useColorSchemeSetValue,
  useColorSchemeValue,
  useComlinkStore,
  useComments,
  useCommentsEnabled,
  useCommentsSelectedPath,
  useCommentsTelemetry,
  useConditionalToast,
  useConfigContextFromSource,
  useConnectionState,
  useConnectionStatusStore,
  useCopyErrorDetails,
  useCopyPaste,
  useCurrentLocale,
  useCurrentUser,
  useDataset,
  useDateTimeFormat,
  useDialogStack,
  useDidUpdate,
  useDiffAnnotationColor,
  useDivergenceNavigator,
  useDocumentChange,
  useDocumentDivergences,
  useDocumentForm,
  useDocumentIdStack,
  useDocumentLimitsUpsellContext,
  useDocumentOperation,
  useDocumentOperationEvent,
  useDocumentPairPermissions,
  useDocumentPairPermissionsFromHookFactory,
  useDocumentPresence,
  useDocumentPreviewStore,
  useDocumentPreviewValues,
  useDocumentStore,
  useDocumentSyncState,
  useDocumentType,
  useDocumentValuePermissions,
  useDocumentValues,
  useDocumentVersionInfo,
  useDocumentVersionTypeSortedList,
  useDocumentVersions,
  useEditState,
  useEnhancedObjectDialog,
  useEvents,
  useEventsStore,
  useExcludedPerspective,
  useFeatureEnabled,
  useFeedback,
  useFieldActions,
  useFilteredReleases,
  useFormBuilder,
  useFormCallbacks,
  useFormState,
  useFormValue,
  useFormatRelativeLocalePublishDate,
  useFormattedDuration,
  useGetDefaultPerspective,
  useGetFormValue,
  useGetI18nText,
  useGlobalCopyPasteElementHandler,
  useGlobalPresence,
  useGrantsStore,
  useHistoryStore,
  useHoveredField,
  useI18nText,
  useInStudioFeedback,
  useInitialValue,
  useInitialValueResolverContext,
  useIsReleaseActive,
  useKeyValueStore,
  useListFormat,
  useLoadable,
  useLocale,
  useManageFavorite,
  useMiddlewareComponents,
  useNavigateToCanvasDoc,
  useNumberFormat,
  useOnScroll,
  useOnlyHasVersions,
  useParseErrorForPath,
  useParseErrors,
  usePausedScheduledDraft,
  usePerspective,
  usePresenceStore,
  usePreviewCard,
  useProject,
  useProjectDatasets,
  useProjectId,
  useProjectStore,
  useReconnectingToast,
  useReferenceInputOptions,
  useReferringDocuments,
  useRelativeTime,
  useReleasesIds,
  useRenderingContextStore,
  useReportParseError,
  useResolveInitialValueForType,
  useResourceCache,
  useRetryCountdown,
  useReviewChanges,
  useRovingFocus,
  useScheduledDraftDocument,
  useScheduledDraftsEnabled,
  useSchema,
  useSearchMaxFieldDepth,
  useSearchState,
  useSetPerspective,
  useSetVariant,
  useSingleDocRelease,
  useSource,
  useStudioErrorHandler,
  useStudioFeedbackTags,
  useStudioUrl,
  useSyncState,
  useTargetDocumentState,
  useTelemetryConsent,
  useTemplatePermissions,
  useTemplatePermissionsFromHookFactory,
  useTemplates,
  useThrottledCallback,
  useTimeAgo,
  useTimelineSelector,
  useTimelineStore,
  useTools,
  useTrackerStore,
  useTrackerStoreReporter,
  useTranslation,
  useUnique,
  useUnitFormatter,
  useUnstableObserveDocument,
  useUser,
  useUserColor,
  useUserColorManager,
  useUserListWithPermissions,
  useUserStore,
  useValidationStatus,
  useValuePreview,
  useVariantDocumentOperations,
  useVersionOperations,
  useVersionRelease,
  useVirtualizerScrollInstance,
  useVisibleWorkspaces,
  useWorkspace,
  useWorkspaceLoader,
  useWorkspaceSchemaId,
  useWorkspaces,
  useZIndex,
  userHasRole,
  validateBasePaths,
  validateDocument,
  validateNames,
  validateWorkspaces,
  validation,
  visitDiff
};
