// @sanity/workbench-cli/dist/actions/build/views/artifact.js
import { VIEW_COMPONENTS } from '../../../contract.js';
import { renderRemote } from '../render-remote.js';
/** Subdirectory under the federation runtime dir where view artifacts are written. */ const VIEWS_DIR_NAME = 'views';
/**
 * Expand each view into one generated artifact per component it exposes. A
 * panel's `title` and `panel` each become their own render-contract module and
 * module-federation expose, so the host renders each as an independent island.
 *
 * Each artifact binds its component as the `App` the render contract renders —
 * a single-component view exports a bare function, a multi-component one keys by
 * name — behind an HMR boundary so view edits re-render through the new module.
 */ export function viewArtifacts(views) {
    const artifacts = [];
    for (const view of views){
        for (const component of VIEW_COMPONENTS[view.type]){
            artifacts.push({
                expose: `./${VIEWS_DIR_NAME}/${view.name}/${component}`,
                path: `${VIEWS_DIR_NAME}/${view.name}/${component}.js`,
                source: ({ resolveImport })=>renderRemote({
                        app: `typeof view.components === 'function' ? view.components : view.components[${JSON.stringify(component)}]`,
                        hmr: true,
                        preamble: `import view from ${JSON.stringify(resolveImport(view.src))}`,
                        version: `view.version`
                    })
            });
        }
    }
    return artifacts;
}
