// @sanity/cli-build/dist/actions/build/buildStaticFiles.js
import path from 'node:path';
import { build, createBuilder } from 'vite';
import { copyDir } from '../../util/copyDir.js';
import { buildDebug } from './buildDebug.js';
import { getAppEnvironmentVariables, getStudioEnvironmentVariables } from './getEnvironmentVariables.js';
import { extendViteConfigWithUserConfig, finalizeViteConfig, getViteConfig } from './getViteConfig.js';
import { writeFavicons } from './writeFavicons.js';
import { resolveEntries, writeSanityRuntime } from './writeSanityRuntime.js';
/**
 * Builds static files
 *
 * @internal
 */ export async function buildStaticFiles(options) {
    const { appTitle, autoUpdates, basePath, cwd, entry, exposes, isApp, isWorkbenchApp, minify = true, outputDir, reactCompiler, schemaExtraction, sourceMap = false, vite: extendViteConfig, workbenchAppId } = options;
    const mode = 'production';
    /* Federation builds only produce the federation environment
   * (remote-entry, mf-manifest) — skip client-specific steps like
   * runtime generation, static file copies, and favicons.
   */ if (isWorkbenchApp) {
        // A workbench remote can also serve itself standalone. When
        // `SANITY_INTERNAL_IS_WORKBENCH_REMOTE` is set, emit an SPA (index.html +
        // bootstrap + favicons) into the same `dist` as the federation output via a
        // dedicated `client` environment (see plugin-sanity-environment). Skipped for
        // a dock-only app (no `./App` to mount), matching `exposesApp` in plugin.ts.
        const emitSpa = process.env.SANITY_INTERNAL_IS_WORKBENCH_REMOTE === 'true' && !(isApp && !entry);
        let entries;
        if (emitSpa) {
            buildDebug('Writing Sanity runtime files (workbench remote SPA)');
            ({ entries } = await writeSanityRuntime({
                appTitle,
                basePath,
                cwd,
                entry,
                isApp,
                isWorkbenchApp,
                reactStrictMode: false,
                watch: false
            }));
        } else {
            buildDebug('Resolving entries for federation build');
            entries = await resolveEntries({
                cwd,
                entry,
                isApp,
                isWorkbenchApp
            });
        }
        buildDebug('Resolving vite config (federation)');
        let viteConfig = await getViteConfig({
            basePath,
            cwd,
            entries,
            exposes,
            getEnvironmentVariables,
            isApp,
            isWorkbenchApp,
            minify,
            mode,
            outputDir,
            reactCompiler,
            // Schema extraction is a build-time artifact, not a client-specific step,
            // so a federated studio extracts its schema like the legacy studio build.
            schemaExtraction,
            sourceMap,
            workbenchAppId
        });
        // Apply the user's Vite config so plugins like `@vanilla-extract/vite-plugin`
        // transform source files before the federation environment is bundled.
        // `finalizeViteConfig` is intentionally skipped: the federation environment
        // has its own entry and does not use `.sanity/runtime/app.js`.
        if (extendViteConfig) {
            viteConfig = await extendViteConfigWithUserConfig({
                command: 'build',
                mode
            }, viteConfig, extendViteConfig);
        }
        if (emitSpa) {
            const fromPath = path.join(cwd, 'static');
            const staticPath = path.join(outputDir, 'static');
            buildDebug(`Copying static files from ${fromPath} to output dir`);
            await copyDir(fromPath, staticPath);
            buildDebug('Writing favicons to output dir');
            const faviconBasePath = `${basePath.replace(/\/+$/, '')}/static`;
            await writeFavicons(faviconBasePath, staticPath);
        }
        buildDebug('Bundling federation environment');
        const builder = await createBuilder(viteConfig);
        await builder.buildApp();
        buildDebug('Bundling complete');
        // TODO: add stats here
        return {
            chunks: []
        };
    }
    buildDebug('Writing Sanity runtime files');
    const { entries } = await writeSanityRuntime({
        appTitle,
        basePath,
        cwd,
        entry,
        isApp,
        isWorkbenchApp,
        reactStrictMode: false,
        watch: false
    });
    function getEnvironmentVariables() {
        return isApp ? getAppEnvironmentVariables({
            jsonEncode: true,
            prefix: 'process.env.'
        }) : getStudioEnvironmentVariables({
            jsonEncode: true,
            prefix: 'process.env.'
        });
    }
    buildDebug('Resolving vite config');
    let viteConfig = await getViteConfig({
        autoUpdates,
        basePath,
        cwd,
        entries,
        getEnvironmentVariables,
        isApp,
        isWorkbenchApp,
        minify,
        mode,
        outputDir,
        reactCompiler,
        schemaExtraction,
        sourceMap
    });
    if (extendViteConfig) {
        const defaultViteConfig = viteConfig;
        viteConfig = await extendViteConfigWithUserConfig({
            command: 'build',
            mode
        }, viteConfig, extendViteConfig);
        viteConfig = await finalizeViteConfig(viteConfig, defaultViteConfig);
    }
    const fromPath = path.join(cwd, 'static');
    // Copy files placed in /static to the built /static
    buildDebug(`Copying static files from ${fromPath} to output dir`);
    const staticPath = path.join(outputDir, 'static');
    await copyDir(fromPath, staticPath);
    // Write favicons, not overwriting ones that already exist, to static folder
    buildDebug('Writing favicons to output dir');
    const faviconBasePath = `${basePath.replace(/\/+$/, '')}/static`;
    await writeFavicons(faviconBasePath, staticPath);
    buildDebug('Bundling using vite');
    const bundle = await build(viteConfig);
    buildDebug('Bundling complete');
    // For typescript only - this shouldn't ever be the case given we're not watching
    if (Array.isArray(bundle) || !('output' in bundle)) {
        return {
            chunks: []
        };
    }
    const stats = [];
    for (const chunk of bundle.output){
        if (chunk.type !== 'chunk') {
            continue;
        }
        stats.push({
            modules: Object.entries(chunk.modules).map(([rawFilePath, chunkModule])=>{
                const filePath = rawFilePath.startsWith('\u0000') ? rawFilePath.slice('\u0000'.length) : rawFilePath;
                return {
                    name: path.isAbsolute(filePath) ? path.relative(cwd, filePath) : filePath,
                    renderedLength: chunkModule.renderedLength
                };
            }),
            name: chunk.name
        });
    }
    return {
        chunks: stats
    };
}
