// @sanity/cli/dist/util/checks.js
import { debug as cliDebug } from '@sanity/cli-core';
import { getErrorMessage } from '@sanity/cli-core/errors';
import { logSymbols } from '@sanity/cli-core/ux';
export function createFailFastReporter(output) {
    return {
        report (check) {
            // Fixes surface in both modes: appended after the message here, and in the
            // dry-run report, so the problem and its fix never drift apart.
            const text = check.solution ? `${check.message}: ${check.solution}` : check.message;
            if (check.status === 'fail') {
                output.error(text, {
                    exit: check.exitCode ?? 1
                });
            } else if (check.status === 'warn') {
                output.warn(text);
            }
        }
    };
}
export function createCollectingReporter() {
    const results = [];
    return {
        report (check) {
            results.push(check);
        },
        results
    };
}
/**
 * Runs a fallible step and turns a throw into a `fail` check. In a real run
 * that fail exits (aborting the sequence); in a dry run it's recorded and
 * `null` comes back so the caller can skip the rest of the step. `name` labels
 * the step in debug logs.
 */ export async function runStep(reporter, step) {
    const { debug = cliDebug, formatError = getErrorMessage, name, solution, work } = step;
    try {
        return await work();
    } catch (err) {
        debug(`${name} step failed`, err);
        reporter.report({
            message: formatError(err),
            solution,
            status: 'fail'
        });
        return null;
    }
}
export function checkStatusIcon(status) {
    switch(status){
        case 'fail':
            {
                return logSymbols.error;
            }
        case 'skip':
            {
                return logSymbols.info;
            }
        case 'warn':
            {
                return logSymbols.warning;
            }
        default:
            {
                return logSymbols.success;
            }
    }
}
/**
 * Indents continuation lines of a multi-line check message past the status
 * icon so its items nest under the heading.
 */ export function nestLines(text) {
    return text.replaceAll('\n', '\n    ');
}
/** Renders problem/warning checks under a title, each with its fix appended. */ export function renderIssues(output, title, checks) {
    if (checks.length === 0) return;
    output.log(`\n${title}`);
    for (const check of checks){
        const fix = check.solution ? `: ${check.solution}` : '';
        output.log(nestLines(`  ${checkStatusIcon(check.status)} ${check.message}${fix}`));
    }
}
