// @sanity/runtime-cli/dist/server/handlers/invoke.js
import { readLocalBlueprint } from '../../actions/blueprints/blueprint.js';
import config from '../../config.js';
import { resolveDeployedResources } from '../../utils/blueprints/resolve-deployed-resources.js';
import { findFunctionInResources } from '../../utils/functions/find.js';
import { runFunction } from '../../utils/functions/invoke/nested.js';
import { Logger } from '../../utils/logger.js';
const silentLogger = Logger(() => { });
export async function handleInvokeRequest(functionName, event, metadata, context, validateResources, executionOptions, depth = 0, callStack = [], onNestedInvoke) {
    const start = performance.now();
    const { resources, stackId, scopeType, scopeId } = await readLocalBlueprint(silentLogger, {
        resources: validateResources,
    });
    const resource = findFunctionInResources(resources, functionName);
    const readBlueprintTime = performance.now() - start;
    const resolvedResources = await resolveDeployedResources({
        resources,
        stackId,
        auth: config.token && scopeType && scopeId ? { token: config.token, scopeType, scopeId } : undefined,
        logger: silentLogger,
    });
    const payload = {
        payload: event,
        ...metadata,
    };
    const response = await runFunction(resource, payload, { ...context, resources: resolvedResources }, {
        forceColor: executionOptions?.forceColor ?? false,
        timeout: executionOptions?.timeout,
        depth,
        callStack,
        onNestedInvoke,
    });
    const timings = { ...response.timings, 'blueprint:read': readBlueprintTime };
    return { ...response, timings };
}
