// @sanity/workbench-cli/dist/services/applications.js
import { PassThrough } from 'node:stream';
import { getGlobalCliClient } from '@sanity/cli-core';
import { isStaging } from '@sanity/cli-core/util';
import FormData from 'form-data';
import { APP_WORKBENCH_API_VERSION } from './apiVersion.js';
export function getWorkbenchUrl(organizationId) {
    return `https://${organizationId}.${isStaging() ? 'run.sanity.work' : 'sanity.run'}`;
}
/** Where a deployed application is served on its organization's workbench. */ export function getApplicationUrl(application) {
    const segment = application.type === 'studio' ? 'studio' : 'application';
    return `${getWorkbenchUrl(application.organizationId)}/${segment}/${application.id}`;
}
async function getClient() {
    return getGlobalCliClient({
        apiVersion: APP_WORKBENCH_API_VERSION,
        requireUser: true
    });
}
export async function getApplication(applicationId) {
    const client = await getClient();
    try {
        return await client.request({
            uri: `/applications/${applicationId}`
        });
    } catch (err) {
        if (err?.statusCode === 404) return null;
        throw err;
    }
}
/**
 * Create an application record (no deployment), so the CLI can build with the
 * returned id, then ship it via {@link createDeployment}.
 */ export async function createApplication(options) {
    const { isSingleton, organizationId, projectId, slug, title, type, visibility } = options;
    const client = await getClient();
    return client.request({
        body: {
            organizationId,
            slug,
            title,
            type,
            ...isSingleton === undefined ? {} : {
                isSingleton
            },
            ...visibility ? {
                visibility
            } : {},
            // Studio config is set once, at create — it's immutable on redeploy.
            ...projectId ? {
                config: {
                    studio: {
                        projectId
                    }
                }
            } : {}
        },
        method: 'POST',
        uri: `/applications`
    });
}
// Patch an application's mutable fields.
export async function updateApplication(applicationId, update) {
    const client = await getClient();
    await client.request({
        body: update,
        method: 'PATCH',
        uri: `/applications/${applicationId}`
    });
}
/** Deploy a new active version to an existing application. */ export async function createDeployment(options) {
    const { applicationId, interfaces, isAutoUpdating, tarball, version, workspaces } = options;
    const formData = new FormData();
    formData.append('isAutoUpdating', isAutoUpdating.toString());
    appendDeploymentParts(formData, {
        interfaces,
        tarball,
        version,
        workspaces
    });
    return request(`/applications/${applicationId}/deployments`, formData);
}
/** Soft-deletes the application and all its deployments; already deleted counts as done. */ export async function deleteApplication(applicationId) {
    const client = await getClient();
    try {
        await client.request({
            method: 'DELETE',
            uri: `/applications/${applicationId}`
        });
    } catch (err) {
        if (err?.statusCode !== 404) throw err;
    }
}
function appendDeploymentParts(formData, { interfaces, tarball, version, workspaces }) {
    formData.append('version', version);
    appendJson(formData, 'interfaces', interfaces);
    // Studio-only — the server rejects a workspaces part on non-studio types.
    if (workspaces?.length) appendJson(formData, 'workspaces', workspaces);
    formData.append('tarball', tarball, {
        contentType: 'application/gzip',
        filename: 'app.tar.gz'
    });
}
/** Structured parts must arrive as JSON so the server parses them. */ function appendJson(formData, name, value) {
    formData.append(name, JSON.stringify(value), {
        contentType: 'application/json'
    });
}
async function request(uri, formData) {
    const client = await getClient();
    return client.request({
        body: formData.pipe(new PassThrough()),
        headers: formData.getHeaders(),
        method: 'POST',
        uri
    });
}
