// @sanity/workbench-cli/dist/services/installations.js
import { PassThrough } from 'node:stream';
import { getGlobalCliClient } from '@sanity/cli-core';
import FormData from 'form-data';
import { APP_WORKBENCH_API_VERSION } from './apiVersion.js';
async function getClient() {
    return getGlobalCliClient({
        apiVersion: APP_WORKBENCH_API_VERSION,
        requireUser: true
    });
}
/** The org's active singleton installation, matched on its slug. */ export async function resolveSingletonInstallationId(organizationId, slug) {
    const client = await getClient();
    // `limit=none` returns every installation in one response, no pagination.
    const { data } = await client.request({
        query: {
            limit: 'none',
            organizationId
        },
        uri: '/installations'
    });
    return data.find((item)=>item.application?.slug === slug)?.id;
}
/** Upload a config snapshot to the installation as a multipart tarball. */ export async function createConfig(installationId, { tarball, version }) {
    const formData = new FormData();
    formData.append('version', version);
    formData.append('tarball', tarball, {
        contentType: 'application/gzip',
        filename: 'installation-config.tar.gz'
    });
    const client = await getClient();
    await client.request({
        body: formData.pipe(new PassThrough()),
        headers: formData.getHeaders(),
        method: 'POST',
        uri: `/installations/${installationId}/configs`
    });
}
/** The installation's deployed config snapshots, newest first. */ export async function listConfigs(installationId) {
    const client = await getClient();
    const { data } = await client.request({
        query: {
            limit: 'none'
        },
        uri: `/installations/${installationId}/configs`
    });
    return data;
}
/**
 * Soft-deletes one config snapshot; its content is purged once superseded.
 * Already deleted counts as done, so a partially-failed undeploy can re-run.
 */ export async function deleteConfig(installationId, configId) {
    const client = await getClient();
    try {
        await client.request({
            method: 'DELETE',
            uri: `/installations/${installationId}/configs/${configId}`
        });
    } catch (err) {
        if (err?.statusCode !== 404) throw err;
    }
}
