// @portabletext/plugin-table/dist/_chunks-es/get-table-selection.js
import { getEnclosingBlock, getParent } from "@portabletext/editor/traversal";
const defaultTableConfig = {
  tableType: "table",
  rowsField: "rows",
  rowType: "row",
  cellsField: "cells",
  cellType: "cell",
  valueField: "value"
}, registeredConfigs = [];
function registerTableConfig(config) {
  const existingIndex = registeredConfigs.findIndex((candidate) => candidate.tableType === config.tableType);
  existingIndex === -1 ? registeredConfigs.push(config) : registeredConfigs[existingIndex] = config;
}
function resolveTableConfig(type) {
  return registeredConfigs.find((config) => config.tableType === type || config.rowType === type || config.cellType === type) ?? defaultTableConfig;
}
function createTableGuards(config) {
  return {
    isTable: (node) => node._type === config.tableType && Array.isArray(readField(node, config.rowsField)),
    isRow: (node) => node._type === config.rowType && Array.isArray(readField(node, config.cellsField)),
    isCell: (node) => node._type === config.cellType
  };
}
function tableRows(config, table) {
  return readField(table, config.rowsField) ?? [];
}
function rowCells(config, row) {
  return readField(row, config.cellsField) ?? [];
}
function cellValue(config, cell) {
  return readField(cell, config.valueField) ?? [];
}
function readField(node, field) {
  return node[field];
}
function resolveCell(snapshot, path, config = defaultTableConfig) {
  const {
    isCell,
    isRow,
    isTable
  } = createTableGuards(config), cell = getEnclosingBlock(snapshot, path, {
    match: isCell
  }), row = cell && getParent(snapshot, cell.path, {
    match: isRow
  }), table = row && getParent(snapshot, row.path, {
    match: isTable
  });
  if (!(!cell || !row || !table))
    return {
      cell,
      row,
      table
    };
}
function getTableSelection(snapshot, config = defaultTableConfig) {
  const selection = snapshot.context.selection;
  if (!selection)
    return;
  const anchor = resolveCell(snapshot, selection.anchor.path, config), focus = resolveCell(snapshot, selection.focus.path, config);
  if (!anchor || !focus || anchor.cell.node._key === focus.cell.node._key || anchor.table.node._key !== focus.table.node._key)
    return;
  const rows = tableRows(config, anchor.table.node), anchorRowIndex = rows.findIndex((row) => row._key === anchor.row.node._key), focusRowIndex = rows.findIndex((row) => row._key === focus.row.node._key), anchorColIndex = rowCells(config, anchor.row.node).findIndex((cell) => cell._key === anchor.cell.node._key), focusColIndex = rowCells(config, focus.row.node).findIndex((cell) => cell._key === focus.cell.node._key);
  if (!(anchorRowIndex === -1 || focusRowIndex === -1 || anchorColIndex === -1 || focusColIndex === -1))
    return {
      tablePath: anchor.table.path,
      rowRange: [Math.min(anchorRowIndex, focusRowIndex), Math.max(anchorRowIndex, focusRowIndex)],
      colRange: [Math.min(anchorColIndex, focusColIndex), Math.max(anchorColIndex, focusColIndex)]
    };
}
function resolveTableSelection(snapshot, config = defaultTableConfig) {
  const tableSelection = getTableSelection(snapshot, config);
  if (!tableSelection)
    return;
  const table = getEnclosingBlock(snapshot, tableSelection.tablePath, {
    match: createTableGuards(config).isTable
  });
  if (table)
    return {
      tableSelection,
      table
    };
}
function* memberCells(tableSelection, table, config = defaultTableConfig) {
  const [rowStart, rowEnd] = tableSelection.rowRange, [colStart, colEnd] = tableSelection.colRange;
  for (let rowIndex = rowStart; rowIndex <= rowEnd; rowIndex++) {
    const row = tableRows(config, table)[rowIndex];
    if (row)
      for (let colIndex = colStart; colIndex <= colEnd; colIndex++) {
        const cell = rowCells(config, row)[colIndex];
        cell && (yield {
          node: cell,
          path: [...tableSelection.tablePath, config.rowsField, {
            _key: row._key
          }, config.cellsField, {
            _key: cell._key
          }]
        });
      }
  }
}
export {
  cellValue,
  createTableGuards,
  getTableSelection,
  memberCells,
  registerTableConfig,
  resolveCell,
  resolveTableConfig,
  resolveTableSelection,
  rowCells,
  tableRows
};
