// @sanity/workbench/src/_internal/render.ts
/// <reference types="vite/client" />

import { BehaviorSubject, type Observable } from "rxjs";

import { getSanityEnv } from "../core";
import type { LocalApplication } from "../core/applications/application";
import type { LocalAppConfig } from "../core/applications/config";
import type { Config, RemoteModuleRenderOptions } from "../core/config";
import { logger } from "../core/log";
import { createInstance, log } from "../system/module-federation";

/**
 * Options for rendering the workbench.
 *
 * @public
 */
export interface RenderWorkbenchOptions extends RemoteModuleRenderOptions {}

declare global {
  interface ImportMetaEnv {
    readonly SANITY_INTERNAL_WORKBENCH_REMOTE_URL: string;
  }
  interface ImportMeta {
    readonly env: ImportMetaEnv;
  }
}

type RemoteModule<TProps extends any> = {
  render: (
    rootElement: HTMLElement,
    props: TProps,
    options?: RenderWorkbenchOptions,
  ) => () => void;
};

/**
 * Module defining the remote workbench application.
 *
 * @internal
 */
type WorkbenchRemoteModule = RemoteModule<{
  config: Config;
  localApplications?: Observable<LocalApplication[]>;
  appConfigs?: Observable<LocalAppConfig[]>;
}>;

const REMOTE_NAME = "workbench-remote";
const REMOTE_MODULE = "App";

const LOCAL_APPS_HMR_EVENT = "sanity:workbench:local-applications";
const LOCAL_APPS_HMR_REQUEST = "sanity:workbench:get-local-applications";

/**
 * Creates a Module Federation instance, loads a remote workbench
 * application, and renders it into the provided root element.
 *
 * @param rootElement - The DOM element to render into
 * @param config - Workbench configuration, forwarded to the remote. The CLI
 *   sources `organizationId` from `unstable_defineApp` (required there).
 * @param options - Rendering options forwarded to the remote
 * @returns A cleanup function that unmounts the remote application
 *
 * @public
 */
export async function renderWorkbench(
  rootElement: HTMLElement,
  config: Config,
  options?: RenderWorkbenchOptions,
) {
  if (!rootElement) {
    throw new Error("Missing root element to mount application into");
  }

  const deployedWorkbenchUrl =
    getSanityEnv() === "staging"
      ? "workbench-apps-of5p8qpku.run.sanity.work"
      : "workbench-apps-oSyH1iET5.sanity.run";

  const remoteUrl =
    import.meta.env.SANITY_INTERNAL_WORKBENCH_REMOTE_URL ??
    `https://${deployedWorkbenchUrl}/mf-manifest.json`;

  const mf = createInstance({
    name: "sanity-workbench",
    plugins: [log(logger.debug)],

    remotes: [
      {
        name: REMOTE_NAME,
        entry: remoteUrl,
      },
    ],
  });

  let remoteModule = await mf.loadRemote<WorkbenchRemoteModule>(
    `${REMOTE_NAME}/${REMOTE_MODULE}`,
  );

  if (!remoteModule || typeof remoteModule.render !== "function") {
    throw new Error(
      `Remote module "${REMOTE_NAME}/${REMOTE_MODULE}" did not expose a render function`,
    );
  }

  let localApplications = undefined;
  let appConfigs = undefined;
  let cleanupHmr = () => {};

  if (import.meta.hot) {
    const localApps$ = new BehaviorSubject<LocalApplication[]>([]);
    const configs$ = new BehaviorSubject<LocalAppConfig[]>([]);

    const handler = (payload: {
      applications: LocalApplication[];
      configs?: LocalAppConfig[];
    }) => {
      localApps$.next(payload.applications);
      configs$.next(payload.configs ?? []);
    };

    import.meta.hot.on(LOCAL_APPS_HMR_EVENT, handler);
    import.meta.hot.send(LOCAL_APPS_HMR_REQUEST);

    localApplications = localApps$;
    appConfigs = configs$;

    cleanupHmr = () => import.meta.hot?.off(LOCAL_APPS_HMR_EVENT, handler);
  }

  const unmount = remoteModule.render(
    rootElement,
    { config, appConfigs, localApplications },
    options,
  );

  return () => {
    cleanupHmr();
    unmount();
  };
}
