// @sanity/cli/dist/services/organizations.js
import { getGlobalCliClient } from '@sanity/cli-core';
export const ORGANIZATIONS_API_VERSION = 'v2025-05-14';
/**
 * List all organizations the user has access to
 */ export async function listOrganizations(query) {
    const client = await getGlobalCliClient({
        apiVersion: ORGANIZATIONS_API_VERSION,
        requireUser: true
    });
    return client.request({
        query,
        uri: '/organizations'
    });
}
/**
 * Create a new organization
 */ export async function createOrganization(name, defaultRoleName) {
    const client = await getGlobalCliClient({
        apiVersion: ORGANIZATIONS_API_VERSION,
        requireUser: true
    });
    return client.request({
        body: {
            name,
            ...defaultRoleName ? {
                defaultRoleName
            } : {}
        },
        method: 'post',
        uri: '/organizations'
    });
}
/**
 * Get organization grants for a specific organization
 */ export async function getOrganizationGrants(organizationId) {
    const client = await getGlobalCliClient({
        apiVersion: ORGANIZATIONS_API_VERSION,
        requireUser: true
    });
    return client.request({
        uri: `/organizations/${organizationId}/grants`
    });
}
/**
 * Get a single organization by ID
 */ export async function getOrganization(organizationId) {
    const client = await getGlobalCliClient({
        apiVersion: ORGANIZATIONS_API_VERSION,
        requireUser: true
    });
    return client.request({
        query: {
            includeFeatures: 'false',
            includeMembers: 'false'
        },
        uri: `/organizations/${organizationId}`
    });
}
/**
 * Update an organization
 */ export async function updateOrganization(organizationId, params) {
    const client = await getGlobalCliClient({
        apiVersion: ORGANIZATIONS_API_VERSION,
        requireUser: true
    });
    return client.request({
        body: params,
        method: 'patch',
        uri: `/organizations/${organizationId}`
    });
}
/**
 * Delete an organization
 */ export async function deleteOrganization(organizationId) {
    const client = await getGlobalCliClient({
        apiVersion: ORGANIZATIONS_API_VERSION,
        requireUser: true
    });
    return client.request({
        method: 'delete',
        uri: `/organizations/${organizationId}`
    });
}
