// @sanity/workbench/dist/_chunks-es/bus.js
import { map, filter, Subject, firstValueFrom, BehaviorSubject } from "rxjs";
import { logger } from "./index.js";
class MessageBusError extends Error {
  code;
  constructor(code, message, options) {
    super(message ?? code, options), this.name = "MessageBusError", this.code = code;
  }
}
function createMeta(appId) {
  return { appId, timestamp: Date.now() };
}
const topicMigrations = {}, BUS = /* @__PURE__ */ Symbol.for("sanity.os.bus"), PROTOCOL = /* @__PURE__ */ Symbol.for("sanity.os.protocol"), REGISTRY = /* @__PURE__ */ Symbol.for("sanity.os.registry"), REQUEST = /* @__PURE__ */ Symbol.for("sanity.os.request"), OS_PROTOCOL = 1, log = logger.child("os"), DEFAULT_APP_ID = "workbench", DEFAULT_TIMEOUT_MS = 5e3, NO_VALUE = /* @__PURE__ */ Symbol.for("sanity.os.no-value");
function resolveStateSubject(registry, type) {
  let subject = registry.stateSubjects.get(type);
  return subject || (log.warn(`state topic "${type}" read before any value was published`), subject = new BehaviorSubject(NO_VALUE), registry.stateSubjects.set(type, subject)), subject;
}
function toStateSource(values, getCurrent) {
  const source = values;
  return source.getCurrent = getCurrent, source.firstValue = firstValueFrom(values), source;
}
function resolveStateSource(registry, type) {
  let source = registry.stateSources.get(type);
  if (!source) {
    const subject = resolveStateSubject(registry, type);
    source = toStateSource(
      subject.pipe(filter((value) => value !== NO_VALUE)),
      () => {
        const current = subject.getValue();
        return current === NO_VALUE ? void 0 : current;
      }
    ), registry.stateSources.set(type, source);
  }
  return source;
}
function resolveEventSubject(registry, type) {
  let subject = registry.eventSubjects.get(type);
  return subject || (subject = new Subject(), registry.eventSubjects.set(type, subject)), subject;
}
function settle(request, outcome) {
  request.settled || (request.settled = !0, request.deliver ? request.deliver(outcome) : request.capturedOutcome = outcome);
}
function createMessage(appId, type, payload, request) {
  const message = {
    type,
    payload,
    meta: createMeta(appId),
    reply: (value) => {
      if (request.settled) {
        log.warn(
          `reply ignored for "${type}": no waiting caller or already replied`
        );
        return;
      }
      settle(request, { ok: !0, value });
    },
    get signal() {
      return request.responderAbort.signal;
    }
  };
  return message[REQUEST] = request, message;
}
function createReplyPromise(request, options) {
  return new Promise((resolve, reject) => {
    const deliver = (outcome) => outcome.ok ? resolve(outcome.value) : reject(outcome.error);
    if (request.capturedOutcome) {
      request.responderAbort.abort(), deliver(request.capturedOutcome);
      return;
    }
    if (!options.hadResponder) {
      reject(new MessageBusError("NO_RESPONDER"));
      return;
    }
    const timeoutMs = options.timeout === void 0 ? DEFAULT_TIMEOUT_MS : options.timeout;
    let timer;
    const onAbort = () => settle(request, { ok: !1, error: new MessageBusError("ABORTED") });
    request.deliver = (outcome) => {
      timer !== void 0 && clearTimeout(timer), options.signal?.removeEventListener("abort", onAbort), request.responderAbort.abort(), deliver(outcome);
    }, timeoutMs !== null && (timer = setTimeout(
      () => settle(request, { ok: !1, error: new MessageBusError("TIMEOUT") }),
      timeoutMs
    )), options.signal && (options.signal.aborted ? onAbort() : options.signal.addEventListener("abort", onAbort, { once: !0 }));
  });
}
function emitEvent(registry, type, payload, options, appId) {
  const hadResponder = (registry.responderCounts.get(type) ?? 0) > 0, request = {
    responderAbort: new AbortController(),
    settled: !1
  };
  resolveEventSubject(registry, type).next(
    createMessage(appId, type, payload, request)
  );
  const awaitReply = () => request.replyPromise ??= createReplyPromise(request, {
    ...options,
    hadResponder
  });
  return {
    // oxlint-disable-next-line unicorn/no-thenable -- intentional PromiseLike: awaiting it arms the failure machinery; not awaiting is fire-and-forget.
    then: (onFulfilled, onRejected) => awaitReply().then(onFulfilled, onRejected),
    catch: (onRejected) => awaitReply().then(void 0, onRejected),
    finally: (onFinally) => awaitReply().finally(onFinally)
  };
}
function emit(registry, type, payload, options, appId) {
  if (registry.stateSubjects.has(type)) {
    const subject = resolveStateSubject(registry, type);
    Object.is(subject.getValue(), payload) || subject.next(payload);
    return;
  }
  return emitEvent(registry, type, payload, options, appId);
}
function query(registry, type, options) {
  const source = resolveStateSource(registry, type), current = source.getCurrent();
  if (current !== void 0) return Promise.resolve(current);
  const signal = options?.signal;
  return signal?.aborted ? Promise.reject(new MessageBusError("ABORTED")) : new Promise((resolve, reject) => {
    const timer = setTimeout(
      () => reject(new MessageBusError("TIMEOUT", `query("${type}") timed out`)),
      DEFAULT_TIMEOUT_MS
    ), onAbort = () => {
      clearTimeout(timer), reject(new MessageBusError("ABORTED"));
    };
    signal?.addEventListener("abort", onAbort, { once: !0 }), source.firstValue.then((value) => {
      clearTimeout(timer), signal?.removeEventListener("abort", onAbort), resolve(value);
    });
  });
}
function unsubscribeOnAbort(subscription, signal) {
  if (signal) {
    if (signal.aborted) {
      subscription.unsubscribe();
      return;
    }
    signal.addEventListener("abort", () => subscription.unsubscribe(), {
      once: !0
    });
  }
}
function invokeResponder(handler, message) {
  const request = message[REQUEST], fail = (error) => {
    request && settle(request, {
      ok: !1,
      error: new MessageBusError("HANDLER_THREW", void 0, {
        cause: error
      })
    });
  };
  try {
    const result = handler(message);
    result instanceof Promise && result.catch(fail);
  } catch (error) {
    fail(error);
  }
}
function subscribe(registry, type, handler, options) {
  const isState = registry.stateSubjects.has(type);
  if (!handler)
    return isState ? resolveStateSource(registry, type) : resolveEventSubject(registry, type).pipe(
      map((message) => message.payload)
    );
  if (isState) {
    const subscription2 = resolveStateSource(registry, type).subscribe(
      handler
    );
    unsubscribeOnAbort(subscription2, options?.signal);
    return;
  }
  registry.responderCounts.set(
    type,
    (registry.responderCounts.get(type) ?? 0) + 1
  );
  const subscription = resolveEventSubject(registry, type).subscribe(
    (message) => invokeResponder(
      handler,
      message
    )
  );
  subscription.add(() => {
    const count = registry.responderCounts.get(type) ?? 0;
    count > 0 && registry.responderCounts.set(type, count - 1);
  }), unsubscribeOnAbort(subscription, options?.signal);
}
const bundledMigrations = () => new Map(Object.entries(topicMigrations));
function createBus(appId = DEFAULT_APP_ID, config = {}) {
  const registry = {
    appId,
    stateSubjects: /* @__PURE__ */ new Map(),
    stateSources: /* @__PURE__ */ new Map(),
    eventSubjects: /* @__PURE__ */ new Map(),
    responderCounts: /* @__PURE__ */ new Map(),
    appAborts: /* @__PURE__ */ new Map(),
    migrations: config.migrations ?? bundledMigrations()
  }, instance = {
    emit: (type, payload, options) => emit(registry, type, payload, options, registry.appId),
    query: (type, options) => query(registry, type, options),
    subscribe: (type, handler, options) => subscribe(registry, type, handler, options)
  };
  return instance[REGISTRY] = registry, instance[PROTOCOL] = OS_PROTOCOL, instance;
}
function topicVersions(migrations) {
  const versions = /* @__PURE__ */ new Map();
  for (const [topic, steps] of migrations) {
    let version = 1;
    for (const step of steps) step.to > version && (version = step.to);
    versions.set(topic, version);
  }
  return versions;
}
function topicAdapters(coreMigrations, clientMigrations) {
  const coreVersions = topicVersions(coreMigrations), clientVersions = topicVersions(clientMigrations), clientVersion = (type) => clientVersions.get(type) ?? 1, coreVersion = (type) => coreVersions.get(type) ?? 1, chainFor = (type) => (clientVersion(type) > coreVersion(type) ? clientMigrations : coreMigrations).get(type);
  return {
    toCore: (type, value) => applyChain(chainFor(type), value, clientVersion(type), coreVersion(type)),
    toClient: (type, value) => applyChain(chainFor(type), value, coreVersion(type), clientVersion(type))
  };
}
function applyChain(steps, value, from, to) {
  if (!steps || steps.length === 0 || from === to) return value;
  let current = value;
  if (from < to)
    for (let version = from; version < to; version++) {
      const step = steps.find((candidate) => candidate.from === version);
      step && (current = step.up(current));
    }
  else
    for (let version = from; version > to; version--) {
      const step = steps.find((candidate) => candidate.from === version - 1);
      step && (current = step.down(current));
    }
  return current;
}
function projectCurrent(input, project) {
  let lastInput, lastOutput, cached = !1;
  return () => {
    const current = input();
    if (current !== void 0)
      return (!cached || current !== lastInput) && (lastInput = current, lastOutput = project(current), cached = !0), lastOutput;
  };
}
function mapStateSource(source, project) {
  return toStateSource(
    source.pipe(map(project)),
    projectCurrent(source.getCurrent, project)
  );
}
function adaptMessage(message, downPayload) {
  return {
    type: message.type,
    payload: downPayload(message.payload),
    meta: message.meta,
    reply: message.reply,
    get signal() {
      return message.signal;
    }
  };
}
function createInertBus(installed) {
  const fail = () => {
    throw new MessageBusError(
      "PROTOCOL_MISMATCH",
      `installed OS bus speaks protocol ${String(installed)}, this copy speaks ${OS_PROTOCOL}`
    );
  };
  return { emit: fail, query: fail, subscribe: fail };
}
function resolveAppAbort(registry, appId) {
  let controller = registry.appAborts.get(appId);
  return controller || (controller = new AbortController(), registry.appAborts.set(appId, controller)), controller;
}
function disconnectApp(bus, appId) {
  const registry = bus[REGISTRY], controller = registry.appAborts.get(appId);
  controller && (registry.appAborts.delete(appId), controller.abort());
}
function connect(core, config = {}) {
  const appId = config.appId ?? DEFAULT_APP_ID, installedProtocol = core[PROTOCOL];
  if (installedProtocol !== OS_PROTOCOL)
    return log.error(
      `OS bus protocol mismatch for "${appId}": installed ${String(installedProtocol)}, this copy speaks ${OS_PROTOCOL}`
    ), createInertBus(installedProtocol);
  const clientMigrations = config.migrations ?? bundledMigrations(), registry = core[REGISTRY], { toCore, toClient } = topicAdapters(
    registry.migrations,
    clientMigrations
  ), isState = (type) => registry.stateSubjects.has(type), appAbort = resolveAppAbort(registry, appId), scoped = (signal) => signal ? AbortSignal.any([signal, appAbort.signal]) : appAbort.signal, clientStreams = /* @__PURE__ */ new Map(), instance = {
    emit: (type, payload, options) => emit(
      registry,
      type,
      toCore(type, payload),
      { ...options, signal: scoped(options?.signal) },
      appId
    ),
    query: (type, options) => query(registry, type, { signal: scoped(options?.signal) }).then(
      (value) => toClient(type, value)
    ),
    subscribe: (type, handler, options) => {
      if (!handler) {
        let stream = clientStreams.get(type);
        if (!stream) {
          const raw = subscribe(registry, type, void 0, options);
          stream = isState(type) ? mapStateSource(
            raw,
            (value) => toClient(type, value)
          ) : raw.pipe(
            map((payload) => toClient(type, payload))
          ), clientStreams.set(type, stream);
        }
        return stream;
      }
      const adapted = isState(type) ? (value) => handler(toClient(type, value)) : (message) => handler(
        adaptMessage(message, (value) => toClient(type, value))
      );
      return subscribe(registry, type, adapted, {
        signal: scoped(options?.signal)
      });
    }
  };
  return instance[REGISTRY] = registry, instance[PROTOCOL] = OS_PROTOCOL, instance;
}
function installBus() {
  const globals = globalThis, existing = globals[BUS];
  return existing && PROTOCOL in existing ? existing : (existing && log.warn("replaced a foreign value at the OS bus install key"), globals[BUS] = createBus());
}
const os = connect(installBus(), {
  appId: typeof __SANITY_APP_ID__ == "string" ? __SANITY_APP_ID__ : DEFAULT_APP_ID
});
function defineStateTopics(target, topics) {
  const registry = target[REGISTRY];
  for (const [name, seed] of Object.entries(topics)) {
    const existing = registry.stateSubjects.get(name);
    existing ? seed !== void 0 && !Object.is(existing.getValue(), seed) && existing.next(seed) : registry.stateSubjects.set(
      name,
      new BehaviorSubject(seed === void 0 ? NO_VALUE : seed)
    );
  }
}
export {
  MessageBusError,
  connect,
  defineStateTopics,
  disconnectApp,
  os
};
