// @sanity/workbench/dist/_chunks-es/index.js
const LEVELS = ["none", "error", "warn", "info", "debug"];
function createLogger({
  namespace,
  context: baseContext,
  logLevel = "info"
} = {}) {
  function isLevelEnabled(level) {
    return LEVELS.indexOf(level) <= LEVELS.indexOf(logLevel);
  }
  function logAtLevel(level, message, context) {
    if (!isLevelEnabled(level)) return;
    const merged = baseContext ?? context ? { ...baseContext, ...context } : void 0, args = [
      ...namespace ? [`[${namespace}]`] : [],
      message,
      ...merged ? [merged] : []
    ];
    level === "error" ? console.error(...args) : level === "warn" ? console.warn(...args) : level === "info" ? console.info(...args) : console.debug(...args);
  }
  return {
    error: (message, context) => logAtLevel("error", message, context),
    warn: (message, context) => logAtLevel("warn", message, context),
    info: (message, context) => logAtLevel("info", message, context),
    debug: (message, context) => logAtLevel("debug", message, context),
    child: (domain, context) => createLogger({
      logLevel,
      namespace: namespace ? `${namespace}:${domain}` : domain,
      context: baseContext || context ? { ...baseContext, ...context } : void 0
    })
  };
}
const logger = createLogger({
  namespace: "sanity-workbench",
  logLevel: "debug"
});
export {
  createLogger,
  logger
};
