// @sanity/workbench/dist/_chunks-es/module-federation.js
import { createInstance as createInstance$1 } from "@module-federation/runtime";
function createInstance(options) {
  const instance = createInstance$1({
    remotes: [],
    ...options
  });
  return {
    registerRemotes: (remotes) => instance.registerRemotes(remotes, { force: !1 }),
    loadRemote: async (id) => {
      let remoteModule;
      try {
        remoteModule = await instance.loadRemote(id);
      } catch (error) {
        throw new Error(`Failed to load remote module "${id}"`, {
          cause: error
        });
      }
      return remoteModule;
    }
  };
}
function log(onDebug) {
  const logEvent = (eventName, args) => {
    const { origin: _origin, ...data } = typeof args == "object" && args !== null ? args : { args };
    onDebug(`[Lifecycle] ${eventName}`, data);
  };
  return {
    name: "sanity-logger",
    beforeInit(args) {
      return logEvent("beforeInit", args), args;
    },
    beforeRegisterRemote(args) {
      return logEvent("beforeRegisterRemote", args), args;
    },
    beforePreloadRemote(args) {
      logEvent("beforePreloadRemote", args);
    },
    beforeRequest(args) {
      return logEvent("beforeRequest", args), args;
    },
    afterResolve(args) {
      return logEvent("afterResolve", args), args;
    },
    onLoad(args) {
      return logEvent("onLoad", args), args;
    },
    loadShare(args) {
      logEvent("loadShare", args);
    },
    beforeLoadShare(args) {
      return logEvent("beforeLoadShare", args), args;
    }
  };
}
export {
  createInstance,
  log
};
