// @sanity/workbench/dist/_chunks-es/studio.js
import { z } from "zod";
import { coerce, isValid, isLess, isGreater, isGreaterOrEqual, sortReversed } from "verkit";
const MediaLibraryConfigFieldSchema = z.object({
  name: z.string(),
  title: z.string(),
  public: z.boolean().optional()
}), MediaLibraryConfigModuleSchema = z.object({
  config: z.object({
    fields: z.array(z.object({ name: z.string(), config: z.unknown() }))
  })
}), AppConfigSchema = z.discriminatedUnion("appType", [
  z.object({
    appType: z.literal("media-library"),
    config: z.object({
      fields: z.array(MediaLibraryConfigFieldSchema)
    })
  })
]);
function resolveAppConfig(appConfig, module) {
  const parsed = MediaLibraryConfigModuleSchema.safeParse(module);
  if (!parsed.success) return null;
  const valuesByName = new Map(
    parsed.data.config.fields.map((field) => [field.name, field.config])
  );
  return {
    appType: appConfig.appType,
    fields: appConfig.config.fields.map((field) => ({
      ...field,
      config: valuesByName.get(field.name)
    }))
  };
}
const InterfaceBase = z.object({
  id: z.string(),
  name: z.string(),
  title: z.string(),
  version: z.string().optional(),
  moduleId: z.string()
}), AppInterfaceMetadata = z.object({ group: z.string().optional(), priority: z.number().optional() }).nullable(), PanelInterfaceSchema = InterfaceBase.extend({ type: z.literal("panel") }), WorkerInterfaceSchema = InterfaceBase.extend({
  type: z.literal("worker")
}), AppInterfaceSchema = InterfaceBase.extend({
  type: z.literal("app"),
  metadata: AppInterfaceMetadata
}), InterfaceSchema = z.discriminatedUnion("type", [
  PanelInterfaceSchema,
  WorkerInterfaceSchema,
  AppInterfaceSchema
]), LocalInterfaceSchema = z.discriminatedUnion("type", [
  PanelInterfaceSchema.extend({ src: z.string() }),
  WorkerInterfaceSchema.extend({ src: z.string() }),
  AppInterfaceSchema.extend({ src: z.string() })
]), LocalApplicationBase = z.object({
  host: z.string(),
  port: z.number(),
  /**
   * Interfaces the app exposes — dock panels (`type: "panel"`) and
   * background workers (`type: "worker"`), forwarded from the dev
   * server in one list.
   */
  interfaces: z.array(LocalInterfaceSchema).optional(),
  /** The `deployment.appId` from the application's `sanity.cli.ts`, when set. */
  id: z.string().optional(),
  /**
   * The `api.projectId` from the application's `sanity.cli.ts`. Available
   * synchronously at dev-server startup (no manifest extraction required), so
   * the studio's primary project is resolvable from the very first local
   * application event.
   */
  projectId: z.string().optional()
}), LocalApplication = z.discriminatedUnion("type", [
  LocalApplicationBase.extend({
    type: z.literal("studio"),
    manifest: z.custom().optional()
  }),
  LocalApplicationBase.extend({
    type: z.literal("coreApp"),
    manifest: z.custom().optional()
  })
]);
class AbstractApplication {
  type;
  constructor(type) {
    this.type = type;
  }
  /**
   * Whether the application is served by a local CLI dev server rather than a
   * deployed remote. Defaults to `false`; applications flip this when
   * constructed from a `LocalApplication` payload.
   */
  get isLocal() {
    return !1;
  }
  interfaces(_type) {
    return [];
  }
  get config() {
    throw new Error("App configs are not supported yet");
  }
  /**
   * Whether the application has a navigable full-page `app` view (US5, spec 002-workbench-extension-api). Defaults
   * to `true` — studios, Canvas, and Media Library are always navigable. An SDK
   * app overrides this to derive it from whether it declares an `app` interface.
   */
  get hasAppView() {
    return !0;
  }
  /**
   * Federation module id of the app's full-page view, or `null` when it can't
   * be federation-loaded: the app isn't federated (Canvas, Media Library,
   * deployed apps shown in an iframe) or has no app view. Only a non-null
   * `moduleId` is fetched and rendered by the remotes machine.
   */
  get moduleId() {
    return this.isFederated && this.hasAppView ? `${this.id}/App` : null;
  }
  /**
   * Federation module id of one of this app's panel view components, or `null`
   * when the app isn't federated (a non-federated app's panels can't be loaded
   * as remotes).
   */
  resolveViewModuleId(view, component) {
    return this.isFederated ? `${this.id}/views/${view.name}/${component}` : null;
  }
  get initials() {
    const SYMBOLS = /[^\p{Alpha}\p{N}\p{White_Space}]/gu, WHITESPACE = new RegExp("\\p{White_Space}+", "u"), ALPHANUMERIC_SEGMENTS = new RegExp("(\\p{N}+|\\p{Alpha}+)", "gu"), IS_NUMERIC = new RegExp("^\\p{N}+$", "u");
    if (!this.title) return "";
    const namesArray = this.title.replace(SYMBOLS, "").split(WHITESPACE).filter(Boolean);
    if (namesArray.length === 0) return "";
    if (namesArray.length === 1) {
      const word = namesArray[0], segments = word.match(ALPHANUMERIC_SEGMENTS) || [];
      return segments.length === 0 ? "" : segments.length === 1 ? word.length === 1 ? word.toUpperCase() : IS_NUMERIC.test(word) ? `${word.charAt(0)}${word.charAt(1)}`.toUpperCase() : word.charAt(0).toUpperCase() : `${segments[0].charAt(0)}${segments[1].charAt(0)}`.toUpperCase();
    }
    return `${namesArray[0].charAt(0)}${namesArray[namesArray.length - 1].charAt(0)}`.toUpperCase();
  }
}
function getSanityDomain() {
  return getSanityEnv() === "staging" ? "sanity.work" : "sanity.io";
}
function getApiHost() {
  return `https://api.${getSanityDomain()}`;
}
function getApplicationDomain() {
  return getSanityEnv() === "production" ? "sanity.run" : "run.sanity.work";
}
function getSanityEnv() {
  return globalThis.__SANITY_STAGING__ === !0 ? "staging" : "production";
}
const OrganizationId = z.string().nonempty().brand("OrganizationId");
function brandOrganizationId(id) {
  return OrganizationId.parse(id);
}
const OrganizationMemberSchema = z.object({
  sanityUserId: z.string(),
  isCurrentUser: z.boolean(),
  user: z.object({
    id: z.string(),
    displayName: z.string(),
    familyName: z.string(),
    givenName: z.string(),
    middleName: z.string().nullable(),
    imageUrl: z.string().nullable(),
    email: z.string(),
    loginProvider: z.string()
  }),
  roles: z.array(
    z.object({
      name: z.string(),
      title: z.string(),
      description: z.string().optional()
    })
  )
}), Organization = z.object({
  id: OrganizationId,
  name: z.string(),
  slug: z.string().nullable(),
  createdAt: z.string(),
  updatedAt: z.string(),
  dashboardStatus: z.enum(["enabled", "disabled"]),
  aiFeaturesStatus: z.enum(["enabled", "disabled"]),
  defaultRoleName: z.string().nullable()
});
function parseOrganization(data, options) {
  const includeMembers = options?.includeMembers ?? !0, includeFeatures = options?.includeFeatures ?? !0, extensions = {
    ...includeMembers && {
      members: z.array(OrganizationMemberSchema)
    },
    ...includeFeatures && {
      features: z.array(z.string())
    }
  };
  return (Object.keys(extensions).length > 0 ? Organization.extend(extensions) : Organization).parse(data);
}
const ProjectId = z.string().nonempty().brand("ProjectId");
function brandProjectId(id) {
  return ProjectId.parse(id);
}
const ProjectMemberSchema = z.object({
  id: z.string(),
  createdAt: z.string(),
  updatedAt: z.string(),
  isCurrentUser: z.boolean(),
  isRobot: z.boolean(),
  roles: z.array(
    z.object({
      name: z.string(),
      title: z.string(),
      description: z.string()
    })
  )
}), Project = z.object({
  id: ProjectId,
  displayName: z.string(),
  studioHost: z.string().nullable(),
  organizationId: OrganizationId,
  metadata: z.object({
    color: z.string().optional(),
    externalStudioHost: z.string().optional().nullable(),
    initialTemplate: z.string().optional(),
    cliInitializedAt: z.string().optional(),
    integration: z.string().optional()
  }),
  isBlocked: z.boolean(),
  isDisabled: z.boolean(),
  isDisabledByUser: z.boolean(),
  activityFeedEnabled: z.boolean(),
  createdAt: z.string(),
  updatedAt: z.string()
});
function parseProject(data, options) {
  const includeMembers = options?.includeMembers ?? !0, includeFeatures = options?.includeFeatures ?? !0, extensions = {
    ...includeMembers && { members: z.array(ProjectMemberSchema) },
    ...includeFeatures && { features: z.array(z.string()) }
  };
  return (Object.keys(extensions).length > 0 ? Project.extend(extensions) : Project).parse(data);
}
const ApplicationId = z.string().nonempty().brand("ApplicationId");
function brandApplicationId(id) {
  return ApplicationId.parse(id);
}
const BrettWorkspace = z.object({
  id: z.string(),
  name: z.string(),
  title: z.string().nullable(),
  subtitle: z.string().nullable(),
  projectId: ProjectId,
  dataset: z.string(),
  schemaDescriptorId: z.string().nullable(),
  basePath: z.string().nullable(),
  icon: z.string().nullable()
}), ActiveDeployment = z.object({
  id: z.string(),
  applicationId: ApplicationId,
  size: z.number().nullable(),
  version: z.string().nullable(),
  isAutoUpdating: z.boolean().nullable(),
  isActiveDeployment: z.boolean(),
  deployedBy: z.string().nullable(),
  createdAt: z.string(),
  updatedAt: z.string(),
  interfaces: z.array(InterfaceSchema).optional(),
  workspaces: z.array(BrettWorkspace).nullable().optional()
}), applicationAttributes = z.object({
  id: ApplicationId,
  organizationId: OrganizationId,
  isSingleton: z.boolean(),
  createdAt: z.string(),
  updatedAt: z.string(),
  activeDeployment: ActiveDeployment.nullable().optional()
}), InternalApplication = applicationAttributes.extend({
  slug: z.string(),
  externalUrl: z.null()
}), ExternalApplication = applicationAttributes.extend({
  slug: z.null(),
  externalUrl: z.string()
}), ApplicationBase = z.union([
  InternalApplication,
  ExternalApplication
]);
class BrettApplication extends AbstractApplication {
  application;
  id;
  /**
   * For local applications (`isLocal === true`), the deployed application
   * that shares the same `id` — if one was passed at construction. `null`
   * for deployed applications or when no remote twin was provided.
   */
  remoteApplication;
  #isLocal;
  constructor(application, type, options = {}) {
    super(type), this.application = application, this.id = brandApplicationId(application.id), this.#isLocal = options.isLocal ?? !1, this.remoteApplication = options.remoteApplication ?? null;
  }
  /**
   * Local dev servers and deployed federation apps (they serve an
   * mf-manifest) render as federated remotes; everything else in an iframe.
   */
  get isFederated() {
    return this.isLocal || !!this.application.config?.mfManifest;
  }
  get isLocal() {
    return this.#isLocal;
  }
  interfaces(type) {
    const all = this.application.activeDeployment?.interfaces ?? [];
    return type ? all.filter((iface) => iface.type === type) : all;
  }
  /**
   * @returns A fully resolved URL instance for the application.
   * For Sanity-hosted applications, constructs a URL from the `slug` using the
   * studio domain pattern resolved from the environment at the consuming
   * app's build time. For externally hosted applications (including local dev
   * servers), returns the `externalUrl` as-is.
   */
  get url() {
    const application = this.application;
    return application.externalUrl !== null ? new URL(application.externalUrl) : this.isFederated && application.isSingleton ? new URL(
      `https://${application.slug}-apps-${application.organizationId}.${getApplicationDomain()}`
    ) : getSanityEnv() === "production" ? new URL(`https://${application.slug}.sanity.studio`) : new URL(`https://${application.slug}.studio.${getSanityDomain()}`);
  }
}
const CanvasIdSchema = z.string().nonempty().brand("CanvasId");
function brandCanvasId(id) {
  return CanvasIdSchema.parse(id);
}
const CanvasSchema = z.object({
  id: CanvasIdSchema,
  organizationId: OrganizationId,
  status: z.enum(["active", "provisioning"])
});
function parseCanvas(data) {
  return CanvasSchema.parse(data);
}
class CanvasApplication extends AbstractApplication {
  canvas;
  constructor(canvas) {
    super("canvas"), this.canvas = canvas;
  }
  get id() {
    return this.canvas.id;
  }
  get href() {
    return "canvas";
  }
  get title() {
    return "Canvas";
  }
  get url() {
    return new URL(
      `https://canvas.${getSanityDomain()}/o/${this.canvas.organizationId}`
    );
  }
  get isFederated() {
    return !1;
  }
}
const InstallationId = z.string().nonempty().brand("InstallationId"), InstallationSchema = z.object({
  id: InstallationId,
  applicationId: ApplicationId,
  organizationId: OrganizationId,
  application: z.object({
    title: z.string(),
    slug: z.string().nullable()
  }),
  access: z.array(
    z.object({ resourceType: z.string(), resourceId: z.string() })
  )
});
function parseInstalledApplication(data) {
  return InstallationSchema.parse(data);
}
class InstalledApplication extends AbstractApplication {
  id;
  organizationId;
  slug;
  constructor({
    type,
    slug,
    id,
    organizationId
  }) {
    super(type), this.slug = slug, this.id = id, this.organizationId = organizationId;
  }
  get isFederated() {
    return !1;
  }
}
const joinUrlPaths = (...paths) => {
  let nextPath = null;
  const safeJoinSegments = (segment1, segment2) => segment1 ? segment2 ? segment1.endsWith("/") && segment2.startsWith("/") ? segment1 + segment2.slice(1) : segment1.endsWith("/") || segment2.startsWith("/") ? segment1 + segment2 : `${segment1}/${segment2}` : segment1 : segment2 || "", validPaths = paths.filter(
    (path) => path != null
  );
  for (const path of validPaths)
    nextPath = safeJoinSegments(
      nextPath,
      path instanceof URL ? path.pathname : path
    );
  return nextPath ?? "";
};
function normalizePath(pathname) {
  return pathname.startsWith("/") || (pathname = `/${pathname}`), pathname !== "/" && pathname.endsWith("/") && (pathname = pathname.slice(0, -1)), pathname;
}
const LocalCoreApplicationManifest = z.object({
  version: z.string(),
  icon: z.string().optional(),
  title: z.string().optional()
}), internalCoreApplication = InternalApplication.extend({
  type: z.literal("coreApp"),
  title: z.string(),
  config: z.object({ mfManifest: z.unknown().optional() }).optional(),
  manifest: LocalCoreApplicationManifest.nullable().optional()
});
ExternalApplication.extend({
  type: z.literal("coreApp"),
  title: z.string(),
  config: z.object({ mfManifest: z.unknown().optional() }).optional(),
  manifest: LocalCoreApplicationManifest.nullable().optional()
});
function parseCoreApplication(data) {
  return internalCoreApplication.parse(data);
}
class CoreApplication extends BrettApplication {
  activeDeployment;
  constructor(application, options = {}) {
    super(application, "coreApp", options), this.activeDeployment = application.activeDeployment;
  }
  // Typed `string | null` so the UI-aware remote subclass can override it to
  // `null` for a non-navigable app (US5, spec 002-workbench-extension-api); the core route itself is always set.
  get href() {
    return this.isLocal ? `/local/${this.id}` : `/application/${this.application.id}`;
  }
  get title() {
    return this.manifest?.title ?? this.application.title;
  }
  /**
   * The manifest a local CLI dev server forwards. Deployed apps carry none —
   * their dock metadata rides the `app` interface (see {@link group} and
   * {@link priority}), and icons have no slot on brett's payload yet, so
   * deployed apps fall back to the avatar.
   */
  get manifest() {
    return this.application.manifest ?? null;
  }
  get subtitle() {
  }
  get updatedAt() {
    return this.application.updatedAt;
  }
  get(attr) {
    if (!(attr in this.application))
      throw new Error(
        `Attribute ${attr.toString()} does not exist on application ${this.application.id}`
      );
    return this.application[attr];
  }
}
const WorkspaceSchema = z.object({
  name: z.string(),
  title: z.string(),
  subtitle: z.string().nullable().optional(),
  basePath: z.string(),
  projectId: ProjectId,
  dataset: z.string().optional(),
  icon: z.string().nullable().optional()
}), LocalStudioApplicationManifest = z.object({
  version: z.number(),
  createdAt: z.string(),
  studioVersion: z.string().optional(),
  workspaces: z.array(
    WorkspaceSchema.extend({
      dataset: z.string(),
      schema: z.string(),
      tools: z.string().optional()
    })
  )
}), internalStudioApplication = InternalApplication.extend({
  type: z.literal("studio"),
  title: z.string(),
  config: z.object({
    studio: z.object({
      projectId: ProjectId,
      autoUpdatingVersion: z.string().nullable(),
      dashboardStatus: z.enum(["default", "disabled"])
    }),
    mfManifest: z.unknown().optional()
  })
}), externalStudioApplication = ExternalApplication.extend({
  type: z.literal("studio"),
  title: z.string(),
  config: z.object({
    studio: z.object({
      projectId: ProjectId,
      autoUpdatingVersion: z.string().nullable(),
      dashboardStatus: z.enum(["default", "disabled"])
    }),
    mfManifest: z.unknown().optional()
  })
}), Studio = z.union([
  internalStudioApplication,
  externalStudioApplication
]);
function parseStudioApplication(data) {
  return Studio.parse(data);
}
class StudioWorkspace extends AbstractApplication {
  /**
   * Workspaces always belong to a studio application.
   * They do not exist on their own & therefore can access
   * information about the studio they're in via this property.
   */
  studioApplication;
  workspace;
  project;
  isDefaultWorkspace;
  constructor(studioApplication, workspace, project, isDefaultWorkspace) {
    super("workspace"), this.studioApplication = studioApplication, this.workspace = workspace, this.project = project, this.isDefaultWorkspace = isDefaultWorkspace;
  }
  /**
   * The studio application that this workspace belongs to.
   */
  get studio() {
    return this.studioApplication;
  }
  get id() {
    return StudioWorkspace.makeId(this.studio.id, this.workspace.name);
  }
  get href() {
    return joinUrlPaths(this.studio.href, this.workspace.name);
  }
  get title() {
    return this.isDefaultWorkspace ? this.project.displayName : this.workspace.title;
  }
  get subtitle() {
    return this.isDefaultWorkspace ? this.studio.url.hostname : this.get("subtitle");
  }
  get(attr) {
    return this.workspace[attr];
  }
  get isFederated() {
    return this.studio.isFederated;
  }
  /**
   * @returns the URL to the workspace of a studio application.
   */
  get url() {
    const studioUrl = new URL(this.studio.url), normalizedUrlPath = normalizePath(studioUrl.pathname);
    let finalBasePath = normalizePath(this.get("basePath"));
    return finalBasePath.startsWith(normalizedUrlPath) && (finalBasePath = finalBasePath.slice(normalizedUrlPath.length)), studioUrl.pathname = joinUrlPaths(normalizedUrlPath, finalBasePath), studioUrl;
  }
  static makeId(applicationId, workspaceName) {
    return `${applicationId}-${workspaceName}`;
  }
  static splitId(id) {
    return id.split(new RegExp(/-(.*)/)).slice(0, 2);
  }
}
const DEFAULT_WORKSPACE_DATA = {
  name: "default",
  title: "Default",
  basePath: "/"
};
class StudioApplication extends BrettApplication {
  /**
   * Returns a list of studio workspaces based on the application manifest.
   * If there is no manifest, or alternatively it is not valid, then we create a default workspace.
   */
  workspaces = [];
  project;
  /**
   * The projects actually referenced by this studio — the application's own
   * project plus any project used by a workspace. Preserved so the instance
   * can be re-constructed (e.g. by dock wrappers) without having to thread
   * the full organization project list through again.
   */
  projects;
  /**
   * @param application - The studio application to create a list of workspaces for
   * @param projects - The projects available in the organization. It's not enough to just pass
   * the project that associates with the application because that is the project the app is deployed in relation to.
   * The workspaces may have different projects completely.
   */
  constructor(application, projects, options = {}) {
    super(application, "studio", options);
    let workspaces = [];
    const deployedWorkspaces = application.activeDeployment?.workspaces;
    deployedWorkspaces?.length && (workspaces = deployedWorkspaces.map((workspace) => ({
      name: workspace.name,
      title: workspace.title ?? workspace.name,
      subtitle: workspace.subtitle,
      basePath: workspace.basePath ?? "/",
      projectId: workspace.projectId,
      dataset: workspace.dataset,
      icon: workspace.icon
    })));
    const workspacesWithProjectsMap = workspaces.reduce((acc, workspace) => {
      const project2 = projects.find((p) => p.id === workspace.projectId);
      return project2 ? acc.set(workspace, project2) : console.warn(
        `Project not found for application ${application.id} and workspace ${workspace.name}. This workspace has been omitted.`
      ), acc;
    }, /* @__PURE__ */ new Map()), projectId = application.config.studio.projectId, project = projects.find((p) => p.id === projectId);
    if (!project)
      throw new Error(`Project not found for application ${application.id}`);
    workspacesWithProjectsMap.size === 0 && workspacesWithProjectsMap.set(
      {
        ...DEFAULT_WORKSPACE_DATA,
        projectId
      },
      project
    ), this.workspaces = Object.freeze(
      Array.from(workspacesWithProjectsMap.entries()).map(([workspace, p]) => {
        const isDefaultWorkspace = workspace.name === DEFAULT_WORKSPACE_DATA.name && workspace.basePath === DEFAULT_WORKSPACE_DATA.basePath && workspace.title === DEFAULT_WORKSPACE_DATA.title;
        return this.createWorkspace(workspace, p, isDefaultWorkspace);
      })
    ), this.project = project;
    const usedProjects = /* @__PURE__ */ new Set([project]);
    for (const workspace of this.workspaces)
      usedProjects.add(workspace.project);
    this.projects = Array.from(usedProjects);
  }
  /**
   * Factory hook called for each workspace during construction. Subclasses
   * override this to substitute a `StudioWorkspace` subclass (e.g. a UI-aware
   * variant in a downstream package) without re-implementing the constructor's
   * workspace-derivation logic.
   */
  createWorkspace(workspace, project, isDefaultWorkspace) {
    return new StudioWorkspace(this, workspace, project, isDefaultWorkspace);
  }
  get href() {
    return this.isLocal ? `/local/${this.id}` : `/studio/${this.application.id}`;
  }
  get title() {
    return this.get("title") || (this.workspaces.length > 1 && this.get("externalUrl") === null ? this.project.displayName : this.workspaces[0].title);
  }
  get subtitle() {
    return new URL(this.url).hostname;
  }
  get(attr) {
    if (!(attr in this.application))
      throw new Error(
        `Attribute ${attr.toString()} does not exist on studio ${this.application.id}`
      );
    return this.application[attr];
  }
  get hasManifest() {
    return !!this.application.activeDeployment?.workspaces?.length;
  }
  get hasSchema() {
    const deployedWorkspaces = this.application.activeDeployment?.workspaces;
    return deployedWorkspaces?.length ? deployedWorkspaces.every((w) => w.schemaDescriptorId !== null) : !1;
  }
  resolveVersion() {
    const version = this.get("activeDeployment")?.version, coerced = version ? coerce(version) : null;
    return !coerced || !isValid(coerced) ? null : coerced;
  }
  get version() {
    return this.resolveVersion();
  }
  get compatibilityStatus() {
    return StudioApplication.resolveCompatibilityStatus(this);
  }
  /**
   * Used to calculate the compatibility status of a given studio application.
   * Optionally if you've resolved the version elsewhere provide that value to
   * get the new compatibility status without mutating the application.
   */
  static resolveCompatibilityStatus(application, version = application.version) {
    return version === null || isLess(version, StudioApplication.MinimumStudioVersion) ? StudioApplication.CompatibilityStatuses.UNKNOWN : !application.hasSchema || !application.hasManifest || StudioApplication.resolveIssues(application, version).length > 0 ? StudioApplication.CompatibilityStatuses.PARTIALLY_COMPATIBLE : StudioApplication.CompatibilityStatuses.FULLY_COMPATIBLE;
  }
  get isAutoRedirecting() {
    return StudioApplication.resolveIsAutoRedirecting(this);
  }
  /**
   * Mirrors the `isRedirectable` function from Saison defined in
   * https://github.com/sanity-io/saison/blob/83556405d23e07f6d3a71c76249c67e33fe1101f/src/utils/applications.ts
   *
   * Returns whether a studio application is auto-redirecting, meaning it can only be accessed in the context of
   * Dashboard.
   */
  static resolveIsAutoRedirecting(application, versionArg = application.version) {
    let version = versionArg;
    if (application.get("externalUrl") !== null || !version)
      return !1;
    if (!application.get("activeDeployment")?.isAutoUpdating)
      return isGreater(version, "3.92.0");
    const autoUpdatingVersion = application.get("config").studio.autoUpdatingVersion;
    if (autoUpdatingVersion) {
      if (["next", "stable", "latest"].includes(autoUpdatingVersion))
        return !0;
      const autoUpdatingVersionPinnedVersion = coerce(autoUpdatingVersion);
      autoUpdatingVersionPinnedVersion && (version = autoUpdatingVersionPinnedVersion);
    }
    return isGreater(version, StudioApplication.MinimumStudioVersion);
  }
  /**
   * Returns a list of issues that prevent the studio from functioning properly in the Dashboard.
   * This static value depends on the version of the studio that comes from the `activeDeployment` property.
   * As such, if the studio is auto-updating, this list will be incorrect & instead you should use
   * the static method `resolveIssues` to get the correct issues by passing the resolved version.
   */
  get issues() {
    return StudioApplication.resolveIssues(this);
  }
  static resolveIssues(application, version = application.version) {
    const issues = StudioApplication.Features.filter(
      (feature) => !application.isFeatureSupported(feature.id, version)
    );
    return application.hasManifest || issues.push({
      id: StudioApplication.StudioIssues.ISSUE_MANIFEST
    }), issues;
  }
  isFeatureSupported(feature, version = this.version) {
    const featureVersion = StudioApplication.Features.find(
      (_) => _.id === feature
    )?.version;
    return !featureVersion || !version ? !1 : isGreaterOrEqual(version, featureVersion);
  }
  static CompatibilityStatuses = {
    UNKNOWN: "unknown",
    PARTIALLY_COMPATIBLE: "partially-compatible",
    FULLY_COMPATIBLE: "fully-compatible"
  };
  static StudioIssues = {
    ISSUE_ACTIVITY: "ACTIVITY",
    ISSUE_AGENT: "AGENT",
    ISSUE_FAVORITES: "FAVORITES",
    ISSUE_URL_SYNCING: "URL_SYNCING",
    ISSUE_UI_ADJUSTMENT: "UI_ADJUSTMENT",
    ISSUE_CONTENT_MAPPING: "CONTENT_MAPPING",
    ISSUE_LOGIN: "LOGIN",
    ISSUE_MANIFEST: "MANIFEST"
  };
  static Features = [
    {
      id: StudioApplication.StudioIssues.ISSUE_AGENT,
      version: "5.1.0"
    },
    {
      id: StudioApplication.StudioIssues.ISSUE_FAVORITES,
      version: "3.88.1"
    },
    {
      id: StudioApplication.StudioIssues.ISSUE_ACTIVITY,
      version: "3.88.1"
    },
    {
      id: StudioApplication.StudioIssues.ISSUE_URL_SYNCING,
      version: "3.75.0"
    },
    {
      id: StudioApplication.StudioIssues.ISSUE_UI_ADJUSTMENT,
      version: "3.78.1"
    },
    {
      id: StudioApplication.StudioIssues.ISSUE_CONTENT_MAPPING,
      version: "3.68.0"
    },
    {
      id: StudioApplication.StudioIssues.ISSUE_LOGIN,
      version: "2.28.0"
    }
  ];
  static MinimumStudioVersion = "2.28.0";
  static MinimumStudioVersionWithNoIssues = sortReversed(
    StudioApplication.Features.map((feature) => feature.version)
  ).at(0);
}
export {
  AbstractApplication,
  ActiveDeployment,
  AppConfigSchema,
  ApplicationBase,
  ApplicationId,
  BrettApplication,
  BrettWorkspace,
  CanvasApplication,
  CoreApplication,
  ExternalApplication,
  InstallationId,
  InstalledApplication,
  InterfaceSchema,
  InternalApplication,
  LocalApplication,
  LocalCoreApplicationManifest,
  LocalInterfaceSchema,
  LocalStudioApplicationManifest,
  Organization,
  OrganizationId,
  Project,
  ProjectId,
  Studio,
  StudioApplication,
  StudioWorkspace,
  brandApplicationId,
  brandCanvasId,
  brandOrganizationId,
  brandProjectId,
  getApiHost,
  getApplicationDomain,
  getSanityDomain,
  getSanityEnv,
  joinUrlPaths,
  parseCanvas,
  parseCoreApplication,
  parseInstalledApplication,
  parseOrganization,
  parseProject,
  parseStudioApplication,
  resolveAppConfig
};
