// @portabletext/editor/lib/traversal/index.js
import { getTextBlock } from "../_chunks-es/get-container.js";
import { getAncestor, getContainer, getFirstChild, getLastChild, getLeaf, getSpan, getText, getUnionSchema, isLeafObject, pathContains, rangeIntersects } from "../_chunks-es/get-container.js";
import { comparePoints, getAncestors, getBlock, getEnclosingBlock, getPathSubSchema, getSibling, hasNode, isBlock, isInline, isObject } from "../_chunks-es/get-path-sub-schema.js";
import { isKeyedSegment } from "../_chunks-es/get-parent.js";
import { getChildren, getContainerChildren, getNode, getParent } from "../_chunks-es/get-parent.js";
function getAnnotation(snapshot, path) {
  for (let i = 0; i < path.length - 1; i++) {
    if (path[i] !== "markDefs")
      continue;
    const annotationSegment = path[i + 1];
    if (!isKeyedSegment(annotationSegment))
      continue;
    const block = getTextBlock(snapshot, path.slice(0, i));
    if (!block)
      continue;
    const annotation = block.node.markDefs?.find((markDef) => markDef._key === annotationSegment._key);
    return annotation ? {
      node: annotation,
      path: [...block.path, "markDefs", {
        _key: annotation._key
      }]
    } : void 0;
  }
}
export {
  comparePoints,
  getAncestor,
  getAncestors,
  getAnnotation,
  getBlock,
  getChildren,
  getContainer,
  getContainerChildren,
  getEnclosingBlock,
  getFirstChild,
  getLastChild,
  getLeaf,
  getNode,
  getParent,
  getPathSubSchema,
  getSibling,
  getSpan,
  getText,
  getTextBlock,
  getUnionSchema,
  hasNode,
  isBlock,
  isInline,
  isLeafObject,
  isObject,
  pathContains,
  rangeIntersects
};
